---
title: "Upload Tab"
subtitle: "Single Cell Toolkit"
author:
- name: David Jenkins
  affiliation: 
  - The Section of Computational Biomedicine, Boston University School of Medicine, Boston, MA
  - Program in Bioinformatics, Boston University, Boston, MA
  email: dfj@bu.edu
- name: Tyler Faits
  affiliation: 
  - The Section of Computational Biomedicine, Boston University School of Medicine, Boston, MA
  - Program in Bioinformatics, Boston University, Boston, MA
- name: W. Evan Johnson
  affiliation: 
  - The Section of Computational Biomedicine, Boston University School of Medicine, Boston, MA
  - Program in Bioinformatics, Boston University, Boston, MA
package: singleCellTK
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{Primer}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

![](img/upload_01.png)

# Introduction

The upload tab is the first tab displayed when the singleCellTK is loaded. Here,
you can upload a matrix of data along with annotation information to explore
within the toolkit.

Under the hood, the SCTK created a SCtkExperiment object which stores one or
more matrix of data along with a column annotation data frame and a row
annotation data frame. The SCtkExperiment is an extension of the
SingleCellExperiment object.

# Input Data Format

The SCTK accepts tab-delimited files on the upload tab. You must provide a
matrix of count data, and you can optionally provide both sample and feature
annotation files as well. All files must have the same cells or genes and must
be in the same order.

## Assay

The input assay should be a matrix with genes as rows and samples (cells) as
columns. For most users, the input assay will be a matrix of counts, but if the
assay data being uploaded is not a count matrix, you can change the input assay
type with the "Input Assay Type" drop-down. Download an example count file [here](https://drive.google.com/open?id=1n0CtM6phfkWX0O6xRtgPPg6QuPFP6pY8).

| Gene  | Cell1 | Cell2 | ... | CellN |
|-------|-------|-------|-----|-------|
| Gene1 | 0     | 0     | ... | 0     |
| Gene2 | 5     | 6     | ... | 0     |
| Gene3 | 4     | 3     | ... | 8     |
| ...   | ...   | ...   | ... | ...   |
| GeneM | 10    | 10    | ... | 10    |

## Sample Annotations (optional)

Sample annotations should be a table with samples (cells) as rows and annotations
as columns. Download an example annotation file [here](https://drive.google.com/open?id=10IDmZQUiASN4wnzO4-WRJQopKvxCNu6J).

| Cell  | Annot1 |  ... |
|-------|--------|------|
| Cell1 | a      |  ... |
| Cell2 | a      |  ... |
| Cell3 | b      |  ... |
| ...   | ...    |  ... |
| CellN | b      |  ... |

## Feature Annotations (optional)

Feature annotations should be a table with genes as rows and gene annotations
as columns. Download an example feature file [here](https://drive.google.com/open?id=1gxXaZPq5Wrn2lNHacEVaCN2a_FHNvs4O).

| Gene  | Annot2 |  ... |
|-------|--------|------|
| Gene1 | a      |  ... |
| Gene2 | a      |  ... |
| Gene3 | b      |  ... |
| ...   | ...    |  ... |
| GeneM | b      |  ... |

# Example Data

You can also load example data sets on the upload tab for testing, see the
package documentation for details about the available datasets.

# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
