<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>enrichR • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="enrichR">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../articles/articles/introduction_to_SCTK.html">General Workflow</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shell Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../../articles/articles/cmd_qc.html">Quality Control</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    R Console Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/cnsl_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_dropletqc.html">Quality Control of droplet data</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_cellqc.html">Quality Control of cell data</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_batch_correction.html">Batch Correction</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_feature_selection.html">Feature Selection</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_find_marker.html">Find Marker via DE</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_heatmap.html">Heatmap Visualization</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shiny UI Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_qc.html">Quality Control</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/ui_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_enrichR.html">EnrichR</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/ui_sample_size_calc.html">Sample Size Calculator</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_cellViewer.html">cellViewer</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="cnsl_enrichR_files/header-attrs-2.5/header-attrs.js"></script><link href="cnsl_enrichR_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="cnsl_enrichR_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>enrichR</h1>
                        <h4 class="author">Irzam Sarfraz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/master/vignettes/articles/cnsl_enrichR.Rmd"><code>vignettes/articles/cnsl_enrichR.Rmd</code></a></small>
      <div class="hidden name"><code>cnsl_enrichR.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><strong><em>EnrichR</em></strong> (<a href="https://dx.doi.org/10.1093%2Fnar%2Fgkw377">Kuleshov MV, Jones MR, Rouillard AD, et al. Enrichr: a comprehensive gene set enrichment analysis web server 2016 update. Nucleic Acids Res. 2016;44(W1):W90-W97.</a> and <a href="https://doi.org/10.1186/1471-2105-14-128">Chen, E.Y., Tan, C.M., Kou, Y. et al. Enrichr: interactive and collaborative HTML5 gene list enrichment analysis tool. BMC Bioinformatics 14, 128 (2013).</a>) is a <strong><em>GSE</em></strong> (Gene Set Enrichment) method that infers biological knowledge by performing enrichment of input gene sets with curated biologically relevant prior databases that represent biological functions or functional pathways. The <strong><em>singleCellTK</em></strong> offers an easy to use interface to the <strong><em>EnrichR</em></strong> method and only requires the selection of <strong><em>input genes</em></strong> and the <strong><em>databases</em></strong> against which the genes should be tested.</p>
<p>To view detailed instructions on how to use the method from the toolkit, please select ‘Interactive Analysis’ for using enrichR in shiny application or ‘Console Analysis’ for using this method on R console from the tabs below:</p>

<meta name="viewport" content="width=device-width, initial-scale=1">
</div>
<div id="workflow-guide" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-guide" class="anchor"></a>Workflow Guide</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p>An overall workflow guide on how to use <strong><em>EnrichR</em></strong> sub-tab is described below. <strong><em>This workflow guide assumes that the </em>SingleCell Data* has already been uploaded through the <em>Import Single Cell Data</em></strong> sub-tab. <br></p>
<p><img src="ui_screenshots/pathway_analysis/enrichR_1.png"></p>
<ol style="list-style-type: decimal">
<li>Select <strong><em>Cell Annotation &amp; Pathway Analysis</em></strong> tab from the top menu. <br>
</li>
<li>Select <strong><em>EnrichR</em></strong> option from the drop-down menu. <br>
</li>
</ol>
<p><img src="ui_screenshots/pathway_analysis/enrichR_2.png"></p>
<ol start="3" style="list-style-type: decimal">
<li><p>Select if you want to upload a gene symbol file using the option <strong><em>Upload file</em></strong> or choose the <strong><em>Select Gene(s)</em></strong> option if you want to select genes from the already uploaded data. <br></p></li>
<li><p>If <strong><em>Select Gene(s)</em></strong> option is selected, you can select the genes from the available data here. Alternatively, you can upload a gene file using the options specified in <strong>(9)</strong> and <strong>(10)</strong>. <br></p></li>
<li><p>Select <strong><em>databases</em></strong> to test the input genes against. Selecting <strong><em>ALL</em></strong> will test against all available <strong><em>databases</em></strong> and may take long time to compute. <br></p></li>
<li><p>Press <strong><em>Run</em></strong> to start computation of enrichment scores. <br></p></li>
<li><p>Once <strong><em>EnrichR</em></strong> computation is complete, results can be saved to local drive by pressing the <strong><em>Download Results</em></strong> button. <br></p></li>
<li><p>Results are displayed in the table for genes that are significant (adjusted p value &lt; 0.05). <br></p></li>
</ol>
<p><img src="ui_screenshots/pathway_analysis/enrichR_3.png"></p>
<ol start="9" style="list-style-type: decimal">
<li><p>If <strong><em>Upload file</em></strong> option is selected in <strong>(3)</strong>, additional options are displayed to upload the <strong><em>gene symbols</em></strong> file. The file path can be selected here. <br></p></li>
<li><p>Once filed is selected in <strong>(9)</strong>, options to select the parameters of the input file are displayed so the input file is correctly parsed.</p></li>
</ol>
</div>
<div id="console" class="tabcontent">
<p><strong>1. Select genes and databases:</strong></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HES1"</span>, <span class="st">"PON2"</span>, <span class="st">"FBXO44"</span>, <span class="st">"DEK"</span>, <span class="st">"SOX4"</span>, <span class="st">"FBXL16"</span><span class="op">)</span>
<span class="va">databases</span> <span class="op">&lt;-</span> <span class="fu">enrichR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichR/man/listEnrichrDbs.html">listEnrichrDbs</a></span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">libraryName</span></code></pre></div>
<p><strong>2. Run ‘enrichR’ using <code>enrichRSCE</code> function:</strong></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">enrichRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/enrichRSCE.html">enrichRSCE</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>,
                        glist <span class="op">=</span> <span class="va">genes</span>,
                        db <span class="op">=</span> <span class="va">databases</span><span class="op">)</span></code></pre></div>
<p>The <code>enrichRSCE</code> function above, takes the following parameters: <br><code>inSCE</code>: input <code>SingleCellExperiment</code> object <br><code>glist</code>: vector of input gene names <br><code>db</code>: a list of databases to test <br></p>
<p><strong>3. Print results:</strong></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">enrichRes</span><span class="op">)</span></code></pre></div>
<details><summary><b>Example</b></summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="co"># Load singleCellTK &amp; pbmc3k example data</span>
  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://compbiomed.github.io/sctk_docs/">singleCellTK</a></span><span class="op">)</span>
  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"pbmc3k"</span><span class="op">)</span>
  <span class="co"># Select genes</span>
  <span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HES1"</span>, <span class="st">"PON2"</span>, <span class="st">"FBXO44"</span>, <span class="st">"DEK"</span>, <span class="st">"SOX4"</span>, <span class="st">"FBXL16"</span><span class="op">)</span>
  <span class="co"># Fetch databases</span>
  <span class="va">databases</span> <span class="op">&lt;-</span> <span class="fu">enrichR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichR/man/listEnrichrDbs.html">listEnrichrDbs</a></span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">libraryName</span>
  <span class="co"># Run enrichR</span>
  <span class="va">enrichRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/enrichRSCE.html">enrichRSCE</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, glist <span class="op">=</span> <span class="va">genes</span>, db <span class="op">=</span> <span class="va">databases</span><span class="op">)</span></code></pre></div>
<pre><code>## Uploading data to Enrichr... Done.
##   Querying Genome_Browser_PWMs... Done.
##   Querying TRANSFAC_and_JASPAR_PWMs... Done.
##   Querying Transcription_Factor_PPIs... Done.
##   Querying ChEA_2013... Done.
##   Querying Drug_Perturbations_from_GEO_2014... Done.
##   Querying ENCODE_TF_ChIP-seq_2014... Done.
##   Querying BioCarta_2013... Done.
##   Querying Reactome_2013... Done.
##   Querying WikiPathways_2013... Done.
##   Querying Disease_Signatures_from_GEO_up_2014... Done.
##   Querying KEGG_2013... Done.
##   Querying TF-LOF_Expression_from_GEO... Done.
##   Querying TargetScan_microRNA... Done.
##   Querying PPI_Hub_Proteins... Done.
##   Querying GO_Molecular_Function_2015... Done.
##   Querying GeneSigDB... Done.
##   Querying Chromosome_Location... Done.
##   Querying Human_Gene_Atlas... Done.
##   Querying Mouse_Gene_Atlas... Done.
##   Querying GO_Cellular_Component_2015... Done.
##   Querying GO_Biological_Process_2015... Done.
##   Querying Human_Phenotype_Ontology... Done.
##   Querying Epigenomics_Roadmap_HM_ChIP-seq... Done.
##   Querying KEA_2013... Done.
##   Querying NURSA_Human_Endogenous_Complexome... Done.
##   Querying CORUM... Done.
##   Querying SILAC_Phosphoproteomics... Done.
##   Querying MGI_Mammalian_Phenotype_Level_3... Done.
##   Querying MGI_Mammalian_Phenotype_Level_4... Done.
##   Querying Old_CMAP_up... Done.
##   Querying Old_CMAP_down... Done.
##   Querying OMIM_Disease... Done.
##   Querying OMIM_Expanded... Done.
##   Querying VirusMINT... Done.
##   Querying MSigDB_Computational... Done.
##   Querying MSigDB_Oncogenic_Signatures... Done.
##   Querying Disease_Signatures_from_GEO_down_2014... Done.
##   Querying Virus_Perturbations_from_GEO_up... Done.
##   Querying Virus_Perturbations_from_GEO_down... Done.
##   Querying Cancer_Cell_Line_Encyclopedia... Done.
##   Querying NCI-60_Cancer_Cell_Lines... Done.
##   Querying Tissue_Protein_Expression_from_ProteomicsDB... Done.
##   Querying Tissue_Protein_Expression_from_Human_Proteome_Map... Done.
##   Querying HMDB_Metabolites... Done.
##   Querying Pfam_InterPro_Domains... Done.
##   Querying GO_Biological_Process_2013... Done.
##   Querying GO_Cellular_Component_2013... Done.
##   Querying GO_Molecular_Function_2013... Done.
##   Querying Allen_Brain_Atlas_up... Done.
##   Querying ENCODE_TF_ChIP-seq_2015... Done.
##   Querying ENCODE_Histone_Modifications_2015... Done.
##   Querying Phosphatase_Substrates_from_DEPOD... Done.
##   Querying Allen_Brain_Atlas_down... Done.
##   Querying ENCODE_Histone_Modifications_2013... Done.
##   Querying Achilles_fitness_increase... Done.
##   Querying Achilles_fitness_decrease... Done.
##   Querying MGI_Mammalian_Phenotype_2013... Done.
##   Querying BioCarta_2015... Done.
##   Querying HumanCyc_2015... Done.
##   Querying KEGG_2015... Done.
##   Querying NCI-Nature_2015... Done.
##   Querying Panther_2015... Done.
##   Querying WikiPathways_2015... Done.
##   Querying Reactome_2015... Done.
##   Querying ESCAPE... Done.
##   Querying HomoloGene... Done.
##   Querying Disease_Perturbations_from_GEO_down... Done.
##   Querying Disease_Perturbations_from_GEO_up... Done.
##   Querying Drug_Perturbations_from_GEO_down... Done.
##   Querying Genes_Associated_with_NIH_Grants... Done.
##   Querying Drug_Perturbations_from_GEO_up... Done.
##   Querying KEA_2015... Done.
##   Querying Gene_Perturbations_from_GEO_up... Done.
##   Querying Gene_Perturbations_from_GEO_down... Done.
##   Querying ChEA_2015... Done.
##   Querying dbGaP... Done.
##   Querying LINCS_L1000_Chem_Pert_up... Done.
##   Querying LINCS_L1000_Chem_Pert_down... Done.
##   Querying GTEx_Tissue_Sample_Gene_Expression_Profiles_down... Done.
##   Querying GTEx_Tissue_Sample_Gene_Expression_Profiles_up... Done.
##   Querying Ligand_Perturbations_from_GEO_down... Done.
##   Querying Aging_Perturbations_from_GEO_down... Done.
##   Querying Aging_Perturbations_from_GEO_up... Done.
##   Querying Ligand_Perturbations_from_GEO_up... Done.
##   Querying MCF7_Perturbations_from_GEO_down... Done.
##   Querying MCF7_Perturbations_from_GEO_up... Done.
##   Querying Microbe_Perturbations_from_GEO_down... Done.
##   Querying Microbe_Perturbations_from_GEO_up... Done.
##   Querying LINCS_L1000_Ligand_Perturbations_down... Done.
##   Querying LINCS_L1000_Ligand_Perturbations_up... Done.
##   Querying L1000_Kinase_and_GPCR_Perturbations_down... Done.
##   Querying L1000_Kinase_and_GPCR_Perturbations_up... Done.
##   Querying Reactome_2016... Done.
##   Querying KEGG_2016... Done.
##   Querying WikiPathways_2016... Done.
##   Querying ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X... Done.
##   Querying Kinase_Perturbations_from_GEO_down... Done.
##   Querying Kinase_Perturbations_from_GEO_up... Done.
##   Querying BioCarta_2016... Done.
##   Querying HumanCyc_2016... Done.
##   Querying NCI-Nature_2016... Done.
##   Querying Panther_2016... Done.
##   Querying DrugMatrix... Done.
##   Querying ChEA_2016... Done.
##   Querying huMAP... Done.
##   Querying Jensen_TISSUES... Done.
##   Querying RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO... Done.
##   Querying MGI_Mammalian_Phenotype_2017... Done.
##   Querying Jensen_COMPARTMENTS... Done.
##   Querying Jensen_DISEASES... Done.
##   Querying BioPlex_2017... Done.
##   Querying GO_Cellular_Component_2017... Done.
##   Querying GO_Molecular_Function_2017... Done.
##   Querying GO_Biological_Process_2017... Done.
##   Querying GO_Cellular_Component_2017b... Done.
##   Querying GO_Molecular_Function_2017b... Done.
##   Querying GO_Biological_Process_2017b... Done.
##   Querying ARCHS4_Tissues... Done.
##   Querying ARCHS4_Cell-lines... Done.
##   Querying ARCHS4_IDG_Coexp... Done.
##   Querying ARCHS4_Kinases_Coexp... Done.
##   Querying ARCHS4_TFs_Coexp... Done.
##   Querying SysMyo_Muscle_Gene_Sets... Done.
##   Querying miRTarBase_2017... Done.
##   Querying TargetScan_microRNA_2017... Done.
##   Querying Enrichr_Libraries_Most_Popular_Genes... Done.
##   Querying Enrichr_Submissions_TF-Gene_Coocurrence... Done.
##   Querying Data_Acquisition_Method_Most_Popular_Genes... Done.
##   Querying DSigDB... Done.
##   Querying GO_Biological_Process_2018... Done.
##   Querying GO_Cellular_Component_2018... Done.
##   Querying GO_Molecular_Function_2018... Done.
##   Querying TF_Perturbations_Followed_by_Expression... Done.
##   Querying Chromosome_Location_hg19... Done.
##   Querying NIH_Funded_PIs_2017_Human_GeneRIF... Done.
##   Querying NIH_Funded_PIs_2017_Human_AutoRIF... Done.
##   Querying Rare_Diseases_AutoRIF_ARCHS4_Predictions... Done.
##   Querying Rare_Diseases_GeneRIF_ARCHS4_Predictions... Done.
##   Querying NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions... Done.
##   Querying NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions... Done.
##   Querying Rare_Diseases_GeneRIF_Gene_Lists... Done.
##   Querying Rare_Diseases_AutoRIF_Gene_Lists... Done.
##   Querying SubCell_BarCode... Done.
##   Querying GWAS_Catalog_2019... Done.
##   Querying WikiPathways_2019_Human... Done.
##   Querying WikiPathways_2019_Mouse... Done.
##   Querying TRRUST_Transcription_Factors_2019... Done.
##   Querying KEGG_2019_Human... Done.
##   Querying KEGG_2019_Mouse... Done.
##   Querying InterPro_Domains_2019... Done.
##   Querying Pfam_Domains_2019... Done.
##   Querying DepMap_WG_CRISPR_Screens_Broad_CellLines_2019... Done.
##   Querying DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019... Done.
##   Querying MGI_Mammalian_Phenotype_Level_4_2019... Done.
##   Querying UK_Biobank_GWAS_v1... Done.
##   Querying BioPlanet_2019... Done.
##   Querying ClinVar_2019... Done.
##   Querying PheWeb_2019... Done.
##   Querying DisGeNET... Done.
##   Querying HMS_LINCS_KinomeScan... Done.
##   Querying CCLE_Proteomics_2020... Done.
##   Querying ProteomicsDB_2020... Done.
##   Querying lncHUB_lncRNA_Co-Expression... Done.
##   Querying Virus-Host_PPI_P-HIPSTer_2020... Done.
##   Querying Elsevier_Pathway_Collection... Done.
##   Querying Table_Mining_of_CRISPR_Studies... Done.
##   Querying COVID-19_Related_Gene_Sets... Done.
##   Querying MSigDB_Hallmark_2020... Done.
##   Querying Enrichr_Users_Contributed_Lists_2020... Done.
##   Querying TG_GATES_2020... Done.
##   Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.
##   Querying Descartes_Cell_Types_and_Tissue_2021... Done.
##   Querying KEGG_2021_Human... Done.
##   Querying WikiPathway_2021_Human... Done.
##   Querying HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression... Done.
##   Querying GO_Biological_Process_2021... Done.
##   Querying GO_Cellular_Component_2021... Done.
##   Querying GO_Molecular_Function_2021... Done.
##   Querying MGI_Mammalian_Phenotype_Level_4_2021... Done.
##   Querying CellMarker_Augmented_2021... Done.
## Parsing results... Done.</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="co"># Print the first five results</span>
  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">enrichRes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##                     Database_selected                               Term
## 7750       LINCS_L1000_Chem_Pert_down LJP006 MCF10A 3H-withaferin-a-3.33
## 7751       LINCS_L1000_Chem_Pert_down      LJP006 MCF10A 3H-radicicol-10
## 5530         LINCS_L1000_Chem_Pert_up    LJP006 A375 24H-KIN001-043-1.11
## 4156 Genes_Associated_with_NIH_Grants                        R01GM072915
## 7752       LINCS_L1000_Chem_Pert_down       LJP006 A549 24H-BI-2536-0.37
##      Overlap P.value Adjusted.P.value Old.P.value Old.Adjusted.P.value
## 7750    3/68 7.0e-07        0.0013233           0                    0
## 7751    3/82 1.3e-06        0.0013233           0                    0
## 5530    3/91 1.8e-06        0.0040058           0                    0
## 4156     2/8 2.1e-06        0.0002791           0                    0
## 7752   3/119 4.1e-06        0.0027163           0                    0
##      Odds.Ratio Combined.Score          Genes
## 7750   306.6000       4325.556 PON2;HES1;SOX4
## 7751   252.0886       3413.374 PON2;HES1;SOX4
## 5530   226.2045       2991.617 PON2;HES1;SOX4
## 4156  1665.6667      21777.601       DEK;SOX4
## 7752   171.3621       2127.588 PON2;HES1;SOX4
##                                         link
## 7750                        https://clue.io/
## 7751                        https://clue.io/
## 5530                        https://clue.io/
## 4156 https://grants.nih.gov/grants/oer.htm\n
## 7752                        https://clue.io/</code></pre>
</details>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
