<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comprehensive and Interactive Analysis of Single Cell RNA-Seq Data • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<script src="extra.js"></script><meta property="og:title" content="Comprehensive and Interactive Analysis of Single Cell RNA-Seq Data">
<meta property="og:description" content='The Single Cell Toolkit (SCTK) in the singleCellTK package provides an interface to popular tools for importing, quality control, analysis, and visualization of single cell RNA-seq data. SCTK allows users to seamlessly integrate tools from various packages at different stages of the analysis workflow. A general "a la carte" workflow gives users the ability access to multiple methods for data importing, calculation of general QC metrics, doublet detection, ambient RNA estimation and removal, filtering, normalization, batch correction or integration, dimensionality reduction, 2-D embedding, clustering, marker detection, differential expression, cell type labeling, pathway analysis, and data exporting. Curated workflows can be used to run Seurat and Celda. Streamlined quality control can be performed on the command line using the SCTK-QC pipeline. Users can analyze their data using commands in the R console or by using an interactive Shiny Graphical User Interface (GUI). Specific analyses or entire workflows can be summarized and shared with comprehensive HTML reports generated by Rmarkdown. Additional documentation and vignettes can be found at camplab.net/sctk.'>
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.17.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>

  </a>
</li>
<li>
  <a href="articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/01_import_and_qc_tutorial.html">Beginners Guide</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="articles/import_data.html">Import</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Clustering Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
        </li>
        <li>
          <a href="articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
        </li>
        <li>
          <a href="articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Detailed Documentation</a>
      <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Import and Export</a>
          <ul class="dropdown-menu" role="menu">
<li>
              <a href="articles/import_genesets.html">Import Gene Sets</a>
            </li>
            <li>
              <a href="articles/import_annotation.html">Annotation</a>
            </li>
            <li>
              <a href="articles/export_data.html">Export Data</a>
            </li>
            <li>
              <a href="articles/delete_data.html">Delete Data</a>
            </li>
          </ul>
</li>
        <li class="dropdown-submenu">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
          <ul class="dropdown-menu" role="menu">
<li>
              <a href="articles/feature_selection.html">Feature Selection</a>
            </li>
            <li>
              <a href="articles/dimensionality_reduction.html">Dimensionality Reduction</a>
            </li>
            <li>
              <a href="articles/2d_embedding.html">2D Embedding</a>
            </li>
          </ul>
</li>
        <li>
          <a href="articles/clustering.html">Clustering</a>
        </li>
        <li class="dropdown-submenu">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
          <ul class="dropdown-menu" role="menu">
<li>
              <a href="articles/differential_expression.html">Differential Expression</a>
            </li>
            <li>
              <a href="articles/find_marker.html">Find Marker</a>
            </li>
            <li>
              <a href="articles/cell_type_labeling.html">Cell Type Labeling</a>
            </li>
          </ul>
</li>
        <li class="dropdown-submenu">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
          <ul class="dropdown-menu" role="menu">
<li>
              <a href="articles/enrichR.html">EnrichR</a>
            </li>
            <li>
              <a href="articles/pathwayAnalysis.html">Pathway Analysis</a>
            </li>
          </ul>
</li>
        <li class="dropdown-submenu">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
          <ul class="dropdown-menu" role="menu">
<li>
              <a href="articles/trajectoryAnalysis.html">TSCAN</a>
            </li>
          </ul>
</li>
        <li class="dropdown-submenu">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
          <ul class="dropdown-menu" role="menu">
<li>
              <a href="articles/visualization.html">General Visualization</a>
            </li>
            <li>
              <a href="articles/heatmap.html">Generic Heatmap</a>
            </li>
          </ul>
</li>
        <li class="dropdown-submenu">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
          <ul class="dropdown-menu" role="menu">
<li>
              <a href="articles/webApp.html">FAQs</a>
            </li>
          </ul>
</li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="articles/cmd_qc.html">SCTK-QC (command line)</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="the-single-cell-tool-kit">The Single Cell Tool Kit<a class="anchor" aria-label="anchor" href="#the-single-cell-tool-kit"></a>
</h1></div>
<p>The Single Cell Toolkit (SCTK) in the <em>singleCellTK</em> R package is an analysis platform that provides an <strong>R interface to several popular single-cell RNA-sequencing (scRNAseq) data preprocessing, quality control, analysis, and visualization tools</strong>. SCTK imports raw or filtered counts from various scRNAseq preprocessing tools such as 10x CellRanger, BUStools, Optimus, STARSolo, and more. By integrating several publicly available tools written in R or Python, SCTK can be used to perform extensive quality control including doublet detection, ambient RNA removal. SCTK integrates analysis workflows from popular tools such as Seurat and Bioconductor/OSCA into a single unified framework. Results from various workflows can be summarized and easily shared using comprehensive HTML reports. Lastly, data can be exported to Seurat or AnnData object to allow for seamless integration with other downstream analysis workflows.</p>
<p><img src="SingleCellTK_summary2.png" alt="import"><br></p>
<div class="section level2">
<h2 id="features">Features<a class="anchor" aria-label="anchor" href="#features"></a>
</h2>
<p>SCTK offers multiple ways to analyze your scRNAseq data through the R console, the command line, and a graphical user interface (GUI) with the ability to use a large number of algorithms from both R &amp; Python integrated within the toolkit.</p>
<ul>
<li><p><strong>Interactive Analysis:</strong> The Shiny APP allows users without programming experience to easily analyze their scRNAseq data with a GUI. Try it out at <a href="https://sctk.bu.edu/" class="external-link uri">https://sctk.bu.edu/</a>.</p></li>
<li><p><strong>Console Analysis:</strong> Traditional analysis of scRNAseq data can be performed in the R console using wrapper functions for a multitude of tools and algorithms.</p></li>
<li><p><strong>Reports:</strong> Comprehensive HTML reports developed with RMarkdown allows users to document, explore, and share their analyses.</p></li>
<li><p><strong>Interoperability:</strong> Tools from both R and Python package can be seamlessly integrated within the same analysis workflow without the need for manual conversion between different data objects and file formats.</p></li>
<li><p><strong>Number of tools:</strong> SCTK provides access to the largest number of tools within the same platform streamlining end-to-end analysis workflows. Curated workflows include those from Seurat, Scanpy, Scater/Scran (Bioconductor), and Celda.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="tutorials">Tutorials<a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h2>
<ul>
<li>
<a href="https://camplab.net/sctk/current/articles/import_data.html" class="external-link">Import and QC:</a> The Import and QC workflows allow users to import data from multiple formats and perform comprehensive quality control and filtering.</li>
<li>
<a href="https://camplab.net/sctk/current/articles/02_a_la_carte_workflow.html" class="external-link">“<em>A la carte</em>” workflow:</a> The “A la carte” workflow lets users choose from a variety of options during each step of the analysis workflow including normalization, batch correction (optional), dimensionality reduction, 2-D embedding, and clustering.</li>
<li>
<a href="https://camplab.net/sctk/current/articles/seurat_curated_workflow.html" class="external-link">Seurat curated workflow:</a> This curated workflows recapitulates the steps for clustering and integration from the Seurat package (R).</li>
<li>
<a href="https://camplab.net/sctk/current/articles/scanpy_curated_workflow.html" class="external-link">Scanpy curated workflow:</a> This curated workflows recapitulates the steps for clustering from the Scanpy package (Python).</li>
<li>
<a href="https://camplab.net/sctk/current/articles/celda_curated_workflow.html" class="external-link">Celda curated workflow:</a> The curated Celda workflow performs matrix factorization by clustering genes into co-expression modules, cells into subpopulations, and estimating the amount of each module in each cell population.</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>R package <code>singleCellTK</code> is available on <a href="https://bioconductor.org/packages/release/bioc/html/singleCellTK.html" class="external-link">Bioconductor</a> and can be installed with the following commands:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"singleCellTK"</span><span class="op">)</span></span></code></pre></div>
<p>Additional information on how to install from GitHub, install Python dependencies, and for troubleshooting is available on the <a href="https://camplab.net/sctk/current/articles/installation.html" class="external-link">Installation</a> page.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>If you use SCTK for quality control, please cite our <em>Nature Communication</em> paper:</p>
<blockquote>
<p>Rui Hong, Yusuke Koga, Shruthi Bandyadka, Anastasia Leshchyk, Yichen Wang, Vidya Akavoor, Xinyun Cao, Irzam Sarfraz, Zhe Wang, Salam Alabdullatif, Frederick Jansen, Masanao Yajima, W. Evan Johnson &amp; Joshua D. Campbell, “Comprehensive generation, visualization, and reporting of quality control metrics for single-cell RNA sequencing data,” <em>Nature Communications</em>, vol. 13, no. 1688, 2022, doi: 10.1038/s41467-022-29212-9.</p>
</blockquote>
<p>If you use SCTK for analysis in the Rconsole or the interactive graphical user interface, please cite our <em>Patterns</em> paper:</p>
<blockquote>
<p>Yichen Wang, Irzam Sarfraz, Nida Pervaiz, Rui Hong, Yusuke Koga, Vidya Akavoor, Xinyun Cao, Salam Alabdullatif, Syed Ali Zaib, Zhe Wang, Frederick Jansen, Masanao Yajima, W Evan Johnson, Joshua D Campbell, “Interactive analysis of single-cell data using flexible workflows with SCTK2”, <em>Patterns (N Y)</em>, Aug 3;4(8):100814, 2023, doi: 10.1016/j.patter.2023.100814.</p>
</blockquote>
<p><a href="https://github.com/compbiomed/singleCellTK/actions/workflows/BioC-check.yaml" class="external-link"><img src="https://github.com/compbiomed/singleCellTK/actions/workflows/BioC-check.yaml/badge.svg?branch=master" alt="BioC-check"></a> <a href="https://github.com/compbiomed/singleCellTK/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/compbiomed/singleCellTK/actions/workflows/R-CMD-check.yaml/badge.svg?branch=master" alt="R-CMD-check"></a> <a href="https://codecov.io/gh/compbiomed/singleCellTK" class="external-link"><img src="https://codecov.io/gh/compbiomed/singleCellTK/branch/devel/graph/badge.svg" alt="codecov"></a></p>
</div>
<div class="section level2">
<h2 id="report-issues">Report Issues<a class="anchor" aria-label="anchor" href="#report-issues"></a>
</h2>
<p>If you face any difficulty in installing or have identified a bug in the toolkit, please feel free to open up an <a href="https://github.com/compbiomed/singleCellTK/issues" class="external-link">Issue</a> on GitHub. Questions about how to best analyze your scRNA-seq data can be asked in the <a href="https://github.com/compbiomed/singleCellTK/discussions" class="external-link">Discussions</a> page on GitHub.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/singleCellTK" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/compbiomed/singleCellTK/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/compbiomed/singleCellTK/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing singleCellTK</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yichen Wang <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-4347-5199" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Irzam Sarfraz <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-8121-792X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Rui Hong <br><small class="roles"> Author </small>  </li>
<li>Yusuke Koga <br><small class="roles"> Author </small>  </li>
<li>Salam Alabdullatif <br><small class="roles"> Author </small>  </li>
<li>Nida Pervaiz <br><small class="roles"> Author </small>  </li>
<li>David Jenkins <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-7451-4288" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Vidya Akavoor <br><small class="roles"> Author </small>  </li>
<li>Xinyun Cao <br><small class="roles"> Author </small>  </li>
<li>Shruthi Bandyadka <br><small class="roles"> Author </small>  </li>
<li>Anastasia Leshchyk <br><small class="roles"> Author </small>  </li>
<li>Tyler Faits <br><small class="roles"> Author </small>  </li>
<li>Mohammed Muzamil Khan <br><small class="roles"> Author </small>  </li>
<li>Zhe Wang <br><small class="roles"> Author </small>  </li>
<li>W. Evan Johnson <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6247-6595" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Ming Liu <br><small class="roles"> Author </small>  </li>
<li>Joshua David Campbell <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-0780-8662" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
