<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comprehensive and Interactive Analysis of Single Cell RNA-Seq Data • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Comprehensive and Interactive Analysis of Single Cell RNA-Seq Data">
<meta property="og:description" content="Run common single cell analysis in the R console or directly through your browser. Includes many functions for import, quality control, normalization, batch correction, clustering, differential expression, and visualization..">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/introduction_to_SCTK.html">General Workflow</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shell Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/cmd_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="articles/cmd_qc.html">Quality Control</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    R Console Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/cnsl_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="articles/cnsl_qc.html">Quality Control</a>
    </li>
    <li>
      <a href="articles/cnsl_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="articles/cnsl_batch_correction.html">Batch Correction</a>
    </li>
    <li>
      <a href="articles/cnsl_feature_selection.html">Feature Selection</a>
    </li>
    <li>
      <a href="articles/cnsl_clustering.html">Clustering</a>
    </li>
    <li>
      <a href="articles/cnsl_differential_expression.html">Differential Expression</a>
    </li>
    <li>
      <a href="articles/cnsl_find_marker.html">Find Marker via DE</a>
    </li>
    <li>
      <a href="articles/cnsl_visualization.html">Visualization</a>
    </li>
    <li>
      <a href="articles/cnsl_heatmap.html">Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shiny UI Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Import</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/ui_import_data.html">Data</a>
        </li>
        <li>
          <a href="articles/ui_import_genesets.html">Gene Sets</a>
        </li>
        <li>
          <a href="articles/ui_import_cell_annotation.html">Cell Annotation</a>
        </li>
        <li>
          <a href="articles/ui_import_feature_annotation.html">Feature Annotation</a>
        </li>
        <li>
          <a href="articles/ui_export_data.html">Export Data</a>
        </li>
        <li>
          <a href="articles/ui_delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/ui_qc.html">Quality Control</a>
        </li>
        <li>
          <a href="articles/ui_filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/ui_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="articles/ui_batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/ui_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="articles/ui_dimred.html">Dimensionality Reduction</a>
        </li>
      </ul>
</li>
    <li>
      <a href="articles/ui_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/ui_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="articles/ui_find_marker.html">Find Marker</a>
        </li>
      </ul>
</li>
    <li>
      <a href="articles/ui_heatmap.html">Generic Heatmap</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="single-cell-tk" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#single-cell-tk" class="anchor"></a>Single Cell TK</h1></div>

<p>The Single Cell ToolKit (SCTK) is an analysis platform that provides an <strong>R interface to several popular scRNA-seq preprocessing, quality control, and visualization tools</strong>. SCTK imports raw or filtered counts from various single cell sequencing technologies and upstream tools such as 10x CellRanger, BUStools, Optimus, STARSolo, and more. By integrating several publicly available tools written in R as well as Python, SCTK performs extensive quality control measures including doublet detection and batch effect correction. Additionally, SCTK summarizes results and related visualizations in a comprehensive R markdown and/or HTML report. SCTK provides a standardized single cell analysis workflow by representing the counts data and the results using the <a href="https://www.bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">SingleCellExperiment</a> R object. Furthermore, SCTK enables seamless downstream analysis by exporting data and results in flat .txt and Python Anndata formats.</p>
<p>A comprehensive list of available functions is listed in the Reference section.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="system-setup" class="section level3">
<h3 class="hasAnchor">
<a href="#system-setup" class="anchor"></a>System setup</h3>
<p>If you are the first time to install R, please don’t install 32 bit R. Make sure to uncheck the ‘32-bit Files’ box when you see the following window:</p>
<p><img src="exec/png/32bit-R.png"></p>
<div id="windows-user" class="section level4">
<h4 class="hasAnchor">
<a href="#windows-user" class="anchor"></a>Window’s user</h4>
<p>For window’s users, please install <a href="https://cran.r-project.org/bin/windows/Rtools/history.html">rtools</a> based on your R version. Make sure to click ‘Edit the system PATH’ box when you see this window:</p>
<p><img src="exec/png/rtools.png"></p>
<p>After installing rtools, install ‘devtools’ package with the following command. If it asks whether install the package that requires compilation, type ‘y’.</p>
<pre><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages('devtools')</a></code></pre>
</div>
<div id="macos-user" class="section level4">
<h4 class="hasAnchor">
<a href="#macos-user" class="anchor"></a>macOS user</h4>
<p>For macbook’s users, please install gfortran with brew. If you have not installed brew, please check <a href="https://brew.sh/">this link</a> to set up brew on your machine.</p>
<pre><code>brew install gcc</code></pre>
<p>After that, install ‘devtools’ package with the following command.</p>
<pre><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages('devtools')</a></code></pre>
</div>
</div>
<div id="release-version" class="section level3">
<h3 class="hasAnchor">
<a href="#release-version" class="anchor"></a>Release Version</h3>
<p>You can download the release version of the Single Cell Toolkit in <a href="https://bioconductor.org/packages/release/bioc/html/singleCellTK.html">Bioconductor v3.10</a>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"singleCellTK"</span><span class="op">)</span></code></pre></div>
</div>
<div id="devel-version" class="section level3">
<h3 class="hasAnchor">
<a href="#devel-version" class="anchor"></a>Devel Version</h3>
<p>You can download the development version of the Single Cell Toolkit in <a href="https://bioconductor.org/packages/devel/bioc/html/singleCellTK.html">Bioconductor v3.11</a> or from this repository:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"compbiomed/singleCellTK"</span><span class="op">)</span></code></pre></div>
</div>
<div id="r-34-version" class="section level3">
<h3 class="hasAnchor">
<a href="#r-34-version" class="anchor"></a>R 3.4 Version</h3>
<p>If you are still running an earlier version of R than 3.5, you can install the following version from this repository:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"compbiomed/singleCellTK"</span>, ref<span class="op">=</span><span class="st">"r_3_4"</span><span class="op">)</span></code></pre></div>
<div id="troubleshooting-installation" class="section level4">
<h4 class="hasAnchor">
<a href="#troubleshooting-installation" class="anchor"></a>Troubleshooting Installation</h4>
<p>For the majority of users, the commands above will install the latest version of the singleCellTK without any errors. Rarely, you may encounter an error due to previously installed versions of some packages that are required for the singleCellTK. If you encounter an error during installation, use the commands below to check the version of Bioconductor that is installed:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/version.html">version</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>If the version number is not 3.6 or higher, you must upgrade Bioconductor to install the toolkit:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>After you install Bioconductor 3.6 or higher, you should be able to install the toolkit using <code><a href="https://devtools.r-lib.org//reference/remote-reexports.html">devtools::install_github("compbiomed/singleCellTK")</a></code>. If you still encounter an error, ensure your Bioconductor packages are up to date by running the following command.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/valid.html">valid</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>If the command above does not return <code>TRUE</code>, run the following command to update your R packages:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Then, try to install the toolkit again:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"compbiomed/singleCellTK"</span><span class="op">)</span></code></pre></div>
<p>If you still encounter an error, please <a href="mailto:dfj@bu.edu">contact us</a> and we’d be happy to help.</p>
</div>
</div>
</div>
<div id="develop-singlecelltk" class="section level2">
<h2 class="hasAnchor">
<a href="#develop-singlecelltk" class="anchor"></a>Develop singleCellTK</h2>
<p>To contribute to singleCellTK, follow these steps:</p>
<p><strong>Note</strong>: Development of the singleCellTK is done using the latest version of R.</p>
<ol>
<li>Fork the repo using the “Fork” button above.</li>
<li>Download a local copy of your forked repository “<code>git clone https://github.com/{username}/singleCellTK.git</code>”</li>
<li>Open Rstudio</li>
<li>Go to “File” -&gt; “New Project” -&gt; “Existing Directory” and select your git repository directory</li>
</ol>
<p>You can then make your changes and test your code using the Rstudio build tools. There is a lot of information about building packages available here: <a href="http://r-pkgs.had.co.nz/" class="uri">http://r-pkgs.had.co.nz/</a>.</p>
<p>Information about building shiny packages is available here: <a href="http://shiny.rstudio.com/tutorial/" class="uri">http://shiny.rstudio.com/tutorial/</a>.</p>
<p>When you are ready to upload your changes, commit them locally, push them to your forked repo, and make a pull request to the compbiomed repository.</p>
<p>Report bugs and request features on our <a href="https://github.com/compbiomed/singleCellTK/issues">GitHub issue tracker</a>.</p>
<p>Join us on <a href="https://compbiomed.slack.com/">slack</a>!</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/singleCellTK">https://​www.bioconductor.org/​packages/​singleCellTK</a>
</li>
<li>Browse source code at <br><a href="https://github.com/compbiomed/singleCellTK/">https://​github.com/​compbiomed/​singleCellTK/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/compbiomed/singleCellTK/issues">https://​github.com/​compbiomed/​singleCellTK/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>David Jenkins <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-7451-4288" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Vidya Akavoor <br><small class="roles"> Author </small>  </li>
<li>Salam Alabdullatif <br><small class="roles"> Author </small>  </li>
<li>Shruthi Bandyadka <br><small class="roles"> Author </small>  </li>
<li>Emma Briars <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-9350-5523" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Xinyun Cao <br><small class="roles"> Author </small>  </li>
<li>Sebastian Carrasco Pro <br><small class="roles"> Author </small>  </li>
<li>Tyler Faits <br><small class="roles"> Author </small>  </li>
<li>Rui Hong <br><small class="roles"> Author </small>  </li>
<li>Mohammed Muzamil Khan <br><small class="roles"> Author </small>  </li>
<li>Yusuke Koga <br><small class="roles"> Author, maintainer </small>  </li>
<li>Anastasia Leshchyk <br><small class="roles"> Author </small>  </li>
<li>Irzam Sarfraz <br><small class="roles"> Author </small>  </li>
<li>Yichen Wang <br><small class="roles"> Author </small>  </li>
<li>Zhe Wang <br><small class="roles"> Author </small>  </li>
<li>W. Evan Johnson <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6247-6595" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Joshua David Campbell <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/compbiomed/singleCellTK"><img src="https://travis-ci.org/compbiomed/singleCellTK.svg?branch=master" alt="Travis build status"></a></li>
<li><a href="https://codecov.io/gh/compbiomed/singleCellTK"><img src="https://codecov.io/gh/compbiomed/singleCellTK/branch/master/graph/badge.svg" alt="codecov"></a></li>
<li><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/singleCellTK"><img src="https://www.bioconductor.org/shields/build/release/bioc/singleCellTK.svg" alt="BioC status"></a></li>
<li><a href="https://www.tidyverse.org/lifecycle/#stable"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" alt="lifecycle"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
