<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Processing and Visualizing Data in the Single Cell Toolkit • singleCellTK</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">singleCellTK</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/01-Introduction_to_singleCellTK.html">Introduction to singleCellTK</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Interactive Shiny Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/03-tab01_Upload.html">1: Upload</a>
    </li>
    <li>
      <a href="../articles/04-tab02_Data-Summary-and-Filtering.html">2: Data Summary and Filtering</a>
    </li>
    <li>
      <a href="../articles/05-tab03_Dimensionality-Reduction-and-Clustering.html">3: Dimensionality Reduction and Clustering</a>
    </li>
    <li>
      <a href="../articles/06-tab04_Batch-Correction.html">4: Batch Correction</a>
    </li>
    <li>
      <a href="../articles/07-tab05_Differential-Expression.html">5: Differential Expression</a>
    </li>
    <li>
      <a href="../articles/08-tab06_Pathway-Activity-Analysis.html">6: Pathway Activity Analysis</a>
    </li>
    <li>
      <a href="../articles/09-tab07_Sample-Size.html">7: Sample Size</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/02-Processing_and_Visualizing_Data_in_the_SingleCellTK.html">Processing and Visualizing Data</a>
    </li>
    <li>
      <a href="../articles/10-Aligning_and_Quantifying_scRNA-Seq_Data.html">Aligning and Quantifying scRNA-Seq Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><script src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/htmlwidgets-1.0/htmlwidgets.js"></script><script src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/jquery-1.12.4/jquery.min.js"></script><link href="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/datatables-binding-0.4/datatables.js"></script><link href="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/dt-core-1.10.16/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/dt-core-1.10.16/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/dt-core-1.10.16/js/jquery.dataTables.min.js"></script><link href="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet">
<script src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/crosstalk-1.0.0/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Processing and Visualizing Data in the Single Cell Toolkit</h1>
                        <h4 class="author">David Jenkins</h4>
            <address class="author_afil">
      The Section of Computational Biomedicine, Boston University School of Medicine, Boston, MA<br><a class="author_email" href="mailto:#"></a><a href="mailto:dfj@bu.edu">dfj@bu.edu</a>
      </address>
                              <h4 class="author">Tyler Faits</h4>
            <address class="author_afil">
      The Section of Computational Biomedicine, Boston University School of Medicine, Boston, MA<br><h4 class="author">W. Evan Johnson</h4>
            <address class="author_afil">
      The Section of Computational Biomedicine, Boston University School of Medicine, Boston, MA<br>
</address>
</address>
</div>

    
    
<div class="contents">
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Intro</p>
</div>
<div id="maits-example" class="section level1">
<h1 class="hasAnchor">
<a href="#maits-example" class="anchor"></a>MAITS Example</h1>
<p>MAITS intro</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(singleCellTK)
<span class="kw">library</span>(DT)
<span class="kw">data</span>(<span class="st">"maits_sce"</span>)</code></pre></div>
<div id="summarizetable" class="section level2">
<h2 class="hasAnchor">
<a href="#summarizetable" class="anchor"></a>summarizeTable</h2>
<p>You can get summary metrics with the <code>summarizeTable</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/DT/topics/datatable">datatable</a></span>(<span class="kw"><a href="../reference/summarizeTable.html">summarizeTable</a></span>(maits_sce))</code></pre></div>
<div id="htmlwidget-0bd5825284c96075e3ec" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-0bd5825284c96075e3ec">{"x":{"filter":"none","data":[["1","2","3","4","5","6"],["Number of Samples","Number of Genes","Average number of reads per cell","Average number of genes per cell","Samples with &lt;1700 detected genes","Genes with no expression across all samples"],[96,16302,997336,6833,5,0]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Metric<\/th>\n      <th>Value<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":2},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="filtering-by-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-by-annotation" class="anchor"></a>Filtering by Annotation</h2>
<p>Explore the available annotations in the data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(<span class="kw">colData</span>(maits_sce))</code></pre></div>
<pre><code>##  [1] "wellKey"          "condition"        "nGeneOn"         
##  [4] "libSize"          "PercentToHuman"   "MedianCVCoverage"
##  [7] "PCRDuplicate"     "exonRate"         "pastFastqc"      
## [10] "ncells"           "ngeneson"         "cngeneson"       
## [13] "TRAV1"            "TRBV6"            "TRBV4"           
## [16] "TRBV20"           "alpha"            "beta"            
## [19] "ac"               "bc"               "ourfilter"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">colData</span>(maits_sce)<span class="op">$</span>ourfilter)</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##    22    74</code></pre>
<p>The data has a filtered dataset with 74 ‘pass filter’ samples, let’s subset the data to include the pass filter samples</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">maits_subset &lt;-<span class="st"> </span>maits_sce[, <span class="kw">colData</span>(maits_sce)<span class="op">$</span>ourfilter]
<span class="kw">table</span>(<span class="kw">colData</span>(maits_subset)<span class="op">$</span>ourfilter)</code></pre></div>
<pre><code>## 
## TRUE 
##   74</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/DT/topics/datatable">datatable</a></span>(<span class="kw"><a href="../reference/summarizeTable.html">summarizeTable</a></span>(maits_subset))</code></pre></div>
<div id="htmlwidget-8f38141a6ccf65b0b4f2" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-8f38141a6ccf65b0b4f2">{"x":{"filter":"none","data":[["1","2","3","4","5","6"],["Number of Samples","Number of Genes","Average number of reads per cell","Average number of genes per cell","Samples with &lt;1700 detected genes","Genes with no expression across all samples"],[74,16302,999245,7539,0,157]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Metric<\/th>\n      <th>Value<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":2},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#visualization" class="anchor"></a>Visualization</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">reducedDims</span>(maits_subset)</code></pre></div>
<pre><code>## List of length 4
## names(4): PCA_counts PCA_logcounts TSNE_counts TSNE_logcounts</code></pre>
<div id="pca" class="section level3">
<h3 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotPCA.html">plotPCA</a></span>(maits_subset, <span class="dt">reducedDimName =</span> <span class="st">"PCA_logcounts"</span>, <span class="dt">colorBy =</span> <span class="st">"condition"</span>)</code></pre></div>
<p><img src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/figure-html/maits_pca-1.png" width="672"></p>
</div>
<div id="t-sne" class="section level3">
<h3 class="hasAnchor">
<a href="#t-sne" class="anchor"></a>t-SNE</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotTSNE.html">plotTSNE</a></span>(maits_subset, <span class="dt">reducedDimName =</span> <span class="st">"TSNE_logcounts"</span>, <span class="dt">colorBy =</span> <span class="st">"condition"</span>)</code></pre></div>
<p><img src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/figure-html/maits_tsne-1.png" width="672"></p>
</div>
</div>
<div id="differential-expression-with-mast" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression-with-mast" class="anchor"></a>Differential Expression with MAST</h2>
<div id="adaptive-thresholding" class="section level3">
<h3 class="hasAnchor">
<a href="#adaptive-thresholding" class="anchor"></a>Adaptive Thresholding</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">thresholds &lt;-<span class="st"> </span><span class="kw"><a href="../reference/thresholdGenes.html">thresholdGenes</a></span>(maits_subset)</code></pre></div>
<pre><code>## `fData` has no primerid.  I'll make something up.</code></pre>
<pre><code>## (0.0144,0.163]  (0.163,0.334]   (0.334,0.53]   (0.53,0.755]   (0.755,1.01] 
##       1.048076       1.048076       1.048076       1.048076       1.048076 
##    (1.01,1.31]    (1.31,1.65]    (1.65,2.04]    (2.04,2.48]    (2.48,2.99] 
##       1.048076       1.048076       1.048076       1.243318       1.515176 
##    (2.99,3.58]    (3.58,4.25]    (4.25,5.02]    (5.02,5.91]    (5.91,6.92] 
##       1.841511       2.293826       2.293826       2.505690       2.909298 
##    (6.92,8.08]    (8.08,9.42]    (9.42,10.9]    (10.9,14.7] 
##       3.091104       3.978445       5.059031       7.900775</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">4</span>))
<span class="kw">plot</span>(thresholds)
<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</code></pre></div>
<p><img src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/figure-html/maits_thresh-1.png" width="672"></p>
</div>
<div id="run-mast" class="section level3">
<h3 class="hasAnchor">
<a href="#run-mast" class="anchor"></a>Run MAST</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mast_results &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MAST.html">MAST</a></span>(maits_subset, <span class="dt">condition =</span> <span class="st">"condition"</span>, <span class="dt">usethresh =</span> <span class="ot">TRUE</span>,
                     <span class="dt">use_assay =</span> <span class="st">"logcounts"</span>)</code></pre></div>
<pre><code>## (0.0144,0.163]  (0.163,0.334]   (0.334,0.53]   (0.53,0.755]   (0.755,1.01] 
##       1.048076       1.048076       1.048076       1.048076       1.048076 
##    (1.01,1.31]    (1.31,1.65]    (1.65,2.04]    (2.04,2.48]    (2.48,2.99] 
##       1.048076       1.048076       1.048076       1.243318       1.515176 
##    (2.99,3.58]    (3.58,4.25]    (4.25,5.02]    (5.02,5.91]    (5.91,6.92] 
##       1.841511       2.293826       2.293826       2.505690       2.909298 
##    (6.92,8.08]    (8.08,9.42]    (9.42,10.9]    (10.9,14.7] 
##       3.091104       3.978445       5.059031       7.900775</code></pre>
<pre><code>## ....................................................................................................................................................................................................................................................................................................................................................................................................................................................................
## Done!
## Combining coefficients and standard errors
## Calculating log-fold changes
## Calculating likelihood ratio tests
## Refitting on reduced model...
## ....................................................................................................................................................................................................................................................................................................................................................................................................................................................................
## Done!</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/MASTviolin.html">MASTviolin</a></span>(maits_subset, <span class="dt">fcHurdleSig =</span> mast_results, <span class="dt">threshP =</span> <span class="ot">TRUE</span>, <span class="dt">variable =</span> <span class="st">"condition"</span>)</code></pre></div>
<pre><code>## `fData` has no primerid.  I'll make something up.</code></pre>
<pre><code>## (0.0144,0.163]  (0.163,0.334]   (0.334,0.53]   (0.53,0.755]   (0.755,1.01] 
##       1.048076       1.048076       1.048076       1.048076       1.048076 
##    (1.01,1.31]    (1.31,1.65]    (1.65,2.04]    (2.04,2.48]    (2.48,2.99] 
##       1.048076       1.048076       1.048076       1.243318       1.515176 
##    (2.99,3.58]    (3.58,4.25]    (4.25,5.02]    (5.02,5.91]    (5.91,6.92] 
##       1.841511       2.293826       2.293826       2.505690       2.909298 
##    (6.92,8.08]    (8.08,9.42]    (9.42,10.9]    (10.9,14.7] 
##       3.091104       3.978445       5.059031       7.900775</code></pre>
<p><img src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/figure-html/maits_violin-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/MASTregression.html">MASTregression</a></span>(maits_subset, <span class="dt">fcHurdleSig =</span> mast_results, <span class="dt">threshP =</span> <span class="ot">TRUE</span>, <span class="dt">variable =</span> <span class="st">"condition"</span>)</code></pre></div>
<pre><code>## `fData` has no primerid.  I'll make something up.</code></pre>
<pre><code>## (0.0144,0.163]  (0.163,0.334]   (0.334,0.53]   (0.53,0.755]   (0.755,1.01] 
##       1.048076       1.048076       1.048076       1.048076       1.048076 
##    (1.01,1.31]    (1.31,1.65]    (1.65,2.04]    (2.04,2.48]    (2.48,2.99] 
##       1.048076       1.048076       1.048076       1.243318       1.515176 
##    (2.99,3.58]    (3.58,4.25]    (4.25,5.02]    (5.02,5.91]    (5.91,6.92] 
##       1.841511       2.293826       2.293826       2.505690       2.909298 
##    (6.92,8.08]    (8.08,9.42]    (9.42,10.9]    (10.9,14.7] 
##       3.091104       3.978445       5.059031       7.900775</code></pre>
<p><img src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/figure-html/maits_lm-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_DiffEx.html">plot_DiffEx</a></span>(maits_subset, <span class="dt">use_assay =</span> <span class="st">"logcounts"</span>, <span class="dt">condition =</span> <span class="st">"condition"</span>,
                 <span class="dt">geneList =</span> mast_results<span class="op">$</span>Gene[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>],
                 <span class="dt">annotationColors =</span> <span class="st">"auto"</span>,
            <span class="dt">displayRowLabels =</span> <span class="ot">FALSE</span>, <span class="dt">displayColumnLabels =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p><img src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/figure-html/maits_heatmap-1.png" width="672"></p>
</div>
<div id="pathway-activity-with-gsva" class="section level3">
<h3 class="hasAnchor">
<a href="#pathway-activity-with-gsva" class="anchor"></a>Pathway Activity with GSVA</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gsva_res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/GSVA_sce.html">GSVA_sce</a></span>(maits_subset, <span class="st">"logcounts"</span>,
                     <span class="st">"MSigDB c2 (Human, Entrez ID only)"</span>,
                     <span class="kw">c</span>(<span class="st">"KEGG_PROTEASOME"</span>,
                       <span class="st">"REACTOME_VIF_MEDIATED_DEGRADATION_OF_APOBEC3G"</span>,
                       <span class="st">"REACTOME_P53_INDEPENDENT_DNA_DAMAGE_RESPONSE"</span>,
                       <span class="st">"BIOCARTA_PROTEASOME_PATHWAY"</span>,
                       <span class="st">"REACTOME_METABOLISM_OF_AMINO_ACIDS"</span>,
                       <span class="st">"REACTOME_REGULATION_OF_ORNITHINE_DECARBOXYLASE"</span>,
                       <span class="st">"REACTOME_CYTOSOLIC_TRNA_AMINOACYLATION"</span>,
                       <span class="st">"REACTOME_STABILIZATION_OF_P53"</span>,
                       <span class="st">"REACTOME_SCF_BETA_TRCP_MEDIATED_DEGRADATION_OF_EMI1"</span>),
                     <span class="dt">parallel.sz=</span><span class="dv">1</span>)</code></pre></div>
<pre><code>## Loading required package: GSEABase</code></pre>
<pre><code>## Loading required package: annotate</code></pre>
<pre><code>## Loading required package: AnnotationDbi</code></pre>
<pre><code>## Loading required package: XML</code></pre>
<pre><code>## Loading required package: graph</code></pre>
<pre><code>## 
## Attaching package: 'graph'</code></pre>
<pre><code>## The following object is masked from 'package:XML':
## 
##     addNode</code></pre>
<pre><code>## Warning in .local(expr, gset.idx.list, ...): 157genes with constant
## expression values throuhgout the samples.</code></pre>
<pre><code>## Warning in .local(expr, gset.idx.list, ...): Since argument method!
## ="ssgsea", genes with constant expression values are discarded.</code></pre>
<pre><code>## Estimating GSVA scores for 9 gene sets.
## Computing observed enrichment scores
## Estimating ECDFs in microarray data with Gaussian kernels
## Using parallel with 1 cores
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |=======                                                          |  11%
  |                                                                       
  |==============                                                   |  22%
  |                                                                       
  |======================                                           |  33%
  |                                                                       
  |=============================                                    |  44%
  |                                                                       
  |====================================                             |  56%
  |                                                                       
  |===========================================                      |  67%
  |                                                                       
  |===================================================              |  78%
  |                                                                       
  |==========================================================       |  89%
  |                                                                       
  |=================================================================| 100%</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/GSVA_plot.html">GSVA_plot</a></span>(maits_subset, gsva_res, <span class="st">"Violin"</span>, <span class="st">"condition"</span>)</code></pre></div>
<p><img src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/figure-html/maits_gsva-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/GSVA_plot.html">GSVA_plot</a></span>(maits_subset, gsva_res, <span class="st">"Heatmap"</span>, <span class="st">"condition"</span>)</code></pre></div>
<p><img src="02-Processing_and_Visualizing_Data_in_the_SingleCellTK_files/figure-html/maits_gsva-2.png" width="672"></p>
</div>
</div>
</div>
<div id="pbmc---4k-example" class="section level1">
<h1 class="hasAnchor">
<a href="#pbmc---4k-example" class="anchor"></a>PBMC - 4k Example</h1>
<p>PBMC - 4k example</p>
</div>
<div id="batch-effects-example" class="section level1">
<h1 class="hasAnchor">
<a href="#batch-effects-example" class="anchor"></a>Batch Effects Example</h1>
<p>It is possible to use ComBat within the Single Cell Toolkit. This support is experimental, since ComBat was not designed for scRNA-Seq. Here, we will load the bladderbatch example data into a SingleCellExperiment object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(bladderbatch)
<span class="kw">data</span>(bladderdata)
dat &lt;-<span class="st"> </span>bladderEset[<span class="dv">1</span><span class="op">:</span><span class="dv">50</span>,]

pheno =<span class="st"> </span><span class="kw">pData</span>(dat)
edata =<span class="st"> </span><span class="kw">exprs</span>(dat)
bladder_sctke &lt;-<span class="st"> </span><span class="kw"><a href="../reference/createSCE.html">createSCE</a></span>(<span class="dt">assayfile =</span> edata,
                           <span class="dt">annotfile =</span> pheno,
                           <span class="dt">assay_name =</span> <span class="st">"microarray"</span>,
                           <span class="dt">inputdataframes =</span> <span class="ot">TRUE</span>,
                           <span class="dt">create_logcounts =</span> <span class="ot">FALSE</span>)
<span class="kw">assay</span>(bladder_sctke, <span class="st">"combat"</span>) &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ComBat_SCE.html">ComBat_SCE</a></span>(<span class="dt">SCEdata =</span> bladder_sctke,
                                             <span class="dt">batch =</span> <span class="st">"batch"</span>,
                                             <span class="dt">use_assay =</span> <span class="st">"microarray"</span>,
                                             <span class="dt">covariates =</span> <span class="st">"cancer"</span>)</code></pre></div>
<pre><code>## Found5batches</code></pre>
<pre><code>## Adjusting for2covariate(s) or covariate level(s)</code></pre>
<pre><code>## Standardizing Data across genes</code></pre>
<pre><code>## Fitting L/S model and finding priors</code></pre>
<pre><code>## Finding parametric adjustments</code></pre>
<pre><code>## Adjusting the Data</code></pre>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<pre><code>## R version 3.4.3 (2017-11-30)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS High Sierra 10.13.3
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] bladderbatch_1.16.0        GSEABase_1.40.1           
##  [3] graph_1.56.0               annotate_1.56.1           
##  [5] XML_3.98-1.10              AnnotationDbi_1.40.0      
##  [7] DT_0.4                     singleCellTK_0.4.6        
##  [9] SingleCellExperiment_1.0.0 SummarizedExperiment_1.8.1
## [11] DelayedArray_0.4.1         matrixStats_0.53.1        
## [13] Biobase_2.38.0             GenomicRanges_1.30.3      
## [15] GenomeInfoDb_1.14.0        IRanges_2.12.0            
## [17] S4Vectors_0.16.0           BiocGenerics_0.24.0       
## 
## loaded via a namespace (and not attached):
##  [1] splines_3.4.3          bit64_0.9-7            jsonlite_1.5          
##  [4] shiny_1.0.5            blob_1.1.0             GenomeInfoDbData_1.0.0
##  [7] yaml_2.1.16            pillar_1.2.1           RSQLite_2.0           
## [10] backports_1.1.2        lattice_0.20-35        limma_3.34.9          
## [13] digest_0.6.15          RColorBrewer_1.1-2     XVector_0.18.0        
## [16] colorspace_1.3-2       htmltools_0.3.6        httpuv_1.3.5          
## [19] Matrix_1.2-12          plyr_1.8.4             GetoptLong_0.1.6      
## [22] genefilter_1.60.0      zlibbioc_1.24.0        xtable_1.8-2          
## [25] scales_0.5.0           BiocParallel_1.12.0    tibble_1.4.2          
## [28] GSVA_1.26.0            mgcv_1.8-23            ggplot2_2.2.1         
## [31] lazyeval_0.2.1         survival_2.41-3        magrittr_1.5          
## [34] mime_0.5               memoise_1.1.0          evaluate_0.10.1       
## [37] nlme_3.1-131.1         MAST_1.4.1             tools_3.4.3           
## [40] data.table_1.10.4-3    GlobalOptions_0.0.12   ComplexHeatmap_1.17.1 
## [43] stringr_1.3.0          munsell_0.4.3          compiler_3.4.3        
## [46] rlang_0.2.0            grid_3.4.3             RCurl_1.95-4.10       
## [49] rjson_0.2.15           htmlwidgets_1.0        circlize_0.4.3        
## [52] crosstalk_1.0.0        bitops_1.0-6           labeling_0.3          
## [55] rmarkdown_1.8          gtable_0.2.0           abind_1.4-5           
## [58] DBI_0.7                reshape2_1.4.3         R6_2.2.2              
## [61] knitr_1.20             bit_1.1-12             shinythemes_1.1.1     
## [64] rprojroot_1.3-2        shape_1.4.4            stringi_1.1.6         
## [67] sva_3.26.0             Rcpp_0.12.15           geneplotter_1.56.0</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#maits-example">MAITS Example</a><ul class="nav nav-pills nav-stacked">
<li><a href="#summarizetable">summarizeTable</a></li>
      <li><a href="#filtering-by-annotation">Filtering by Annotation</a></li>
      <li><a href="#visualization">Visualization</a></li>
      <li><a href="#differential-expression-with-mast">Differential Expression with MAST</a></li>
      </ul>
</li>
      <li><a href="#pbmc---4k-example">PBMC - 4k Example</a></li>
      <li><a href="#batch-effects-example">Batch Effects Example</a></li>
      <li><a href="#session-info">Session info</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by David Jenkins, Tyler Faits, Emma Briars, Sebastian Carrasco Pro, W. Evan Johnson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
