<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Aligning and Quantifying scRNA-Seq Data • singleCellTK</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">singleCellTK</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Interactive Shiny Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tab01_Upload.html">1: Upload</a>
    </li>
    <li>
      <a href="../articles/tab02_Data-Summary-and-Filtering.html">2: Data Summary and Filtering</a>
    </li>
    <li>
      <a href="../articles/tab03_Dimensionality-Reduction-and-Clustering.html">3: Dimensionality Reduction and Clustering</a>
    </li>
    <li>
      <a href="../articles/tab04_Batch-Correction.html">4: Batch Correction</a>
    </li>
    <li>
      <a href="../articles/tab05_Differential-Expression.html">5: Differential Expression</a>
    </li>
    <li>
      <a href="../articles/tab06_Pathway-Activity-Analysis.html">6: Pathway Activity Analysis</a>
    </li>
    <li>
      <a href="../articles/tab07_Sample-Size.html">7: Sample Size</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Aligning_and_Quantifying_scRNA-Seq_Data.html">Aligning and Quantifying scRNA-Seq Data</a>
    </li>
    <li>
      <a href="../articles/Processing_and_Visualizing_Data_in_the_SingleCellTK.html">Processing and Visualizing Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Aligning and Quantifying scRNA-Seq Data</h1>
                        <h4 class="author">David Jenkins</h4>
            <address class="author_afil">
      The Section of Computational Biomedicine, Boston University School of Medicine, Boston, MA<br><a class="author_email" href="mailto:#"></a><a href="mailto:dfj@bu.edu">dfj@bu.edu</a>
      </address>
                  
          </div>

    
    
<div class="contents">
<p>In this vignette, we will suggest a workflow for processing single cell RNA-Seq data to produce an SCtkExperiment object that can be used in the single cell toolkit.</p>
<div id="suggested-software" class="section level2">
<h2 class="hasAnchor">
<a href="#suggested-software" class="anchor"></a>Suggested Software</h2>
<p>Some of the steps in this workflow are performed outside of R and are optional, but they produce useful metrics that can be explored using the single cell toolkit. Please see the individual tool websites for installation instructions.</p>
<ul>
<li>
<a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> - Command line or GUI program for performing quality control on sequencing files</li>
<li>
<a href="http://multiqc.info/">MultiQC</a> - Command line tool that creates a combined report from FastQC results run on individual files.</li>
<li>
<a href="https://www.bioconductor.org/packages/release/bioc/html/Rsubread.html">Rsubread</a> - R package for alignment of fastq files to a reference genome</li>
</ul>
</div>
<div id="tutorial-single-cell-rna-seq-reads-to-single-cell-object" class="section level2">
<h2 class="hasAnchor">
<a href="#tutorial-single-cell-rna-seq-reads-to-single-cell-object" class="anchor"></a>Tutorial, Single Cell RNA-Seq Reads to Single Cell Object</h2>
<div id="step-1-fastqc-quality-control" class="section level3">
<h3 class="hasAnchor">
<a href="#step-1-fastqc-quality-control" class="anchor"></a>Step 1: FastQC Quality Control</h3>
<p>Quality control performed on the FASTQ files can identify low quality or failed samples that you may want to exclude from downstream analysis. FastQC is the standard tool used for quality control of fastq files. Run FastQC on the command line on each individual file:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">fastqc</span> input_read_1.fastq.gz</code></pre></div>
<p>This command will create a fastqc HTML report and a zip file of fastqc result files. To combine individual sample reports into a single report, use multiqc:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">multiqc</span> .</code></pre></div>
</div>
<div id="step-2-gather-sample-annotations-and-feature-annotations" class="section level3">
<h3 class="hasAnchor">
<a href="#step-2-gather-sample-annotations-and-feature-annotations" class="anchor"></a>Step 2: Gather sample annotations and feature annotations</h3>
<p>The single cell object contains raw read counts, normalized expression data, sample annotations, and feature annotations along with downstream analysis. While you can make a single cell object with just a count matrix, to take advantage of some of the analyses available in the single cell toolkit, sample annotation information and feature information may be helpful.</p>
<div id="sample-annotation-data-frame" class="section level4">
<h4 class="hasAnchor">
<a href="#sample-annotation-data-frame" class="anchor"></a>Sample Annotation Data Frame</h4>
<p>The sample annotation data frame should contain columns of annotation information and row names that match the sample names in the count matrix. The output table <code>multiqc_fastqc.txt</code> can be used as a sample annotation data frame.</p>
</div>
<div id="feature-annotation-data-frame" class="section level4">
<h4 class="hasAnchor">
<a href="#feature-annotation-data-frame" class="anchor"></a>Feature Annotation Data Frame</h4>
<p>The feature annotation data frame should contain columns of annotation information and row names that match the row names in the count matrix.</p>
</div>
</div>
<div id="step-3-align-the-data-with-alignsinglecelldata" class="section level3">
<h3 class="hasAnchor">
<a href="#step-3-align-the-data-with-alignsinglecelldata" class="anchor"></a>Step 3: Align the data with alignSingleCellData()</h3>
<p>For basic alignment and feature counting you can use the <code><a href="../reference/alignSingleCellData.html">alignSingleCellData()</a></code> function in the single cell toolkit to align fastq data to a reference genome, count the number of reads per gene, and create a single cell object that contains annotation information. After running the <code><a href="../reference/alignSingleCellData.html">alignSingleCellData()</a></code> function you can take the single cell object directly into the shiny app for downstream analysis. Detailed information about the options available in the <code><a href="../reference/alignSingleCellData.html">alignSingleCellData()</a></code> function can be found in the function help, but here is an example command:</p>
<blockquote>
<p><strong>NOTE</strong>: Alignment to large genomes (the human genome) can take more than 8GB of memory. Make sure to run the <code><a href="../reference/alignSingleCellData.html">alignSingleCellData()</a></code> function from a computer with sufficient memory.</p>
</blockquote>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">singlecellobject &lt;-<span class="st"> </span><span class="kw"><a href="../reference/alignSingleCellData.html">alignSingleCellData</a></span>(<span class="dt">inputfile1 =</span> <span class="kw">c</span>(
  <span class="st">"/path/to/sample1_1.fastq.gz"</span>,<span class="st">"/path/to/sample2_1.fastq.gz"</span>),
  <span class="dt">inputfile2 =</span> <span class="kw">c</span>(<span class="st">"/path/to/sample1_2.fastq.gz"</span>,
                 <span class="st">"/path/to/sample2_2.fastq.gz"</span>),
  <span class="dt">index_path =</span> <span class="st">"/path/to/genome/index"</span>,
  <span class="dt">gtf_annotation =</span> <span class="st">"/path/to/gene/annotations.gtf"</span>,
  <span class="dt">sample_annotations =</span> sample.annotation.df,
  <span class="dt">threads=</span><span class="dv">4</span>)</code></pre></div>
<div id="parallelization" class="section level4">
<h4 class="hasAnchor">
<a href="#parallelization" class="anchor"></a>Parallelization</h4>
<p>While setting the number of threads in the <code><a href="../reference/alignSingleCellData.html">alignSingleCellData()</a></code> function will increase the speed of each individual alignment, the function will align each sample sequentially. To save time, you may want to perform alignment and feature counting on each file in a parallel computing environment and combine the samples to create an SCtkExperiment object. The <code><a href="../reference/alignSingleCellData.html">alignSingleCellData()</a></code> function can be run on an individual file. Once individual sample objects have been created, combine them using the <code>mergeSCESet()</code> function.</p>
</div>
</div>
</div>
<div id="creating-a-single-cell-object-from-aligned-bam-reads" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-a-single-cell-object-from-aligned-bam-reads" class="anchor"></a>Creating a Single Cell Object from aligned BAM reads</h2>
<p>If your data requires additional processing steps (UMI normalization, etc), you can use the <code><a href="../reference/alignSingleCellData.html">alignSingleCellData()</a></code> function to perform feature counting and create a single cell object without the alignment step if ‘.bam’ files are provided in the <code>inputfile1</code> parameter. Note that if your read data was sequenced using paired-end reads, set the <code>isPairedEnd</code> parameter <code>TRUE</code>.</p>
</div>
<div id="manually-creating-a-single-cell-object-with-createsce" class="section level2">
<h2 class="hasAnchor">
<a href="#manually-creating-a-single-cell-object-with-createsce" class="anchor"></a>Manually Creating a Single Cell Object with createSCE()</h2>
<p>If you choose to align and quantify your data using alternative tools, you can use the <code><a href="../reference/createSCE.html">createSCE()</a></code> function to create a single cell object that can be used in the single cell toolkit by providing a count data frame and optionally providing sample and feature annotations. See the <code><a href="../reference/createSCE.html">createSCE()</a></code> function help page for details.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#suggested-software">Suggested Software</a></li>
      <li><a href="#tutorial-single-cell-rna-seq-reads-to-single-cell-object">Tutorial, Single Cell RNA-Seq Reads to Single Cell Object</a></li>
      <li><a href="#creating-a-single-cell-object-from-aligned-bam-reads">Creating a Single Cell Object from aligned BAM reads</a></li>
      <li><a href="#manually-creating-a-single-cell-object-with-createsce">Manually Creating a Single Cell Object with createSCE()</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by David Jenkins, Tyler Faits, Emma Briars, Sebastian Carrasco Pro, W. Evan Johnson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
