<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Batch Correction • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Batch Correction">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.14.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Batch Correction</h1>
                        <h4 data-toc-skip class="author">Yichen
Wang</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/batch_correction.Rmd" class="external-link"><code>vignettes/articles/batch_correction.Rmd</code></a></small>
      <div class="hidden name"><code>batch_correction.Rmd</code></div>

    </div>

    
    

<meta name="viewport" content="width=device-width, initial-scale=1">
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This guide helps users with performing batch correction when
necessary. SingleCellTK provides 11 methods that are already published
including <code>BBKNN</code>, <code>ComBatSeq</code>,
<code>FastMNN</code>, <code>MNN</code>, <code>Harmony</code>,
<code>LIGER</code>, <code>Limma</code>, <code>Scanorama</code>,
<code>scMerge</code>, <code>Seurat</code> integration and
<code>ZINBWaVE</code>. All of these methods are available to use through
our shiny ui application as well as through the R console environment
through our wrapper functions.</p>
<p>These methods accept various type of input expression matrices
(e.g. raw counts or log-normalized counts), and generate either a new
corrected expression matrix or a low-dimensional representation. Users
should choose a function according to their needs. More detailed
instructions on how to use these methods either through the shiny ui
application (select “Interactive Analysis”) or through the R console
(select “Console Analysis”) are described below:</p>
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p><strong>Entry of the Panel</strong></p>
<p>From anywhere of the UI, the panel for normalization and batch
correction can be accessed from the top navigation panel at the boxed
tab. After this, click on the second boxed tab to enter the sub-page for
batch correction.</p>
<p><img src="ui_screenshots/batch_correction/bc_ui_entry.png" alt="Entry"><br></p>
<p><strong>Perform the Correction</strong></p>
<p>The UI is constructed in a sidebar style, where the left-sided
sidebar works for selecting method, setting the parameters and running
the batch correction, and the right part main panel is for visualization
checking.</p>
<p><img src="ui_screenshots/batch_correction/bc_ui_param.png" alt="Parameters"><br></p>
<p>For running any types of batch correction, there are always three
essential inputs that users should be sure with:</p>
<ul>
<li>The batch correction method - selection input <strong>“Select Batch
Correction Method”</strong>. All methods supported are listed as
options, in the alphabetic order.</li>
<li>The data matrix to correct - selection input <strong>“Select
Assay”</strong>. In terms of batch correction, SCTK always requires a
full-sized feature expression data (i.e. assay) as a valid input.</li>
<li>The annotation for batches - selection input <strong>“Select Batch
Annotation”</strong>. Users will choose from the cell annotation list.
The chosen annotation class should be able to define from which batch
each cell sampled. Usually, <code>"Sample"</code> or
<code>"sample"</code>, as the preprocessing steps generate this
annotation by default.</li>
</ul>
<p>After the batch correction method is confirmed, the lower part will
dynamically switch to the method specific settings. Different types of
method have various parameter requirements, and various forms of output.
For the detailed help information of each method, users can go to <a href="../reference/index.html#batch-effect-correction">reference</a>
page and click on the function that stands for the method chosen, or
click on the “(help for {method})” link in the UI.</p>
<p><strong>Visualization</strong></p>
<p><img src="ui_screenshots/batch_correction/bc_ui_vis.png" alt="Visualization"><br></p>
<p>SCTK adopts two approaches to demonstrate the effect of batch
correction:</p>
 <style>
div.offset { background-color:inherit ; border-radius: 5px; margin-left: 25px;}
</style>
<div class="offset">
1. SCTK attempts to plot the variance explained by annotations calculated with the assay to correct and the output data matrix. 
<details><summary><b>Explanation: Explained Variance Plot</b></summary><p>There will be three columns in the plots for the variance explained.
The third one is for the variance explained by batch division, the
second is for that of an additional condition, which is optional, and
the first one is the variance explained by combining both annotations.
The additional condition is usually an annotation that indicates the
differences between cells that should not be eliminated, such as cell
types and treatments. In the first plot, the combined variance should
ideally be close to the sum of the two types of variance. (i.e. the
height of the first column should be closed to the sum of the heights of
the other two) As a result that makes sense, the changes between the two
plots in the second column should not be too much, while the third
column should be largely eliminated after correction.</p>
</details>
</div>
<br><div class="offset">
2. SCTK plots the PCA projections of the assay selected for correction, and the output corrected data matrix. For both cases, user should use the same batch annotation as used when running the algorithm to see a plausible result.  
<details><summary><b>Explanation: PCA plots</b></summary><p>In the PCA plots, cells (i.e. dots) will be labeled with different
colors to mark the batch belonging, and different shapes to mark for the
additional condition, which was mentioned in the previous paragraph.
Usually in the result, in the plot for the output, users can observe
that different colors of dots are mixed up together, while different
shapes are still grouped into separate clusters.</p>
</details>
</div>
<br><details><summary><b>Visualization Parameters</b></summary><ul>
<li>The data matrix used for correction - selection input
<strong>“Original Assay”</strong>. User should choose the one that was
used when generating the result they want to visualize.</li>
<li>The batch annotation used for correction - selection input
<strong>“Batch Annotation”</strong>. User should choose the one that was
used when generating the result they want to visualize.</li>
<li>The additional condition - selection input <strong>“Additional
Condition (optional)”</strong>. See the explanation in the previous
paragraphs. Such a condition that makes the most sense does not always
exist, so it becomes optional in SCTK.</li>
<li>The result to visualize - selection input <strong>“Corrected
Matrix”</strong>. Only the results generated in the current SCTK UI
session will be gathered in the option list here.</li>
</ul></details>
</div>

<div id="console" class="tabcontent">
<p><strong>Before Correction</strong></p>
<p>To run the pipeline, the most basic requirements are:</p>
<ul>
<li>An assay of expression value, usually pre-processed.</li>
<li>The annotation of the batches.</li>
</ul>
<p>As we adopt <a href="https://rdrr.io/bioc/SingleCellExperiment/" class="external-link"><code>SingleCellExperiment</code></a>
(<code>sce</code>) object through out the whole SCTK for a collection of
all matrices and metadatas of the dataset of interests, the assay to be
corrected, called <code>"assayToCorr"</code>, has to be saved at
<code>assay(sce, "assayToCorr")</code>. Meanwhile, the batch annotation
information has to be saved in a column of
<code>colData(sce)</code>.</p>
<blockquote>
<p>Note that the batch annotation should better be saved as a
<code>factor</code> in the <code>colData</code>, especially when the
batches are represented by integer numbers, because some downstream
analysis are likely to parse the non-character and non-logical
information as continuous values instead of categorical values.</p>
</blockquote>
<p><strong>Choosing a Method</strong></p>
<p>As mentioned in the introduction, our 11 methods vary in type of
input and output. Input type for the expression matrix could be raw
counts, log-transformed counts or scaled counts. Users need to make sure
the assay being used is of the correct type. The output matrix can be an
expression matrix (<code>assay</code>), a low-dimensional representation
(<code>reducedDim</code>), or a subset features of expression matrix
(<code>altExp</code>). Users need to choose a method that produce a
matrix that fits in their analysis.</p>
<details><summary><b>Example</b></summary><p><strong>1. Prepare an SCE object with multiple batches</strong></p>
<p>Here we present an example dataset that is combined from “pbmc3k” and
“pbmc4k”, originally available from package <a href="https://bioconductor.org/packages/release/data/experiment/html/TENxPBMCData.html" class="external-link">TENxPBMCData</a>,
which you can import by a function called
<code><a href="../reference/importExampleData.html">importExampleData()</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span><span class="va">pbmc6k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">'pbmc6k'</span><span class="op">)</span></span>
<span><span class="va">pbmc8k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">'pbmc8k'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmc6k</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'genes'</span>, <span class="st">'cells'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmc8k</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'genes'</span>, <span class="st">'cells'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "32738 genes" "5419 cells" </span></span>
<span><span class="co">## [1] "33694 genes" "8381 cells"</span></span></code></pre>
<p>SCTK has a function called <code><a href="../reference/combineSCE.html">combineSCE()</a></code>, which accepts a
list of SCE objects as input and returns a combined SCE object. This
function requires that the number of genes in each SCE object has to be
the same, and the gene metadata (i.e. <code>rowData</code>) has to match
with each other if the same fields exist. Therefore, we need some
pre-process for the combination. Users do not necessarily have to follow
the same way, depending on how the raw datasets are provided.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Combine the two SCE objects</span></span>
<span><span class="va">sce.combine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineSCE.html">combineSCE</a></span><span class="op">(</span>sceList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pbmc6k <span class="op">=</span> <span class="va">pbmc6k</span>, pbmc8k <span class="op">=</span> <span class="va">pbmc8k</span><span class="op">)</span>, by.r <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">pbmc6k</span><span class="op">)</span><span class="op">)</span>, by.c <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc6k</span><span class="op">)</span><span class="op">)</span>, combined <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce.combine</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## pbmc6k pbmc8k </span></span>
<span><span class="co">##   5419   8381</span></span></code></pre>
<p>In this manual, we only present a toy example instead of a best
practice for real data. Therefore, QC and filtering are skipped.<br>
Additionally, most of the batch correction methods provided
<strong>require a log-normalized assay</strong> as input, users can
check the required assay type of each method by looking at its default
setting. (i.e. if by default <code>useAssay = "logcounts"</code>, then
it requires log-normalized assay; else if by default
<code>useAssay = "counts"</code>, then the raw count assay.)</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Simply filter out the genes that are expressed in less than 1% of all cells.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce.combine</span><span class="op">)</span><span class="op">$</span><span class="va">expPerc</span> <span class="op">&lt;-</span> <span class="fu">rowSums</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce.combine</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce.combine</span><span class="op">)</span></span>
<span><span class="va">sce.filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetSCERows.html">subsetSCERows</a></span><span class="op">(</span><span class="va">sce.combine</span>, rowData <span class="op">=</span> <span class="st">"expPerc &gt;= 0.01"</span>, returnAsAltExp <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sce.filter</span><span class="op">)</span></span>
<span><span class="va">sce.small</span> <span class="op">&lt;-</span> <span class="va">sce.filter</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sce.filter</span><span class="op">)</span>, <span class="fl">800</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce.filter</span><span class="op">)</span>, <span class="fl">800</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sce.small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaterlogNormCounts.html">scaterlogNormCounts</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce.small</span>, useAssay <span class="op">=</span> <span class="st">'counts'</span>, assayName <span class="op">=</span> <span class="st">'logcounts'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sce.small</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 10440 13800 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(10440): AP006222.2 RP11-206L10.9 ... SLC5A3 AC007325.4</span></span>
<span><span class="co">## rowData names(5): rownames ENSEMBL_ID Symbol_TENx Symbol expPerc</span></span>
<span><span class="co">## colnames(13800): pbmc6k_AAACATACAACCAC-1 pbmc6k_AAACATACACCAGT-1 ...</span></span>
<span><span class="co">##   pbmc8k_TTTGTCATCTCGAGTA-1 pbmc8k_TTTGTCATCTGCTTGC-1</span></span>
<span><span class="co">## colData names(13): rownames Sample ... Date_published sample</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 800 800 </span></span>
<span><span class="co">## metadata(1): assayType</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(800): CHPT1 SLC15A4 ... ZFAND2B LINC01473</span></span>
<span><span class="co">## rowData names(5): rownames ENSEMBL_ID Symbol_TENx Symbol expPerc</span></span>
<span><span class="co">## colnames(800): pbmc6k_ACGTTACTAGCCAT-1 pbmc8k_ATTGGACTCTCGAGTA-1 ...</span></span>
<span><span class="co">##   pbmc8k_GCAAACTCAGGTCCAC-1 pbmc6k_GCGAGCACCTCAAG-1</span></span>
<span><span class="co">## colData names(14): rownames Sample ... sample sizeFactor</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p><strong>2. Run a batch correction method on the prepared SCE
object</strong></p>
<p>The basic way to run a batch correction method from SingleCellTK is
to select a function for the corresponding <strong>method</strong>,
input the <strong>SCE object</strong>, specify the
<strong>assay</strong> to correct, and the <strong>batch
annotation</strong>.</p>
<p>For example, here we will try the batch correction method provided by
<a href="https://rdrr.io/bioc/limma/man/removeBatchEffect.html" class="external-link">Limma</a>,
which fits a linear model to the data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runLimmaBC.html">runLimmaBC</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce.small</span>, useAssay <span class="op">=</span> <span class="st">'logcounts'</span>, batch <span class="op">=</span> <span class="st">'sample'</span>, assayName <span class="op">=</span> <span class="st">'LIMMA'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sce.small</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 800 800 </span></span>
<span><span class="co">## metadata(2): assayType batchCorr</span></span>
<span><span class="co">## assays(3): counts logcounts LIMMA</span></span>
<span><span class="co">## rownames(800): CHPT1 SLC15A4 ... ZFAND2B LINC01473</span></span>
<span><span class="co">## rowData names(5): rownames ENSEMBL_ID Symbol_TENx Symbol expPerc</span></span>
<span><span class="co">## colnames(800): pbmc6k_ACGTTACTAGCCAT-1 pbmc8k_ATTGGACTCTCGAGTA-1 ...</span></span>
<span><span class="co">##   pbmc8k_GCAAACTCAGGTCCAC-1 pbmc6k_GCGAGCACCTCAAG-1</span></span>
<span><span class="co">## colData names(14): rownames Sample ... sample sizeFactor</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p><strong>3. Visualization</strong></p>
<p>In this documentation, we provide three ways to examine the removal
of batch effect, in terms of visualization.</p>
<ol style="list-style-type: decimal">
<li>Plot the variation explained by the batch annotation and another
condition.</li>
</ol>
<p>This functionality is implemented in
<code><a href="../reference/plotBatchVariance.html">plotBatchVariance()</a></code>. It plots a violin plot of the
variation explained by the given batch annotation, an additional
condition, and the variation explained by combining these two
conditions.<br>
This plot would be useful for examining whether the existing batch
effect is different than another condition (e.g. subtype) or is
confounded by that. However, the additional condition labels (e.g. cell
types) do not necessarily exist when batch effect removal is wanted, so
only plotting the variation explained by batches is also supported.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBatchVariance.html">plotBatchVariance</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce.small</span>, useAssay <span class="op">=</span> <span class="st">'logcounts'</span>, batch <span class="op">=</span> <span class="st">'sample'</span>, title <span class="op">=</span> <span class="st">'Variation Before Correction'</span><span class="op">)</span></span></code></pre></div>
<p><img src="batch_correction_files/figure-html/batchVarBefore-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBatchVariance.html">plotBatchVariance</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce.small</span>, useAssay <span class="op">=</span> <span class="st">'LIMMA'</span>, batch <span class="op">=</span> <span class="st">'sample'</span>, title <span class="op">=</span> <span class="st">'Variation After Correction'</span><span class="op">)</span></span></code></pre></div>
<p><img src="batch_correction_files/figure-html/batchVarAfter-1.png" width="700"></p>
<ol start="2" style="list-style-type: decimal">
<li>Plot the mean expression level of each gene separately for each
batch.</li>
</ol>
<p>This functionality is implemented in
<code><a href="../reference/plotSCEBatchFeatureMean.html">plotSCEBatchFeatureMean()</a></code>. The methodology is straight
forward, which plots violin plots for all the batches, and within each
batch, the plot illustrates the distribution of the mean expression
level of each gene. Thus the batch effect can be observed from the mean
and standard deviation of each batch.</p>
<ol start="3" style="list-style-type: decimal">
<li>Plot an embedding to see the grouping of cells</li>
</ol>
<p>There is no function special for batch correction, but this can be
achieved simply by using the <a href="dimensionality_reduction.html">dimension reduction</a>/<a href="2d_embedding.html">2D embedding</a> calculation functions
(e.g. <code><a href="../reference/scaterPCA.html">scaterPCA()</a></code>, <code><a href="../reference/runUMAP.html">runUMAP()</a></code> and
<code>run</code>TSNE()<code>) and</code>plotSCEDimReduceColData()`.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce.small</span>, initialDims <span class="op">=</span> <span class="fl">25</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>, useReducedDim <span class="op">=</span> <span class="cn">NULL</span>, logNorm <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSCEDimReduceColData.html">plotSCEDimReduceColData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce.small</span>, colorBy <span class="op">=</span> <span class="st">'sample'</span>, reducedDimName <span class="op">=</span> <span class="st">'UMAP'</span>, title <span class="op">=</span> <span class="st">'UMAP Before Correction'</span><span class="op">)</span></span></code></pre></div>
<p><img src="batch_correction_files/figure-html/umapBefore-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce.small</span>, useAssay <span class="op">=</span> <span class="st">'LIMMA'</span>, useReducedDim <span class="op">=</span> <span class="cn">NULL</span>, initialDims <span class="op">=</span> <span class="fl">25</span>, logNorm <span class="op">=</span> <span class="cn">FALSE</span>, reducedDimName <span class="op">=</span> <span class="st">'UMAP_LIMMA'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSCEDimReduceColData.html">plotSCEDimReduceColData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce.small</span>, colorBy <span class="op">=</span> <span class="st">'sample'</span>, reducedDimName <span class="op">=</span> <span class="st">'UMAP_LIMMA'</span>, title <span class="op">=</span> <span class="st">'UMAP After Correction'</span><span class="op">)</span></span></code></pre></div>
<p><img src="batch_correction_files/figure-html/umapAfter-1.png" width="700"></p>
<p>If the cell typing is already given, it is strongly recommended to
specify <code>shape = {colData colname for cell typing}</code> in
<code><a href="../reference/plotSCEDimReduceColData.html">plotSCEDimReduceColData()</a></code> to visualize the grouping
simultaneously.</p>
</details><details><summary><b>Method Reference</b></summary><table class="table">
<colgroup>
<col width="25%">
<col width="25%">
<col width="25%">
<col width="25%">
</colgroup>
<thead><tr class="header">
<th align="center">Method</th>
<th align="center">Function</th>
<th align="center">Input Type</th>
<th align="center">Output Type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">BBKNN <span class="citation">[1]</span>
</td>
<td align="center"><code><a href="../reference/runBBKNN.html">runBBKNN()</a></code></td>
<td align="center">Scaled</td>
<td align="center"><code>reducedDim</code></td>
</tr>
<tr class="even">
<td align="center">ComBatSeq <span class="citation">[2]</span>
</td>
<td align="center"><code><a href="../reference/runComBatSeq.html">runComBatSeq()</a></code></td>
<td align="center">raw counts</td>
<td align="center"><code>assay</code></td>
</tr>
<tr class="odd">
<td align="center">FastMNN <span class="citation">[3]</span>
</td>
<td align="center"><code><a href="../reference/runFastMNN.html">runFastMNN()</a></code></td>
<td align="center">log-normalized</td>
<td align="center"><code>reducedDim</code></td>
</tr>
<tr class="even">
<td align="center">MNN <span class="citation">[3]</span>
</td>
<td align="center"><code><a href="../reference/runMNNCorrect.html">runMNNCorrect()</a></code></td>
<td align="center">log-normalized</td>
<td align="center"><code>assay</code></td>
</tr>
<tr class="odd">
<td align="center">Harmony <span class="citation">[4]</span>
</td>
<td align="center"><code><a href="../reference/runHarmony.html">runHarmony()</a></code></td>
<td align="center">PCA</td>
<td align="center"><code>reducedDim</code></td>
</tr>
<tr class="even">
<td align="center">LIGER <span class="citation">[5]</span>
</td>
<td align="center"><code>runLIGER()</code></td>
<td align="center">raw counts</td>
<td align="center"><code>reducedDim</code></td>
</tr>
<tr class="odd">
<td align="center">Limma <span class="citation">[6]</span>
</td>
<td align="center"><code><a href="../reference/runLimmaBC.html">runLimmaBC()</a></code></td>
<td align="center">log-normalized</td>
<td align="center"><code>assay</code></td>
</tr>
<tr class="even">
<td align="center">Scanorama <span class="citation">[7]</span>
</td>
<td align="center"><code><a href="../reference/runSCANORAMA.html">runSCANORAMA()</a></code></td>
<td align="center">log-normalized</td>
<td align="center"><code>assay</code></td>
</tr>
<tr class="odd">
<td align="center">scMerge <span class="citation">[8]</span>
</td>
<td align="center"><code><a href="../reference/runSCMerge.html">runSCMerge()</a></code></td>
<td align="center">log-normalized</td>
<td align="center"><code>assay</code></td>
</tr>
<tr class="even">
<td align="center">Seurat Integration <span class="citation">[9]</span>
</td>
<td align="center"><code><a href="../reference/runSeuratIntegration.html">runSeuratIntegration()</a></code></td>
<td align="center">raw counts</td>
<td align="center"><code>altExp</code></td>
</tr>
<tr class="odd">
<td align="center">ZINBWaVE <span class="citation">[10]</span>
</td>
<td align="center"><code><a href="../reference/runZINBWaVE.html">runZINBWaVE()</a></code></td>
<td align="center">raw counts</td>
<td align="center"><code>reducedDim</code></td>
</tr>
</tbody>
</table></details>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body" entry-spacing="0">
<div id="ref-Polanski2020" class="csl-entry">
<div class="csl-left-margin">[1] </div>
<div class="csl-right-inline">K.
Polański, M. D. Young, Z. Miao, K. B. Meyer, S. A. Teichmann, and J. E.
Park, <span>“<span class="nocase">BBKNN: fast batch alignment of single
cell transcriptomes</span>,”</span> <em>Bioinformatics</em>, vol. 36,
pp. 964--965, Aug. 2020, doi: <a href="https://doi.org/10.1093/bioinformatics/btz625" class="external-link">10.1093/bioinformatics/btz625</a>.</div>
</div>
<div id="ref-ZhangY2020" class="csl-entry">
<div class="csl-left-margin">[2] </div>
<div class="csl-right-inline">Y.
Zhang, G. Parmigiani, and W. E. Johnson, <span>“<span class="nocase">ComBat-seq: batch effect adjustment for RNA-seq count
data</span>,”</span> <em>NAR Genomics and Bioinformatics</em>, vol. 2,
p. lqaa078, Sep. 2020, doi: <a href="https://doi.org/10.1093/nargab/lqaa078" class="external-link">10.1093/nargab/lqaa078</a>.</div>
</div>
<div id="ref-Haghverdi2018" class="csl-entry">
<div class="csl-left-margin">[3] </div>
<div class="csl-right-inline">L.
Haghverdi, A. T. L. Lun, M. D. Morgan, and J. C. Marioni, <span>“<span class="nocase">Batch effects in single-cell RNA-sequencing data are
corrected by matching mutual nearest neighbors</span>,”</span>
<em>Nature Biotechnology</em>, vol. 36, pp. 421--427, Apr. 2018, doi: <a href="https://doi.org/10.1038/nbt.4091" class="external-link">10.1038/nbt.4091</a>.</div>
</div>
<div id="ref-Korsunsky2019" class="csl-entry">
<div class="csl-left-margin">[4] </div>
<div class="csl-right-inline">I.
Korsunsky <em>et al.</em>,<span>“<span class="nocase"> Fast, sensitive
and accurate integration of single-cell data with
Harmony</span>,”</span> <em>Nature Methods</em>, vol. 16, pp.
1289--1296, Nov. 2019, doi: <a href="https://doi.org/10.1038/s41592-019-0619-0" class="external-link">10.1038/s41592-019-0619-0</a>.</div>
</div>
<div id="ref-LiuL2020" class="csl-entry">
<div class="csl-left-margin">[5] </div>
<div class="csl-right-inline">J.
Liu, C. Gao, J. Sodicoff, V. Kozareva, E. Z. Macosko, and J. D. Welch,
<span>“<span class="nocase">Jointly defining cell types from multiple
single-cell datasets using LIGER</span>,”</span> <em>Nature
Protocols</em>, vol. 15, pp. 3632--3662, Oct. 2020, doi: <a href="https://doi.org/10.1038/s41596-020-0391-8" class="external-link">10.1038/s41596-020-0391-8</a>.</div>
</div>
<div id="ref-Ritchie2015" class="csl-entry">
<div class="csl-left-margin">[6] </div>
<div class="csl-right-inline">M.
E. Ritchie <em>et al.</em>, <span>“<span class="nocase">limma powers
differential expression analyses for RNA-sequencing and microarray
studies</span>,”</span> <em>Nucleic Acids Research</em>, vol. 43, p.
e47, Apr. 2015, doi: <a href="https://doi.org/10.1093/nar/gkv007" class="external-link">10.1093/nar/gkv007</a>.</div>
</div>
<div id="ref-Hie2019" class="csl-entry">
<div class="csl-left-margin">[7] </div>
<div class="csl-right-inline">B.
Hie, B. Bryson, and B. Berger, <span>“<span class="nocase">Efficient
integration of heterogeneous single-cell transcriptomes using
Scanorama</span>,”</span> <em>Nature Biotechnology</em>, vol. 37, pp.
685--691, May 2019, doi: <a href="https://doi.org/10.1038/s41587-019-0113-3" class="external-link">10.1038/s41587-019-0113-3</a>.</div>
</div>
<div id="ref-LinY2019" class="csl-entry">
<div class="csl-left-margin">[8] </div>
<div class="csl-right-inline">Y.
Lin <em>et al.</em>, <span>“<span class="nocase">scMerge leverages
factor analysis, stable expression, and pseudoreplication to merge
multiple single-cell RNA-seq datasets</span>,”</span> <em>PNAS</em>,
vol. 116, pp. 9775–9784, May 2019, doi: <a href="https://doi.org/10.1073/pnas.1820006116" class="external-link">10.1073/pnas.1820006116</a>.</div>
</div>
<div id="ref-Butler2018a" class="csl-entry">
<div class="csl-left-margin">[9] </div>
<div class="csl-right-inline">A.
Butler, P. Hoffman, P. Smibert, E. Papalexi, and R. Satija, <span>“<span class="nocase">Integrating single-cell transcriptomic data across
different conditions, technologies, and species</span>,”</span>
<em>Nature Biotechnology</em>, vol. 36, no. 5, pp. 411–420, Jun. 2018,
doi: <a href="https://doi.org/10.1038/nbt.4096" class="external-link">10.1038/nbt.4096</a>.</div>
</div>
<div id="ref-Risso2018" class="csl-entry">
<div class="csl-left-margin">[10] </div>
<div class="csl-right-inline">D.
Risso, F. Perraudeau, S. Gribkova, S. Dudoit, and J.-P. Vert,
<span>“<span class="nocase">A general and flexible method for signal
extraction from single-cell RNA-seq data</span>,”</span> <em>Nature
Communications</em>, vol. 9, no. 284, Jan. 2018, doi: <a href="https://doi.org/10.1038/s41467-017-02554-5" class="external-link">10.1038/s41467-017-02554-5</a>.</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
