<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Celda Curated Workflow • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Celda Curated Workflow">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.13.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Celda Curated Workflow</h1>
                        <h4 data-toc-skip class="author">Nida
Pervaiz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/celda_curated_workflow.Rmd" class="external-link"><code>vignettes/articles/celda_curated_workflow.Rmd</code></a></small>
      <div class="hidden name"><code>celda_curated_workflow.Rmd</code></div>

    </div>

    
    

<meta name="viewport" content="width=device-width, initial-scale=1">
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Celda is a Bayesian hierarchical model that can perform bi-clustering
of features into modules and observations into subpopulations.</p>
<p>To view instructions on how to apply Celda to single-cell RNA
sequencing (scRNA-seq) datasets, please select ‘Interactive Analysis’
for using Celda in shiny application or ‘Console Analysis’ for using
this package on R console from the tabs below. The detailed tutorial of
Celda can be found at <a href="https://www.camplab.net/celda/" class="external-link">Celda’s
official website</a>.</p>
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p><strong>Entry of The Panel</strong></p>
<p>In this tutorial example, we illustrate all the steps of the curated
workflow and focus on the options available to manipulate and customize
the steps of the workflow as per user requirements. From anywhere of the
UI, the panel for celda can be accessed from the top navigation panel at
the circled tab shown below by simply clicking on the ‘Curated
Workflows’ from the top menu and then select <code>Celda</code>:</p>
<p><img src="ui_screenshots/celdaAnalysis/entry.PNG" alt="celdaEntry"><br></p>
<p>NOTE: This tutorial assumes that the data has already been uploaded
via the upload tab of the toolkit and filtered before using the
workflow.</p>
<p><strong>Run Celda Analysis</strong></p>
<p><strong>1. Identify number of feature modules</strong></p>
<p>To identify number of feature modules, there are always five
essential inputs that users should be sure with:</p>
<ul>
<li>The data matrix to use - selection input <strong>“Choose an
Assay”</strong>. In terms of celda analysis, SCTK always requires a
full-sized feature expression data (i.e. assay) as a valid input.</li>
<li>Feature selection method - selection input <strong>“Choose Feature
Selection Method”</strong>. Select feature selection method from ‘None’
, ‘runSeuratFindHVG’ and ‘Scran_modelGeneVar’.</li>
<li>Minimum number of counts required for feature selection - integer
<strong>“Keep features with this many counts”</strong>.</li>
<li>Minimum number of cells required for feature selection - integer
<strong>“In at least this many cells”</strong>.</li>
<li>Number of initial feature modules - integer <strong>“Select Number
of Initial Feature Modules”</strong>.</li>
<li>Number of maximum feature modules - integer <strong>“Select Number
of Maximum Feature Modules”</strong>.</li>
</ul>
<p><img src="ui_screenshots/celdaAnalysis/module1.PNG" alt="celdamodule1"><br></p>
<p>After the feature selection method is confirmed, the lower part will
dynamically switch to the method specific settings.</p>
 <style>
div.offset { background-color:inherit ; border-radius: 5px; margin-left: 15px;}
</style>
<div class="offset">
<details><summary><b>runSeuratFindHVG</b></summary><p>The method specific parameters for <code>runSeuratFindHVG</code>
includes:</p>
<ul>
<li>Select method for computation of highly variable genes - selection
input <strong>“Select HVG method”</strong>. <code>Seurat</code> provides
three methods for variable genes identification i.e. <code>vst</code>
(uses local polynomial regression to fit a relationship between log of
variance and log of mean), <code>mean.var.plot</code> (uses mean and
dispersion to divide features into bins) and <code>dispersion</code>
(uses highest dispersion values only).</li>
<li>Select variable features - selection input <strong>“Select number of
variable features”</strong>.Input the number of genes that should be
identified as variable. Default is 2500.</li>
</ul></details><details><summary><b>Scran_modelGeneVar</b></summary><p>The method specific parameters for <code>Scran_modelGeneVar</code>
includes:</p>
<ul>
<li>Select variable features - selection input <strong>“Select number of
variable features”</strong>.Input the number of genes that should be
identified as variable. Default is 2500.</li>
</ul></details>
</div>
<br><p>After feature selection method and it’s parameters are set, user can
click recursive module split button to run
<code>recursiveSplitModule</code> function which is used used to cluster
features into modules for a range of feature modules (L). Two plots are
shown when user runs recursive module split; perplexity plot and rate of
perplexity change. To evaluate the performance of individual models we
use perplexity plot by calculating the probability of observing
expression counts given an estimated Celda model. The perplexity alone
often does not show a clear elbow or “leveling off”. However, the rate
of perplexity change (RPC) can be more informative.</p>
<p><img src="ui_screenshots/celdaAnalysis/module2.PNG" alt="celdamodule2"><br></p>
<p><img src="ui_screenshots/celdaAnalysis/module3.PNG" alt="celdamodule3"><br></p>
<p><strong>2. Identify number of cell clusters</strong></p>
<p>To identify number of cell clusters, there are always two essential
inputs that users should be sure with:</p>
<ul>
<li>Number of initial cell modules - integer <strong>“Select Number of
Initial Cell Modules”</strong>. Initial number of cell populations to
try. Default 5.</li>
<li>Number of maximum cell modules - integer <strong>“Select Number of
Maximum Cell Modules”</strong>. Maximum number of cell populations to
try. Default 25.</li>
</ul>
<p><img src="ui_screenshots/celdaAnalysis/cluster1.PNG" alt="celdacluster1"><br></p>
<p>After these two parameters are set, user can click recursive cell
split button to run <code>recursiveSplitCell</code> function to cluster
cells into population for range of possible sub populations(K). Three
plots are shown when user runs recursive cell split; perplexity plot,
rate of perplexity change and preliminiary UMAP plot. To evaluate the
performance of individual models we use perplexity plot by calculating
the probability of observing expression counts given an estimated Celda
model. The perplexity alone often does not show a clear elbow or
“leveling off”. However, the rate of perplexity change (RPC) can be more
informative.</p>
<p><img src="ui_screenshots/celdaAnalysis/cluster2.PNG" alt="celdacluster2"><br></p>
<p><img src="ui_screenshots/celdaAnalysis/cluster3.PNG" alt="celdacluster3"><br></p>
<p>UMAP plot is useful for visualizing the relationship between
cells.</p>
<p><img src="ui_screenshots/celdaAnalysis/cluster4.PNG" alt="celdacluster4"><br></p>
<p><strong>3. Visualization</strong></p>
<p>After selecting a celda model with specific values of L and K, we can
then perform additional exploratory and downstream analyses to
understand the biology of the transcriptional modules and cell
populations.</p>
 <style>
div.offset { background-color:inherit ; border-radius: 5px; margin-left: 15px;}
</style>
<div class="offset">
<details><summary><b>Embeddings</b></summary><p>To visualize the relationships between the cells in a 2-D embedding
we can generate a dimension reduction plot with the Uniform Manifold
Approximation and Projection (UMAP) or a t-distributed stochastic
neighbor embedding (t-SNE) method. User can select from UMAP and TSNE
that which plot they want to visualize.</p>
<p><img src="ui_screenshots/celdaAnalysis/visualization1.PNG" alt="celdavisualization1"><br></p>
</details><details><summary><b>Module Analysis</b></summary><p>The Module Analysis tab contains the probabilities and heatmaps for
each module. Module probability plots will color cells by the
probability of each module on a 2-D embedding plot. Module heatmaps show
the relative expression for the features in a module across the cells.
Cells within the heatmap will be ordered from the lowest to highest
probability of the module. The parameter topCells can be used to control
the number of cells included in the heatmap.</p>
<p><img src="ui_screenshots/celdaAnalysis/visualization2.PNG" alt="celdavisualization2"><br></p>
</details><details><summary><b>Probability Map</b></summary><p>Celda has the ability to identify modules of co-expressed features
and quantify the probability of these modules in each cell population.
An overview of the relationships between modules and cell subpopulations
can be explored with the function celdaProbabilityMap.The “Absolute
probability” map gives insights into the absolute abundance of a module
within a given cell subpopulation. The absolute heatmap can be used to
explore which modules are higher than other modules within a cell
population. The “Relative expression” map shows the standardized
(z-scored) module probabilities across cell subpopulations. The relative
heatmap can be used to explore which modules are relatively higher than
other modules across cell populations.</p>
<p><img src="ui_screenshots/celdaAnalysis/visualization3.PNG" alt="celdavisualization3"><br></p>
</details>
</div>
<br><p><strong>4. Downstream Analysis</strong></p>
<p>Once all steps of the celda workflow are completed, users can further
analyze the data by directly going to the various downstream analysis
options (Differential Expression, Marker Selection &amp; Pathway
Analysis) from within the celda workflow.</p>
<p><img src="ui_screenshots/celdaAnalysis/downstream.PNG" alt="celdadownstream"><br></p>
<div class="offset">

  

</div>
</div>

<div id="console" class="tabcontent">
<p>All methods provided by SCTK for Celda workflow use a
<code>SingleCellExperiment</code> object both as an input and output. To
demonstrate simple and clear examples, here we use the “PBMC-3k” dataset
from “10X” which can be easily imported with SCTK functions.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span></code></pre></div>
<p>After the data has been uploaded, user may perform basic QC and
filtering via QC &amp; Filtering tab.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mito.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCellQC.html">runCellQC</a></span><span class="op">(</span><span class="va">sce</span>, sample <span class="op">=</span> <span class="cn">NULL</span>, algorithms <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"QCMetrics"</span>, <span class="st">"scDblFinder"</span>, <span class="st">"decontX"</span><span class="op">)</span>, geneSetList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mito<span class="op">=</span><span class="va">mito.genes</span><span class="op">)</span>, geneSetListLocation <span class="op">=</span> <span class="st">"rownames"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter SCE</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">sce</span>, colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"total &gt; 600"</span>, <span class="st">"detected &gt; 300"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># See number of cells after filtering</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<div class="offset">
<details><summary><b>Finding number of modules</b></summary><p>In general, removing features with low numbers of counts across all
cells is recommended to reduce computational run time. A simple
selection can be performed by removing features with a minimum number of
counts in a minimum number of cells using the selectFeatures function.
Other than this options such as runSeuratFindHVG and Scran_modelGeneVar
also exist for feature selection.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">celda</span><span class="op">)</span></span>
<span></span>
<span><span class="va">useAssay</span> <span class="op">&lt;-</span> <span class="st">"counts"</span></span>
<span><span class="va">altExpName</span> <span class="op">&lt;-</span> <span class="st">"featureSubset"</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/celda/man/selectFeatures.html" class="external-link">selectFeatures</a></span><span class="op">(</span><span class="va">sce</span>, minCount <span class="op">=</span> <span class="fl">3</span>, minCell <span class="op">=</span> <span class="fl">3</span>, useAssay <span class="op">=</span> <span class="va">useAssay</span>, altExpName <span class="op">=</span> <span class="va">altExpName</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># See number of features after filtering</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span>, <span class="va">altExpName</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">moduleSplit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/celda/man/recursiveSplitModule.html" class="external-link">recursiveSplitModule</a></span><span class="op">(</span><span class="va">sce</span>, useAssay <span class="op">=</span> <span class="va">useAssay</span>, altExpName <span class="op">=</span> <span class="va">altExpName</span>, initialL <span class="op">=</span> <span class="fl">10</span>, maxL <span class="op">=</span> <span class="fl">150</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#plots to decide value of L</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/celda/man/plotGridSearchPerplexity.html" class="external-link">plotGridSearchPerplexity</a></span><span class="op">(</span><span class="va">moduleSplit</span>, altExpName <span class="op">=</span> <span class="va">altExpName</span>, sep <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/celda/man/plotRPC.html" class="external-link">plotRPC</a></span><span class="op">(</span><span class="va">moduleSplit</span>, altExpName <span class="op">=</span> <span class="va">altExpName</span>, sep <span class="op">=</span> <span class="fl">10</span>, n <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p>Using the plots visualization we can then decide the value for L
i.e. the number of modules. The point at which RPC curve tends to level
off can be taken as the best value for L.</p>
</details><details><summary><b>Finding number of clusters</b></summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/celda/man/subsetCeldaList.html" class="external-link">subsetCeldaList</a></span><span class="op">(</span><span class="va">moduleSplit</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>L <span class="op">=</span> <span class="va">L</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/celda/man/recursiveSplitCell.html" class="external-link">recursiveSplitCell</a></span><span class="op">(</span><span class="va">sce</span>, useAssay <span class="op">=</span> <span class="va">useAssay</span>, altExpName <span class="op">=</span> <span class="va">altExpName</span>, initialK <span class="op">=</span> <span class="fl">3</span>, maxK <span class="op">=</span> <span class="fl">25</span>, yInit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/celda/man/celdaModules.html" class="external-link">celdaModules</a></span><span class="op">(</span><span class="va">temp</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#plots to decide value of K</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/celda/man/plotGridSearchPerplexity.html" class="external-link">plotGridSearchPerplexity</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/celda/man/plotRPC.html" class="external-link">plotRPC</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p>Using the plots visualization we can then decide the value for K
i.e. the number of clusters. The point at which RPC curve tends to level
off can be taken as the best value for K. The following code selects the
final celda_CG model with L and K value chosen by the user.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/celda/man/subsetCeldaList.html" class="external-link">subsetCeldaList</a></span><span class="op">(</span><span class="va">sce</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>L <span class="op">=</span> <span class="va">L</span>, K <span class="op">=</span> <span class="va">K</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><details><summary><b>Visualization</b></summary><p>After selecting a celda model with specific values of L and K, we can
then perform additional exploratory and downstream analyses to
understand the biology of the transcriptional modules and cell
populations. We can start by generating a dimension reduction plot with
the Uniform Manifold Approximation and Projection (UMAP) method to
visualize the relationships between the cells in a 2-D embedding. This
can be done with function celdaUmap.Alternatively, a t-distributed
stochastic neighbor embedding (t-SNE) can be generated using function
celdaTsne. The UMAP and t-SNE plots generated by celdaUmap and celdaTsne
are computed based on the module probabilities (analogous to using PCs
from PCA). The calculated dimension reduction coordinates for the cells
are stored under the reducedDim slot of the altExp slot in the original
SCE object.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/celda/man/celdaUmap.html" class="external-link">celdaUmap</a></span><span class="op">(</span><span class="va">sce</span>, useAssay <span class="op">=</span> <span class="va">useAssay</span>, altExpName <span class="op">=</span> <span class="va">altExpName</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/celda/man/celdaTsne.html" class="external-link">celdaTsne</a></span><span class="op">(</span><span class="va">sce</span>, useAssay <span class="op">=</span> <span class="va">useAssay</span>, altExpName <span class="op">=</span> <span class="va">altExpName</span><span class="op">)</span></span></code></pre></div>
<p>The function plotDimReduceCluster can be used to plot the cluster
labels for cell populations identified by celda on the UMAP.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/celda/man/plotDimReduceCluster.html" class="external-link">plotDimReduceCluster</a></span><span class="op">(</span><span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"celda_UMAP"</span>, labelClusters <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/celda/man/plotDimReduceCluster.html" class="external-link">plotDimReduceCluster</a></span><span class="op">(</span><span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"celda_tSNE"</span>, labelClusters <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The function moduleHeatmap can be used to view the expression of
features across cells for a specific module. The featureModule parameter
denotes the module(s) to be displayed. Cells are ordered from those with
the lowest probability of the module on the left to the highest
probability on the right. Similarly, features are ordered from those
with the highest probability within the module on the top to the lowest
probability on the bottom.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/celda/man/moduleHeatmap.html" class="external-link">moduleHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, featureModule <span class="op">=</span> <span class="fl">27</span>, useAssay <span class="op">=</span> <span class="va">useAssay</span>, altExpName <span class="op">=</span> <span class="va">altExpName</span><span class="op">)</span></span></code></pre></div>
<p>The function plotDimReduceModule can be used visualize the
probabilities of a particular module or sets of modules on a reduced
dimensional plot such as a UMAP. This can be another quick method to see
how modules are expressed across various cells in 2-D space.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/celda/man/plotDimReduceModule.html" class="external-link">plotDimReduceModule</a></span><span class="op">(</span><span class="va">sce</span>, modules <span class="op">=</span> <span class="fl">27</span>, useAssay <span class="op">=</span> <span class="va">useAssay</span>, altExpName <span class="op">=</span> <span class="va">altExpName</span>, reducedDimName <span class="op">=</span> <span class="st">"celda_UMAP"</span><span class="op">)</span></span></code></pre></div>
<p>Celda has the ability to identify modules of co-expressed features
and quantify the probability of these modules in each cell population.
An overview of the relationships between modules and cell subpopulations
can be explored with the function celdaProbabilityMap.The “Absolute
probability” map gives insights into the absolute abundance of a module
within a given cell subpopulation. The absolute heatmap can be used to
explore which modules are higher than other modules within a cell
population. The “Relative expression” map shows the standardized
(z-scored) module probabilities across cell subpopulations. The relative
heatmap can be used to explore which modules are relatively higher than
other modules across cell populations.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/celda/man/celdaProbabilityMap.html" class="external-link">celdaProbabilityMap</a></span><span class="op">(</span><span class="va">sce</span>, useAssay <span class="op">=</span> <span class="va">useAssay</span>, altExpName <span class="op">=</span> <span class="va">altExpName</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
