<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cell Type Labeling • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Cell Type Labeling">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.12.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Cell Type Labeling</h1>
                        <h4 data-toc-skip class="author">Yichen
Wang</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/cell_type_labeling.Rmd" class="external-link"><code>vignettes/articles/cell_type_labeling.Rmd</code></a></small>
      <div class="hidden name"><code>cell_type_labeling.Rmd</code></div>

    </div>

    
    

<meta name="viewport" content="width=device-width, initial-scale=1">
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The methods for labeling the cell type for single-cell RNA-sequencing
(scRNAseq) data varies a lot. Including using a labeled reference
dataset as the training set and classifying the test data by cluster or
individual cells, or using a prior knowledge of the projection between
marker gene sets and specific cell types and labeling the cells basing
on detected markers. Currently, SCTK only supports using
<code>SingleR</code><span class="citation">[1]</span> for cell type
labeling, which works with the first approach as previously mentioned.
We are planning to add more methods that annotate cells in various
ways.</p>
<p><code>SingleR</code> is an automatic annotation method for scRNAseq
data. Given a reference dataset of samples (single-cell or bulk) with
known labels, it labels new cells from a test dataset based on
similarity to the reference. The only prerequisite for using this method
is a labeled reference dataset, and we have already enabled automated
caching of online datasets, while users can always use their own data in
the console analysis.</p>
<p>To view detailed instructions on how to use this method, please
select ‘Interactive Analysis’ for using <code>SingleR</code> in shiny
application or ‘Console Analysis’ for using this method on R console
from the tabs below: <br></p>
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p><strong>Entry</strong></p>
<p><img src="ui_screenshots/cell_type_labeling/entry.jpg" alt="entry"><br></p>
<p>Users should enter the page as instructed in the screenshot above, by
clicking on “Differential Expression &amp; Cell Type Labeling” first and
choosing “Cell Type Labeling” sub-option next.</p>
<p><strong>Usage</strong></p>
<p><img src="ui_screenshots/cell_type_labeling/panel.jpg" alt="panel"><br><img src="ui_screenshots/cell_type_labeling/panel2.jpg" alt="panel2"><br>
There are 6 basic widgets that users need to work with before reaching
to the result:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Select assay</strong> - Select the feature expression matrix
for the comparison against the reference dataset. This matrix has to be
a log-normalized assay. <a href="normalization.html">(How to normalize
an assay with SCTK)</a>
</li>
<li>
<strong>Choose a reference</strong> - Select the curated online
reference dataset. There are multiple options that cover the need of
analyzing data from human and mouse and from different tissues. The link
at the bottom will pop users to the detail of the reference, which is
currently chosen.</li>
<li>
<strong>Labeling level</strong> - For most of the provided
references, there will be 3 levels of annotations -
<strong>“main”</strong> for main cell types, <strong>“fine”</strong> for
more subtypes, and with <strong>“ont”</strong>, subtypes are mapped to
Cell Ontology.</li>
<li>
<strong>Feature type</strong> - Choose the identifier type that
matches with users’ dataset. Unfortunately, the references are only
annotated with symbols and ensembl IDs. Therefore, users working with
other types of IDs might need to do the conversion prior to using this
functionality. Automated conversion maybe added in future versions of
SCTK.</li>
<li>
<strong>Label by</strong> - Whether to label each cell individually
or label each pre-identified cluster. <a href="clustering.html">(How to
cluster cells)</a>
</li>
</ol>
<ul>
<li>5.1. <strong>Use cluster label</strong> - Choose the cell level
annotation that represent the clustering result. This option will only
pop out when users choose “Clusters” in “Label by”.</li>
</ul>
<ol start="6" style="list-style-type: decimal">
<li>
<strong>Label</strong> - The button to trigger the algorithm. Click
on this to run the cell type labeling process.</li>
</ol>
<p>The result will not be pop out directly, but three levels of cell
annotation will be inserted to the background object. They will all have
the same prefix as
<code>"SingleR_{reference abbr}_{annotation level}_"</code>, then
<code>"score"</code>, <code>"labels"</code>, <code>"delta.next"</code>
and <code>"pruned.labels"</code>, respectively. <code>"labels"</code>
contains the predicted label, basing only on the maximum entry in
scores. <code>"delta.next"</code> contains difference between the best
and next-best score. <code>"pruned.labels"</code> contains predictions
where “low-quality” labels are replaced with <code>NA</code>s.</p>
<blockquote>
<p>With older versions of SingleR, the output variables ends with
<code>"score"</code>, <code>"first.labels"</code>, <code>"labels"</code>
and <code>"pruned.labels"</code>, respectively.
<code>"first.labels"</code> refers to the labeling initially indicated
by the scores, <code>"labels"</code> is fine-tuned, and
<code>"pruned.labels"</code> is the pruned result.</p>
</blockquote>
<p><strong>Visualization</strong></p>
<p>SCTK does not have a quick visualization at this point. But users can
go to the <a href="visualization.html">CellViewer</a> tab to create a
scatter plot using the pre-generated embedding and color it with the
cell type labeling result.</p>
</div>

<div id="console" class="tabcontent">
<p><strong>Preprocessing</strong></p>
<p>As mentioned previously in this article, performing cell type
labeling with SingleR does not require sophisticated preprocessing, but
only a reference dataset that covers the expected cell types. Now that
we have already incorporated many curated online dataset that SingleR
recommends, users only need to log-normalize the feature expression
matrix to match what is in the reference.</p>
<p>Here, we use “PBMC3k” dataset provided by CellRanger as an
example.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runNormalization.html">runNormalization</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, normalizationMethod <span class="op">=</span> <span class="st">"logNormCounts"</span>, outAssayName <span class="op">=</span> <span class="st">"logcounts"</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Running SingleR</strong></p>
<p>Then users can use <code><a href="../reference/runSingleR.html">runSingleR()</a></code> to invoke the algorithm.
Here we choose to use <code>useBltinRef = "hpca"</code> to set the
reference, which refers to Human Primary Cell Atlas Data <span class="citation">[2]</span> and this will be automatically cached to
users’ local. We set <code>level = "main"</code> to label the PBMC3k
dataset with general cell types. Users can also set it to
<code>"fine"</code> for more subtypes, and <code>"ont"</code> for Gene
Ontology information. For more information, please click on the function
name and see the reference page of this function.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSingleR.html">runSingleR</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>, useBltinRef <span class="op">=</span> <span class="st">"hpca"</span>, level <span class="op">=</span> <span class="st">"main"</span>, featureType <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p>Four results from SingleR will be stored in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>
slot with the same prefix as
<code>"SingleR_{reference abbr}_{annotation level}_"</code>, then
<code>"score"</code>, <code>"labels"</code>, <code>"delta.next"</code>
and <code>"pruned.labels"</code>, respectively. <code>"labels"</code>
contains the predicted label, basing only on the maximum entry in
scores. <code>"delta.next"</code> contains difference between the best
and next-best score. <code>"pruned.labels"</code> contains predictions
where “low-quality” labels are replaced with <code>NA</code>s.</p>
<blockquote>
<p>With older versions of SingleR, the output variables ends with
<code>"score"</code>, <code>"first.labels"</code>, <code>"labels"</code>
and <code>"pruned.labels"</code>, respectively.
<code>"first.labels"</code> refers to the labeling initially indicated
by the scores, <code>"labels"</code> is fine-tuned, and
<code>"pruned.labels"</code> is the pruned result.</p>
</blockquote>
<p>Besides, users can also use their own labeled reference datasets
wrapped in a <a href="https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>
object. Refer to argument <code>useSCERef</code> and
<code>labelColName</code>. Additionally, the labeling can also be done
on cluster label if users have already performed clustering on their
dataset and have the result stored in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> slot. Refer
to argument <code>labelByCluster</code>.</p>
<p><strong>Visualization</strong></p>
<p>Users can choose to visualize the labeling result on a scatter plot.
Here, we provide an example to create a UMAP embedding that can clearly
reflect the relationship between cells in 2D, and then color the cells
by the labels.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratFindHVG.html">runSeuratFindHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getTopHVG.html">setTopHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, method <span class="op">=</span> <span class="st">"vst"</span>, hvgNumber <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDimReduce.html">runDimReduce</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, method <span class="op">=</span> <span class="st">"scaterPCA"</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>, useFeatureSubset <span class="op">=</span> <span class="st">"HVG_vst2000"</span>, reducedDimName <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDimReduce.html">runDimReduce</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, method <span class="op">=</span> <span class="st">"scaterUMAP"</span>, useReducedDim <span class="op">=</span> <span class="st">"PCA"</span>, reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSCEDimReduceColData.html">plotSCEDimReduceColData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, colorBy <span class="op">=</span> <span class="st">"SingleR_hpca_main_pruned.labels"</span>, reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cell_type_labeling_files/figure-html/visualization-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body" entry-spacing="0">
<div id="ref-Aran2019" class="csl-entry">
<div class="csl-left-margin">[1] </div>
<div class="csl-right-inline">D.
Aran <em>et al.</em>, <span>“Reference-based analysis of lung
single-cell sequencing reveals a transitional profibrotic
macrophage,”</span> <em>Nature Immunology</em>, vol. 20, pp. 163–172,
2019.</div>
</div>
<div id="ref-Mabbott2013" class="csl-entry">
<div class="csl-left-margin">[2] </div>
<div class="csl-right-inline">N.
A. Mabbott, J. K. Baillie, H. Brown, T. C. Freeman, and D. A. Hume,
<span>“An expression atlas of human primary cells: Inference of gene
function from coexpression networks,”</span> <em>BMC Genomics</em>, vol.
14, no. 632, 2013.</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
