<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Clustering • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Clustering">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.12.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Clustering</h1>
                        <h4 data-toc-skip class="author">Yichen
Wang</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/clustering.Rmd" class="external-link"><code>vignettes/articles/clustering.Rmd</code></a></small>
      <div class="hidden name"><code>clustering.Rmd</code></div>

    </div>

    
    

<meta name="viewport" content="width=device-width, initial-scale=1">
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>SCTK allows users to do clustering on their data with various
methods. The clustering methods that SCTK adopts are mainly graph based
algorithms, such as Louvain and Leiden, which partitions a Shared
Nearest-Neighbor (SNN) graph. The graph can be constructed by Scran
<span class="citation">[1]</span> or Seurat <span class="citation">[2]</span>, and the graph clustering is done by <a href="https://igraph.org/r/" class="external-link">igraph</a> or Seurat, respectively.
Meanwhile, we also support traditional method such as K-Means <span class="citation">[3]</span>.</p>
<p>SCTK allows different types data matrix as input. When using Scran’s
SNN graph based clustering algorithms, users can specify either an
expression matrix or a low-dimension representation as the input, while
for Seurat’s methods and K-Means, users can only use the reduced
dimensions. For any of the options, we recommend using a low-dimension
representation as the input.</p>
<p>To view detailed instructions on how to use these methods, please
select ‘Interactive Analysis’ for using clustering methods in Shiny
application or ‘Console Analysis’ for using these methods in R console
from the tabs below:</p>
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p><strong>Entry of The Panel</strong></p>
<p>From anywhere of the UI, the panel for clustering can be accessed
from the top navigation panel at the circled tab shown below.</p>
<p><img src="ui_screenshots/clustering/clst_ui_entry.png" alt="Entry"><br></p>
<p>The UI consists of the parameter setting panel on the left and the
visualization panel on the right.</p>
<p><img src="ui_screenshots/clustering/clst_ui.png" alt="UI"><br></p>
<p><strong>Run Clustering</strong></p>
<p><img src="ui_screenshots/clustering/clst_ui_algo.png" alt="Algo"><br></p>
<p>User will choose an algorithm to run the clustering at the very first
step. The slide-down option list is constructed in a grouped style. Each
group lists all the algorithm that a dependency (shown with grey text)
supports. By selecting an algorithm that belongs to different groups,
the parameter settings will change.</p>
<p>Method specific parameter settings are shown and explained below:</p>
<p><img src="ui_screenshots/clustering/clst_ui_param.png" alt="scran"><br></p>
 <style>
div.offset { background-color:inherit ; border-radius: 5px; margin-left: 15px;}
</style>
<div class="offset">
<details><summary><b>Scran SNNs</b></summary><p>For the choices in the algorithm list, please refer to
<code>cluster_louvain()</code>, <code>cluster_leiden()</code>,
<code>cluster_walktrap()</code>, <code>cluster_infomap()</code>,
<code>cluster_fast_greedy()</code>, <code>cluster_label_prop()</code>,
and <code>cluster_leading_eigen()</code> in the <code>igraph</code>
package documentation.</p>
<ul>
<li>Data matrix selection - selection input <strong>“Select Input
Matrix”</strong>. Scran SNN method allows various types of data
matrices. Either a full-sized or a subsetted expression matrix,
technically called <code>assay</code> or <code>altExp</code>,
respectively, or a reduced dimension, technically
<code>reducedDim</code>, is allowed.</li>
<li>Resolution related parameters:
<ul>
<li>Numeric input <strong>“K value”</strong>. <code>K</code> is an
integer scalar specifying the number of nearest neighbors to consider
during graph construction. Considering more neighbors results in larger
groups. (See parameter <code>k</code> of
<code><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">scran::buildSNNGraph()</a></code>)</li>
<li>Numeric input <strong>“Resolution”</strong>, only seen when choosing
“leiden” algorithm. Higher resolutions lead to more smaller communities,
while lower resolutions lead to fewer larger communities. (See parameter
<code>resolution_parameter</code> of
<code><a href="https://r.igraph.org/reference/cluster_leiden.html" class="external-link">igraph::cluster_leiden()</a></code>)</li>
<li>Numeric input <strong>“Steps”</strong>, only seen when choosing
“walktrap” algorithm. The length of the random walks to perform. Less
steps lead to fewer larger communities, and take more time for
computation. (See parameter <code>steps</code> of
<code><a href="https://r.igraph.org/reference/cluster_walktrap.html" class="external-link">igraph::cluster_walktrap()</a></code>)</li>
</ul>
</li>
<li>Algorithm detail setting
<ul>
<li>selection input <strong>“Edge Weight Type”</strong>. Users can
specify the type of weighting scheme to use for shared neighbors. (See
parameter <code>type</code> of <code><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">scran::buildSNNGraph()</a></code>)</li>
<li>Selection input <strong>“Objective Function”</strong>, only seen
when choosing “leiden” algorithm. The objective function, “CPM” <span class="citation">[4]</span> or “modularity” <span class="citation">[5]</span>, evaluates the partitioning of the graph,
and the algorithm attempts to maximize the score.</li>
</ul>
</li>
<li>Component number setting - numeric input <strong>“Number of
Components”</strong>. If users specify a low-dimensional matrix as
input, the method will be applied on the top components. When an
<code>assay</code> or <code>altExp</code> is chosen, the algorithm will
perform a PCA on the matrix and obtain the top PCs for downstream use.
This number could be determined with the help of an Elbow plot in the <a href="dimensionality_reduction.html">“Dimension Reduction”</a> section.
For better visualization, we suggest generating <a href="2d_embedding.html">2D embedding</a> with the same number of
components from the selected reduced dimensions.</li>
</ul></details><details><summary><b>K-Means</b></summary><p>When the selected algorithm belongs to “K-Means” group, the parameter
settings will look like the figure above.</p>
<ul>
<li>Data matrix selection - selection input <strong>“Select a
ReducedDim”</strong>. Here the data matrix allowed can only be a reduced
dimension, technically called <code>reducedDim</code> and should be
obtained in advance from <a href="dimensionality_reduction.html">Dimensionality Reduction</a>
tab.</li>
<li>Cluster number determining - numeric input <strong>“Number of
Centers (Clusters)”</strong>. User will decide the exact number of
clusters here. In term of K-means algorithm, the number of cluster
centroids.</li>
<li>Iteration limit setting - numeric input <strong>“Max Number of
Iterations”</strong>. User will here decide the maximum number of
iterations to run.</li>
<li>Algorithm detail setting - numeric input <strong>“Number of Random
Sets”</strong>. Kmeans attempts multiple initial configurations and
reports on the best one. Here users will set the number of initial
configurations to use.</li>
</ul></details><details><summary><b>Seurat</b></summary><p>When the selected algorithm belongs to “Seurat” group, the parameter
settings will look like the figure above.</p>
<ul>
<li>Data matrix selection - selection input <strong>“Select a
ReducedDim”</strong>. Here the data matrix allowed can only be a reduced
dimension, technically called <code>reducedDim</code> and should be
obtained in advance from <a href="dimensionality_reduction.html">“Feature Selection &amp; Dimension
Reduction” tab</a>.</li>
<li>Component number setting - numeric input <strong>“How Many
Dimensions to Use”</strong>.</li>
<li>Algorithm detail setting - checkbox input <strong>“Group
Singletons”</strong>.</li>
<li>Resolution setting - numeric input
<strong>“Resolution”</strong>.</li>
</ul></details>
</div>
<br><p><strong>Visualization</strong></p>
<p><img src="ui_screenshots/clustering/clst_ui_vis.png"><br></p>
<p>The visualization is implemented with a scatter plot of a chosen
low-dimension embedding, colored with a chosen cluster assignment (the
newly generated result by default). SCTK by default uses the matrix used
for clustering calculation for the visualization. However, usually,
users use a PCA for clustering while needing a UMAP/t-SNE embedding for
visualization. To change the embedding used, please click on the blue
settings (cog) button on the left-top corner.</p>
<div class="offset">
<details><summary><b>Settings</b></summary><ul>
<li>Cluster annotation selection - radio button selection
<strong>“Select from Current Results”</strong> and <strong>“Select from
All Present Annotation”</strong>. This selection updates the dropdown
menu with the available annotations. Only a successful run of clustering
in the current SCTK session adds an option for the new result to “Select
from Current Results”, while all the cell annotation stored in
background (i.e. <code>colData(sce)</code>) are accessible if “Select
from All Present Annotation” is chosen.</li>
<li>Dimension Reduction selection - selection input <strong>“Use
Reduction”</strong>. User have to choose a reduction here for plotting.
If there is not yet any option, user can obtain one from <a href="dimensionality_reduction.html">Dimension Reduction</a> tab.</li>
</ul></details>
</div>
</div>

<div id="console" class="tabcontent">
<p>The clustering methods mentioned in the Introduction can be easily
applied to any preprocessed SCE object. Meanwhile, there are other types
of clustering methods supported by SCTK, but usually dependent to a
curated workflow, such as the <a href="celda_curated_workflow.html">Celda Curated Workflow</a>. Here we
present the usage of the three functions only.</p>
<p><strong>Basic Parameters</strong></p>
<p>For each of the three functions, the common things are:</p>
<ul>
<li>The input SCE object - <code>inSCE</code>.</li>
<li>The <code>colData</code> column name to save the cluster labels -
<code>clusterName</code>
</li>
<li>The specific algorithm type supported by the dependency -
<code>algorithm</code>
</li>
</ul>
<p>As for another essential parameter, the data matrix used for running
the algorithm is a must. Here, users should notice that:</p>
<ul>
<li>For <code><a href="../reference/runScranSNN.html">runScranSNN()</a></code>, either a feature expression matrix
or a dimension reduction is acceptable. Users need to use:
<ul>
<li>
<code>useAssay</code> for a full-sized expression data
(<code>assay</code>)</li>
<li>
<code>useAltExp</code> and <code>altExpAssay</code> for a subsetted
expression data (<code>altExp</code>)</li>
<li>
<code>useReducedDim</code> for a dimension reduction.
(<code>reducedDim</code>)</li>
<li>
<code>useAltExp</code> and <code>altExpRedDim</code> for a reduced
dimension stored in the <code>altExp</code> object.</li>
</ul>
</li>
<li>For <code><a href="../reference/runKMeans.html">runKMeans()</a></code>, only a dimension reduction is
acceptable. Users need to use <code>useReducedDim</code> to pass the
argument.</li>
<li>For <code><a href="../reference/runSeuratFindClusters.html">runSeuratFindClusters()</a></code>, it is included within the
Seurat Curated Workflow, yet usable as an independent function. However,
it will be complicated to use this way. It is recommended to invoke the
Seurat clustering functionality in the UI since it is automated there.
We will still present the complicated workflow later.</li>
</ul>
<p>When using <code>useReducedDim</code> as data input, users should be
aware of the number of top components being passed to the underlying
method. This is usually controlled by argument <code>nComp</code>. To
determine this parameter, users can refer to the explanation in <a href="dimensionality_reduction.html">Dimensionality Reduction
Documentation</a>. Meanwhile, for better visualization, we also
recommend users to generate <a href="2d_embedding.html">2D Embedding</a>
using the same number of top components from the same dimension
reduction result.</p>
<p>Other parameters are method specific, please refer to the function
manual pages for the detail. The typical command call for for each
method is shown below:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Scran method</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScranSNN.html">runScranSNN</a></span><span class="op">(</span><span class="va">sce</span>, useReducedDim <span class="op">=</span> <span class="st">"PCA"</span>, clusterName <span class="op">=</span> <span class="st">"scranSNN"</span><span class="op">)</span></span>
<span><span class="co"># K-means method</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runKMeans.html">runKMeans</a></span><span class="op">(</span><span class="va">sce</span>, nCenters <span class="op">=</span> <span class="fl">9</span>, useReducedDim <span class="op">=</span> <span class="st">"PCA"</span>, clusterName <span class="op">=</span> <span class="st">"KMeans"</span><span class="op">)</span></span>
<span><span class="co"># Seurat method</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratFindClusters.html">runSeuratFindClusters</a></span><span class="op">(</span><span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"seuratScaledData"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Example</strong></p>
<p>To demonstrate simple and clear examples, here we use the “PBMC-3k”
dataset from “10X” which can be easily imported with SCTK functions. The
preprocessing only includes necessary steps before getting cluster
labels (i.e. QC and filtering are excluded).</p>
<div class="offset">
<details><summary><b>Preprocessing</b></summary><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runNormalization.html">runNormalization</a></span><span class="op">(</span><span class="va">sce</span>, outAssayName <span class="op">=</span> <span class="st">"logcounts"</span>, normalizationMethod <span class="op">=</span> <span class="st">"logNormCounts"</span><span class="op">)</span></span>
<span><span class="co"># Default HVG method is "vst" from Seurat</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runFeatureSelection.html">runFeatureSelection</a></span><span class="op">(</span><span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getTopHVG.html">setTopHVG</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDimReduce.html">runDimReduce</a></span><span class="op">(</span><span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>, useFeatureSubset <span class="op">=</span> <span class="st">"HVG_vst2000"</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, reducedDimName <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="co"># Optional visualization</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDimReduce.html">runDimReduce</a></span><span class="op">(</span><span class="va">sce</span>, method <span class="op">=</span> <span class="st">"scaterUMAP"</span>, useReducedDim <span class="op">=</span> <span class="st">"PCA"</span>, reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, nComponents <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotDimRed.html">plotDimRed</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"UMAP"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustering_files/figure-html/clst_prep-1.png" width="700" style="display: block; margin: auto;"></p>
</details><details><summary><b>Example with Scran SNN</b></summary><details><summary><b>Method specific parameters</b></summary><ul>
<li>
<code>k</code>, the number of nearest neighbors used to construct
the graph. Smaller value indicates higher resolution and larger number
of clusters.</li>
<li>
<code>nComp</code>, the number of components to use when
<code>useAssay</code> or <code>useAltExp</code> is specified. WON’T work
with <code>useReducedDim</code>.</li>
<li>For <code>weightType</code> and <code>algorithm</code>, users should
choose from a given list of options, which can be found with
<code><a href="../reference/runScranSNN.html">?runScranSNN</a></code>. For the introduction of those options, please
refer to <code><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">scran::buildSNNGraph()</a></code> and <a href="https://rdrr.io/cran/igraph/" class="external-link">igraph</a>.</li>
<li>Different algorithms, supported by <a href="https://rdrr.io/cran/igraph/" class="external-link">igraph</a>, might allows different
parameters, such as <code>objective_function</code> and
<code>resolution_parameter</code> for Leiden <span class="citation">[6]</span>, <code>steps</code> for Walktrap <span class="citation">[7]</span>. Please also refer to function reference for
detail.</li>
</ul></details><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directly use an assay</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScranSNN.html">runScranSNN</a></span><span class="op">(</span><span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>, clusterName <span class="op">=</span> <span class="st">"scranSNN_logcounts"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSCEDimReduceColData.html">plotSCEDimReduceColData</a></span><span class="op">(</span><span class="va">sce</span>, colorBy <span class="op">=</span> <span class="st">"scranSNN_logcounts"</span>, reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustering_files/figure-html/clst_ScranSNN-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directly use a dimension reduction</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScranSNN.html">runScranSNN</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useReducedDim <span class="op">=</span> <span class="st">"PCA"</span>, nComp <span class="op">=</span> <span class="fl">10</span>, clusterName <span class="op">=</span> <span class="st">"scranSNN_PCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSCEDimReduceColData.html">plotSCEDimReduceColData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, colorBy <span class="op">=</span> <span class="st">"scranSNN_PCA"</span>, reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustering_files/figure-html/clst_ScranSNN-2.png" width="700" style="display: block; margin: auto;"></p>
</details><details><summary><b>Example with K-Means</b></summary><details><summary><b>Method specific parameters</b></summary><ul>
<li>
<code>nCenters</code>, the number of final clusters. This is
required.</li>
<li>
<code>nIter</code>, the maximum number of iterations allowed.</li>
<li>
<code>nStart</code>, the number of random sets to choose. Since
K-Means is an algorithm with reasonable randomness, the function allows
attempting multiple initial configurations and reports on the best
one.</li>
<li>
<code>seed</code>, The seed for the random number generator.</li>
<li>For <code>algorithm</code>, users should choose from a given list of
options, which can be found with <code><a href="../reference/runKMeans.html">?runKMeans</a></code>. For the
introduction of those options, please refer to
<code><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">?stats::kmeans</a></code>.</li>
</ul></details><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runKMeans.html">runKMeans</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useReducedDim <span class="op">=</span> <span class="st">"PCA"</span>, nComp <span class="op">=</span> <span class="fl">10</span>, nCenters <span class="op">=</span> <span class="fl">10</span>, clusterName <span class="op">=</span> <span class="st">"kmeans"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSCEDimReduceColData.html">plotSCEDimReduceColData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, colorBy <span class="op">=</span> <span class="st">"kmeans"</span>, reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustering_files/figure-html/clst_kmeans-1.png" width="700" style="display: block; margin: auto;"></p>
</details><details><summary><b>Example with Seurat</b></summary><p>Seurat clustering method is recommended to be used in the <a href="seurat_curated_workflow.html">Seurat Curated Workflow</a> in
console analysis. When users preprocessed the dataset with other
approaches, the clustering with Seurat function can be complicated,
since some sorts of Seurat specific metadata is always needed.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Prepare what seurat function needs</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pca</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="st">"-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pca</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">stdev</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">pca</span>, <span class="st">"percentVar"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">new_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateDimReducObject.html" class="external-link">CreateDimReducObject</a></span><span class="op">(</span>embeddings <span class="op">=</span> <span class="va">pca</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, stdev <span class="op">=</span> <span class="va">stdev</span>, key <span class="op">=</span> <span class="st">"PC_"</span><span class="op">)</span></span>
<span><span class="co"># Then we can take the Seurat Curated Workflow independent dimension </span></span>
<span><span class="co"># reduction as an input. </span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratFindClusters.html">runSeuratFindClusters</a></span><span class="op">(</span><span class="va">sce</span>, externalReduction <span class="op">=</span> <span class="va">new_pca</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSCEDimReduceColData.html">plotSCEDimReduceColData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, colorBy <span class="op">=</span> <span class="st">"Seurat_louvain_Resolution0.8"</span>, reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustering_files/figure-html/clst_seurat2-1.png" width="700" style="display: block; margin: auto;"></p>
</details>
</div>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body" entry-spacing="0">
<div id="ref-Aaron2016" class="csl-entry">
<div class="csl-left-margin">[1] </div>
<div class="csl-right-inline">A.
T. L. Lun, D. J. McCarthy, and J. C. Marioni, <span>“A step-by-step
workflow for low-level analysis of single-cell RNA-seq data with
bioconductor,”</span> <em>F1000Research</em>, vol. 5, p. 2122, 2016,
doi: <a href="https://doi.org/10.12688/f1000research.9501.2" class="external-link">10.12688/f1000research.9501.2</a>.</div>
</div>
<div id="ref-Stuart2019" class="csl-entry">
<div class="csl-left-margin">[2] </div>
<div class="csl-right-inline">T.
Stuart <em>et al.</em>, <span>“<span class="nocase">Comprehensive
Integration of Single-Cell Data</span>,”</span> <em>Cell</em>, vol. 177,
no. 7, pp. 1888–1902.e21, Jun. 2019, doi: <a href="https://doi.org/10.1016/j.cell.2019.05.031" class="external-link">10.1016/j.cell.2019.05.031</a>.</div>
</div>
<div id="ref-Forgy65" class="csl-entry">
<div class="csl-left-margin">[3] </div>
<div class="csl-right-inline">E.
Forgy, <span>“<span class="nocase">Cluster analysis of multivariate
data: efficiency versus interpretability of
classifications</span>,”</span> <em>Biometrics</em>, vol. 21, pp.
768–780, 1965.</div>
</div>
<div id="ref-Ronhovde2010" class="csl-entry">
<div class="csl-left-margin">[4] </div>
<div class="csl-right-inline">P.
Ronhovde and Z. Nussinov, <span>“Local resolution-limit-free potts model
for community detection,”</span> <em>Phys. Rev. E</em>, vol. 81, no.
46114, 4, 2010, doi: <a href="https://doi.org/10.1103/PhysRevE.81.046114" class="external-link">10.1103/PhysRevE.81.046114</a>.</div>
</div>
<div id="ref-Newman2004" class="csl-entry">
<div class="csl-left-margin">[5] </div>
<div class="csl-right-inline">M.
E. J. Newman and M. Girvan, <span>“Finding and evaluating community
structure in networks,”</span> <em>Phys. Rev. E</em>, vol. 69, no.
26113, 2, 2004, doi: <a href="https://doi.org/10.1103/PhysRevE.69.026113" class="external-link">10.1103/PhysRevE.69.026113</a>.</div>
</div>
<div id="ref-Traag2019" class="csl-entry">
<div class="csl-left-margin">[6] </div>
<div class="csl-right-inline">V.
A. Traag, L. Waltman, and N. J. van Eck, <span>“From louvain to leiden:
Guaranteeing well-connected communities,”</span> <em>Scientific
Reports</em>, vol. 9, no. 5233, 2019, doi: <a href="https://doi.org/10.1038/s41598-019-41695-z" class="external-link">10.1038/s41598-019-41695-z</a>.</div>
</div>
<div id="ref-Pons2005" class="csl-entry">
<div class="csl-left-margin">[7] </div>
<div class="csl-right-inline">P.
Pons and M. Latapy, <span>“Computing communities in large networks using
random walks,”</span> in <em>Computer and information sciences - ISCIS
2005</em>, 2005, pp. 284–293. doi: <a href="https://doi.org/10.1007/11569596_31" class="external-link">10.1007/11569596_31</a>.</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
