<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Generation of comprehensive quality control metrics with SCTK • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Generation of comprehensive quality control metrics with SCTK">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.12.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Generation of comprehensive quality control
metrics with SCTK</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/cmd_qc.Rmd" class="external-link"><code>vignettes/articles/cmd_qc.Rmd</code></a></small>
      <div class="hidden name"><code>cmd_qc.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This pipeline will import data from single-cell preprocessing
algorithms (e.g. <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">CellRanger</a>,
<a href="https://data.humancellatlas.org/pipelines/optimus-workflow" class="external-link">HCA
Optimus</a>, <a href="https://salmon.readthedocs.io/en/latest/alevin.html" class="external-link">Alevin</a>),
generate various quality control metrics (e.g. general metrics, doublet
scores, contamination estimates) using multiple tools, and output
results in standard data containers (e.g. <a href="https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>,
<a href="https://satijalab.org/seurat/index.html" class="external-link">Seurat object</a>, <a href="https://github.com/theislab/anndata" class="external-link">AnnData</a>).</p>
<p>For data generated with microfluidic devices, the first major step
after UMI counting is to detect cell barcodes that represent droplets
containing a true cell and exclude empty droplets that only contain
ambient RNA.</p>
<ul>
<li>We use the terms <strong>“Droplet” matrix</strong> to denote a count
matrix that still contains empty droplets;</li>
<li>
<strong>“Cell” matrix</strong> to denote a count matrix of cells
where empty droplets have been excluded but no other filtering has been
performed;</li>
<li>And <strong>“FilteredCell” matrix</strong> to indicate a count
matrix where poor quality cells have also been excluded.</li>
</ul>
<p>The Droplet and Cell matrices have also been called “raw” and
“filtered” matrices, respectively, by tools such as CellRanger. However,
using the term “filtered” can be ambiguous as other forms of cell
filtering can be applied beyond empty droplets (e.g. excluding
poor-quality cells based on low number of UMIs). Both the original
droplet matrix and the filtered cell matrix can be QC’ed in this
pipeline. However, QC of the droplet matrix is specific for single cell
data generated from microfluidic devices (e.g. 10X).</p>
<p>To run the pipeline, users can <a href="installation.html">install
the singleCellTK (SCTK) package</a> along with Python dependencies that
may be potentially used (mainly <a href="https://github.com/swolock/scrublet" class="external-link">Scrublet</a> and <a href="https://github.com/theislab/anndata" class="external-link">AnnData</a>). Alternatively,
users can run the Docker version of the pipeline which is described in
detail further down in this page.</p>
</div>
<div class="section level2">
<h2 id="running-the-pipeline">Running the pipeline<a class="anchor" aria-label="anchor" href="#running-the-pipeline"></a>
</h2>
<div class="section level3">
<h3 id="running-sctk-qc-with-sctk-local-installation">Running SCTK-QC with SCTK local installation<a class="anchor" aria-label="anchor" href="#running-sctk-qc-with-sctk-local-installation"></a>
</h3>
<p>To run the pipeline script, users will need to download the
<code>SCTK_runQC.R</code> script file <a href="https://github.com/compbiomed/singleCellTK/blob/master/exec/SCTK_runQC.R" class="external-link">here</a>.
Also, the tutorial of installing singleCellTK package can found in <a href="installation.html">Installation</a> page.</p>
<p>A simple example to run this pipeline on the ‘Cell’ matrix generated
by Cellranger V3 is shown below:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">Rscript</span> SCTK_runQC.R <span class="dt">\</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>-b /base/path <span class="dt">\</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>-P CellRangerV3 <span class="dt">\</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>-s SampleName <span class="dt">\</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>-o Output_Directory <span class="dt">\</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>-S TRUE <span class="dt">\</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>-F SCE,AnnData,FlatFile,Seurat <span class="dt">\</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>-g /Path_to_gmt/name_of_gmt_file.gmt <span class="dt">\</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>-d Cell <span class="dt">\</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>-n 2 <span class="dt">\</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>-T MulticoreParam </span></code></pre></div>
<p>This pipeline enables different ways to import
CellrangerV2/CellrangerV3 data for flexibility. Also, the pipeline is
compatible with datasets generated by other algorithms. Please refer to
the section <a href="#importing-data-from-different-preprocessing-tools"><em>Importing
data from different preprocessing tools</em></a> for more details.</p>
<p>User can quantify expression of mitochondrial genes by passing a GMT
files containing mitochondrial genes (with <code>-g/--gmt</code>
argument). User can also quantify expression of mitochondrial genes for
human or mouse dataset by setting <code>-M/--detectMitoLevel</code>
argument. Please refer to the section <a href="#gene-sets"><em>Gene
sets</em></a> for more details.</p>
<p>Besides, the pipeline contains various parameters to control the
process of quality control. Please refer to the section <a href="#parameters"><em>Parameters</em></a> for more details.</p>
</div>
<div class="section level3">
<h3 id="running-sctk-qc-with-docker">Running SCTK-QC with Docker<a class="anchor" aria-label="anchor" href="#running-sctk-qc-with-docker"></a>
</h3>
<div class="section level4">
<h4 id="installing-docker">Installing docker<a class="anchor" aria-label="anchor" href="#installing-docker"></a>
</h4>
<p>If you have not used docker before, you can follow the instruction to
install and set up docker in <a href="https://docs.docker.com/desktop/windows/install/" class="external-link">Windows</a>, <a href="https://docs.docker.com/desktop/mac/install/" class="external-link">Mac</a> or <a href="https://runnable.com/docker/install-docker-on-linux" class="external-link">Linux</a>.</p>
</div>
<div class="section level4">
<h4 id="running-sctk-qc-pipeline-using-docker-image">Running SCTK-QC pipeline using docker image<a class="anchor" aria-label="anchor" href="#running-sctk-qc-pipeline-using-docker-image"></a>
</h4>
<p>The Docker image can be obtained by running:</p>
<pre><code>docker pull campbio/sctk_qc:2.12.2</code></pre>
<p>The usage of each argument is the same as running command line
analysis. Here is an example code to perform QC on CellRangerV3 data
with SCTK docker:</p>
<pre><code>docker run --rm -v /path/to/data:/SCTK_docker \
  -it campbio/sctk_qc:2.12.2 \
  -b /SCTK_docker/cellranger_folder \
  -P CellRangerV3 \
  -s SampleName \
  -o /SCTK_docker/Output_Directory \
  -g /SCTK_docker/name_of_gmt_file.gmt \
  -S TRUE \
  -F SCE,AnnData,FlatFile,Seurat</code></pre>
<p>The docker image will not access files in your host file system by
default. To get access to the files on your machine, you can properly
set up a mount volume. Noted that the transcriptome data and GMT file
needed to be accessible to the container through mounted volume. In the
above example, mount volume is enabled for accessing input and output
directory using argument <code>-v</code>. The transcriptome and GMT
files stored in the path <code>/path/to/data</code> of your machine file
system is now available in <code>/SCTK_docker</code> folder inside the
docker. To learn more about mounted volumes, please check out <a href="https://docs.docker.com/storage/volumes/" class="external-link">this post</a>.</p>
<p>Please refer to the section <a href="#parameters"><em>Parameters</em></a> for more details about
parameters.</p>
</div>
<div class="section level4">
<h4 id="running-sctk-qc-pipeline-docker-image-with-singularity">Running SCTK-QC pipeline docker image with singularity<a class="anchor" aria-label="anchor" href="#running-sctk-qc-pipeline-docker-image-with-singularity"></a>
</h4>
<p>Users who have not used <a href="https://sylabs.io/guides/2.6/user-guide/introduction.html" class="external-link">Singularity</a>
before can install it following the instruction <a href="https://sylabs.io/guides/3.1/user-guide/installation.html" class="external-link">here</a>.
The Singularity image for SCTK-QC can be easily built using Docker Hub
as a source:</p>
<pre><code>singularity pull docker://campbio/sctk_qc:2.12.2</code></pre>
<p>The usage of singleCellTK Singularity image is very similar to that
of Docker. In Singularity 3.0+, the mount volume is <a href="https://sylabs.io/guides/3.1/user-guide/bind_paths_and_mounts.html" class="external-link">automatically
overlaid</a>.</p>
<p>It’s recommended to re-set the home directory when you run
singularity. Singularity will mount <code>\$HOME</code> path on your
file system by default, which might contain your personal R/Python
library folder. If we don’t re-set the home to mount, singularity will
try to use R/Python libraries which are not built within the singularity
image and cause some conflicts. You can point to some “sanitized home”,
which is different from <code>\$HOME</code> path on your machine, using
argument <code>-H</code>/<code>--home</code> <a href="https://sylabs.io/guides/3.1/user-guide/bind_paths_and_mounts.html" class="external-link">(see
more information)</a>. Or you can disable the <code>\$HOME</code>
binding by setting the argument <code>--no-home</code>. Besides, you can
use argument <code>--bind</code>/<code>-B</code> to specify your own
mount volume, which is the path that contains the dataset and will be
used to store the output of QC pipeline. The example is shown as
below:</p>
<pre><code>singularity run --home=/PathToSanitizedHome \
  --bind /PathToData:/data sctk_qc_2.12.2.sif \
  -b /SCTK_docker/cellranger_folder \
  -P CellRangerV3 \
  -s SampleName \
  -o /SCTK_docker/Output_Directory \
  -g /SCTK_docker/name_of_gmt_file.gmt \
  -S TRUE \
  -F SCE,AnnData,FlatFile,Seurat</code></pre>
</div>
<div class="section level4">
<h4 id="important-note-about-docker-image">Important note about docker image<a class="anchor" aria-label="anchor" href="#important-note-about-docker-image"></a>
</h4>
<p>One important note about this docker image: please run the docker
image on a machine / node which has a <strong>CPU</strong> with the
following architecture: <strong>broadwell, haswell, skylake, cascadelake
or the latest architecture</strong>. This can avoid having the “illegal
operation” issue from Scrublet package, because this Python package are
compiled by SIMD instructions that are compatible with these CPU
architectures. Please specify the CPU architecture, at the script header
after <code>#$ -l cpu_arch=</code>, as one of the following:
<code>broadwell</code>, <code>haswell</code>, <code>skylake</code>,
<code>cascadelake</code> or latest architecture. One of the example is
shown below:</p>
<pre><code>#!/bin/bash
#$ -cwd
#$ -j y
#$ -pe omp 16
#$ -l cpu_arch=broadwell|haswell|skylake|cascadelake

singularity run --home=/PathToSanitizedHome  ### this also works for 'docker run'
  --bind /PathToData:/data sctk_qc_2.12.2.sif \
  -b /SCTK_docker/cellranger_folder \
  -P CellRangerV3 \
  -s SampleName \
  -o /SCTK_docker/Output_Directory \
  -g /SCTK_docker/name_of_gmt_file.gmt \
  -S TRUE \
  -F SCE,AnnData,FlatFile,Seurat</code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="parameters">Parameters<a class="anchor" aria-label="anchor" href="#parameters"></a>
</h2>
<div class="section level3">
<h3 id="table-of-parameters">Table of Parameters<a class="anchor" aria-label="anchor" href="#table-of-parameters"></a>
</h3>
<p>The pipeline contains various parameters to control the process of
quality control. The function of each parameter is shown below:</p>
<div class="section level4">
<h4 id="required-arguments">Required arguments<a class="anchor" aria-label="anchor" href="#required-arguments"></a>
</h4>
<p>The required arguments are as follows:</p>
<table class="table">
<colgroup>
<col width="10%">
<col width="89%">
</colgroup>
<thead><tr class="header">
<th>Parameter</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>-b</code>, <code>--basePath</code>
(<strong>required</strong>)</td>
<td>Base path for the output from the preprocessing algorithm.</td>
</tr>
<tr class="even">
<td>
<code>-P</code>, <code>--preproc</code>
(<strong>required</strong>)</td>
<td>Algorithm used for preprocessing. One of <code>CellRangerV2</code>,
<code>CellRangerV3</code>, <code>BUStools</code>, <code>STARSolo</code>,
<code>SEQC</code>, <code>Optimus</code>, <code>DropEst</code>,
<code>SceRDS</code>, <code>Seurat</code>, <code>CountMatrix</code> and
<code>AnnData</code>.</td>
</tr>
<tr class="odd">
<td>
<code>-s</code>, <code>--sample</code>
(<strong>required</strong>)</td>
<td>Name of the sample. This will be prepended to the cell
barcodes.</td>
</tr>
<tr class="even">
<td>
<code>-o</code>, <code>--directory</code>
(<strong>required</strong>)</td>
<td>Output directory. A new subdirectory will be created with the name
“sample”. R, Python, and FlatFile directories will be created under the
“sample” directory containing the data containers with QC metrics.
Default <code>.</code>. More information about output directory
structure is explained in <a href="#outputs"><em>Outputs</em></a>
section below.</td>
</tr>
<tr class="odd">
<td>
<code>-F</code>, <code>--outputFormat</code>
(<strong>required</strong>)</td>
<td>The output format of this QC pipeline. Currently, it supports
<code>SCE</code>, <code>Seurat</code>, <code>FlatFile</code>,
<code>AnnData</code> and <code>HTAN</code> (manifest files that meets
HTAN requirement).</td>
</tr>
<tr class="even">
<td>
<code>-S</code>, <code>--splitSample</code>
(<strong>required</strong>)</td>
<td>Save a <code>SingleCellExperiment</code> object for each sample.
Default is <code>TRUE</code>. If <code>FALSE</code>, the data of all
samples will be combined into one <code>SingleCellExperiment</code>
object and this object will be output.</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="batch-processing">Batch processing<a class="anchor" aria-label="anchor" href="#batch-processing"></a>
</h4>
<p>When running the pipeline with more than one sample to process, care
should be exercised with the -S flag, and whether or not one wishes to
combine the samples.</p>
<p>In addition, the following parameters will need multiple inputs,
separated by a <strong>comma</strong> <strong>without any
spacing</strong>:</p>
<table class="table">
<colgroup>
<col width="10%">
<col width="89%">
</colgroup>
<thead><tr class="header">
<th>Parameter</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>-b</code>, <code>--basePath</code>
</td>
<td>Base path for the output from the preprocessing algorithm; one is
required for each sample, even if they are in the same location.</td>
</tr>
<tr class="even">
<td>
<code>-P</code>, <code>--preproc</code>
</td>
<td>In multiple object mode, each entry should have a preprocessing
algorithm associated with it; the pipeline supports serial processing of
different sample types.</td>
</tr>
<tr class="odd">
<td>
<code>-s</code>, <code>--sample</code>
</td>
<td>Names of the samples.</td>
</tr>
</tbody>
</table>
<p>An example execution is shown below:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">Rscript</span> SCTK_runQC.R <span class="dt">\</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>-b /base/path1,/base/path2 <span class="dt">\</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>-P Preprocessing_Algorithm_1,Preprocess_Algorithm_2 <span class="dt">\</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>-s SampleName1,SampleName2 <span class="dt">\</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>-o Output_Directory <span class="dt">\</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>-S FALSE <span class="dt">\</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>-F SCE,AnnData,FlatFile <span class="dt">\</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>-d Droplet <span class="dt">\</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>-D TRUE <span class="dt">\</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="optional-arguments">Optional arguments<a class="anchor" aria-label="anchor" href="#optional-arguments"></a>
</h4>
<p>The optional arguments are as follows. Their usage depend on type of
data and user-defined behaviour.</p>
<table class="table">
<colgroup>
<col width="11%">
<col width="88%">
</colgroup>
<thead><tr class="header">
<th>Parameter</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>-g</code>, <code>--gmt</code>
</td>
<td>GMT file containing gene sets for quality control.</td>
</tr>
<tr class="even">
<td>
<code>-t</code>, <code>--delim</code> (required when -g is
specified)</td>
<td>Delimiter used in GMT file. Default <code>"\t"</code>.</td>
</tr>
<tr class="odd">
<td>
<code>-G</code>, <code>--genome</code>
</td>
<td>The name of genome reference. This is only required for CellRangerV2
data.</td>
</tr>
<tr class="even">
<td>
<code>-y</code>, <code>--yaml</code>
</td>
<td>YAML file used to specify parameters of QC functions called by
SCTK-QC pipeline. Please check <a href="#specify-parameters-using-yaml-file"><em>Specify parameters using
yaml file</em></a> section for details.</td>
</tr>
<tr class="odd">
<td>
<code>-c</code>, <code>--cellData</code>
</td>
<td>The full path of the RDS, H5AD or Matrix file of the cell matrix.
This would be use only when <code>--preproc</code> is
<code>SceRDS</code>, <code>Seurat</code>, <code>AnnData</code>, or
<code>CountMatrix</code>.</td>
</tr>
<tr class="even">
<td>
<code>-r</code>, <code>--rawData</code>
</td>
<td>The full path of the RDS file or Matrix file of the droplet matrix.
This would be provided only when <code>--preproc</code> is
<code>SceRDS</code>, <code>Seurat</code>, <code>AnnData</code>, or
<code>CountMatrix</code>.</td>
</tr>
<tr class="odd">
<td>
<code>-C</code>, <code>--cellPath</code>
</td>
<td>The directory containing <code>matrix.mtx.gz</code>,
<code>features.tsv.gz</code> and <code>barcodes.tsv.gz</code> files
originally generated by 10x CellrangerV2 or CellrangerV3 (files in the
<code>filtered_feature_bc_matrix</code> directory). This argument only
works when <code>--preproc</code> is <code>CellRangerV2</code> or
<code>CellRangerV3</code>. Default is <code>NULL</code>. If
<code>base_path</code> is <code>NULL</code>, <code>cellPath</code> or
<code>rawPath</code> should be specified.</td>
</tr>
<tr class="even">
<td>
<code>-R</code>, <code>--rawPath</code>
</td>
<td>The directory containing <code>matrix.mtx.gz</code>,
<code>features.tsv.gz</code> and <code>barcodes.tsv.gz</code> files
originally generated by 10x CellrangerV2 or CellrangerV3 (files in the
<code>raw_feature_bc_matrix</code> directory). This argument only works
when <code>--preproc</code> is <code>CellRangerV2</code> or
<code>CellRangerV3</code>. Default is <code>NULL</code>. If
<code>base_path</code> is <code>NULL</code>, <code>cellPath</code> or
<code>rawPath</code> should be specified.</td>
</tr>
<tr class="odd">
<td>
<code>-d</code>, <code>--dataType</code>
</td>
<td>Type of data as input. Default is <code>Both</code>, which means
taking both droplet and cell matrix as input. If set as
<code>Droplet</code>, it will only processes droplet data. If set as
<code>Cell</code>, it will only processes cell data.</td>
</tr>
<tr class="even">
<td>
<code>-D</code>, <code>--detectCells</code>
</td>
<td>Detect cells from droplet matrix. Default is <code>FALSE</code>.
This argument is only evaluated when <code>-d</code> is
<code>Droplet</code>. If set as <code>TRUE</code>, cells will be
detected and cell matrix will be subset from the droplet matrix. Also,
QC will be performed on the detected cell matrix.</td>
</tr>
<tr class="odd">
<td>
<code>-m</code>, <code>--cellDetectMethod</code>
</td>
<td>Methods to detect cells from droplet matrix. Default is
<code>EmptyDrops</code>. This argument is only evaluated when
<code>-D</code> is <code>TRUE</code>. Other options could be
<code>Knee</code> or <code>Inflection</code>. More information is
provided in the <a href="cnsl_dropletqc.html"><em>Droplet QC</em>
documentation</a>.</td>
</tr>
<tr class="even">
<td>
<code>-n</code>, <code>--numCores</code>
</td>
<td>Number of cores used to run the pipeline. By default is
<code>1</code>. Parallel computing is enabled if <code>-n</code> is
greater than 1.</td>
</tr>
<tr class="odd">
<td>
<code>-T</code>, <code>--parallelType</code>
</td>
<td>Type of parallel computing used for parallel computing. Parallel
computing used in this pipeline depends on <a href="https://bioconductor.org/packages/release/bioc/html/BiocParallel.html" class="external-link">BiocParallel</a>
package. Default is <code>MulticoreParam</code>. It can be
<code>MulticoreParam</code> or <code>SnowParam</code>. This argument
will be evaluated only when <code>--numCores</code> is greater than
1.</td>
</tr>
<tr class="even">
<td>
<code>-i</code>, <code>--studyDesign</code>
</td>
<td>The TXT file containing the description of the study design. Default
is <code>NULL</code>. This would be shown at the beginning of the HTML
report of cell and droplet QC.</td>
</tr>
<tr class="odd">
<td>
<code>-L</code>, <code>--subTitle</code>
</td>
<td>The subtitle used in the cell and droplet QC HTML report. Default is
<code>None</code>. The subtitle can contain information of the sample,
like sample name, etc. If <code>-S</code> is set as <code>TRUE</code>,
the length of subsitle should be the same as the number of samples. If
<code>-S</code> is set as <code>FALSE</code>, the length of subtitle
should be one or NULL.</td>
</tr>
<tr class="even">
<td>
<code>-M</code>, <code>--detectMitoLevel</code>
</td>
<td>Detect mitochondrial gene expression level. If <code>TRUE</code>,
the pipeline will examine mitochondrial gene expression level
automatically without the need of importing user defined GMT file.
Default is <code>TRUE</code>.</td>
</tr>
<tr class="odd">
<td>
<code>-E</code>, <code>--mitoType</code>
</td>
<td>Type of mitochondrial gene-set to be loaded when
<code>--detectMitoLevel</code> is set to <code>TRUE</code>. Possible
choices are: <code>human-ensembl</code>, <code>human-symbol</code>,
<code>human-entrez</code>, <code>human-ensemblTranscriptID</code>,
<code>mouse-ensembl</code>, <code>mouse-symbol</code>,
<code>mouse-entrez</code> and
<code>mouse-ensemblTranscriptID</code>.</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level3">
<h3 id="specify-parameters-using-yaml-file">Specify parameters using YAML file<a class="anchor" aria-label="anchor" href="#specify-parameters-using-yaml-file"></a>
</h3>
<p>Users can specify parameters for QC algorithms in this pipeline with
a YAML file (supplied with <code>-y</code>/<code>--yamlFile</code>
argument). The current supported QC algorithms including doublet
detection (<code>bcds</code>, <code>cxds</code>,
<code>cxds_bcds_hybrid</code>, <code>doubletFinder</code>,
<code>doubletCells</code> and <code>scrublet</code>), decontamination
(<code>decontX</code>), emptyDrop detection (<code>emptyDrops</code>)
and barcodeRankDrops (<code>barcodeRanks</code>). A summary of each
function is shown below:</p>
<div class="float">
<img src="qc_yamlParameters.png" alt="QC yaml parameters"><div class="figcaption">QC yaml parameters</div>
</div>
<p>An example of QC parameters YAML file is shown below:</p>
<pre><code><span><span class="op">-</span><span class="op">-</span><span class="op">-</span></span>
<span><span class="va">Params</span><span class="op">:</span>   <span class="co">### should not be omitted</span></span>
<span>  <span class="va">bcds</span><span class="op">:</span></span>
<span>    <span class="va">ntop</span><span class="op">:</span> <span class="fl">600</span></span>
<span></span>
<span>  <span class="va">cxds</span><span class="op">:</span></span>
<span>    <span class="va">ntop</span><span class="op">:</span> <span class="fl">600</span></span>
<span></span>
<span>  <span class="va">cxds_bcds_hybrid</span><span class="op">:</span></span>
<span>    <span class="va">nTop</span><span class="op">:</span> <span class="fl">600</span></span>
<span></span>
<span>  <span class="va">decontX</span><span class="op">:</span></span>
<span>    <span class="va">maxIter</span><span class="op">:</span> <span class="fl">600</span></span>
<span></span>
<span>  <span class="va">emptyDrops</span><span class="op">:</span></span>
<span>    <span class="va">lower</span><span class="op">:</span> <span class="fl">50</span></span>
<span>    <span class="va">niters</span><span class="op">:</span> <span class="fl">5000</span></span>
<span>    <span class="va">testAmbient</span><span class="op">:</span> <span class="va">True</span></span>
<span></span>
<span>  <span class="va">barcodeRanks</span><span class="op">:</span></span>
<span>    <span class="va">lower</span><span class="op">:</span> <span class="fl">50</span></span></code></pre>
<p>The format of YAML file can be found <a href="https://docs.ansible.com/ansible/latest/reference_appendices/YAMLSyntax.html" class="external-link">here</a>.
The parameters should be consistent with the parameters of each QC
function in SCTK. Parameters that are not defined in this YAML file will
use the default value. Please refer <a href="../reference/index.html#quality-control-preprocessing">reference</a>
for detailed information about arguments of each QC function.</p>
</div>
<div class="section level3">
<h3 id="parallel-computing">Parallel computing<a class="anchor" aria-label="anchor" href="#parallel-computing"></a>
</h3>
<p>SCTK-QC pipeline enables parallel computing to speed up the analysis.
Parallel computing is enabled by setting
<code>-n</code>/<code>--numCores</code> greater than 1. The
<code>-n</code>/<code>--numCores</code> is used to set the number of
cores used for the pipeline.</p>
<p>The backend of parallel computing is supported by <a href="https://bioconductor.org/packages/release/bioc/html/BiocParallel.html" class="external-link">BiocParallel</a>
package. Therefore, users can select different types of parallel
evaluation by setting <code>-T</code>/<code>--parallelType</code>
argument. Default is <code>MulticoreParam</code>. Currently,
<code>MulticoreParam</code> and <code>SnowParam</code> are supported for
<code>-T</code> argument. However, <code>MulticoreParam</code> is not
supported by Windows system. Windows user can choose
<code>SnowParam</code> as the backend of parallel computing.</p>
</div>
</div>
<div class="section level2">
<h2 id="description-of-input-data">Description of input data<a class="anchor" aria-label="anchor" href="#description-of-input-data"></a>
</h2>
<div class="section level3">
<h3 id="qc-on-combinations-of-droplet-and-cell-matrix">QC on combinations of droplet and cell matrix<a class="anchor" aria-label="anchor" href="#qc-on-combinations-of-droplet-and-cell-matrix"></a>
</h3>
<p><code>-d</code> argument is used to specify which type of count
matrix is used in the pipeline. Default is <code>Both</code>, which
means the pipeline will run quality control on both droplet and cell
count data.</p>
<p>Users can also choose to run SCTK-QC pipeline on only the droplet
count or cell count matrix, instead of running on both. In this case,
the pipeline will only take the single input and perform QC on it. An
example is shown below:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">Rscript</span> SCTK_runQC.R <span class="dt">\</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>-b /base/path <span class="dt">\</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>-P Preprocessing_Algorithm <span class="dt">\</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>-s SampleName <span class="dt">\</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>-o Output_Directory <span class="dt">\</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>-S TRUE <span class="dt">\</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>-F SCE,AnnData,FlatFile,Seurat <span class="dt">\</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>-d Droplet <span class="dt">\</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>-D TRUE <span class="dt">\</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>-m EmptyDrops</span></code></pre></div>
<p>If <code>-d</code> argument is set as <code>Droplet</code>, the QC
pipeline will only take droplet count matrix as input and perform
quality control. You can choose whether to detect cells from the droplet
matrix by setting <code>-D</code> as <code>TRUE</code>. If yes, cell
count matrix will be detected and the pipeline will also perform quality
control on this matrix and output the result. You could further define
the method used to detect cells from droplet matrix by setting
<code>-m</code> argument. <code>-m</code> could be one of
<code>EmptyDrops</code>, <code>Knee</code> or <code>Inflection</code>.
<code>EmptyDrops</code> will keep cells that pass the
<code><a href="../reference/runEmptyDrops.html">runEmptyDrops()</a></code> function test. <code>Knee</code> and
<code>Inflection</code> will keep cells that pass the knee or inflection
point returned from <code><a href="../reference/runBarcodeRankDrops.html">runBarcodeRankDrops()</a></code> function.</p>
<p>If <code>-d</code> argument is set as <code>Cell</code>, the QC
pipeline will only take cell count matrix as input and perform quality
control. A figure showing the analysis steps and outputs of different
inputs is shown below:</p>
<p><img src="qc_singleInput.png" alt="QC single input"><br></p>
</div>
<div class="section level3">
<h3 id="importing-data-from-different-preprocessing-tools">Importing data from different preprocessing tools<a class="anchor" aria-label="anchor" href="#importing-data-from-different-preprocessing-tools"></a>
</h3>
<div class="section level4">
<h4 id="import-data-from-cellranger">Import data from Cellranger<a class="anchor" aria-label="anchor" href="#import-data-from-cellranger"></a>
</h4>
<p>This pipeline enables different ways to import
CellrangerV2/CellrangerV3 data for flexibility.</p>
<ol style="list-style-type: decimal">
<li>If the cellranger data set is saved in the default <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/overview" class="external-link">cellranger
output directory</a>, you can load the data by running following
code:</li>
</ol>
<p>For CellRangerV3:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">Rscript</span> SCTK_runQC.R <span class="dt">\</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>-b /base/path <span class="dt">\</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>-P CellRangerV3 <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>-s SampleName <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>-o Output_Directory <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>-S TRUE <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>-F SCE,AnnData,FlatFile,Seurat</span></code></pre></div>
<p>For CellRangerV2, the reference used by cellranger needs to be
specified by <code>-G</code>/<code>--genome</code>:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">Rscript</span> SCTK_runQC.R <span class="dt">\</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>-b /base/path <span class="dt">\</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>-P CellRangerV2 <span class="dt">\</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>-s SampleName <span class="dt">\</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>-o Output_Directory <span class="dt">\</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>-S TRUE <span class="dt">\</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>-G GenomeName <span class="dt">\</span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>-F SCE,AnnData,FlatFile,Seurat</span></code></pre></div>
<p><code>-b</code> specifies the base path and usually it’s the output
folder of 10x <code>cellranger-count</code>. <code>-s</code> specifies
the sample name, which has to be the same as the name of the sample
folder under the base folder. The folder layout would look like the
following:</p>
<pre><code>├── BasePath
└── SampleName 
    ├── outs
    |   ├── filtered_feature_bc_matrix
    |   |   ├── barcodes.tsv.gz
    |   |   ├── features.tsv.gz
    |   |   └── matrix.mtx.gz
    |   ├── raw_feature_bc_matrix
    |   |   ├── barcodes.tsv.gz
    |   |   ├── features.tsv.gz
    |   |   └── matrix.mtx.gz
        ...</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>If the <code>cellranger-count</code> output have been moved out of
the default <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/overview" class="external-link">cellranger
output directory</a>, you can specified the path to droplet and cell
count matrix using arguments <code>-R</code> and <code>-C</code>:</li>
</ol>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">Rscript</span> SCTK_runQC.R <span class="dt">\</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>-P CellRangerV2 <span class="dt">\</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>-C /path/to/cell/matrix <span class="dt">\</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>-R /path/to/droplet/matrix <span class="dt">\</span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>-s SampleName <span class="dt">\</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>-o Output_Directory <span class="dt">\</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>-S TRUE <span class="dt">\</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>-F SCE,AnnData,FlatFile,Seurat</span></code></pre></div>
<p>In this case, you <strong>must</strong> skip <code>-b</code>
arguments and you can also skip <code>-G</code> argument for
CellRangerV2 data.</p>
</div>
<div class="section level4">
<h4 id="import-data-from-rds-h5ad-or-matrix-stored-in-a-text-file">Import data from RDS, H5AD or matrix stored in a text file<a class="anchor" aria-label="anchor" href="#import-data-from-rds-h5ad-or-matrix-stored-in-a-text-file"></a>
</h4>
<ol style="list-style-type: decimal">
<li>If your data in stored as a <code>SingleCellExperiment</code> object
in RDS file, singleCellTK also supports this type of input. To run
quality control with RDS file as input, run the following code:</li>
</ol>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">Rscript</span> SCTK_runQC.R <span class="dt">\</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>-P SceRDS <span class="dt">\</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>-s Samplename <span class="dt">\</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>-o Output_Directory <span class="dt">\</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>-S TRUE <span class="dt">\</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>-F SCE,AnnData,FlatFile,Seurat <span class="dt">\</span></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a>-r /path/to/rds/file/droplet.RDS <span class="dt">\</span></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a>-c /path/to/rds/file/cell.RDS</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>If your input is stored as a <code>Seurat</code> object in RDS file,
you may use the following code:</li>
</ol>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">Rscript</span> SCTK_runQC.R <span class="dt">\</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>-P Seurat <span class="dt">\</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>-s Samplename <span class="dt">\</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>-o Output_Directory <span class="dt">\</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>-S TRUE <span class="dt">\</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>-F SCE,AnnData,FlatFile,Seurat <span class="dt">\</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>-r /path/to/rds/file/droplet.RDS <span class="dt">\</span></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a>-c /path/to/rds/file/cell.RDS</span></code></pre></div>
<p><strong>Warning:</strong> Both <code>SingleCellExperiment</code> and
<code>Seurat</code> objects use the same RDS format. Be sure which one
your data is formatted as or the pipeline may not successfully run.</p>
<ol start="3" style="list-style-type: decimal">
<li>If your input as stored as an <code>AnnData</code> object in H5AD
file,</li>
</ol>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">Rscript</span> SCTK_runQC.R <span class="dt">\</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>-P AnnData <span class="dt">\</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>-s Samplename <span class="dt">\</span></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>-o Output_Directory <span class="dt">\</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>-S TRUE <span class="dt">\</span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a>-F SCE,AnnData,FlatFile,Seurat <span class="dt">\</span></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a>-r /path/to/rds/file/droplet.RDS <span class="dt">\</span></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a>-c /path/to/rds/file/cell.RDS</span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>If your input is stored in TXT file as a matrix, which has barcodes
as column names and genes as row names, run the following code to start
the quality control pipeline:</li>
</ol>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="ex">Rscript</span> SCTK_runQC.R <span class="dt">\</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>-P CountMatrix <span class="dt">\</span></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>-s Samplename <span class="dt">\</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>-o Output_Directory <span class="dt">\</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>-S TRUE <span class="dt">\</span></span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>-F SCE,AnnData,FlatFile,Seurat <span class="dt">\</span></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a>-r /path/to/matrix/file/droplet.txt <span class="dt">\</span></span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>-c path/to/matrix/file/cell.txt</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="methods-to-run-pipeline-on-data-set-generated-by-other-algorithms">Methods to run pipeline on data set generated by other
algorithms<a class="anchor" aria-label="anchor" href="#methods-to-run-pipeline-on-data-set-generated-by-other-algorithms"></a>
</h4>
<p>SCTK-QC pipeline allows importing data from the following
pre-processing tools or objects:</p>
<ul>
<li><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">CellRanger</a></li>
<li><a href="https://data.humancellatlas.org/pipelines/optimus-workflow" class="external-link">Optimus</a></li>
<li><a href="https://github.com/hms-dbmi/dropEst" class="external-link">DropEst</a></li>
<li><a href="https://github.com/BUStools/bustools" class="external-link">BUStools</a></li>
<li><a href="https://github.com/ambrosejcarr/seqc" class="external-link">Seqc</a></li>
<li><a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md" class="external-link">STARSolo</a></li>
<li>
<a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>
object saved in RDS file</li>
<li>
<a href="https://github.com/theislab/anndata" class="external-link">AnnData</a> object
saved in HDF5 file</li>
</ul>
<p>If your data is preprocessed by other algorithms, you might want to
make sure the ‘-b’ argument matches the path storing the data and the
‘-P’ argument matches the right preprocessing tools. Basically, the
templated is shown below:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="ex">Rscript</span> SCTK_runQC.R <span class="dt">\</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>-b /base/path <span class="dt">\</span></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>-P Preprocessing_Algorithm <span class="dt">\</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a>-s SampleName <span class="dt">\</span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a>-o Output_Directory <span class="dt">\</span></span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a>-S TRUE <span class="dt">\</span></span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a>-F SCE,AnnData,FlatFile,Seurat</span></code></pre></div>
<p>The following table describes how SCTK expects the inputs to be
structured and passed for each import function. In all cases, SCTK
retains the standard output directory structure from upstream tools. All
the import functions return the imported counts matrix as an
<code>assay</code> in a <code>SingleCellExperiment</code> object, with
associated information in respective <code>colData</code>,
<code>rowData</code>, <code>reducedDims</code>, and
<code>metadata</code> fields.</p>
<p><img src="qc_inputShell.png" alt="QC data format"><br></p>
<p>The table above also shows the R console functions for the QC
algorithms. Detailed information about function parameters and defaults
are available in the <a href="../reference/index.html"><em>Reference</em></a> section.</p>
</div>
</div>
<div class="section level3">
<h3 id="gene-sets">Gene sets<a class="anchor" aria-label="anchor" href="#gene-sets"></a>
</h3>
<p>Quantifying the level of gene sets can be useful quality control. For
example, the percentage of counts from mitochondrial genes can be an
indicator of cell stress or death.</p>
<p>User can quantify the expression of mitochondrial genes for human or
mouse dataset by setting <code>-M</code>/<code>--detectMitoLevel</code>
as <code>TRUE</code>. User needs to specify the correct
<code>--mitoType</code> argument for the dataset. The SCTK-QC pipeline
has built-in mitochondrial gene sets for human and mouse genes. It
supports four different type of gene id: <code>gene symbol</code>,
<code>entrez ID</code>, <code>ensembl ID</code> and
<code>ensembl transcript ID</code>. Therefore, there are eight options
for <code>--mitoType</code> arguments: <code>human-ensembl</code>,
<code>human-symbol</code>, <code>human-entrez</code>,
<code>human-ensemblTranscriptID</code>, <code>mouse-ensembl</code>,
<code>mouse-symbol</code>, <code>mouse-entrez</code> and
<code>mouse-ensemblTranscriptID</code>.</p>
<p>To quantify expression of other gene set, users can pass a <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29" class="external-link">GMT</a>
file (with <code>-g</code>/<code>--gmt</code> argument) to the pipeline
with one row for each gene set. The first column should be the name of
the gene set (e.g. <code>mito</code>).</p>
<p>The second column for each gene set in the GMT file (i.e. the
description) should contain the location of where to look for the
matching IDs in the data. If set to <code>rownames</code>, then the gene
set IDs will be matched with the row IDs of the data matrix. If a
character string or an integer index is supplied, then gene set IDs will
be matched to the IDs in that column of feature table. Gene sets with
mitochondrial genes can be found <a href="https://github.com/compbiomed/singleCellTK/tree/master/exec" class="external-link">here</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="outputs">Outputs<a class="anchor" aria-label="anchor" href="#outputs"></a>
</h2>
<p>The output directory is created under the path specified by
<code>-o</code>/<code>--directory</code> argument. Each sample is stored
in the subdirectory (named by <code>-s</code>/<code>--sample</code>
argument) within this output direcotry. Within each sample directory,
each output format will be separated into subdirectories. The output
file hierarchy is shown below:</p>
<pre><code>(root; output directory)
├── level3Meta.csv
├── level4Meta.csv
├── sample1_dropletQC.html
├── sample1_cellQC.html
└── sample1 
    ├──sample1_cellQC_summary.csv
    ├── R
    |   ├── sample1_Droplets.rds
    |   └── sample1_Cells.rds
    ├── Python
    |   ├── Droplets
    |   |   └── sample1.h5ad
    |   └── Cells
    |       └── sample1.h5ad
    ├── FlatFile
    |   ├── Droplets
    |   |   ├── assays
    |   |   |   └── sample1_counts.mtx.gz
    |   |   ├── metadata
    |   |   |   └── sample1_metadata.rds
    |   |   ├── sample1_colData.txt.gz
    |   |   └── sample1_rowData.txt.gz 
    |   └── Cells
    |       ├── assays
    |       |   └── sample1_counts.mtx.gz
    |       ├── metadata
    |       |   └── sample1_metadata.rds
    |       ├── reducedDims
    |       |   ├──sample1_decontX_UMAP.txt.gz
    |       |   ├──sample1_scrublet_TSNE.txt.gz
    |       |   └──sample1_scrublet_UMAP.txt.gz
    |       ├── sample1_colData.txt.gz
    |       └── sample1_rowData.txt.gz 
    └── sample1_QCparameters.yaml</code></pre>
</div>
<div class="section level2">
<h2 id="documentation-of-available-tools">Documentation of available tools<a class="anchor" aria-label="anchor" href="#documentation-of-available-tools"></a>
</h2>
<div class="section level4">
<h4 id="empty-droplet-detection">Empty droplet detection:<a class="anchor" aria-label="anchor" href="#empty-droplet-detection"></a>
</h4>
<ul>
<li>
<a href="https://rdrr.io/github/MarioniLab/DropletUtils/man/emptyDrops.html" class="external-link">emptyDrops</a>
from the package <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link">DropletUtils</a>
</li>
<li>
<a href="https://rdrr.io/github/MarioniLab/DropletUtils/man/barcodeRanks.html" class="external-link">barcodeRanks</a>
from the package <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link">DropletUtils</a>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="doublet-detection">Doublet Detection<a class="anchor" aria-label="anchor" href="#doublet-detection"></a>
</h4>
<ul>
<li>
<a href="https://rdrr.io/bioc/scDblFinder/man/scDblFinder.html" class="external-link">scDblFinder</a>
from the package <a href="https://bioconductor.org/packages/release/bioc/html/scDblFinder.html" class="external-link">scDblFinder</a>
</li>
<li>
<a href="https://rdrr.io/bioc/scds/man/cxds.html" class="external-link">cxds</a>, <a href="https://rdrr.io/bioc/scds/man/bcds.html" class="external-link">bcds</a>, and <a href="https://rdrr.io/bioc/scds/man/cxds_bcds_hybrid.html" class="external-link">cxds_bcds_hybrid</a>
from the package <a href="http://bioconductor.org/packages/release/bioc/html/scds.html" class="external-link">scds</a>
</li>
<li>
<a href="https://rdrr.io/github/chris-mcginnis-ucsf/DoubletFinder/man/doubletFinder.html" class="external-link">doubletFinder</a>
from the package <a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder" class="external-link">DoubletFinder</a>
</li>
<li>
<a href="https://bioconda.github.io/recipes/scrublet/README.html" class="external-link">Scrublet</a>
from the package <a href="https://github.com/allonkleinlab/scrublet" class="external-link">scrublet</a>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="ambient-rna-detection">Ambient RNA detection<a class="anchor" aria-label="anchor" href="#ambient-rna-detection"></a>
</h4>
<ul>
<li>
<a href="https://rdrr.io/bioc/celda/man/decontX.html" class="external-link">decontX</a>
from the package <a href="https://bioconductor.org/packages/release/bioc/html/celda.html" class="external-link">celda</a>
</li>
<li>
<a href="https://github.com/constantAmateur/SoupX" class="external-link">SoupX</a> from
package <a href="https://github.com/constantAmateur/SoupX" class="external-link">SoupX</a>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
