<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quality Control for droplet data â€¢ singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Quality Control for droplet data">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.18.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>

  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><script src="cnsl_dropletqc_files/kePrint-0.0.1/kePrint.js"></script><link href="cnsl_dropletqc_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quality Control for droplet data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/cnsl_dropletqc.Rmd" class="external-link"><code>vignettes/articles/cnsl_dropletqc.Rmd</code></a></small>
      <div class="hidden name"><code>cnsl_dropletqc.Rmd</code></div>

    </div>

    
    
<style>
body {
text-align: justify}
</style>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Performing comprehensive quality control (QC) is necessary to remove
poor quality cells for downstream analysis of single-cell RNA sequencing
(scRNA-seq) data. Within droplet-based scRNA-seq data, droplets
containing cells must be differentiated from empty droplets. Therefore,
assessment of the data is required, for which various QC algorithms have
been developed. In singleCellTK (SCTK), we have written convenience
functions for several of these tools. In this guide, we will demonstrate
how to use these functions to perform quality control on unfiltered,
droplet data. (For definition of droplet data, please refer <a href="cmd_qc.html#introduction-1">this documentation</a>.)</p>
<p>The package can be loaded using the <code><a href="https://rdrr.io/r/base/library.html" class="external-link">library()</a></code>
command.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="running-qc-on-droplet-raw-data">Running QC on droplet raw data<a class="anchor" aria-label="anchor" href="#running-qc-on-droplet-raw-data"></a>
</h2>
<div class="section level3">
<h3 id="load-pbmc4k-data-from-10x">Load PBMC4k data from 10X<a class="anchor" aria-label="anchor" href="#load-pbmc4k-data-from-10x"></a>
</h3>
<p>SCTK takes in a <code>SingleCellExperiment</code> object from the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>
package. We will utilize the 10X PBMC 4K dataset as an example. For the
QC of droplet-based counts data, we will install the dataset from the
10X Genomics website using the <a href="https://www.bioconductor.org/packages/release/bioc/html/BiocFileCache.html" class="external-link"><code>BiocFileCache</code></a>
package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install BiocFileCache if is it not already</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocFileCache"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"BiocFileCache"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"BiocFileCache"</span><span class="op">)</span></span>
<span><span class="va">bfc</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">BiocFileCache</a></span><span class="op">(</span><span class="st">"raw_data"</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">raw.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcrpath</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span></span>
<span>  <span class="st">"http://cf.10xgenomics.com/samples"</span>,</span>
<span>  <span class="st">"cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz"</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span><span class="va">raw.path</span>, exdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pbmc4k"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fname</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pbmc4k/raw_gene_bc_matrices/GRCh38"</span><span class="op">)</span></span>
<span><span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu">DropletUtils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html" class="external-link">read10xCounts</a></span><span class="op">(</span><span class="va">fname</span>, col.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### change the sample column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample"</span>, <span class="st">"Barcode"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"pbmc4k"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>SCTK also supports the importing of single-cell data from the
following platforms: <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">10X
CellRanger</a>, <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md" class="external-link">STARSolo</a>,
<a href="https://www.kallistobus.tools/" class="external-link">BUSTools</a>, <a href="https://github.com/ambrosejcarr/seqc" class="external-link">SEQC</a>, <a href="https://github.com/hms-dbmi/dropEst" class="external-link">DropEST</a>, <a href="https://salmon.readthedocs.io/en/latest/alevin.html" class="external-link">Alevin</a>,
as well as dataset already stored in <a href="https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>
object and <a href="https://github.com/theislab/anndata" class="external-link">AnnData</a>
object. To load your own input data, please refer
<code>Function Reference for pre-processing tools</code> under
<code>Console Analysis</code> section in <a href="import_data.html">Import data into SCTK</a> for detailed
instruction.</p>
</div>
<div class="section level3">
<h3 id="rundropletqc">runDropletQC<a class="anchor" aria-label="anchor" href="#rundropletqc"></a>
</h3>
<p>All droplet-based QC functions are able to be run under the wrapper
function <code><a href="../reference/runDropletQC.html">runDropletQC()</a></code>. By default all possible QC
algorithms will be run.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDropletQC.html">runDropletQC</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Wed Apr 23 12:13:23 2025 ... Running 'perCellQCMetrics'</span></span></code></pre>
<pre><code><span><span class="co">## Wed Apr 23 12:13:23 2025 ...... Attempting to find mitochondrial genes by identifying features in 'rownames' that match mitochondrial genes from reference 'human' and ID type 'ensembl'.</span></span></code></pre>
<pre><code><span><span class="co">## Wed Apr 23 12:13:26 2025 ... Running 'emptyDrops'</span></span></code></pre>
<pre><code><span><span class="co">## Wed Apr 23 12:13:42 2025 ... Running 'barcodeRanks'</span></span></code></pre>
<p>If users choose to only run a specific set of algorithms, they can
specify which to run with the <code>algorithms</code> parameter.</p>
<p>After running QC functions with SCTK, the output will be stored in
the <code>colData</code> slot of the <code>SingleCellExperiment</code>
object.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 5 rows and 19 columns</span></span>
<span><span class="co">##                         sample            Barcode       sum  detected</span></span>
<span><span class="co">##                    &lt;character&gt;        &lt;character&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## TTTCCTCAGAGCTGGT-1      pbmc4k TTTCCTCAGAGCTGGT-1         0         0</span></span>
<span><span class="co">## CGCGTTTGTCAAACTC-1      pbmc4k CGCGTTTGTCAAACTC-1         0         0</span></span>
<span><span class="co">## CTACGTCTCAGGCAAG-1      pbmc4k CTACGTCTCAGGCAAG-1         0         0</span></span>
<span><span class="co">## CGATCGGCACCATGTA-1      pbmc4k CGATCGGCACCATGTA-1         0         0</span></span>
<span><span class="co">## AGTGTCACAAGTTAAG-1      pbmc4k AGTGTCACAAGTTAAG-1         1         1</span></span>
<span><span class="co">##                    percent.top_50 percent.top_100 percent.top_200</span></span>
<span><span class="co">##                         &lt;numeric&gt;       &lt;numeric&gt;       &lt;numeric&gt;</span></span>
<span><span class="co">## TTTCCTCAGAGCTGGT-1            NaN             NaN             NaN</span></span>
<span><span class="co">## CGCGTTTGTCAAACTC-1            NaN             NaN             NaN</span></span>
<span><span class="co">## CTACGTCTCAGGCAAG-1            NaN             NaN             NaN</span></span>
<span><span class="co">## CGATCGGCACCATGTA-1            NaN             NaN             NaN</span></span>
<span><span class="co">## AGTGTCACAAGTTAAG-1            100             100             100</span></span>
<span><span class="co">##                    percent.top_500  mito_sum mito_detected mito_percent</span></span>
<span><span class="co">##                          &lt;numeric&gt; &lt;numeric&gt;     &lt;integer&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">## TTTCCTCAGAGCTGGT-1             NaN         0             0          NaN</span></span>
<span><span class="co">## CGCGTTTGTCAAACTC-1             NaN         0             0          NaN</span></span>
<span><span class="co">## CTACGTCTCAGGCAAG-1             NaN         0             0          NaN</span></span>
<span><span class="co">## CGATCGGCACCATGTA-1             NaN         0             0          NaN</span></span>
<span><span class="co">## AGTGTCACAAGTTAAG-1             100         0             0            0</span></span>
<span><span class="co">##                        total dropletUtils_emptyDrops_total</span></span>
<span><span class="co">##                    &lt;numeric&gt;                     &lt;integer&gt;</span></span>
<span><span class="co">## TTTCCTCAGAGCTGGT-1         0                             0</span></span>
<span><span class="co">## CGCGTTTGTCAAACTC-1         0                             0</span></span>
<span><span class="co">## CTACGTCTCAGGCAAG-1         0                             0</span></span>
<span><span class="co">## CGATCGGCACCATGTA-1         0                             0</span></span>
<span><span class="co">## AGTGTCACAAGTTAAG-1         1                             1</span></span>
<span><span class="co">##                    dropletUtils_emptyDrops_logprob</span></span>
<span><span class="co">##                                          &lt;numeric&gt;</span></span>
<span><span class="co">## TTTCCTCAGAGCTGGT-1                              NA</span></span>
<span><span class="co">## CGCGTTTGTCAAACTC-1                              NA</span></span>
<span><span class="co">## CTACGTCTCAGGCAAG-1                              NA</span></span>
<span><span class="co">## CGATCGGCACCATGTA-1                              NA</span></span>
<span><span class="co">## AGTGTCACAAGTTAAG-1                              NA</span></span>
<span><span class="co">##                    dropletUtils_emptyDrops_pvalue</span></span>
<span><span class="co">##                                         &lt;numeric&gt;</span></span>
<span><span class="co">## TTTCCTCAGAGCTGGT-1                             NA</span></span>
<span><span class="co">## CGCGTTTGTCAAACTC-1                             NA</span></span>
<span><span class="co">## CTACGTCTCAGGCAAG-1                             NA</span></span>
<span><span class="co">## CGATCGGCACCATGTA-1                             NA</span></span>
<span><span class="co">## AGTGTCACAAGTTAAG-1                             NA</span></span>
<span><span class="co">##                    dropletUtils_emptyDrops_limited dropletUtils_emptyDrops_fdr</span></span>
<span><span class="co">##                                          &lt;logical&gt;                   &lt;numeric&gt;</span></span>
<span><span class="co">## TTTCCTCAGAGCTGGT-1                              NA                          NA</span></span>
<span><span class="co">## CGCGTTTGTCAAACTC-1                              NA                          NA</span></span>
<span><span class="co">## CTACGTCTCAGGCAAG-1                              NA                          NA</span></span>
<span><span class="co">## CGATCGGCACCATGTA-1                              NA                          NA</span></span>
<span><span class="co">## AGTGTCACAAGTTAAG-1                              NA                          NA</span></span>
<span><span class="co">##                    dropletUtils_BarcodeRank_Knee</span></span>
<span><span class="co">##                                        &lt;integer&gt;</span></span>
<span><span class="co">## TTTCCTCAGAGCTGGT-1                             0</span></span>
<span><span class="co">## CGCGTTTGTCAAACTC-1                             0</span></span>
<span><span class="co">## CTACGTCTCAGGCAAG-1                             0</span></span>
<span><span class="co">## CGATCGGCACCATGTA-1                             0</span></span>
<span><span class="co">## AGTGTCACAAGTTAAG-1                             0</span></span>
<span><span class="co">##                    dropletUtils_BarcodeRank_Inflection</span></span>
<span><span class="co">##                                              &lt;integer&gt;</span></span>
<span><span class="co">## TTTCCTCAGAGCTGGT-1                                   0</span></span>
<span><span class="co">## CGCGTTTGTCAAACTC-1                                   0</span></span>
<span><span class="co">## CTACGTCTCAGGCAAG-1                                   0</span></span>
<span><span class="co">## CGATCGGCACCATGTA-1                                   0</span></span>
<span><span class="co">## AGTGTCACAAGTTAAG-1                                   0</span></span></code></pre>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:80%; ">
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
sample
</th>
<th style="text-align:left;">
Barcode
</th>
<th style="text-align:right;">
sum
</th>
<th style="text-align:right;">
detected
</th>
<th style="text-align:right;">
percent.top_50
</th>
<th style="text-align:right;">
percent.top_100
</th>
<th style="text-align:right;">
percent.top_200
</th>
<th style="text-align:right;">
percent.top_500
</th>
<th style="text-align:right;">
mito_sum
</th>
<th style="text-align:right;">
mito_detected
</th>
<th style="text-align:right;">
mito_percent
</th>
<th style="text-align:right;">
total
</th>
<th style="text-align:right;">
dropletUtils_emptyDrops_total
</th>
<th style="text-align:right;">
dropletUtils_emptyDrops_logprob
</th>
<th style="text-align:right;">
dropletUtils_emptyDrops_pvalue
</th>
<th style="text-align:left;">
dropletUtils_emptyDrops_limited
</th>
<th style="text-align:right;">
dropletUtils_emptyDrops_fdr
</th>
<th style="text-align:right;">
dropletUtils_BarcodeRank_Knee
</th>
<th style="text-align:right;">
dropletUtils_BarcodeRank_Inflection
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
TTTCCTCAGAGCTGGT-1
</td>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
TTTCCTCAGAGCTGGT-1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
CGCGTTTGTCAAACTC-1
</td>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
CGCGTTTGTCAAACTC-1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
CTACGTCTCAGGCAAG-1
</td>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
CTACGTCTCAGGCAAG-1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
CGATCGGCACCATGTA-1
</td>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
CGATCGGCACCATGTA-1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AGTGTCACAAGTTAAG-1
</td>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
AGTGTCACAAGTTAAG-1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
</div>
<p><br></p>
<details><summary><b>A summary of all outputs</b></summary><table class="table">
<colgroup>
<col width="27%">
<col width="46%">
<col width="14%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th align="left">QC output</th>
<th align="left">Description</th>
<th align="left">Methods</th>
<th align="left">Package/Tool</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><code>sum</code></td>
<td align="left">Total counts</td>
<td align="left"><code><a href="../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link">scater</a></td>
</tr>
<tr class="even">
<td align="left"><code>detected</code></td>
<td align="left">Total features</td>
<td align="left"><code><a href="../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link">scater</a></td>
</tr>
<tr class="odd">
<td align="left"><code>percent_top</code></td>
<td align="left">% Expression coming from top features</td>
<td align="left"><code><a href="../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link">scater</a></td>
</tr>
<tr class="even">
<td align="left"><code>subsets_*</code></td>
<td align="left">sum, detected, percent_top calculated on specified gene
list</td>
<td align="left"><code><a href="../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link">scater</a></td>
</tr>
<tr class="odd">
<td align="left"><code>dropletUtils_emptyDrops_total</code></td>
<td align="left">Total counts</td>
<td align="left"><code><a href="../reference/runEmptyDrops.html">runEmptyDrops()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link">DropletUtils</a></td>
</tr>
<tr class="even">
<td align="left"><code>dropletUtils_emptyDrops_logprob</code></td>
<td align="left">The log-probability of droplet being empty</td>
<td align="left"><code><a href="../reference/runEmptyDrops.html">runEmptyDrops()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link">DropletUtils</a></td>
</tr>
<tr class="odd">
<td align="left"><code>dropletUtils_emptyDrops_pvalue</code></td>
<td align="left">Monte Carlo p-value of droplet being empty</td>
<td align="left"><code><a href="../reference/runEmptyDrops.html">runEmptyDrops()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link">DropletUtils</a></td>
</tr>
<tr class="even">
<td align="left"><code>dropletUtils_emptyDrops_limited</code></td>
<td align="left">Whether a lower p-value could be obtained by increasing
niters</td>
<td align="left"><code><a href="../reference/runEmptyDrops.html">runEmptyDrops()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link">DropletUtils</a></td>
</tr>
<tr class="odd">
<td align="left"><code>dropletUtils_emptyDrops_fdr</code></td>
<td align="left">p-value of droplet being empty, corrected for false
detection rate</td>
<td align="left"><code><a href="../reference/runEmptyDrops.html">runEmptyDrops()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link">DropletUtils</a></td>
</tr>
<tr class="even">
<td align="left"><code>dropletUtils_BarcodeRank_Knee</code></td>
<td align="left">Whether total UMI count value is higher than knee
point</td>
<td align="left"><code><a href="../reference/runBarcodeRankDrops.html">runBarcodeRankDrops()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link">DropletUtils</a></td>
</tr>
<tr class="odd">
<td align="left"><code>dropletUtils_BarcodeRank_Inflection</code></td>
<td align="left">Whether total UMI count value is higher than inflection
point</td>
<td align="left"><code><a href="../reference/runBarcodeRankDrops.html">runBarcodeRankDrops()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link">DropletUtils</a></td>
</tr>
</tbody>
</table></details>
</div>
<div class="section level3">
<h3 id="running-individual-qc-methods">Running individual QC methods<a class="anchor" aria-label="anchor" href="#running-individual-qc-methods"></a>
</h3>
<p>Instead of running all quality control methods on the dataset at
once, users may elect to execute QC methods individually. The parameters
as well as the outputs to individual QC functions are described in
detail as follows:</p>
<details><summary><b>runEmptyDrops</b></summary><p>It is crucial to distinguish the data occurring from real cells and
empty droplets containing ambient RNA. SCTK employs the <a href="https://rdrr.io/github/MarioniLab/DropletUtils/man/emptyDrops.html" class="external-link">EmptyDrops</a>
algorithm from the <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link">DropletUtils</a>
package to test for empty droplets. The wrapper function
<code><a href="../reference/runEmptyDrops.html">runEmptyDrops()</a></code> can be used to separately run the
EmptyDrops algorithm on its own.</p>
<ul>
<li>
<code>lower</code> is the lower bound of the total UMI count, in
which all barcodes below the lower bound are assumed to be empty
droplets.</li>
<li>
<code>niters</code> is the number of iterations the function will
run for the calculation.</li>
<li>
<code>testAmbient</code> indicates whether results should be
returned for barcodes that have a total UMI count below what is
specified in <code>lower</code>.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runEmptyDrops.html">runEmptyDrops</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc4k.droplet</span>,</span>
<span>  useAssay <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  lower <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  niters <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="op">)</span></span></code></pre></div>
</details><details><summary><b>runBarcodeRankDrops</b></summary><p><a href="https://rdrr.io/bioc/DropletUtils/man/barcodeRanks.html" class="external-link">BarcodeRanks</a>
from the <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link">DropletUtils</a>
package computes barcode rank statistics and identifies the knee and
inflection points on the total count curve. The knee and inflection
points on the curve represent the difference between empty droplets and
cell-containing droplets with much more RNA. The wrapper function
<code><a href="../reference/runBarcodeRankDrops.html">runBarcodeRankDrops()</a></code> can be used to separately run the
BarcodeRanks algorithm on its own.</p>
<ul>
<li>
<code>lower</code> is the lower bound of the total UMI count, in
which all barcodes below the lower bound are assumed to be empty
droplets.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runBarcodeRankDrops.html">runBarcodeRankDrops</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc4k.droplet</span>,</span>
<span>  useAssay <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  fitBounds <span class="op">=</span> <span class="cn">NULL</span>, df <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level3">
<h3 id="plotting-qc-metrics">Plotting QC metrics<a class="anchor" aria-label="anchor" href="#plotting-qc-metrics"></a>
</h3>
<p>Upon running <code><a href="../reference/runDropletQC.html">runDropletQC()</a></code> or any of the individual
droplet QC functions, the QC outputs will need to be plotted. For each
QC method, SCTK provides a specialized plotting function.</p>
<details><summary><b>EmptyDrops</b></summary><p>The wrapper function <code><a href="../reference/plotEmptyDropsResults.html">plotEmptyDropsResults()</a></code> can be used
to plot the results from the EmptyDrops algorithm. This will visualize
the empty droplets, by plotting the total UMI counts against the log
probability for each barcode.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">emptyDropsResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotEmptyDropsResults.html">plotEmptyDropsResults</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc4k.droplet</span>,</span>
<span>  axisLabelSize <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fdrCutoff <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  dotSize <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  defaultTheme <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">emptyDropsResults</span><span class="op">$</span><span class="va">scatterEmptyDrops</span></span></code></pre></div>
<p><img src="cnsl_dropletqc_files/figure-html/scatter_emptydrops-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Data points are colored by FDR values, where we see a small portion
of the dataset contains barcodes that do not meet the threshold.</p>
</details><details><summary><b>BarcodeRanks</b></summary><p>The wrapper function <code><a href="../reference/plotBarcodeRankScatter.html">plotBarcodeRankScatter()</a></code> can be
used to plot the results from the BarcodeRanks algorithm.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBarcodeRankScatter.html">plotBarcodeRankScatter</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc4k.droplet</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"BarcodeRanks Rank Plot"</span>,</span>
<span>  legendSize <span class="op">=</span> <span class="fl">14</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="cnsl_dropletqc_files/figure-html/run_kneeplot_barcoderank-1.png" width="768" style="display: block; margin: auto;"></p>
<p>The total UMI count of each barcode is plotted against its rank. We
can see a steep dropoff of UMI counts around the inflection point, where
we see a presumed separation between cell-containing and empty
droplets.</p>
</details>
</div>
<div class="section level3">
<h3 id="filtering-the-dataset">Filtering the dataset<a class="anchor" aria-label="anchor" href="#filtering-the-dataset"></a>
</h3>
<p><code>SingleCellExperiment</code> objects can be subset by its
<code>colData</code> using <code><a href="../reference/subsetSCECols.html">subsetSCECols()</a></code>. The
<code>colData</code> parameter takes in a character vector of
expression(s) which will be used to identify a subset of cells using
variables found in the <code>colData</code> of the
<code>SingleCellExperiment</code> object. For example, if <code>x</code>
is a numeric vector in <code>colData</code>, then setting
<code>colData = "x &lt; 5"</code> will return a
<code>SingleCellExperiment</code> object where all columns (cells) meet
the condition that <code>x</code> is less than 5. The <code>index</code>
parameter takes in a numeric vector of indices which should be kept,
while <code>bool</code> takes in a logical vector of <code>TRUE</code>
or <code>FALSE</code> which should be of the same length as the number
of columns (cells) in the <code>SingleCellExperiment</code> object.
Please refer to our <a href="filtering.html"><em>Filtering</em>
documentation</a> for detail.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Before filtering:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 33694 60000</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span>, colData <span class="op">=</span> <span class="st">'dropletUtils_BarcodeRank_Inflection == 1'</span><span class="op">)</span></span>
<span><span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span>, colData <span class="op">=</span> <span class="st">'!is.na(pbmc4k.droplet$dropletUtils_emptyDrops_fdr)'</span><span class="op">)</span></span>
<span><span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span>, colData <span class="op">=</span> <span class="st">'pbmc4k.droplet$dropletUtils_emptyDrops_fdr &lt; 0.01'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#After filtering:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 33694   315</span></span></code></pre>
<p>We can compare the average total UMI counts per cell before and after
cell filtration:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSCEViolinColData.html">plotSCEViolinColData</a></span><span class="op">(</span><span class="va">pbmc4k.droplet.prefilt</span>, coldata <span class="op">=</span> <span class="st">"sum"</span>, summary <span class="op">=</span> <span class="st">"mean"</span>, title <span class="op">=</span> <span class="st">"Pre-filter"</span>, ylab <span class="op">=</span> <span class="st">"Total counts"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSCEViolinColData.html">plotSCEViolinColData</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span>, coldata <span class="op">=</span> <span class="st">"sum"</span>, summary <span class="op">=</span> <span class="st">"mean"</span>, title <span class="op">=</span> <span class="st">"Post-filter"</span>, ylab <span class="op">=</span> <span class="st">"Total counts"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="cnsl_dropletqc_files/figure-html/Count_Violin_after_filtering-1.png" width="576" style="display: block; margin: auto;"></p>
<p><br></p>
<blockquote>
<p>For performing QC on cell-filtered count matrix with SCTK, please
refer to our <a href="cnsl_cellqc.html"><em>Cell QC</em>
documentation</a>.</p>
</blockquote>
<p><br></p>
 <details><summary><b>Session Information</b></summary><div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Sequoia 15.4</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] cowplot_1.1.3               BiocFileCache_2.14.0       </span></span>
<span><span class="co">##  [3] dbplyr_2.5.0                dplyr_1.1.4                </span></span>
<span><span class="co">##  [5] singleCellTK_2.18.0         DelayedArray_0.32.0        </span></span>
<span><span class="co">##  [7] SparseArray_1.6.1           S4Arrays_1.6.0             </span></span>
<span><span class="co">##  [9] abind_1.4-8                 Matrix_1.7-2               </span></span>
<span><span class="co">## [11] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0</span></span>
<span><span class="co">## [13] Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">## [15] GenomeInfoDb_1.42.3         IRanges_2.40.1             </span></span>
<span><span class="co">## [17] S4Vectors_0.44.0            BiocGenerics_0.52.0        </span></span>
<span><span class="co">## [19] MatrixGenerics_1.18.1       matrixStats_1.5.0          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3        rstudioapi_0.17.1        </span></span>
<span><span class="co">##   [3] jsonlite_1.8.9            shape_1.4.6.1            </span></span>
<span><span class="co">##   [5] magrittr_2.0.3            ggbeeswarm_0.7.2         </span></span>
<span><span class="co">##   [7] farver_2.1.2              rmarkdown_2.29           </span></span>
<span><span class="co">##   [9] GlobalOptions_0.1.2       fs_1.6.5                 </span></span>
<span><span class="co">##  [11] zlibbioc_1.52.0           ragg_1.3.3               </span></span>
<span><span class="co">##  [13] vctrs_0.6.5               memoise_2.0.1            </span></span>
<span><span class="co">##  [15] DelayedMatrixStats_1.28.1 htmltools_0.5.8.1        </span></span>
<span><span class="co">##  [17] curl_6.2.0                BiocNeighbors_2.0.1      </span></span>
<span><span class="co">##  [19] Rhdf5lib_1.28.0           rhdf5_2.50.2             </span></span>
<span><span class="co">##  [21] sass_0.4.9                bslib_0.9.0              </span></span>
<span><span class="co">##  [23] htmlwidgets_1.6.4         desc_1.4.3               </span></span>
<span><span class="co">##  [25] plyr_1.8.9                cachem_1.1.0             </span></span>
<span><span class="co">##  [27] lifecycle_1.0.4           iterators_1.0.14         </span></span>
<span><span class="co">##  [29] pkgconfig_2.0.3           rsvd_1.0.5               </span></span>
<span><span class="co">##  [31] R6_2.6.1                  fastmap_1.2.0            </span></span>
<span><span class="co">##  [33] GenomeInfoDbData_1.2.13   clue_0.3-66              </span></span>
<span><span class="co">##  [35] digest_0.6.37             colorspace_2.1-1         </span></span>
<span><span class="co">##  [37] AnnotationDbi_1.68.0      scater_1.34.0            </span></span>
<span><span class="co">##  [39] dqrng_0.4.1               irlba_2.3.5.1            </span></span>
<span><span class="co">##  [41] textshaping_1.0.0         RSQLite_2.3.9            </span></span>
<span><span class="co">##  [43] beachmat_2.22.0           labeling_0.4.3           </span></span>
<span><span class="co">##  [45] filelock_1.0.3            httr_1.4.7               </span></span>
<span><span class="co">##  [47] compiler_4.4.2            bit64_4.6.0-1            </span></span>
<span><span class="co">##  [49] withr_3.0.2               doParallel_1.0.17        </span></span>
<span><span class="co">##  [51] BiocParallel_1.40.0       viridis_0.6.5            </span></span>
<span><span class="co">##  [53] DBI_1.2.3                 HDF5Array_1.34.0         </span></span>
<span><span class="co">##  [55] R.utils_2.12.3            rjson_0.2.23             </span></span>
<span><span class="co">##  [57] tools_4.4.2               vipor_0.4.7              </span></span>
<span><span class="co">##  [59] beeswarm_0.4.0            R.oo_1.27.0              </span></span>
<span><span class="co">##  [61] glue_1.8.0                rhdf5filters_1.18.0      </span></span>
<span><span class="co">##  [63] grid_4.4.2                cluster_2.1.8            </span></span>
<span><span class="co">##  [65] reshape2_1.4.4            generics_0.1.3           </span></span>
<span><span class="co">##  [67] gtable_0.3.6              R.methodsS3_1.8.2        </span></span>
<span><span class="co">##  [69] tidyr_1.3.1               xml2_1.3.6               </span></span>
<span><span class="co">##  [71] BiocSingular_1.22.0       ScaledMatrix_1.14.0      </span></span>
<span><span class="co">##  [73] XVector_0.46.0            ggrepel_0.9.6            </span></span>
<span><span class="co">##  [75] foreach_1.5.2             pillar_1.10.1            </span></span>
<span><span class="co">##  [77] stringr_1.5.1             limma_3.62.2             </span></span>
<span><span class="co">##  [79] circlize_0.4.16           GSVAdata_1.42.0          </span></span>
<span><span class="co">##  [81] lattice_0.22-6            bit_4.5.0.1              </span></span>
<span><span class="co">##  [83] annotate_1.84.0           tidyselect_1.2.1         </span></span>
<span><span class="co">##  [85] ComplexHeatmap_2.22.0     locfit_1.5-9.11          </span></span>
<span><span class="co">##  [87] Biostrings_2.74.1         scuttle_1.16.0           </span></span>
<span><span class="co">##  [89] knitr_1.49                gridExtra_2.3            </span></span>
<span><span class="co">##  [91] edgeR_4.4.2               svglite_2.1.3            </span></span>
<span><span class="co">##  [93] xfun_0.50                 statmod_1.5.0            </span></span>
<span><span class="co">##  [95] DropletUtils_1.26.0       eds_1.8.0                </span></span>
<span><span class="co">##  [97] stringi_1.8.4             UCSC.utils_1.2.0         </span></span>
<span><span class="co">##  [99] yaml_2.3.10               kableExtra_1.4.0         </span></span>
<span><span class="co">## [101] evaluate_1.0.3            codetools_0.2-20         </span></span>
<span><span class="co">## [103] tibble_3.2.1              graph_1.84.1             </span></span>
<span><span class="co">## [105] cli_3.6.4                 xtable_1.8-4             </span></span>
<span><span class="co">## [107] reticulate_1.40.0         systemfonts_1.2.1        </span></span>
<span><span class="co">## [109] munsell_0.5.1             jquerylib_0.1.4          </span></span>
<span><span class="co">## [111] Rcpp_1.0.14               png_0.1-8                </span></span>
<span><span class="co">## [113] XML_3.99-0.18             parallel_4.4.2           </span></span>
<span><span class="co">## [115] pkgdown_2.1.1             ggplot2_3.5.1            </span></span>
<span><span class="co">## [117] blob_1.2.4                sparseMatrixStats_1.18.0 </span></span>
<span><span class="co">## [119] viridisLite_0.4.2         GSEABase_1.68.0          </span></span>
<span><span class="co">## [121] scales_1.3.0              purrr_1.0.4              </span></span>
<span><span class="co">## [123] crayon_1.5.3              GetoptLong_1.0.5         </span></span>
<span><span class="co">## [125] rlang_1.1.5               KEGGREST_1.46.0</span></span></code></pre>
</details>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
