<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quality Control for cell-level data • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Quality Control for cell-level data">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.13.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="cnsl_cellqc_files/kePrint-0.0.1/kePrint.js"></script><link href="cnsl_cellqc_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quality Control for cell-level data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/cnsl_cellqc.Rmd" class="external-link"><code>vignettes/articles/cnsl_cellqc.Rmd</code></a></small>
      <div class="hidden name"><code>cnsl_cellqc.Rmd</code></div>

    </div>

    
    
<style>
body {
text-align: justify}
</style>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Performing comprehensive quality control (QC) is necessary to remove
poor quality cells for downstream analysis of single-cell RNA sequencing
(scRNA-seq) data. Therefore, assessment of the data is required, for
which various QC algorithms have been developed. In singleCellTK (SCTK),
we have written convenience functions for several of these tools. In
this guide, we will demonstrate how to use these functions to perform
quality control on cell data. (For definition of cell data, please refer
<a href="cmd_qc.html#introduction-1">this documentation</a>.)</p>
<p>The package can be loaded using the <code>library</code> command.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="running-qc-on-cell-filtered-data">Running QC on cell-filtered data<a class="anchor" aria-label="anchor" href="#running-qc-on-cell-filtered-data"></a>
</h2>
<div class="section level3">
<h3 id="load-pbmc-data-from-10x">Load PBMC data from 10X<a class="anchor" aria-label="anchor" href="#load-pbmc-data-from-10x"></a>
</h3>
<p>We will use a filtered form of the PBMC 3K and 6K dataset from the
package <a href="http://bioconductor.org/packages/release/data/experiment/html/TENxPBMCData.html" class="external-link">TENxPBMCData</a>,
which is available from the <code><a href="../reference/importExampleData.html">importExampleData()</a></code> function.
We will combine these datasets together into a single <a href="https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>
object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="va">pbmc6k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"pbmc6k"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu">BiocGenerics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pbmc3k</span>, <span class="va">pbmc6k</span><span class="op">)</span></span>
<span><span class="va">sample.vector</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span></span></code></pre></div>
<p>SCTK also supports the importing of single-cell data from the
following platforms: <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">10X
CellRanger</a>, <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md" class="external-link">STARSolo</a>,
<a href="https://www.kallistobus.tools/" class="external-link">BUSTools</a>, <a href="https://github.com/ambrosejcarr/seqc" class="external-link">SEQC</a>, <a href="https://github.com/hms-dbmi/dropEst" class="external-link">DropEST</a>, <a href="https://salmon.readthedocs.io/en/latest/alevin.html" class="external-link">Alevin</a>,
as well as dataset already stored in <a href="https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>
object and <a href="https://github.com/theislab/anndata" class="external-link">AnnData</a>
object. To load your own input data, please refer
<code>Function Reference for pre-processing tools</code> under
<code>Console Analysis</code> section in <a href="import_data.html">Import data into SCTK</a> for detailed
instruction.</p>
</div>
<div class="section level3">
<h3 id="run-2d-embedding">Run 2D embedding<a class="anchor" aria-label="anchor" href="#run-2d-embedding"></a>
</h3>
<p>SCTK utilizes 2D embedding techniques such as TSNE and UMAP for
visualizing single-cell data. Users can modify the dimensions by
adjusting the parameters within the function. The <code>logNorm</code>
parameter should be set to <code>TRUE</code> for normalization prior to
generating the 2D embedding.</p>
<p>The <code>sample</code> parameter may be specified if multiple
samples exist in the <code>SingleCellExperiment</code> object. Here, we
will use the sample vector stored in the <code>colData</code> slot of
the <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runQuickUMAP</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-cellqc">Run CellQC<a class="anchor" aria-label="anchor" href="#run-cellqc"></a>
</h3>
<p>All of the droplet-based QC algorithms are able to be run under the
wrapper function <code><a href="../reference/runCellQC.html">runCellQC()</a></code>. By default all possible QC
algorithms will be run.</p>
<p>Users may set a <code>sample</code> parameter if need to compare
between multiple samples. Here, we will use the sample vector stored in
the <code>SingleCellExperiment</code> object.</p>
<p>If users wishes, a list of gene sets can be applied to the function
to determine the expression of a set of specific genes. A gene list
imported into the <code>SingleCellExperiment</code> object using
<code>importGeneSets*</code> functions can be set as
<code>collectionName</code>. Additionally, a pre-made list of genes can
be used to determine the level of gene expression per cell. A list
containing gene identifiers may be set as <code>geneSetList</code>, or
the user may instead use the <code>geneSetCollection</code> parameter to
supply a <code>GeneSetCollection</code> object from the <a href="https://bioconductor.org/packages/release/bioc/html/GSEABase.html" class="external-link">GSEABase</a>
package. Please also refer to <a href="import_genesets.html"><em>Import
Genesets</em> documentation</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importGeneSetsFromGMT.html">importGeneSetsFromGMT</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, collectionName <span class="op">=</span> <span class="st">"mito"</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mito_subset.gmt"</span>, package <span class="op">=</span> <span class="st">"singleCellTK"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCellQC.html">runCellQC</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, </span>
<span>                           algorithms <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"QCMetrics"</span>, <span class="st">"scrublet"</span>, <span class="st">"scDblFinder"</span>, <span class="st">"cxds"</span>, <span class="st">"bcds"</span>, <span class="st">"cxds_bcds_hybrid"</span>, <span class="st">"doubletFinder"</span>, <span class="st">"decontX"</span>, <span class="st">"soupX"</span><span class="op">)</span>, </span>
<span>                           sample <span class="op">=</span> <span class="va">sample.vector</span>, </span>
<span>                           collectionName <span class="op">=</span> <span class="st">"mito"</span><span class="op">)</span></span></code></pre></div>
<p>Users can also specify <code>mitoRef</code>, <code>mitoIDType</code>
and <code>mitoGeneLocation</code> arguments in <code>runCellQC</code>
function to quantify mitochondrial gene expression without the need to
import gene sets. For the details about these arguments, please refer to
<code>runCellQC</code> and <code><a href="../reference/runPerCellQC.html">runPerCellQC()</a></code> .</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCellQC.html">runCellQC</a></span><span class="op">(</span><span class="va">pbmc.combined</span>,</span>
<span>                           algorithms <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"QCMetrics"</span>, <span class="st">"scrublet"</span>, <span class="st">"scDblFinder"</span>, <span class="st">"cxds"</span>, <span class="st">"bcds"</span>, <span class="st">"cxds_bcds_hybrid"</span>, <span class="st">"doubletFinder"</span>, <span class="st">"decontX"</span>, <span class="st">"soupX"</span><span class="op">)</span>,</span>
<span>                           sample <span class="op">=</span> <span class="va">sample.vector</span>,</span>
<span>                           mitoRef <span class="op">=</span> <span class="st">"human"</span>, mitoIDType <span class="op">=</span> <span class="st">"symbol"</span>, mitoGeneLocation <span class="op">=</span> <span class="st">"rownames"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Preprocessing...</span></span>
<span><span class="co">## Simulating doublets...</span></span>
<span><span class="co">## Embedding transcriptomes using PCA...</span></span>
<span><span class="co">## Calculating doublet scores...</span></span>
<span><span class="co">## Automatically set threshold at doublet score = 0.41</span></span>
<span><span class="co">## Detected doublet rate = 1.3%</span></span>
<span><span class="co">## Estimated detectable doublet fraction = 42.1%</span></span>
<span><span class="co">## Overall doublet rate:</span></span>
<span><span class="co">##  Expected   = 10.0%</span></span>
<span><span class="co">##  Estimated  = 3.0%</span></span>
<span><span class="co">## Elapsed time: 2.6 seconds</span></span>
<span><span class="co">## Preprocessing...</span></span>
<span><span class="co">## Simulating doublets...</span></span>
<span><span class="co">## Embedding transcriptomes using PCA...</span></span>
<span><span class="co">## Calculating doublet scores...</span></span>
<span><span class="co">## Automatically set threshold at doublet score = 0.38</span></span>
<span><span class="co">## Detected doublet rate = 2.4%</span></span>
<span><span class="co">## Estimated detectable doublet fraction = 50.4%</span></span>
<span><span class="co">## Overall doublet rate:</span></span>
<span><span class="co">##  Expected   = 10.0%</span></span>
<span><span class="co">##  Estimated  = 4.8%</span></span>
<span><span class="co">## Elapsed time: 6.4 seconds</span></span></code></pre>
<p><br> If users choose to only run a specific set of algorithms, they
can specify which to run with the <code>algorithms</code> parameter. By
default, the <code><a href="../reference/runCellQC.html">runCellQC()</a></code> will run <code>"QCMetrics"</code>,
<code>"scDblFinder"</code>, <code>"cxds"</code>, <code>"bcds"</code>,
<code>"cxds_bcds_hybrid"</code>, <code>"decontX"</code> and
<code>"soupX"</code> algorithms by default. Besides,
<code>"scrublet"</code> and <code>"doubletFinder"</code> are supported
if our users want to run them.</p>
<p>After running QC functions with SCTK, the output will be stored in
the <code>colData</code> slot of the <code>SingleCellExperiment</code>
object.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:80%; ">
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Sample
</th>
<th style="text-align:left;">
Barcode
</th>
<th style="text-align:left;">
Sequence
</th>
<th style="text-align:right;">
Library
</th>
<th style="text-align:left;">
Cell_ranger_version
</th>
<th style="text-align:left;">
Tissue_status
</th>
<th style="text-align:left;">
Barcode_type
</th>
<th style="text-align:left;">
Chemistry
</th>
<th style="text-align:left;">
Sequence_platform
</th>
<th style="text-align:left;">
Individual
</th>
<th style="text-align:left;">
Date_published
</th>
<th style="text-align:left;">
sample
</th>
<th style="text-align:right;">
sum
</th>
<th style="text-align:right;">
detected
</th>
<th style="text-align:right;">
percent.top_50
</th>
<th style="text-align:right;">
percent.top_100
</th>
<th style="text-align:right;">
percent.top_200
</th>
<th style="text-align:right;">
percent.top_500
</th>
<th style="text-align:right;">
mito_sum
</th>
<th style="text-align:right;">
mito_detected
</th>
<th style="text-align:right;">
mito_percent
</th>
<th style="text-align:right;">
total
</th>
<th style="text-align:right;">
scrublet_score
</th>
<th style="text-align:left;">
scrublet_call
</th>
<th style="text-align:left;">
scDblFinder_sample
</th>
<th style="text-align:left;">
scDblFinder_doublet_call
</th>
<th style="text-align:right;">
scDblFinder_doublet_score
</th>
<th style="text-align:right;">
scDblFinder_weighted
</th>
<th style="text-align:right;">
scDblFinder_cxds_score
</th>
<th style="text-align:right;">
doubletFinder_doublet_score_resolution_1.5
</th>
<th style="text-align:left;">
doubletFinder_doublet_label_resolution_1.5
</th>
<th style="text-align:right;">
scds_cxds_score
</th>
<th style="text-align:left;">
scds_cxds_call
</th>
<th style="text-align:right;">
scds_bcds_score
</th>
<th style="text-align:left;">
scds_bcds_call
</th>
<th style="text-align:right;">
scds_hybrid_score
</th>
<th style="text-align:left;">
scds_hybrid_call
</th>
<th style="text-align:right;">
decontX_contamination
</th>
<th style="text-align:left;">
decontX_clusters
</th>
<th style="text-align:right;">
soupX_nUMIs
</th>
<th style="text-align:left;">
soupX_clusters
</th>
<th style="text-align:right;">
soupX_contamination
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
pbmc3k_AAACATACAACCAC-1
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:left;">
AAACATACAACCAC-1
</td>
<td style="text-align:left;">
AAACATACAACCAC
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
v1.1.0
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
GemCode
</td>
<td style="text-align:left;">
Chromium_v1
</td>
<td style="text-align:left;">
NextSeq500
</td>
<td style="text-align:left;">
HealthyDonor2
</td>
<td style="text-align:left;">
2016-05-26
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:right;">
2421
</td>
<td style="text-align:right;">
781
</td>
<td style="text-align:right;">
47.74886
</td>
<td style="text-align:right;">
63.27964
</td>
<td style="text-align:right;">
74.96902
</td>
<td style="text-align:right;">
88.39323
</td>
<td style="text-align:right;">
73
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
3.015283
</td>
<td style="text-align:right;">
2421
</td>
<td style="text-align:right;">
0.1606218
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
0.0573566
</td>
<td style="text-align:right;">
0.7966514
</td>
<td style="text-align:right;">
0.1272208
</td>
<td style="text-align:right;">
0.0138889
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
22794.14
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
0.0149839
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
0.2109824
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
0.0319956
</td>
<td style="text-align:left;">
pbmc3k-1
</td>
<td style="text-align:right;">
2421
</td>
<td style="text-align:left;">
pbmc3k-6
</td>
<td style="text-align:right;">
0.059
</td>
</tr>
<tr>
<td style="text-align:left;">
pbmc3k_AAACATTGAGCTAC-1
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:left;">
AAACATTGAGCTAC-1
</td>
<td style="text-align:left;">
AAACATTGAGCTAC
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
v1.1.0
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
GemCode
</td>
<td style="text-align:left;">
Chromium_v1
</td>
<td style="text-align:left;">
NextSeq500
</td>
<td style="text-align:left;">
HealthyDonor2
</td>
<td style="text-align:left;">
2016-05-26
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:right;">
4903
</td>
<td style="text-align:right;">
1352
</td>
<td style="text-align:right;">
45.50275
</td>
<td style="text-align:right;">
61.02386
</td>
<td style="text-align:right;">
71.81318
</td>
<td style="text-align:right;">
82.62288
</td>
<td style="text-align:right;">
186
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
3.793596
</td>
<td style="text-align:right;">
4903
</td>
<td style="text-align:right;">
0.1089109
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
0.4663909
</td>
<td style="text-align:right;">
0.5632351
</td>
<td style="text-align:right;">
0.3518394
</td>
<td style="text-align:right;">
0.1527778
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
35689.05
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
0.9710556
</td>
<td style="text-align:left;">
Doublet
</td>
<td style="text-align:right;">
0.8471188
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
0.1196180
</td>
<td style="text-align:left;">
pbmc3k-2
</td>
<td style="text-align:right;">
4903
</td>
<td style="text-align:left;">
pbmc3k-9
</td>
<td style="text-align:right;">
0.059
</td>
</tr>
</tbody>
</table>
</div>
<p><br></p>
<details><summary><b>A summary of all outputs</b></summary><table class="table">
<colgroup>
<col width="29%">
<col width="45%">
<col width="13%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th>QC output</th>
<th align="left">Description</th>
<th align="left">Methods</th>
<th align="left">Package/Tool</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>sum</code></td>
<td align="left">Total counts</td>
<td align="left"><code><a href="../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link">scater</a></td>
</tr>
<tr class="even">
<td><code>detected</code></td>
<td align="left">Total features</td>
<td align="left"><code><a href="../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link">scater</a></td>
</tr>
<tr class="odd">
<td><code>percent_top</code></td>
<td align="left">% Expression coming from top features</td>
<td align="left"><code><a href="../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link">scater</a></td>
</tr>
<tr class="even">
<td><code>subsets_*</code></td>
<td align="left">sum, detected, percent_top calculated on specified gene
list</td>
<td align="left"><code><a href="../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link">scater</a></td>
</tr>
<tr class="odd">
<td><code>scrublet_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../reference/runScrublet.html">runScrublet()</a></code></td>
<td align="left"><a href="https://github.com/swolock/scrublet" class="external-link">scrublet</a></td>
</tr>
<tr class="even">
<td><code>scrublet_call</code></td>
<td align="left">Doublet classification based on threshold</td>
<td align="left"><code><a href="../reference/runScrublet.html">runScrublet()</a></code></td>
<td align="left"><a href="https://github.com/swolock/scrublet" class="external-link">scrublet</a></td>
</tr>
<tr class="odd">
<td><code>scDblFinder_doublet_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../reference/runScDblFinder.html">runScDblFinder()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scDblFinder.html" class="external-link">scDblFinder</a></td>
</tr>
<tr class="even">
<td><code>doubletFinder_doublet_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../reference/runDoubletFinder.html">runDoubletFinder()</a></code></td>
<td align="left"><a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder" class="external-link">DoubletFinder</a></td>
</tr>
<tr class="odd">
<td><code>doubletFinder_doublet_label_resolution</code></td>
<td align="left">Doublet classification based on threshold</td>
<td align="left"><code><a href="../reference/runDoubletFinder.html">runDoubletFinder()</a></code></td>
<td align="left"><a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder" class="external-link">DoubletFinder</a></td>
</tr>
<tr class="even">
<td><code>scds_cxds_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../reference/runCxds.html">runCxds()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html" class="external-link">SCDS</a></td>
</tr>
<tr class="odd">
<td><code>scds_cxds_call</code></td>
<td align="left">Doublet classification based on threshold</td>
<td align="left"><code><a href="../reference/runCxds.html">runCxds()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html" class="external-link">SCDS</a></td>
</tr>
<tr class="even">
<td><code>scds_bcds_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../reference/runBcds.html">runBcds()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html" class="external-link">SCDS</a></td>
</tr>
<tr class="odd">
<td><code>scds_bcds_call</code></td>
<td align="left">Doublet classification based on threshold</td>
<td align="left"><code><a href="../reference/runBcds.html">runBcds()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html" class="external-link">SCDS</a></td>
</tr>
<tr class="even">
<td><code>scds_hybrid_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../reference/runCxdsBcdsHybrid.html">runCxdsBcdsHybrid()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html" class="external-link">SCDS</a></td>
</tr>
<tr class="odd">
<td><code>scds_hybrid_call</code></td>
<td align="left">Doublet classification based on threshold</td>
<td align="left"><code><a href="../reference/runCxdsBcdsHybrid.html">runCxdsBcdsHybrid()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html" class="external-link">SCDS</a></td>
</tr>
<tr class="even">
<td><code>decontX_contamination</code></td>
<td align="left">Ambient RNA contamination</td>
<td align="left"><code><a href="../reference/runDecontX.html">runDecontX()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/celda.html" class="external-link">celda</a></td>
</tr>
<tr class="odd">
<td><code>decontX_clusters</code></td>
<td align="left">Clusters determined in dataset based on underlying
algorithm</td>
<td align="left"><code><a href="../reference/runDecontX.html">runDecontX()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/celda.html" class="external-link">celda</a></td>
</tr>
<tr class="even">
<td><code>soupX_nUMIs</code></td>
<td align="left">Total number of UMI per cell</td>
<td align="left"><code><a href="../reference/runSoupX.html">runSoupX()</a></code></td>
<td align="left"><a href="https://github.com/constantAmateur/SoupX" class="external-link">SoupX</a></td>
</tr>
<tr class="odd">
<td><code>soupX_clusters</code></td>
<td align="left">Quick clustering label if clustering not provided by
users</td>
<td align="left"><code><a href="../reference/runSoupX.html">runSoupX()</a></code></td>
<td align="left"><a href="https://rdrr.io/bioc/scran/man/quickCluster.html" class="external-link">scran</a></td>
</tr>
<tr class="even">
<td><code>soupX_contamination</code></td>
<td align="left">Ambient RNA contamination</td>
<td align="left"><code><a href="../reference/runSoupX.html">runSoupX()</a></code></td>
<td align="left"><a href="https://github.com/constantAmateur/SoupX" class="external-link">SoupX</a></td>
</tr>
</tbody>
</table></details><p><br></p>
<p>The names of the 2D embedding and dimension reduction matrices are
stored in the <code>reducedDims</code> slot of the
<code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of length 8</span></span>
<span><span class="co">## names(8): UMAP scrublet_TSNE ... SoupX_UMAP_pbmc3k SoupX_UMAP_pbmc6k</span></span></code></pre>
<div class="section level4">
<h4 id="generating-a-summary-statistic-table">Generating a summary statistic table<a class="anchor" aria-label="anchor" href="#generating-a-summary-statistic-table"></a>
</h4>
<p>The function <code><a href="../reference/sampleSummaryStats.html">sampleSummaryStats()</a></code> may be used to
generate a table containing the mean and median of the data per sample,
which is stored within the <code>qc_table</code> table under
<code>metadata</code>. The table can then be returned using
<code>getSampleSummaryStatsTable</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sampleSummaryStats.html">sampleSummaryStats</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="va">sample.vector</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/getSampleSummaryStatsTable.html">getSampleSummaryStatsTable</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, statsName <span class="op">=</span> <span class="st">"qc_table"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                           pbmc3k  pbmc6k All Samples</span></span>
<span><span class="co">## Number of Cells          2700.00 5419.00     8119.00</span></span>
<span><span class="co">## Mean counts              2366.90 2027.60     2140.50</span></span>
<span><span class="co">## Median counts            2197.00 1873.00     1988.00</span></span>
<span><span class="co">## Mean features detected    846.99  748.06      780.96</span></span>
<span><span class="co">## Median features detected  817.00  716.00      750.00</span></span></code></pre>
<p>If users choose to generate a table for all QC metrics generated
through <code><a href="../reference/runCellQC.html">runCellQC()</a></code>, they may set the <code>simple</code>
parameter to <code>FALSE</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sampleSummaryStats.html">sampleSummaryStats</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="va">sample.vector</span>, simple <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/getSampleSummaryStatsTable.html">getSampleSummaryStatsTable</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, statsName <span class="op">=</span> <span class="st">"qc_table"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                           pbmc3k    pbmc6k</span></span>
<span><span class="co">## Number of Cells                                        2700.0000 5419.0000</span></span>
<span><span class="co">## Mean counts                                            2366.9000 2027.6000</span></span>
<span><span class="co">## Median counts                                          2197.0000 1873.0000</span></span>
<span><span class="co">## Mean features detected                                  846.9900  748.0600</span></span>
<span><span class="co">## Median features detected                                817.0000  716.0000</span></span>
<span><span class="co">## Scrublet - Number of doublets                             0.0000    0.0000</span></span>
<span><span class="co">## Scrublet - Percentage of doublets                         0.0000    0.0000</span></span>
<span><span class="co">## scDblFinder - Number of doublets                         95.0000  263.0000</span></span>
<span><span class="co">## scDblFinder - Percentage of doublets                      3.5200    4.8500</span></span>
<span><span class="co">## DoubletFinder - Number of doublets, Resolution 1.5      202.0000  406.0000</span></span>
<span><span class="co">## DoubletFinder - Percentage of doublets, Resolution 1.5    7.4800    7.4900</span></span>
<span><span class="co">## CXDS - Number of doublets                               132.0000  294.0000</span></span>
<span><span class="co">## CXDS - Percentage of doublets                             4.8900    5.4300</span></span>
<span><span class="co">## BCDS - Number of doublets                               151.0000  257.0000</span></span>
<span><span class="co">## BCDS - Percentage of doublets                             5.5900    4.7400</span></span>
<span><span class="co">## SCDS Hybrid - Number of doublets                        173.0000  301.0000</span></span>
<span><span class="co">## SCDS Hybrid - Percentage of doublets                      6.4100    5.5500</span></span>
<span><span class="co">## DecontX - Mean contamination                              0.0810    0.0613</span></span>
<span><span class="co">## DecontX - Median contamination                            0.0535    0.0365</span></span>
<span><span class="co">##                                                        All Samples</span></span>
<span><span class="co">## Number of Cells                                          8119.0000</span></span>
<span><span class="co">## Mean counts                                              2140.5000</span></span>
<span><span class="co">## Median counts                                            1988.0000</span></span>
<span><span class="co">## Mean features detected                                    780.9600</span></span>
<span><span class="co">## Median features detected                                  750.0000</span></span>
<span><span class="co">## Scrublet - Number of doublets                               0.0000</span></span>
<span><span class="co">## Scrublet - Percentage of doublets                           0.0000</span></span>
<span><span class="co">## scDblFinder - Number of doublets                          358.0000</span></span>
<span><span class="co">## scDblFinder - Percentage of doublets                        4.4100</span></span>
<span><span class="co">## DoubletFinder - Number of doublets, Resolution 1.5        608.0000</span></span>
<span><span class="co">## DoubletFinder - Percentage of doublets, Resolution 1.5      7.4900</span></span>
<span><span class="co">## CXDS - Number of doublets                                 426.0000</span></span>
<span><span class="co">## CXDS - Percentage of doublets                               5.2500</span></span>
<span><span class="co">## BCDS - Number of doublets                                 408.0000</span></span>
<span><span class="co">## BCDS - Percentage of doublets                               5.0300</span></span>
<span><span class="co">## SCDS Hybrid - Number of doublets                          474.0000</span></span>
<span><span class="co">## SCDS Hybrid - Percentage of doublets                        5.8400</span></span>
<span><span class="co">## DecontX - Mean contamination                                0.0679</span></span>
<span><span class="co">## DecontX - Median contamination                              0.0413</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="running-individual-qc-methods">Running individual QC methods<a class="anchor" aria-label="anchor" href="#running-individual-qc-methods"></a>
</h3>
<p>Instead of running all quality control methods on the dataset at
once, users may elect to execute QC methods individually. The parameters
as well as the outputs to individual QC functions are described in
detail as follows:</p>
<details><summary><b>General QC metrics</b></summary><details><summary><b>runPerCellQC</b></summary><p>SingleCellTK utilizes the <a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link">scater</a>
package to compute cell-level QC metrics. The wrapper function
<code><a href="../reference/runPerCellQC.html">runPerCellQC()</a></code> can be used to separately compute general QC
metrics on its own.</p>
<ul>
<li>
<code>inSCE</code> parameter is the input SingleCellExperiment
object.</li>
<li>
<code>useAssay</code> is the assay object that in the
SingleCellExperiment object the user wishes to use.</li>
</ul>
<p>A list of gene sets can be applied to the function to determine the
expression of a set of specific genes, as mentioned before. Please also
refer to <a href="import_genesets.html"><em>Import Genesets</em>
documentation</a>.</p>
<p>The QC outputs are <code>sum</code>, <code>detected</code>, and
<code>percent_top_X</code>, stored as variables in
<code>colData</code>.</p>
<ul>
<li>
<code>sum</code> contains the total number of counts for each
cell.</li>
<li>
<code>detected</code> contains the total number of features for each
cell.</li>
<li>
<code>percent_top_X</code> contains the percentage of the total
counts that is made up by the expression of the top X genes for each
cell.</li>
<li>The <code>subsets_</code> columns contain information for the
specific gene list that was used. For instance, if a gene list
containing ribosome genes named <code>"ribosome"</code> was used,
<code>subsets_ribosome_sum</code> would contain the total number of
ribosome gene counts for each cell.</li>
<li>
<code>mito_sum</code>, <code>mito_detected</code> and
<code>mito_percent</code> contains number of counts, number of mito
features and percentage of mito gene expression of each cells. These
columns will show up only if you specify arguments related to mito genes
quantification in <code>runCellQC</code> function. Please refer to
<code>runCellQC</code> and <code>runPerCellQC</code> documentation for
more details.</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPerCellQC.html">runPerCellQC</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,</span>
<span>  useAssay <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  collectionName <span class="op">=</span> <span class="st">"ribosome"</span>,</span>
<span>  mitoRef <span class="op">=</span> <span class="st">"human"</span>, mitoIDType <span class="op">=</span> <span class="st">"symbol"</span>, mitoGeneLocation <span class="op">=</span> <span class="st">"rownames"</span><span class="op">)</span></span></code></pre></div>
  </details></details><details><summary><b>Doublet Detection</b></summary><p>Doublets hinder cell-type identification by appearing as a distinct
transcriptomic state, and need to be removed for downstream analysis.
SCTK contains various doublet detection tools that the user may choose
from.</p>
  <details><summary><b>runScrublet</b></summary><p><a href="https://github.com/swolock/scrublet" class="external-link">Scrublet</a> aims to
detect doublets by creating simulated doublets from combining
transcriptomic profiles of existing cells in the dataset. The wrapper
function <code><a href="../reference/runScrublet.html">runScrublet()</a></code> can be used to separately run the
Scrublet algorithm on its own.</p>
<ul>
<li>
<code>sample</code> indicates what sample each cell originates from.
It can be set to <code>NULL</code> if all cells in the dataset came from
the same sample.</li>
</ul>
<p>Scrublet also has a large set of parameters that the user can adjust,
please see the function reference for detail, by clicking on the
function name.</p>
<p>The Scrublet outputs include the following <code>colData</code>
variables:</p>
<ul>
<li>
<code>scrublet_score</code>, which is a numeric variable of the
likelihood that a cell is a doublet</li>
<li>
<code>scrublet_call</code>, which is the assignment of whether the
cell is a doublet.</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScrublet.html">runScrublet</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,</span>
<span>  sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  useAssay <span class="op">=</span> <span class="st">"counts"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</details><details><summary><b>runScDblFinder</b></summary><p><a href="https://rdrr.io/bioc/scDblFinder/" class="external-link">ScDblFinder</a> is a
doublet detection algorithm. ScDblFinder aims to detect doublets by
creating a simulated doublet from existing cells and projecting it to
the same PCA space as the cells. The wrapper function
<code><a href="../reference/runScDblFinder.html">runScDblFinder()</a></code> can be used to separately run the
ScDblFinder algorithm on its own.</p>
<ul>
<li>
<code>nNeighbors</code> is the number of nearest neighbor used to
calculate the density for doublet detection.</li>
<li>
<code>simDoublets</code> is used to determine the number of
simulated doublets used for doublet detection.</li>
</ul>
<p>The output of ScDblFinder is a
<code>scDblFinder_doublet_score</code>, which will be stored as a
<code>colData</code> variable. The doublet score of a droplet will be
higher if the it is deemed likely to be a doublet.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScDblFinder.html">runScDblFinder</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span></code></pre></div>
</details><details><summary><b>runDoubletFinder</b></summary><p><a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder" class="external-link">DoubletFinder</a>
is a doublet detection algorithm which depends on the single cell
analysis package <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>. The
wrapper function <code><a href="../reference/runDoubletFinder.html">runDoubletFinder()</a></code> can be used to
separately run the DoubletFinder algorithm on its own.</p>
<ul>
<li>
<code>seuratRes</code> - <code><a href="../reference/runDoubletFinder.html">runDoubletFinder()</a></code> relies on a
parameter (in Seurat) called “resolution” to determine cells that may be
doublets. Users will be able to manipulate the resolution parameter
through <code>seuratRes</code>. If multiple numeric vectors are stored
in <code>seuratRes</code>, there will be multiple label/scores.</li>
<li>
<code>seuratNfeatures</code> determines the number of features that
is used in the <a href="https://satijalab.org/seurat/reference/findvariablefeatures" class="external-link"><code>FindVariableFeatures</code></a>
function in Seurat.</li>
<li>
<code>seuratPcs</code> determines the number of dimensions used in
the <a href="https://satijalab.org/seurat/reference/findneighbors" class="external-link"><code>FindNeighbors</code></a>
function in Seurat.</li>
<li>
<code>formationRate</code> is the estimated doublet detection rate
in the dataset. It aims to detect doublets by creating simulated
doublets from combining transcriptomic profiles of existing cells in the
dataset.</li>
</ul>
<p>The DoubletFinder outputs include the following <code>colData</code>
variable:</p>
<ul>
<li>
<code>doubletFinder_doublet_score</code>, which is a numeric
variable of the likelihood that a cell is a doublet</li>
<li>
<code>doubletFinder_doublet_label</code>, which is the assignment of
whether the cell is a doublet.</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runDoubletFinder.html">runDoubletFinder</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  seuratRes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.0</span><span class="op">)</span>, seuratPcs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>,</span>
<span>  seuratNfeatures <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  formationRate <span class="op">=</span> <span class="fl">0.075</span>, seed <span class="op">=</span> <span class="fl">12345</span></span>
<span><span class="op">)</span></span></code></pre></div>
</details><details><summary><b>runCXDS</b></summary><p>CXDS, or co-expression based doublet scoring, is an algorithm in the
<a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html" class="external-link">SCDS</a>
package which employs a binomial model for the co-expression of pairs of
genes to determine doublets. The wrapper function <code><a href="../reference/runCxds.html">runCxds()</a></code>
can be used to separately run the CXDS algorithm on its own.</p>
<ul>
<li>
<code>ntop</code> is the number of top variance genes to
consider.</li>
<li>
<code>binThresh</code> is the minimum counts a gene needs to have to
be included in the analysis.</li>
<li>
<code>verb</code> determines whether progress messages will be
displayed or not.</li>
<li>
<code>retRes</code> will determine whether the gene pair results
should be returned or not.</li>
<li>
<code>estNdbl</code> is the user estimated number of doublets.</li>
</ul>
<p>The output of <code><a href="../reference/runCxds.html">runCxds()</a></code> is the doublet score,
<code>scds_cxds_score</code>, which will be stored as a
<code>colData</code> variable.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCxds.html">runCxds</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  ntop <span class="op">=</span> <span class="fl">500</span>, binThresh <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  verb <span class="op">=</span> <span class="cn">FALSE</span>, retRes <span class="op">=</span> <span class="cn">FALSE</span>, estNdbl <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</details><details><summary><b>runBCDS</b></summary><p>BCDS, or binary classification based doublet scoring, is an algorithm
in the <a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html" class="external-link">SCDS</a>
package which uses a binary classification approach to determine
doublets. The wrapper function <code><a href="../reference/runBcds.html">runBcds()</a></code> can be used to
separately run the BCDS algorithm on its own.</p>
<ul>
<li>
<code>ntop</code> is the number of top variance genes to
consider.</li>
<li>
<code>srat</code> is the ratio between original number of cells and
simulated doublets.</li>
<li>
<code>nmax</code> is the maximum number of cycles that the algorithm
should run through. If set to <code>"tune"</code>, this will be
automatic.</li>
<li>
<code>varImp</code> determines if the variable importance should be
returned or not.</li>
</ul>
<p>The output of <code><a href="../reference/runBcds.html">runBcds()</a></code> is <code>scds_bcds_score</code>,
which is the likelihood that a cell is a doublet and will be stored as a
<code>colData</code> variable.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runBcds.html">runBcds</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, seed <span class="op">=</span> <span class="fl">12345</span>, sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  ntop <span class="op">=</span> <span class="fl">500</span>, srat <span class="op">=</span> <span class="fl">1</span>, nmax <span class="op">=</span> <span class="st">"tune"</span>, varImp <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</details><details><summary><b>runCxdsBcdsHybrid</b></summary><p>The CXDS-BCDS hybrid algorithm, uses both CXDS and BCDS algorithms
from the <a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html" class="external-link">SCDS</a>
package. The wrapper function <code><a href="../reference/runCxdsBcdsHybrid.html">runCxdsBcdsHybrid()</a></code> can be
used to separately run the CXDS-BCDS hybrid algorithm on its own.</p>
<p>All parameters from the <code><a href="../reference/runCxds.html">runCxds()</a></code> and
<code><a href="../reference/runBcds.html">runBcds()</a></code> functions may be applied to this function in the
<code>cxdsArgs</code> and <code>bcdsArgs</code> parameters,
respectively.</p>
<p>The output of <code><a href="../reference/runCxdsBcdsHybrid.html">runCxdsBcdsHybrid()</a></code> is the doublet score,
<code>scds_hybrid_score</code>, which will be stored as a
<code>colData</code> variable.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCxdsBcdsHybrid.html">runCxdsBcdsHybrid</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  seed <span class="op">=</span> <span class="fl">12345</span>, nTop <span class="op">=</span> <span class="fl">500</span></span>
<span><span class="op">)</span></span></code></pre></div>
  </details></details><details><summary><b>Ambient RNA Detection</b></summary><details><summary><b>runDecontX</b></summary><p>In droplet-based single cell technologies, ambient RNA that may have
been released from apoptotic or damaged cells may get incorporated into
another droplet, and can lead to contamination. <a href="https://rdrr.io/bioc/celda/man/decontX.html" class="external-link">decontX</a>,
available from <a href="https://bioconductor.org/packages/release/bioc/html/celda.html" class="external-link">celda</a>,
is a Bayesian method for the identification of the contamination level
at a cellular level. The wrapper function <code><a href="../reference/runDecontX.html">runDecontX()</a></code> can
be used to separately run the DecontX algorithm on its own.</p>
<p>The outputs of <code><a href="../reference/runDecontX.html">runDecontX()</a></code> are
<code>decontX_contamination</code> and
<code>decontX_clusters</code>.</p>
<ul>
<li>
<code>decontX_contamination</code> is a numeric vector which
characterizes the level of contamination in each cell.</li>
<li>Clustering is performed as part of the <code><a href="../reference/runDecontX.html">runDecontX()</a></code>
algorithm. <code>decontX_clusters</code> is the resulting cluster
assignment, which can also be labeled on the plot. For performing
fine-tuned clustering in SCTK, please refer to <a href="clustering.html"><em>Clustering</em> documentation</a>
</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>pbmc.combined <span class="ot">&lt;-</span> <span class="fu">runDecontX</span>(</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>  <span class="at">inSCE =</span> pbmc.combined, <span class="at">useAssay =</span> <span class="st">"counts"</span></span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>  <span class="at">sample =</span> <span class="fu">colData</span>(pbmc.combined)<span class="sc">$</span>sample</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>)</span></code></pre></div>
  </details><details><summary><b>runSoupX</b></summary><p>In droplet-based single cell technologies, ambient RNA that may have
been released from apoptotic or damaged cells may get incorporated into
another droplet, and can lead to contamination. <a href="https://github.com/constantAmateur/SoupX" class="external-link">SoupX</a> uses
non-expressed genes to estimates a global contamination fraction. The
wrapper function <code><a href="../reference/runSoupX.html">runSoupX()</a></code> can be used to separately run
the SoupX algorithm on its own.</p>
<p>he main outputs of <code>runSoupX</code> are
<code>soupX_contamination</code>, <code>soupX_clusters</code>, and the
corrected assay <code>SoupX</code>, together with other intermediate
metrics that SoupX generates.</p>
<ul>
<li>
<code>soupX_contamination</code> is a numeric vector which
characterizes the level of contamination in each cell. SoupX generates
one global contamination estimate per sample, instead of returning
cell-specific estimation.</li>
<li>Clustering is required for SoupX algorithm. It will be performed if
users do not provide the label as input. <code>quickCluster()</code>
method from package <a href="https://rdrr.io/bioc/scran/man/quickCluster.html" class="external-link">scran</a> is
adopted for this purpose. <code>soupX_clusters</code> is the resulting
cluster assignment, which can also be labeled on the plot. For
performing fine-tuned clustering in SCTK, please refer to <a href="clustering.html"><em>Clustering</em> documentation</a>
</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a>pbmc.combined <span class="ot">&lt;-</span> <span class="fu">runSoupX</span>(</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>  <span class="at">inSCE =</span> pbmc.combined, <span class="at">useAssay =</span> <span class="st">"counts"</span></span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>  <span class="at">sample =</span> <span class="fu">colData</span>(pbmc.combined)<span class="sc">$</span>sample</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>)</span></code></pre></div>
  </details></details>
</div>
<div class="section level3">
<h3 id="plotting-qc-metrics">Plotting QC metrics<a class="anchor" aria-label="anchor" href="#plotting-qc-metrics"></a>
</h3>
<p>Upon running <code><a href="../reference/runCellQC.html">runCellQC()</a></code> or any individual QC methods,
the QC outputs will need to be plotted. For each QC method, SCTK
provides a specialized plotting function.</p>
<details><summary><b>General QC metrics</b></summary><details><summary><b>runPerCellQC</b></summary><p>The wrapper function <code><a href="../reference/plotRunPerCellQCResults.html">plotRunPerCellQCResults()</a></code> can be
used to plot the general QC outputs.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">runpercellqc.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotRunPerCellQCResults.html">plotRunPerCellQCResults</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="va">sample.vector</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>, axisSize <span class="op">=</span> <span class="fl">8</span>, axisLabelSize <span class="op">=</span> <span class="fl">9</span>, titleSize <span class="op">=</span> <span class="fl">20</span>, labelSamples<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">runpercellqc.results</span></span></code></pre></div>
<p><img src="cnsl_cellqc_files/figure-html/unnamed-chunk-3-1.png" width="960"></p>
  </details></details><details><summary><b>Doublet Detection</b></summary><details><summary><b>Scrublet</b></summary><p>The wrapper function <code><a href="../reference/plotScrubletResults.html">plotScrubletResults()</a></code> can be used
to plot the results from the Scrublet algorithm. Here, we will use the
UMAP coordinates generated from <code><a href="../reference/runUMAP.html">runQuickUMAP()</a></code> in previous
sections.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of length 8</span></span>
<span><span class="co">## names(8): UMAP scrublet_TSNE ... SoupX_UMAP_pbmc3k SoupX_UMAP_pbmc6k</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scrublet.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotScrubletResults.html">plotScrubletResults</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,</span>
<span>  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  combinePlot <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  titleSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  axisLabelSize <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  axisSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  legendSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  legendTitleSize <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scrublet.results</span> </span></code></pre></div>
<p><img src="cnsl_cellqc_files/figure-html/plot_umap_scrublet-1.png" width="960"></p>
</details><details><summary><b>ScDblFinder</b></summary><p>The wrapper function <code><a href="../reference/plotScDblFinderResults.html">plotScDblFinderResults()</a></code> can be
used to plot the QC outputs from the ScDblFinder algorithm.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scDblFinder.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotScDblFinderResults.html">plotScDblFinderResults</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  titleSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  axisSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  legendSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  legendTitleSize <span class="op">=</span> <span class="fl">13</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</details><details><summary><b>DoubletFinder</b></summary><p>The wrapper function <code><a href="../reference/plotDoubletFinderResults.html">plotDoubletFinderResults()</a></code> can be
used to plot the QC outputs from the DoubletFinder algorithm.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">doubletFinderResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDoubletFinderResults.html">plotDoubletFinderResults</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,</span>
<span>  sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  combinePlot <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  titleSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  axisSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  legendSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  legendTitleSize <span class="op">=</span> <span class="fl">13</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</details><details><summary><b>SCDS, CXDS</b></summary><p>The wrapper function <code><a href="../reference/plotCxdsResults.html">plotCxdsResults()</a></code> can be used to
plot the QC outputs from the CXDS algorithm.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cxdsResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCxdsResults.html">plotCxdsResults</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,</span>
<span>  sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  titleSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  axisSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  legendSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  legendTitleSize <span class="op">=</span> <span class="fl">13</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</details><details><summary><b>SCDS, BCDS</b></summary><p>The wrapper function <code><a href="../reference/plotBcdsResults.html">plotBcdsResults()</a></code> can be used to
plot the QC outputs from the BCDS algorithm</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bcdsResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotBcdsResults.html">plotBcdsResults</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,</span>
<span>  sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  titleSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  axisSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  legendSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  legendTitleSize <span class="op">=</span> <span class="fl">13</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</details><details><summary><b>SCDS, CXDS-BCDS hybrid</b></summary><p>The wrapper function <code><a href="../reference/plotScdsHybridResults.html">plotScdsHybridResults()</a></code> can be used
to plot the QC outputs from the CXDS-BCDS hybrid algorithm.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bcdsCxdsHybridResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotScdsHybridResults.html">plotScdsHybridResults</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  titleSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  axisSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  legendSize <span class="op">=</span> <span class="fl">13</span>,</span>
<span>  legendTitleSize <span class="op">=</span> <span class="fl">13</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</details></details><details><summary><b>Ambient RNA Detection</b></summary><details><summary><b>DecontX</b></summary><p>The wrapper function <code><a href="../reference/plotDecontXResults.html">plotDecontXResults()</a></code> can be used to
plot the QC outputs from the DecontX algorithm.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">decontxResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotDecontXResults.html">plotDecontXResults</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  titleSize <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  axisLabelSize <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  axisSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  legendSize <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  legendTitleSize <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  relWidths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  sampleRelWidths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  labelSamples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  labelClusters <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">decontxResults</span></span></code></pre></div>
<p><img src="cnsl_cellqc_files/figure-html/plots_umap_decontx-1.png" width="864"></p>
  </details><details><summary><b>SoupX</b></summary><p>The wrapper function <code><a href="../reference/plotSoupXResults.html">plotSoupXResults()</a></code> can be used to
plot the QC outputs from the SoupX algorithm.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soupxResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSoupXResults.html">plotSoupXResults</a></span><span class="op">(</span></span>
<span>  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  titleSize <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  axisLabelSize <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  axisSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  legendSize <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  legendTitleSize <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  labelClusters <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">soupxResults</span></span></code></pre></div>
<p><img src="cnsl_cellqc_files/figure-html/plots_umap_soupx-1.png" width="864"></p>
  </details></details>
</div>
<div class="section level3">
<h3 id="filtering-the-dataset">Filtering the dataset<a class="anchor" aria-label="anchor" href="#filtering-the-dataset"></a>
</h3>
<p><code>SingleCellExperiment</code> objects can be subset by its
<code>colData</code> using <code><a href="../reference/subsetSCECols.html">subsetSCECols()</a></code>. The
<code>colData</code> parameter takes in a character vector of
expression(s) which will be used to identify a subset of cells using
variables found in the <code>colData</code> of the
<code>SingleCellExperiment</code> object. For example, if <code>x</code>
is a numeric vector in <code>colData</code>, then setting
<code>colData = "x &lt; 5"</code> will return a
<code>SingleCellExperiment</code> object where all columns (cells) meet
the condition that <code>x</code> is less than 5. The <code>index</code>
parameter takes in a numeric vector of indices which should be kept,
while <code>bool</code> takes in a logical vector of <code>TRUE</code>
or <code>FALSE</code> which should be of the same length as the number
of columns (cells) in the <code>SingleCellExperiment</code> object.
Please refer to our <a href="filtering.html"><em>Filtering</em>
documentation</a> for detail.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Before filtering:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 32738  8119</span></span></code></pre>
<p>Remove barcodes with high mitochondrial gene expression:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, colData <span class="op">=</span> <span class="st">'mito_percent &lt; 20'</span><span class="op">)</span></span></code></pre></div>
<p>Remove detected doublets from Scrublet:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, colData <span class="op">=</span> <span class="st">'scrublet_call == "Singlet"'</span><span class="op">)</span></span></code></pre></div>
<p>Remove cells with high levels of ambient RNA contamination:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, colData <span class="op">=</span> <span class="st">'decontX_contamination &lt; 0.5'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#After filtering:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 32738  7934</span></span></code></pre>
<p><br></p>
<blockquote>
<p>For performing QC on droplet-level raw count matrix with SCTK, please
refer to our <a href="cnsl_dropletqc.html"><em>Droplet QC</em>
documentation</a>.</p>
</blockquote>
 <details><summary><b>Session Information</b></summary><div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.4.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] TENxPBMCData_1.20.0         HDF5Array_1.30.0           </span></span>
<span><span class="co">##  [3] rhdf5_2.46.1                dplyr_1.1.4                </span></span>
<span><span class="co">##  [5] singleCellTK_2.13.2         DelayedArray_0.28.0        </span></span>
<span><span class="co">##  [7] SparseArray_1.2.4           S4Arrays_1.2.1             </span></span>
<span><span class="co">##  [9] abind_1.4-5                 Matrix_1.6-5               </span></span>
<span><span class="co">## [11] SingleCellExperiment_1.24.0 SummarizedExperiment_1.32.0</span></span>
<span><span class="co">## [13] Biobase_2.62.0              GenomicRanges_1.54.1       </span></span>
<span><span class="co">## [15] GenomeInfoDb_1.38.8         IRanges_2.36.0             </span></span>
<span><span class="co">## [17] S4Vectors_0.40.2            BiocGenerics_0.48.1        </span></span>
<span><span class="co">## [19] MatrixGenerics_1.14.0       matrixStats_1.2.0          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] spatstat.sparse_3.0-3         fs_1.6.3                     </span></span>
<span><span class="co">##   [3] bitops_1.0-7                  doParallel_1.0.17            </span></span>
<span><span class="co">##   [5] httr_1.4.7                    RColorBrewer_1.1-3           </span></span>
<span><span class="co">##   [7] sctransform_0.4.1             tools_4.3.3                  </span></span>
<span><span class="co">##   [9] utf8_1.2.4                    R6_2.5.1                     </span></span>
<span><span class="co">##  [11] lazyeval_0.2.2                uwot_0.1.16                  </span></span>
<span><span class="co">##  [13] rhdf5filters_1.14.1           withr_3.0.0                  </span></span>
<span><span class="co">##  [15] sp_2.1-2                      gridExtra_2.3                </span></span>
<span><span class="co">##  [17] progressr_0.14.0              cli_3.6.2                    </span></span>
<span><span class="co">##  [19] textshaping_0.3.7             spatstat.explore_3.2-5       </span></span>
<span><span class="co">##  [21] fastDummies_1.7.3             enrichR_3.2                  </span></span>
<span><span class="co">##  [23] labeling_0.4.3                eds_1.4.0                    </span></span>
<span><span class="co">##  [25] sass_0.4.8                    Seurat_5.0.1                 </span></span>
<span><span class="co">##  [27] spatstat.data_3.0-4           ggridges_0.5.6               </span></span>
<span><span class="co">##  [29] pbapply_1.7-2                 pkgdown_2.0.7                </span></span>
<span><span class="co">##  [31] Rsamtools_2.18.0              systemfonts_1.0.6            </span></span>
<span><span class="co">##  [33] dbscan_1.1-12                 svglite_2.1.3                </span></span>
<span><span class="co">##  [35] R.utils_2.12.3                scater_1.30.1                </span></span>
<span><span class="co">##  [37] WriteXLS_6.5.0                parallelly_1.36.0            </span></span>
<span><span class="co">##  [39] maps_3.4.2                    limma_3.58.1                 </span></span>
<span><span class="co">##  [41] rstudioapi_0.15.0             RSQLite_2.3.5                </span></span>
<span><span class="co">##  [43] FNN_1.1.4                     generics_0.1.3               </span></span>
<span><span class="co">##  [45] BiocIO_1.12.0                 combinat_0.0-8               </span></span>
<span><span class="co">##  [47] spatstat.random_3.2-2         ica_1.0-3                    </span></span>
<span><span class="co">##  [49] ggbeeswarm_0.7.2              fansi_1.0.6                  </span></span>
<span><span class="co">##  [51] R.methodsS3_1.8.2             lifecycle_1.0.4              </span></span>
<span><span class="co">##  [53] SoupX_1.6.2                   yaml_2.3.8                   </span></span>
<span><span class="co">##  [55] edgeR_4.0.16                  BiocFileCache_2.10.1         </span></span>
<span><span class="co">##  [57] Rtsne_0.17                    grid_4.3.3                   </span></span>
<span><span class="co">##  [59] blob_1.2.4                    promises_1.2.1               </span></span>
<span><span class="co">##  [61] dqrng_0.3.2                   ExperimentHub_2.10.0         </span></span>
<span><span class="co">##  [63] crayon_1.5.2                  miniUI_0.1.1.1               </span></span>
<span><span class="co">##  [65] lattice_0.22-5                beachmat_2.18.1              </span></span>
<span><span class="co">##  [67] cowplot_1.1.3                 annotate_1.80.0              </span></span>
<span><span class="co">##  [69] KEGGREST_1.42.0               pillar_1.9.0                 </span></span>
<span><span class="co">##  [71] knitr_1.45                    metapod_1.10.1               </span></span>
<span><span class="co">##  [73] rjson_0.2.21                  celda_1.18.1                 </span></span>
<span><span class="co">##  [75] xgboost_1.7.7.1               future.apply_1.11.1          </span></span>
<span><span class="co">##  [77] codetools_0.2-19              leiden_0.4.3.1               </span></span>
<span><span class="co">##  [79] glue_1.7.0                    data.table_1.15.4            </span></span>
<span><span class="co">##  [81] vctrs_0.6.5                   png_0.1-8                    </span></span>
<span><span class="co">##  [83] spam_2.10-0                   gtable_0.3.4                 </span></span>
<span><span class="co">##  [85] cachem_1.0.8                  xfun_0.41                    </span></span>
<span><span class="co">##  [87] mime_0.12                     DropletUtils_1.22.0          </span></span>
<span><span class="co">##  [89] RcppEigen_0.3.4.0.0           GSVAdata_1.38.0              </span></span>
<span><span class="co">##  [91] survival_3.5-8                iterators_1.0.14             </span></span>
<span><span class="co">##  [93] fields_15.2                   statmod_1.5.0                </span></span>
<span><span class="co">##  [95] bluster_1.12.0                interactiveDisplayBase_1.40.0</span></span>
<span><span class="co">##  [97] ellipsis_0.3.2                fitdistrplus_1.1-11          </span></span>
<span><span class="co">##  [99] ROCR_1.0-11                   nlme_3.1-164                 </span></span>
<span><span class="co">## [101] bit64_4.0.5                   scds_1.18.0                  </span></span>
<span><span class="co">## [103] filelock_1.0.3                RcppAnnoy_0.0.22             </span></span>
<span><span class="co">## [105] rprojroot_2.0.4               bslib_0.6.1                  </span></span>
<span><span class="co">## [107] irlba_2.3.5.1                 vipor_0.4.7                  </span></span>
<span><span class="co">## [109] KernSmooth_2.23-22            colorspace_2.1-0             </span></span>
<span><span class="co">## [111] DBI_1.2.1                     tidyselect_1.2.0             </span></span>
<span><span class="co">## [113] bit_4.0.5                     compiler_4.3.3               </span></span>
<span><span class="co">## [115] curl_5.2.1                    graph_1.80.0                 </span></span>
<span><span class="co">## [117] BiocNeighbors_1.20.2          xml2_1.3.6                   </span></span>
<span><span class="co">## [119] desc_1.4.3                    plotly_4.10.4                </span></span>
<span><span class="co">## [121] rtracklayer_1.62.0            scales_1.3.0                 </span></span>
<span><span class="co">## [123] lmtest_0.9-40                 rappdirs_0.3.3               </span></span>
<span><span class="co">## [125] goftest_1.2-3                 stringr_1.5.1                </span></span>
<span><span class="co">## [127] digest_0.6.35                 spatstat.utils_3.0-4         </span></span>
<span><span class="co">## [129] rmarkdown_2.25                XVector_0.42.0               </span></span>
<span><span class="co">## [131] htmltools_0.5.7               pkgconfig_2.0.3              </span></span>
<span><span class="co">## [133] sparseMatrixStats_1.14.0      highr_0.10                   </span></span>
<span><span class="co">## [135] dbplyr_2.4.0                  fastmap_1.1.1                </span></span>
<span><span class="co">## [137] rlang_1.1.3                   htmlwidgets_1.6.4            </span></span>
<span><span class="co">## [139] shiny_1.8.0                   DelayedMatrixStats_1.24.0    </span></span>
<span><span class="co">## [141] farver_2.1.1                  jquerylib_0.1.4              </span></span>
<span><span class="co">## [143] zoo_1.8-12                    jsonlite_1.8.8               </span></span>
<span><span class="co">## [145] BiocParallel_1.36.0           R.oo_1.26.0                  </span></span>
<span><span class="co">## [147] BiocSingular_1.18.0           RCurl_1.98-1.14              </span></span>
<span><span class="co">## [149] magrittr_2.0.3                kableExtra_1.4.0             </span></span>
<span><span class="co">## [151] scuttle_1.12.0                GenomeInfoDbData_1.2.11      </span></span>
<span><span class="co">## [153] dotCall64_1.1-1               patchwork_1.2.0              </span></span>
<span><span class="co">## [155] Rhdf5lib_1.24.1               munsell_0.5.1                </span></span>
<span><span class="co">## [157] Rcpp_1.0.12                   viridis_0.6.5                </span></span>
<span><span class="co">## [159] reticulate_1.34.0             pROC_1.18.5                  </span></span>
<span><span class="co">## [161] stringi_1.8.3                 MCMCprecision_0.4.0          </span></span>
<span><span class="co">## [163] zlibbioc_1.48.2               MASS_7.3-60.0.1              </span></span>
<span><span class="co">## [165] AnnotationHub_3.10.0          plyr_1.8.9                   </span></span>
<span><span class="co">## [167] parallel_4.3.3                listenv_0.9.0                </span></span>
<span><span class="co">## [169] ggrepel_0.9.5                 deldir_2.0-2                 </span></span>
<span><span class="co">## [171] scDblFinder_1.16.0            Biostrings_2.70.1            </span></span>
<span><span class="co">## [173] splines_4.3.3                 tensor_1.5                   </span></span>
<span><span class="co">## [175] locfit_1.5-9.9                igraph_2.0.3                 </span></span>
<span><span class="co">## [177] spatstat.geom_3.2-8           RcppHNSW_0.6.0               </span></span>
<span><span class="co">## [179] reshape2_1.4.4                ScaledMatrix_1.10.0          </span></span>
<span><span class="co">## [181] BiocVersion_3.18.1            XML_3.99-0.16.1              </span></span>
<span><span class="co">## [183] evaluate_0.23                 SeuratObject_5.0.1           </span></span>
<span><span class="co">## [185] scran_1.30.2                  BiocManager_1.30.22          </span></span>
<span><span class="co">## [187] foreach_1.5.2                 httpuv_1.6.14                </span></span>
<span><span class="co">## [189] polyclip_1.10-6               tidyr_1.3.1                  </span></span>
<span><span class="co">## [191] RANN_2.6.1                    purrr_1.0.2                  </span></span>
<span><span class="co">## [193] scattermore_1.2               future_1.33.1                </span></span>
<span><span class="co">## [195] ggplot2_3.5.0                 rsvd_1.0.5                   </span></span>
<span><span class="co">## [197] xtable_1.8-4                  restfulr_0.0.15              </span></span>
<span><span class="co">## [199] RSpectra_0.16-1               later_1.3.2                  </span></span>
<span><span class="co">## [201] viridisLite_0.4.2             ragg_1.3.0                   </span></span>
<span><span class="co">## [203] tibble_3.2.1                  memoise_2.0.1                </span></span>
<span><span class="co">## [205] beeswarm_0.4.0                AnnotationDbi_1.64.1         </span></span>
<span><span class="co">## [207] GenomicAlignments_1.38.2      cluster_2.1.6                </span></span>
<span><span class="co">## [209] globals_0.16.2                GSEABase_1.64.0              </span></span>
<span><span class="co">## [211] here_1.0.1</span></span></code></pre>
</details>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
