<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Feature Selection • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Feature Selection">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.18.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>

  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Feature Selection</h1>
                        <h4 data-toc-skip class="author">Irzam
Sarfraz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/feature_selection.Rmd" class="external-link"><code>vignettes/articles/feature_selection.Rmd</code></a></small>
      <div class="hidden name"><code>feature_selection.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><strong>singleCellTK</strong> offers a convenient way to compute and
select the most variable features that show the highest biological
variability to use them in the downstream analysis. Feature selection
methods available with the toolkit include <code>vst</code>,
<code>mean.var.plot</code> &amp; <code>dispersion</code> from
<code>Seurat</code> [<span class="citation">[1]</span>]<span class="citation">[2]</span>[<span class="citation">[3]</span>]<span class="citation">[4]</span> package and <code>modelGeneVar</code> from
<code>Scran</code> <span class="citation">[5]</span> package. Users can
additionally use visualization options and methods to visualize the top
variable genes interactively as well as from R console. <br></p>
<p>To view detailed instructions on how to use these methods, please
select ‘Interactive Analysis’ for using feature selection in shiny
application or ‘Console Analysis’ for using these methods on R console
from the tabs below:</p>

<meta name="viewport" content="width=device-width, initial-scale=1">
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p><strong>1.</strong> Select <strong>Feature Selection &amp;
Dimensionality Reduction</strong> tab from the top menu. This workflow
guide assumes that the data as been previously uploaded, filtered and
normalized before proceeding with this tab.</p>
<p><img src="ui_screenshots/feature_selection/FS_1_select_1.PNG"><br></p>
<p><strong>2.</strong> Select <strong>Feature Selection</strong> sub-tab
(selected by default) to open up the feature selection
user-interface.</p>
<p><img src="ui_screenshots/feature_selection/FS_1_select_2.png"><br></p>
<p><strong>3.</strong> The <strong>Feature Selection</strong> sub-tab is
divided into three panels namely, <strong>a) Compute variability
metric</strong>, <strong>b) Select number of variable features</strong>,
and <strong>c) Plot</strong>.</p>
<p><img src="ui_screenshots/feature_selection/fs_1.png"><br></p>
<p>The working of sections <em>a</em>, <em>b</em> and <em>c</em> are
described below:</p>
<p><strong>a) Compute variability metric</strong> <br></p>
<p><img src="ui_screenshots/feature_selection/fs_2.png"><br></p>
<p>The <em>Compute variability metric</em> panel performs the
calculation of the variability of all the features present in the
dataset. Users need to select a method either from Seurat
(<code>vst</code>, <code>mean.var.plot</code>, <code>dispersion</code>)
or Scran (<code>modelGeneVar</code>) packages. A expression matrix must
be selected for the calculation. Note that <code>Seurat - vst</code>
method requires an raw count matrix, while other methods requires
normalized expression. After selection, click on <strong>Run</strong> to
start the calculation.</p>
<p>After computation, a scatter plot of all features will be shown in
the <strong>Plot</strong> panel, where the axes are from the computed
metrics. The variability metrics are stored as feature metadata
variables, and could be used for ranking the features.</p>
<p><strong>b) Select number of variable features</strong></p>
<p><img src="ui_screenshots/feature_selection/fs_4.png"><br></p>
<p>Once step (a) is done, subset of highly variable features can be
selected for downstream analysis basing on available metrics. Users need
to select a metric that has been calculated, enter the number of the top
variable features to select, and name the subset. After confirming all
the setting, click on <strong>Run</strong> to create the subset.</p>
<p>Upon the creation of a new subset, the <strong>Plot</strong> will be
updated: the scatter plot will be based on the selected metric, all the
selected highly variable features will be labeled in red, and a few top
features will be highlighted with text labels.</p>
<p><strong>c) Plot</strong></p>
<p><img src="ui_screenshots/feature_selection/fs_6.png"><br></p>
<p>From the dropdown panel at the top of <strong>Plot</strong> panel,
users are allowed to:</p>
<ul>
<li>Update plot with different metrics, i.e. coordinates of points.</li>
<li>Label features with different subsets. Note that a subset created
from a certain method can be labeled on the coordinates from a different
method, for comparison of the variability ranking.</li>
<li>Adjust the number of top features to be highlighted with text
labels.</li>
<li>Change the feature ID type. This should be available in feature
metadata. We recommend preset the feature display ID type when importing
the data set, so that this setting applies globally through out the
toolkit.</li>
</ul>
</div>
<div id="console" class="tabcontent">
<p><strong>1.</strong> Compute statistics for the highly variable
features using the wrapper function <code><a href="../reference/runFeatureSelection.html">runFeatureSelection()</a></code>
as below:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runFeatureSelection.html">runFeatureSelection</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></span></code></pre></div>
<p>In the function above, we recommend using a raw count matrix for
method <code>"vst"</code> from Seurat, or a normalized assay for method
<code>"mean.var.plot"</code>, <code>"dispersion"</code> from Seurat and
<code>"modelGeneVar"</code> from Scran pacakge.</p>
<p><strong>2.</strong> Get names of top genes using the
<code><a href="../reference/getTopHVG.html">getTopHVG()</a></code> function with specifying the same method which
was used for computation in the step 1:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">topGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getTopHVG.html">getTopHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, method <span class="op">=</span> <span class="st">"vst"</span>, hvgNumber <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span></code></pre></div>
<p>Users can also visualize top genes using the
<code><a href="../reference/plotTopHVG.html">plotTopHVG()</a></code> function with specifying the same method which
was used previously:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotTopHVG.html">plotTopHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, method <span class="op">=</span> <span class="st">"vst"</span>, hvgNumber <span class="op">=</span> <span class="fl">1000</span>, labelsCount <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><strong>3.</strong> Set feature subset using the
<code><a href="../reference/getTopHVG.html">setTopHVG()</a></code> function with specifying the same method which
was used previously:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getTopHVG.html">setTopHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, method <span class="op">=</span> <span class="st">"vst"</span>, hvgNumber <span class="op">=</span> <span class="fl">1000</span>, featureSubsetName <span class="op">=</span> <span class="st">"HVG_vst1000"</span><span class="op">)</span></span></code></pre></div>
<p>The name for the subset could be automatically generated by the
function while can be specified by users. The subset will be used in
downstream processing such as performing dimension reduction with the
HVG subset.</p>
<details><summary><b>Example</b></summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Load singleCellTK &amp; pbmc3k example data</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span>  <span class="co"># Compute metrics with 'vst' feature selection method</span></span>
<span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runFeatureSelection.html">runFeatureSelection</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></span>
<span>  <span class="co"># Get the names of the top 1000 highly variable genes</span></span>
<span>  <span class="va">topGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getTopHVG.html">getTopHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, method <span class="op">=</span> <span class="st">"vst"</span>, hvgNumber <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">topGenes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "PPBP"   "LYZ"    "S100A9" "IGLL5"  "GNLY"   "FTL"    "PF4"    "FTH1"  </span></span>
<span><span class="co">##  [9] "GNG11"  "FCER1A"</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Visualize the top variable genes and label the first top 10 genes</span></span>
<span>  <span class="fu"><a href="../reference/plotTopHVG.html">plotTopHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, method <span class="op">=</span> <span class="st">"vst"</span>, hvgNumber <span class="op">=</span> <span class="fl">1000</span>, labelsCount <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="feature_selection_files/figure-html/fs_console_example_1-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Set subset of variable features</span></span>
<span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getTopHVG.html">setTopHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, method <span class="op">=</span> <span class="st">"vst"</span>, hvgNumber <span class="op">=</span> <span class="fl">1000</span>, featureSubsetName <span class="op">=</span> <span class="st">"HVG_vst1000"</span><span class="op">)</span></span></code></pre></div>
</details><details><summary><b>Individual Functions</b></summary><p>While the <code><a href="../reference/runFeatureSelection.html">runFeatureSelection()</a></code> wrapper function can be
used to run all available feature selection methods, separate functions
are also available for all of the included methods. The following
functions can be used for specific feature selection methods:</p>
 <style>
div.offset { background-color:inherit ; border-radius: 5px; margin-left: 15px;}
</style>
<div class="offset">
<details><summary><code>vst</code>, <code>mean.var.plot</code> or <code>dispersion</code>
from Seurat package:
</summary><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratFindHVG.html">runSeuratFindHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, method <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></span></code></pre></div>
<p>The parameters to the above function include:</p>
<ul>
<li>
<code>inSCE</code>: The input <code>SingleCellExperiment</code>
object</li>
<li>
<code>useAssay</code>: Specify the name of the assay to use for
feature selection</li>
<li>
<code>method</code>: One of the options from <code>"vst"</code>,
<code>"mean.var.plot"</code> or <code>"dispersion"</code>
</li>
<li>
<code>hvgNumber</code>: The number of top variable features to
subset ONLY for <a href="seurat_curated_workflow.html">Seurat Curated
Workflow</a>.</li>
<li>
<code>altExp</code>: A logical value indicating if the input object
is an altExperiment</li>
<li>
<code>verbose</code>: A logical value indicating if progress should
be printed</li>
</ul></details><details><summary><code>modelGeneVar</code> from Scran package:
</summary><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runModelGeneVar.html">runModelGeneVar</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span></code></pre></div>
<p>The parameters to the above function include:</p>
<ul>
<li>
<code>inSCE</code>: The input <code>SingleCellExperiment</code>
object</li>
<li>
<code>useAssay</code>: Specify the name of the assay to use for
feature selection</li>
</ul></details>
</div>
    </details>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body" entry-spacing="0">
<div id="ref-Butler2018" class="csl-entry">
<div class="csl-left-margin">[1] </div>
<div class="csl-right-inline">A.
Butler, P. Hoffman, P. Smibert, E. Papalexi, and R. Satija, <span>“<span class="nocase">Integrating single-cell transcriptomic data across
different conditions, technologies, and species</span>,”</span>
<em>Nature Biotechnology</em>, vol. 36, no. 5, pp. 411–420, Jun. 2018,
doi: <a href="https://doi.org/10.1038/nbt.4096" class="external-link">10.1038/nbt.4096</a>.</div>
</div>
<div id="ref-Stuart2019" class="csl-entry">
<div class="csl-left-margin">[2] </div>
<div class="csl-right-inline">T.
Stuart <em>et al.</em>, <span>“<span class="nocase">Comprehensive
Integration of Single-Cell Data</span>,”</span> <em>Cell</em>, vol. 177,
no. 7, pp. 1888–1902.e21, Jun. 2019, doi: <a href="https://doi.org/10.1016/j.cell.2019.05.031" class="external-link">10.1016/j.cell.2019.05.031</a>.</div>
</div>
<div id="ref-Satija2017" class="csl-entry">
<div class="csl-left-margin">[3] </div>
<div class="csl-right-inline">R.
Satija, J. A. Farrell, D. Gennert, A. F. Schier, and A. Regev,
<span>“Spatial reconstruction of single-cell gene expression
data,”</span> <em>Nature Biotechnology</em>, vol. 33, pp. 495–502, 2015,
doi: <a href="https://doi.org/10.1038/nbt.3192" class="external-link">10.1038/nbt.3192</a>.</div>
</div>
<div id="ref-Hao2021" class="csl-entry">
<div class="csl-left-margin">[4] </div>
<div class="csl-right-inline">Y.
Hao <em>et al.</em>, <span>“Integrated analysis of multimodal
single-cell data,”</span> <em>Cell</em>, 2021, doi: <a href="https://doi.org/10.1016/j.cell.2021.04.048" class="external-link">10.1016/j.cell.2021.04.048</a>.</div>
</div>
<div id="ref-Aaron2016" class="csl-entry">
<div class="csl-left-margin">[5] </div>
<div class="csl-right-inline">A.
T. L. Lun, D. J. McCarthy, and J. C. Marioni, <span>“A step-by-step
workflow for low-level analysis of single-cell RNA-seq data with
bioconductor,”</span> <em>F1000Research</em>, vol. 5, p. 2122, 2016,
doi: <a href="https://doi.org/10.12688/f1000research.9501.2" class="external-link">10.12688/f1000research.9501.2</a>.</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
