<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Heatmap Visualization • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Heatmap Visualization">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.18.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>

  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Heatmap Visualization</h1>
                        <h4 data-toc-skip class="author">Yichen
Wang</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/heatmap.Rmd" class="external-link"><code>vignettes/articles/heatmap.Rmd</code></a></small>
      <div class="hidden name"><code>heatmap.Rmd</code></div>

    </div>

    
    

<meta name="viewport" content="width=device-width, initial-scale=1">
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>SingleCellTK (SCTK) is equipped with the generic heatmap plotting
functionality, which helps visualizing the feature expression matrices,
with flexible methods of annotation and grouping on either cells or
features. Our base function is a wrapper of <a href="https://rdrr.io/bioc/ComplexHeatmap/" class="external-link"><code>ComplexHeatmap</code></a>
<span class="citation">[1]</span>.</p>
<p>To view detailed instructions on how to use heatmap, please select
‘Interactive Analysis’ for using heatmap in shiny application or
‘Console Analysis’ for using it on R console from the tabs below:
<br></p>
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p><strong>Entry of the Panel</strong></p>
<p><img src="ui_screenshots/heatmap/hm_ui_entry.png" alt="entry"><br></p>
<p>The UI for plotting a generic heatmap is constructed with 5 sections:
data input, data subsetting, annotation adding, parameter setting, and
finally, plotting.</p>
<p><strong>Basic Workflow</strong></p>
 <style>
div.offset { background-color:inherit ; border-radius: 5px; margin-left: 15px;}
</style>
<div class="offset">
<details><summary><b>1. Data Input</b></summary><p><img src="ui_screenshots/heatmap/hm_ui_main1.png"><br></p>
<p>The data used for plotting a heatmap can only be a feature expression
matrix, technically called <code>assay</code> in R language. This will
be set at the selection input <strong>“Assay to Plot”</strong>.</p>
<p>There is a second selection called <strong>“Import from
analysis”</strong> which is used for the fast setting of plotting a
differential expression (DE) analysis or marker detection specific
style. The detail of this feature will be introduced <a href="#additional-usage">later</a>.</p>
</details><details><summary><b>2. Data Subsetting</b></summary><p><img src="ui_screenshots/heatmap/hm_ui_main2_1.png"><br></p>
<p>Similar to the condition definition in the DE analysis panel, here
SCTK also adopts a data table to do a filtering and select the cells and
features, in order to maintain the maximum flexibility. By default, all
the cell annotation available will be displayed as shown in the
screenshot above, as well as the feature annotation.</p>
<p>Generally, to make a selection, users will first find the annotations
which can define the cells or features of interests, apply the filter,
and then press <strong>“Add all filtered”</strong> to finish the
selection. When there is no applicable annotation, we also allow
manually entering a list of cell/feature identifiers to specify the
contents to be plotted.</p>
<details><summary><b>Filter by Categorical Variables</b></summary><p><img src="ui_screenshots/heatmap/hm_ui_main2_1_2.png"><br></p>
<p>To make the operation cleaner, user can choose to only display the
annotation classes of interests at the multi-selection input
<strong>“Columns to display”</strong>. The outcome will look like the
screenshot above. When the annotation of interests is categorical (in
terms of R language, a <code>character</code> or <code>logical</code>
vector or a <code>factor</code>), filters can be applied by selecting
the categories of interest, at the boxes under the variable names.</p>
</details><details><summary><b>Filter by Continuous Variables</b></summary><p><img src="ui_screenshots/heatmap/hm_ui_main2_1_3.png"><br></p>
<p>The screenshots above demonstrates how to filer by limiting a range
to a continuous data (in R, a <code>numeric</code> or
<code>integer</code> vector).</p>
</details><details><summary><b>Select by Entering A List</b></summary><p><img src="ui_screenshots/heatmap/hm_ui_main2_2.png"><br></p>
<p>Finally, if there is no applicable annotation to use, SCTK provides a
third tab that allows direct pasting a list of unique cell/feature
identifiers. As shown above.</p>
<p>The identifiers pasted to the text box should be able to uniquely
identify the cells or features, and they have to be able to be found in
the background information (i.e. the background
<code>SingleCellExperiment</code> (SCE) object). However, they do not
necessarily have to be the default identifiers. That is to say,
identifiers that can be found in existing annotation are also
acceptable. To accomplish this, users only need to select what class of
annotation they intend to use at the selection input <strong>“Match
input cell/feature identifiers by”</strong>. Note that the option “Row
Names” indicates the default identifier when no annotation is
available.</p>
<p>For example, usually in a “cellRanger” style data, the default
feature identifiers would usually be “Ensembl IDs”, while “gene symbols”
will be embedded in the feature annotation. When users only have a list
of genes symbol, they can directly use this list, and select the
annotation for “gene symbol”, instead of going for some other external
tools to transfer the “symbols” into “IDs”.</p>
</details></details><details><summary><b>3. Annotation Adding</b></summary><p><img src="ui_screenshots/heatmap/hm_ui_main3_1.png"><br></p>
<p>In this sub-panel, users can attach the annotations to be shown in
the heatmap legend, and changes in colors are allowed. As mentioned
previously, there are two types of values - categorical and
continuous.</p>
<details><summary><b>Color Settings for Categorical Annotation</b></summary><p>For categorical information, the color setting will look like the
screenshot above. Each category will be displayed with a color input.
Users can click on the color box to change the color for a specific
category if needed.</p>
<p>Note that when categorical annotation is wanted but too many
categories are detected, SCTK will not display any color selection but
set it with default color scheme, in order to avoid a UI “overflow”.</p>
</details><details><summary><b>Color Settings for Continuous Annotation</b></summary><p><img src="ui_screenshots/heatmap/hm_ui_main3_2.png"><br></p>
<p>While for a continuous value, users will still be asked if the value
is categorical or continuous. The reason is that, technically, sometimes
there are annotations generated in a continuous numeric data type but
indicating categorical information, such as clusters labeled with
integers. If users select “Categorical”, the similar color selections
will be displayed as in the previous screenshot (Refer to the previous
folded paragraph). If “Continuous” is selected, a different color
selection UI will be shown, as the screenshot above. Here users will be
asked to set two colors for a color gradient.</p>
</details></details><details><summary><b>4. Other Parameters</b></summary><p><img src="ui_screenshots/heatmap/hm_ui_main4.png"><br></p>
<p>This is the last sub-panel for the generic heatmap setting before
users can generate a plot.</p>
<ul>
<li>Heatmap splitting - multi-selection inputs <strong>“Split columns
(cell) by”</strong> and <strong>“Split rows (feature) by”</strong>. By
splitting the heatmap in row/column, the clustering of rows/columns will
be performed in each splitted group separately. The procedure is also
called semi-heatmap in some cases.</li>
<li>Label adding - checkbox inputs <strong>“Add cell/feature
labels”</strong>. The labels added here are text identifiers attached
next to each row/column. Not recommanded when there are a large amount
of features/cells to plot.</li>
<li>Dendrogram adding - checkbox inputs <strong>“Show dendrograms
for”</strong>. Adding the dendrogram of cell/feature clustering to the
top/left. The dendrograms will also be splitted if the heatmap splitting
is also requested.</li>
<li>Normalization - checkbox input <strong>“Z-Score SCale”</strong>.
Whether to perform z-score normalization to the data matrix to
plot.</li>
<li>Value trimming - numeric sliding input <strong>“Trim”</strong>. Trim
the value in the data matrix to plot to the closer bound of a range if a
value is out of the range. The numeric slider UI dynamically changes
with the data if normalized.</li>
<li>Heatmap color setting - <strong>“Color Scheme”</strong>. User can
set the colors of the highest value, the lowest value and the middle
value for the heatmap. Meanwhile, some frequently used presets are also
provided. s</li>
</ul></details><details><summary><b>5. Plot Displaying</b></summary><p><img src="ui_screenshots/heatmap/hm_ui_plot1.png"><br></p>
<p>After going over all the settings shown in the screenshots of the
previous sections, users can confirm to generate a heatmap by clicking
on the button <strong>“Plot Heatmap”</strong>. A heatmap in the
screenshot above will be generated at the bottom region of the UI.</p>
</details>
</div>
<br><p><strong>Additional Usage</strong></p>
<p>The documentation above is for the general usage of SCTK Heatmap
page. While heatmap is an important visualization method for many types
of analysis, SCTK allows importing some types analysis results.
Currently, DE analysis and marker detection results are supported. Here,
we will present an example of importing the DE analysis result.</p>
<div class="offset">
<details><summary><b>See Detail</b></summary><p><img src="ui_screenshots/heatmap/hm_ui_de_1.png"><br></p>
<p>First, users need to go back to the top of the Heatmap panel, and
find <strong>“Import from analysis”</strong> selection input. On
selecting <code>"Differential Expression"</code>, the session will scan
the background data and show a list of the analysis information for
users to choose from. After selecting the analysis of interests, click
<strong>“Import”</strong>.</p>
<p>The import functionality will automatically fill most of the options
afterwards, including making up some temporary DE specific annotations
and subset the data accordingly.</p>
<p><img src="ui_screenshots/heatmap/hm_ui_de_2.png"><br><img src="ui_screenshots/heatmap/hm_ui_de_3.png"><br></p>
<p>Note that the features automatically selected are all of the ones
stored to the result. In other words, this is affected by the parameters
used when running the DE analysis. However, sometimes users might not
want to plot all of them but apply some further filters instead. For
example, using features with the absolute value of Log2FC greater than
<code>1</code> usually produces cleaner figure, than setting that bound
to <code>0.25</code>. To apply this filter in the UI, users need to
click on the input box below the column title in the table. If the
column is for categorical information, then just click on the categories
of interests; else if it is numeric, users can either manually drag the
sliders or type a “formula” into the input box. The formula follows the
rule: <code>{low}...{high}</code>, where one of the two bounds can be
omitted for a one-end range. For example if users want to apply a filter
described above for Log2FC values, they need to first click
<strong>“Clear selection”</strong>, type <code>1...</code> , click
<strong>“Add all filtered”</strong>, then type <code>...-1</code>, and
click <strong>“Add all filtered”</strong> again.</p>
<p><img src="ui_screenshots/heatmap/hm_ui_de_8.png"><br></p>
<p>Other heatmap settings will also be automatically filled for a DE
specific heatmap.</p>
<p><img src="ui_screenshots/heatmap/hm_ui_de_4.png"><br><img src="ui_screenshots/heatmap/hm_ui_de_5.png"><br><img src="ui_screenshots/heatmap/hm_ui_de_6.png"><br><img src="ui_screenshots/heatmap/hm_ui_de_7.png"><br></p>
</details>
</div>
</div>

<div id="console" class="tabcontent">
<p>To present the usage of <code><a href="../reference/plotSCEHeatmap.html">plotSCEHeatmap()</a></code>, we would like
to use a small example provided with SCTK.</p>
<p><strong>“Raw” plotting</strong></p>
<p>The minimum setting for <code><a href="../reference/plotSCEHeatmap.html">plotSCEHeatmap()</a></code> is the input
SCE object and the data matrix to plot (default
<code>"logcounts"</code>). In this way, all cells and features will be
presented while no annotation or legend (except the main color scheme)
will be shown.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"scExample"</span><span class="op">)</span> <span class="co"># This imports SCE object "sce"</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 200 390 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(200): ENSG00000251562 ENSG00000205542 ... ENSG00000204472</span></span>
<span><span class="co">##   ENSG00000133872</span></span>
<span><span class="co">## rowData names(2): feature_ID feature_name</span></span>
<span><span class="co">## colnames(390): pbmc_4k_GACCAATTCCTAGAAC-1 pbmc_4k_AAATGCCGTTTCGCTC-1</span></span>
<span><span class="co">##   ... pbmc_4k_CTGAAGTTCCACTCCA-1 pbmc_4k_CATCGAACAGCCAGAA-1</span></span>
<span><span class="co">## colData names(4): cell_barcode column_name sample type</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># QC - Remove empty droplets</span></span>
<span><span class="va">sce2</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">sce</span>, colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"type != 'EmptyDroplet'"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Normalize the counts </span></span>
<span><span class="va">sce2</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/runNormalization.html">runNormalization</a></span><span class="op">(</span><span class="va">sce2</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, outAssayName <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>                        normalizationMethod <span class="op">=</span> <span class="st">"logNormCounts"</span>,scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the data</span></span>
<span><span class="fu"><a href="../reference/plotSCEHeatmap.html">plotSCEHeatmap</a></span><span class="op">(</span><span class="va">sce2</span>,useAssay <span class="op">=</span> <span class="st">"logcounts"</span>,cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>, cluster_columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_files/figure-html/setup-1.png" width="700"></p>
<p><strong>Subsetting</strong></p>
<p>SCTK allows relatively flexible approaches to select the
cells/features to plot.</p>
<p>The basic way to subset the heatmap is to directly use an index
vector that can subset the input SCE object to <code>featureIndex</code>
and <code>cellIndex</code>, including <code>numeric</code>, and
<code>logical</code> vectors, which are widely used, and
<code>character</code> vector containing the row/col names. Of course,
user can directly use a subsetted SCE object as input. First let’s run a
simple clustering workflow to identify clusters and find DE genes for
each cluster. We can subset the heatmap using this list of DE genes</p>
<p><img src="heatmap_files/figure-html/idxSubset-1.png" width="700"></p>

<details><summary><b>Using Identifiers in <code>rowData</code>/<code>colData</code></b></summary><p>In a more complex situation, where users might only have a set of
identifiers which are not inside the row/col names (i.e. unable to
directly subset the SCE object), we provide another approach. The
subset, in this situation, can be accessed via specifying a vector that
contains the identifiers users have, to <code>featureIndexBy</code> or
<code>cellIndexBy</code>. This specification allows directly giving one
column name of <code>rowData</code> or <code>colData</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">list_of_FIDs</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000251562"</span>,<span class="st">"ENSG00000205542"</span>,<span class="st">"ENSG00000177954"</span>,<span class="st">"ENSG00000166710"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSCEHeatmap.html">plotSCEHeatmap</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce2</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>, featureIndexBy <span class="op">=</span> <span class="st">"feature_ID"</span>,  featureIndex <span class="op">=</span> <span class="va">list_of_FIDs</span>, cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>, cluster_columns <span class="op">=</span> <span class="cn">TRUE</span>, rowLabel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_files/figure-html/indexBy-1.png" width="700"></p>
</details><p><strong>Adding Annotations</strong></p>
<p>As introduced before, we allow directly using column names of
<code>rowData</code> or <code>colData</code> to attach color bar
annotations. To make use of this functionality, pass a
<code>character</code> vector to <code>rowDataName</code> or
<code>colDataName</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Creat new annotation for markers </span></span>
<span><span class="fu"><a href="../reference/plotSCEHeatmap.html">plotSCEHeatmap</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce2</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>, featureIndex <span class="op">=</span> <span class="va">topMarkers</span><span class="op">$</span><span class="va">Gene</span>, colDataName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"scranSNN_PCA"</span><span class="op">)</span>,rowLabel <span class="op">=</span> <span class="cn">TRUE</span>, cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>, cluster_columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_files/figure-html/colRowAnn-1.png" width="700"></p>
<details><summary><b>Customized Annotation</b></summary><p>Fully customized annotation is also supported, though it can be
complex for users. For the labeling, it is more recommended to insert
the information into <code>rowData</code> or <code>colData</code> and
then make use. For coloring, information should be passed to
<code>featureAnnotationColor</code> or <code>cellAnnotationColor</code>.
The argument must be a <code>list</code> object with names matching the
annotation classes (such as <code>"randLabel"</code> and
<code>"type"</code>); each inner object under a name must be a named
vector, with colors as the values and existing categories as the names.
The working instance looks like this:</p>
</details><p><strong>Others</strong></p>
<p><strong>1. Grouping/Splitting</strong> In some cases, it might be
better to do a “semi-heatmap” (i.e. split the rows/columns first and
cluster them within each group) to visualize some expression pattern,
such as evaluating the differential expression. For this need, use
<code>rowSplitBy</code> or <code>colSplitBy</code>, and the arguments
must be a <code>character</code> vector that is a subset of the
specified annotation.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a new label in the rowData using the cluster markers</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'dplyr'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     combine</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, union</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:GenomeInfoDb':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     collapse, desc, intersect, setdiff, slice, union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     first, intersect, rename, setdiff, setequal, union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     combine, intersect, setdiff, union</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     count</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter, lag</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, setequal, union</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce2</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">topMarkers</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"feature_name"</span> <span class="op">=</span> <span class="st">"Gene"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="st">"cluster_markers"</span> <span class="op">=</span> <span class="st">"scranSNN_PCA"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">new_row_data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce2</span><span class="op">)</span><span class="op">$</span><span class="va">cluster_markers</span> <span class="op">&lt;-</span><span class="va">new_row_data</span><span class="op">$</span><span class="va">cluster_markers</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSCEHeatmap.html">plotSCEHeatmap</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce2</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>, featureIndex <span class="op">=</span> <span class="va">topMarkers</span><span class="op">$</span><span class="va">Gene</span>, colDataName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"type"</span><span class="op">)</span>, aggregateCol <span class="op">=</span> <span class="st">"scranSNN_PCA"</span>, rowGap <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">'mm'</span><span class="op">)</span>,rowLabel <span class="op">=</span> <span class="cn">TRUE</span>, rowDataName <span class="op">=</span> <span class="st">"cluster_markers"</span>, rowSplitBy <span class="op">=</span> <span class="st">"cluster_markers"</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_files/figure-html/split-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Adding a summary </span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce2</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>summary_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">new_col_data</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce2</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu">DataFrame</span><span class="op">(</span><span class="va">new_col_data</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSCEHeatmap.html">plotSCEHeatmap</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce2</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>, featureIndex <span class="op">=</span> <span class="va">topMarkers</span><span class="op">$</span><span class="va">Gene</span>, colDataName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"type"</span><span class="op">)</span>, aggregateCol <span class="op">=</span> <span class="st">"scranSNN_PCA"</span>, rowGap <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">'mm'</span><span class="op">)</span>,rowLabel <span class="op">=</span> <span class="cn">TRUE</span>, rowDataName <span class="op">=</span> <span class="st">"cluster_markers"</span>, rowSplitBy <span class="op">=</span> <span class="st">"cluster_markers"</span>, addCellSummary <span class="op">=</span> <span class="st">"summary_col"</span> <span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_files/figure-html/split-2.png" width="700"></p>
<p><strong>2. Cell/Feature Labeling</strong> Text labels of features or
cells can be added via <code>rowLabel</code> or <code>colLabel</code>.
Use <code>TRUE</code> or <code>FALSE</code> to specify whether to show
the <code>rownames</code> or <code>colnames</code> of the subsetted SCE
object. Additionally, giving a single string of a column name of
<code>rowData</code> or <code>colData</code> can enable the labeling of
the annotation. Furthermore, users can directly throw a character vector
to the parameter, with the same length of either the full SCE object or
the subsetted.</p>
<p><strong>3. Dendrograms</strong> The dendrograms for features or cells
can be removed by passing <code>FALSE</code> to <code>rowDend</code> or
<code>colDend</code>.</p>
<p><strong>4. Row/Column titles</strong> The row title
(<code>"Genes"</code>) and column title (<code>"Cells"</code>) can be
changed or removed by passing a string or <code>NULL</code> to
<code>rowTitle</code> or <code>colTitle</code>, respectively.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSCEHeatmap.html">plotSCEHeatmap</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce2</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>, featureIndex <span class="op">=</span> <span class="va">topMarkers</span><span class="op">$</span><span class="va">Gene</span>, rowGap <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">'mm'</span><span class="op">)</span>,rowLabel <span class="op">=</span> <span class="cn">TRUE</span>,  rowTitle <span class="op">=</span> <span class="st">"Markers"</span>,colTitle <span class="op">=</span> <span class="st">"Clusters"</span>, cluster_columns <span class="op">=</span> <span class="cn">TRUE</span>, cluster_rows <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_files/figure-html/label-1.png" width="700"></p>
<p>There are still some parameters not mentioned here, but they are not
frequently used. Please refer to <code><a href="../reference/plotSCEHeatmap.html">?plotSCEHeatmap</a></code> as well as
<code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">?ComplexHeatmap::Heatmap</a></code>.</p>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body" entry-spacing="0">
<div id="ref-Gu2016" class="csl-entry">
<div class="csl-left-margin">[1] </div>
<div class="csl-right-inline">Z.
Gu, R. Eils, and M. Schlesner, <span>“<span class="nocase">Complex
heatmaps reveal patterns and correlations in multidimensional genomic
data</span>,”</span> <em>Bioinformatics</em>, vol. 32, pp. 2847--2849,
Sep. 2016, doi: <a href="https://doi.org/10.1093/bioinformatics/btw313" class="external-link">10.1093/bioinformatics/btw313</a>.</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
