<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Manage Annotation of Cell and Features • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Manage Annotation of Cell and Features">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.14.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Manage Annotation of Cell and Features</h1>
                        <h4 data-toc-skip class="author">Yichen
Wang</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/import_annotation.Rmd" class="external-link"><code>vignettes/articles/import_annotation.Rmd</code></a></small>
      <div class="hidden name"><code>import_annotation.Rmd</code></div>

    </div>

    
    

<meta name="viewport" content="width=device-width, initial-scale=1">
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This documentation mainly focuses on the metadata variable that can
be projected to each cell or feature, such as IDs and clinical
conditions. Single Cell Toolkit (singleCellTK, SCTK) allows users to
import and manage cell level and feature level annotations to help with
analysis such as filtering the features for less noise, grouping the
cells for differential expression analysis, and etc. Many of the
analysis steps that SCTK does would also produce annotations, such as
the QC metrics, clustering labels and etc.</p>
<p>To view detailed instructions on how to manipulate the annotation,
please select ‘Interactive Analysis’ for manipulation through shiny
application or ‘Console Analysis’ for manipulation through the R console
from the tabs below: <br></p>
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p>After successfully uploading single-cell dataset, users are allowed
to manipulate the cell and feature level annotations in the UI. <em>The
operation for the two types of annotation is distributed to two
separated tabs, which work in the same way. This documentation will only
describe the usage for the cell annotation tab.</em></p>
<p><img src="ui_screenshots/cell_annotation/cell_ann_ui_entry.png" alt="entry"><br></p>
<p>Users can enter the interface as instructed by the screenshot above.
With this interface, users can either upload extra annotations or edit
existing annotations. The whole tab consists of three parts:</p>
<ol style="list-style-type: decimal">
<li>The editing area within “<strong>Options for editing and importing
Cell Annotation data</strong>” panel;</li>
<li>Save/reset button;</li>
<li>The demonstration area within <strong>“Table of Cell
Annotations”</strong> panel.</li>
</ol>
<p>Here we will take the example PBMC3K dataset as an example. We
imported this dataset through the <strong>“Data” -&gt; “Import Single
Cell Data” tab</strong>, by selecting <strong>“Import example
datasets”</strong> first and using the default “PBMC 3K (10X)” option
then.</p>
<p><strong>General Workflow</strong></p>
<ol style="list-style-type: decimal">
<li>Users can start with making any type of operation within
“<strong>Options for editing and importing Cell Annotation
data</strong>” panel. The detail of each type of operation will be
described later.</li>
<li>After clicking on <strong>“Confirm XX”</strong> button that follows
the specific inputs, the effect will be immediately shown in the
<strong>“Table of Cell Annotations”</strong>. <em>It is important to
know that this table is just for demonstration of the result. Simply
clicking on a “Confirm” button does not take any direct effect to the
background data object.</em>
</li>
<li>After any changes are confirmed, users must click on
<strong>“Save”</strong> button to take them into effect. A
<strong>“Reset”</strong> button is also provided in case any unwanted
operation happens. <em>When users click on the <strong>“Reset”</strong>
button, the annotation table will be reset to the status after the last
time users “save” any change. Or if no change has been “saved”, it will
be reset to the original status.</em>
</li>
</ol>
<p><strong>Detailed Demonstration</strong></p>
 <style>
div.offset { background-color:inherit ; border-radius: 5px; margin-left: 15px;}
</style>
<div class="offset">
<details><summary><b>Import New Annotation from File</b></summary><p>Users can upload a metadata table that annotates the cells in the
dataset with multiple variables. The file has to be in Comma-Separated
Values (CSV) format, and have the first header row for the variable
names. All cells in the file have to match with the existing dataset in
both order and amount, so that SCTK can correctly identify the
information and merge it to the background data container.</p>
<p>Users can choose to replace all existing annotation with the new
table, by selecting the radio button <strong>“Replace Cell
Annotations”</strong>, or merge the new table into the existing table,
by selecting <strong>“Add to existing Cell Annotations”</strong>. Note
that by choosing to merge, SCTK will overwrite the the variables with
the same name.</p>
</details><details><summary><b>Edit Existing Annotation</b></summary><div class="offset">
  <details><summary><b>Bins</b></summary><p>Users can modify an existing column of annotation by changing all
values that meet a criteria (alphabetically ordered comparison by “=”,
“&gt;”, “&lt;”, “&gt;=” or “&lt;=”) to a new value. In the UI, users can
choose the annotation at the selection input <strong>“select
attribute”</strong>, and then set the binning point by selecting an
existing value at the selection input <strong>“select attribute
value”</strong> or entering the desired value at <strong>“criteria
parameter”</strong>. The criteria type (comparison symbol) can be
selected at <strong>“select comparison”</strong>. The new value to be
filled in to the bin can be entered at <strong>“bin name”</strong>.
Finally, press <strong>“Confirm Bin”</strong> button to temporarily save
the result and preview it in the table at the bottom of the page.</p>
<blockquote>
<p>For example: There is an annotation named <code>"ann1"</code>, with
values <code>"a", "b", "c", "d", "e", ...</code> in the dataset, and
users want to change all the values in <code>"ann1"</code>, that come
after <code>"d"</code>, to <code>"afterD"</code>. The steps would be: 1.
Choose <code>"ann1"</code> at <strong>“select attribute”</strong>; 2.
Select <code>"d"</code> at <strong>“select attribute value”</strong> or
enter <code>"d"</code> at <strong>“criteria parameter”</strong>; 3.
Select <code>&gt;</code> at <strong>“select comparison”</strong>; 4.
Enter <code>"afterD"</code> at <strong>“bin name”</strong>.</p>
</blockquote>
</details><details><summary><b>Merge columns</b></summary><p><img src="ui_screenshots/cell_annotation/cell_ann_ui_merge.png" alt="entry"><br></p>
<p>Users can merge any two columns that exist in the cell annotation
table. SCTK merges them in a direct concatenate manner, with allowing
setting a customized separator. As shown in the screenshot above, user
should enter the <strong>“Merge Columns” tab</strong> to access this
functionality. Different values should be selected for the
<strong>“Select first column”</strong> and <strong>Select second
column”</strong> input. The order of these selections matters as it will
determine how the values are ordered in the merged field. It is
recommended that users set an <code>_</code> (underscore) or a
<code>-</code> (dash) as the input for <strong>“add separator between
merged values”</strong>. Setting a <code></code> (space) here will not
trigger any direct error but is not a best practice. After determining
all the three inputs, users should click on <strong>“Confirm
Merge”</strong> button to make the change. The effect can be immediately
observed in the <strong>“Table of Cell Annotations”</strong> at the
bottom, as shown in the lower part of the screenshot above.</p>
</details><details><summary><b>Magic Fill</b></summary><p>Users can modify a column of annotation (column A) according to
another column (column B). The rows that has a certain value in column B
will be selected and these rows in column A will be modified to a
user-entered value. Select column B at <strong>“select attribute
column”</strong>, and column A at <strong>“select column to
fill”</strong>. The value to be found in column B should be selected at
<strong>“select attribute value”</strong>, and the new value to fill
into column A should be entered at <strong>“new value”</strong>.</p>
</details><details><summary><b>Clean</b></summary><p><img src="ui_screenshots/cell_annotation/cell_ann_ui_clean.png" alt="entry"><br></p>
<p>Users can clean up a selected column by removing any “alphabets”,
“digits”, “spaces”, or “symbols” from all the values in this column.
First users will need to specify the column at <strong>“Select attribute
column”</strong>, and choose what type of clean-up should be performed
at <strong>“select removal criteria”</strong>. Then users must click on
<strong>“Confirm Clean”</strong> to enable the change. The effect can be
immediately observed in the <strong>“Table of Cell Annotations”</strong>
at the bottom, as shown in the lower part of the screenshot above.</p>
</details><details><summary><b>Add column</b></summary><p><img src="ui_screenshots/cell_annotation/cell_ann_ui_newCol.png" alt="entry"><br></p>
<p>Users can add a new column where all values are the same. Users will
need to define the name of the new column at <strong>“enter new empty
column name”</strong>, and the value of this column at <strong>“default
value to fill”</strong>. Then users need to click on <strong>“Create
Column”</strong> make the change. The effect can be immediately observed
in the <strong>“Table of Cell Annotations”</strong> at the bottom, as
shown in the screenshot above.</p>
</details><details><summary><b>Delete column</b></summary><p><img src="ui_screenshots/cell_annotation/cell_ann_ui_delete.png" alt="entry"><br></p>
<p>User can remove an entire column. The unwanted column can be deleted
by simply selecting the name of the column at <strong>“select attribute
to delete”</strong> and click on the <strong>“Delete”</strong> button.
The effect can be immediately observed in the <strong>“Table of Cell
Annotations”</strong> at the bottom, as shown in the screenshot
above.</p>
  </details>
</div>
</details>
</div>
</div>

<div id="console" class="tabcontent">
<p>SCTK uses the <a href="https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>
object as the container of users’ datasets. Based on this approach, the
cell level and feature level annotations are all stored in the
<code>colData</code> slot and <code>rowData</code> slot,
respectively.</p>
<p>Here, we present some of the most basic operations on cell
annotations, with <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> method. For operations on
feature annotations, replace <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> with
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a dataset</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Add new information to SCE object</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span> <span class="co"># A numeric vector of the same size as the number of cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">n</span> <span class="op">&lt;-</span> <span class="va">n</span></span>
<span></span>
<span><span class="co">## Extract information from SCE</span></span>
<span><span class="va">barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Barcode</span></span>
<span></span>
<span><span class="co">## Remove an annotation</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Tissue_status</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></code></pre></div>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
