<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Normalization • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Normalization">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.12.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Normalization</h1>
                        <h4 data-toc-skip class="author">Irzam
Sarfraz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/normalization.Rmd" class="external-link"><code>vignettes/articles/normalization.Rmd</code></a></small>
      <div class="hidden name"><code>normalization.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><strong>singleCellTK</strong> offers a convenient way to normalize
data for downstream analysis using a number of methods available through
the toolkit. The integrated normalization methods and other
transformation options can be used through both R console and
interactive shiny application. Normalization methods available with the
toolkit include <code>"LogNormalize"</code>, <code>"CLR"</code>,
<code>"RC"</code> &amp; <code>"SCTransform"</code> from
<code>Seurat</code> [<span class="citation">[1]</span>]<span class="citation">[2]</span>[<span class="citation">[3]</span>]<span class="citation">[4]</span> package and <code>"logNormCounts"</code> or
<code>"CPM"</code> from <code>Scater</code> <span class="citation">[5]</span> package. Additional transformation options
are available for users including ‘log’ &amp; ‘log1p’, trimming of data
assays and Z-Score scaling. <br></p>
<p>To view detailed instructions on how to use these methods, please
select ‘Interactive Analysis’ for using normalization in shiny
application or ‘Console Analysis’ for using these methods on R console
from the tabs below:</p>

<meta name="viewport" content="width=device-width, initial-scale=1">
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p>Normalization tab can be opened up by clicking on the
<strong>Normalization &amp; Batch Correction</strong> from the top menu
and further selecting the <strong>Normalization</strong> sub-tab in the
subsequent window as shown below.</p>
<p><img src="ui_screenshots/normalization/normalization_1_select_menu.PNG"><br><strong>Normalization tab</strong> is divided up into 4 sections as
shown in the screenshot below: <br></p>
<p><img src="ui_screenshots/normalization/normalization_1.PNG"><br>
The section <strong>a</strong> lets you choose a normalization method,
whereas the section <strong>b</strong> allows for various options to be
modified and selected for a particular normalization method selected
from section <strong>a</strong>. The section <strong>c</strong> displays
the overall methods/options selected for this instance of normalization
process. The section <strong>d</strong> displays the output data type
once the normalization is run. This output data type represents the
nature of the data that will be output based on the selected options.
For example, output data type maybe <strong>transformed</strong> if
transformation options are selected including normalization methods, or
maybe <strong>scaled</strong> if the <strong>scale</strong> option is
selected. The section <strong>c</strong> and <strong>d</strong> update
automatically based on normalization method and options selected by the
user. <br></p>
<p><img src="ui_screenshots/normalization/normalization_2.png"><br><strong>1.</strong> Select a normalization method or select
<strong>Custom Normalization</strong> to manually select the options for
normalization including transformation, trimming &amp; scaling.<br></p>
<p><img src="ui_screenshots/normalization/normalization_3.PNG"><br><strong>2a.</strong> Options when a normalization method from a
particular package is selected: <br> i. Select the assay from the data
to normalize. <br> ii. Set scaling factor for this particular method.
These options may vary between different normalization methods. See
method specific documentation for more information. <br> iii. Specify
the name of the output data assay. <br> iv. Select if data should be
scaled after running the normalization method. <br> v. Select if data
should be trimmed between two values running the normalization method.
<br> vi &amp; vii. If trim data option is selected, specify the upper
and lower trim values. The output data is then trimmed between these two
values (inclusive). <br> viii: The user-interface displays the selected
options or methods based on the user selected normalization method. <br>
ix. The output data type based on the selected user options is
displayed. <br> x. Click the button to run the normalization process.
<br></p>
<p><img src="ui_screenshots/normalization/normalization_4.PNG"><br><strong>2b.</strong> Options when a custom normalization is selected:
<br> i. Select the input data assay to use as an input with selected
normalization/transformation options. <br> ii. Specify the name of the
put data assay. <br> iii. Select if data should be normalized (multiple
normalization methods will display once this option is selected). <br>
iv. Select if a pseudo value should be added to the data. <br> v. Select
if the data should be scaled. <br> vi. Select if the data should be
trimmed between two values. <br> vii. The user-interface displays the
selected options or methods based on the user selected normalization
method. <br> viii. The output data type based on the selected user
options is displayed. <br> ix. Click the button to run the normalization
process. <br></p>
</div>
<div id="console" class="tabcontent">
<p>The <code>singleCellTK</code> allows the users to run all
normalization and transformation methods on the input data by using a
single <code>runNormalization</code> function. The
<code>runNormalization</code> function takes in input a
<code>SingleCellExperiment</code> object and a series of parameters that
define the normalization/transformation options to run on the specified
<code>assay</code>. The output of this function is a
<code>SingleCellExperiment</code> object which now contains the
normalized/transformed assay.</p>
<p>To use the function, input a <code>SingleCellExperiment</code> object
that contains the data assay and specify the required parameters:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runNormalization.html">runNormalization</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, normalizationMethod <span class="op">=</span> <span class="st">"RC"</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, outAssayName <span class="op">=</span> <span class="st">"RCLogScaledCounts"</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, transformation <span class="op">=</span> <span class="st">"log2"</span>, pseudocountsBeforeTransform <span class="op">=</span> <span class="fl">1</span>, trim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="op">-</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<details><summary><b>Example</b></summary><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co"># Load singleCellTK &amp; pbmc3k example data</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span>  <span class="co"># Perform normalization</span></span>
<span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runNormalization.html">runNormalization</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, normalizationMethod <span class="op">=</span> <span class="st">"RC"</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, outAssayName <span class="op">=</span> <span class="st">"RCLogScaledCounts"</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, transformation <span class="op">=</span> <span class="st">"log2"</span>, pseudocountsBeforeTransform <span class="op">=</span> <span class="fl">1</span>, trim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="op">-</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><details><summary><b>Parameters</b></summary><p>The <code>runNormalization</code> function specifies the following
parameters:</p>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Parameter</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>inSCE</td>
<td>Input <code>SingleCellExperiment</code> object.</td>
</tr>
<tr class="even">
<td>useAssay</td>
<td>Specify the input <code>assay</code> to use for
normalization/transformation.</td>
</tr>
<tr class="odd">
<td>outAssayName</td>
<td>A <code>character</code> value indicating the name of the new output
assay.</td>
</tr>
<tr class="even">
<td>normalizationMethod</td>
<td>Specify a normalization method from <code>"LogNormalize"</code>,
<code>"CLR"</code>, <code>"RC"</code>, <code>"SCTransform"</code>,
<code>"logNormCounts"</code> or <code>"CPM"</code>. If no method is
specified, normalization will not be performed.</td>
</tr>
<tr class="odd">
<td>scale</td>
<td>Logical value indicating if Z.Score scaling should be performed or
not.</td>
</tr>
<tr class="even">
<td>seuratScaleFactor</td>
<td>Specify the <code>scaleFactor</code> parameter if any of the
<strong>seurat</strong> normalization method is selected.</td>
</tr>
<tr class="odd">
<td>transformation</td>
<td>Specify if a transformation should be applied to the input
<code>assay</code> (if normalization is selected, this transformation is
applied after normalization). Available transformation options include
<code>"log2"</code>, <code>"log1p"</code>, <code>"sqrt"</code>.</td>
</tr>
<tr class="even">
<td>pseudocountsBeforeNorm</td>
<td>A <code>numeric</code> value to add to the input <code>assay</code>
before performing normalization.</td>
</tr>
<tr class="odd">
<td>pseudocountsBeforeTransform</td>
<td>A <code>numeric</code> value to add to the input <code>assay</code>
before performing a transformation.</td>
</tr>
<tr class="even">
<td>trim</td>
<td>A <code>numeric(2)</code> vector that specifies the upper and the
lower trim values between (exclusive) which the input <code>assay</code>
should be trimmed.</td>
</tr>
<tr class="odd">
<td>verbose</td>
<td>A <code>logical</code> value indicating if informative/progress
messages should be displayed on the console.</td>
</tr>
</tbody>
</table></details><details><summary><b>Individual Functions</b></summary><p>While the <code>runNormalization</code> wrapper function can be used
to perform all available normalization and transformation options,
separate functions are also available for all of the included
normalization methods. The following functions can be used for specific
normalization methods:</p>
 <style>
div.offset { background-color:inherit ; border-radius: 5px; margin-left: 15px;}
</style>
<div class="offset">
<details><summary><code>LogNormalize</code>, <code>CLR</code> or <code>RC</code> from
<strong>Seurat</strong> [<span class="citation">[1]</span>]<span class="citation">[2]</span> package:
</summary><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratNormalizeData.html">runSeuratNormalizeData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, normalizationMethod <span class="op">=</span> <span class="st">"CLR"</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, normAssayName <span class="op">=</span> <span class="st">"CLRCounts"</span>, scaleFactor <span class="op">=</span> <span class="fl">10000</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The parameters to the above function include: <br><br><code>inSCE</code>: an input <code>SingleCellExperiment</code> object
<br><code>normalizationMethod</code>: one of the options from
“LogNormalize”, “CLR” or “RC” <br><code>useAssay</code>: name of the
assay to use for normalization <br><code>normAssayName</code>: name of
the output normalized assay <br><code>scaleFactor</code>: a numeric
value that represents the scaling factor <br><code>verbose</code>: a
logical value indicating if progress should be printed <br><br></p>
</details><details><summary><code>SCTransform</code> from <strong>Seurat</strong> [<span class="citation">[1]</span>]<span class="citation">[2]</span><span class="citation">[6]</span> package:
</summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratSCTransform.html">runSeuratSCTransform</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, normAssayName <span class="op">=</span> <span class="st">"SCTCounts"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The parameters to the above function include: <br><br><code>inSCE</code>: an input <code>SingleCellExperiment</code> object
<br><code>useAssay</code>: name of the assay to use for normalization
<br><code>normAssayName</code>: name of the output normalized assay
<br><code>verbose</code>: a logical value indicating if progress should
be printed <br><br></p>
</details><details><summary><code>logNormCounts</code> from <strong>Scater</strong> <span class="citation">[5]</span> package:
</summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaterlogNormCounts.html">scaterlogNormCounts</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, assayName <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span></code></pre></div>
<p>The parameters to the above function include: <br><br><code>inSCE</code>: an input <code>SingleCellExperiment</code> object
<br><code>useAssay</code>: name of the assay to use for normalization
<br><code>assayName</code>: name of the output normalized assay
<br><br></p>
</details><details><summary><code>CPM</code> from <strong>Scater</strong> <span class="citation">[5]</span> package:
</summary><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaterCPM.html">scaterCPM</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, assayName <span class="op">=</span> <span class="st">"countsCPM"</span><span class="op">)</span></span></code></pre></div>
<p>The parameters to the above function include: <br><br><code>inSCE</code>: an input <code>SingleCellExperiment</code> object
<br><code>useAssay</code>: name of the assay to use for normalization
<br><code>assayName</code>: name of the output normalized assay
<br><br></p>
</details><details><summary>Compute <b>Z-Score</b>:</summary><div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"countsZScore"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeZScore.html">computeZScore</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The parameters to the above function include: <br><br><code>counts</code>: input matrix to scale with Z-Score <br><br></p>
</details><details><summary>Matrix <b>trimming</b>:</summary><div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"countsTrimmed"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trimCounts.html">trimCounts</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"counts"</span><span class="op">)</span>, trimValue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="op">-</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The parameters to the above function include: <br><br><code>counts</code>: an input <code>SingleCellExperiment</code> object
<br><code>trimValue</code>: a <code>numeric(2)</code> vector with the
first value indicating the upper trim value and the second value
indicating the lower trim value <br><br></p>
</details>
</div>
    </details>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body" entry-spacing="0">
<div id="ref-Butler2018" class="csl-entry">
<div class="csl-left-margin">[1] </div>
<div class="csl-right-inline">A.
Butler, P. Hoffman, P. Smibert, E. Papalexi, and R. Satija, <span>“<span class="nocase">Integrating single-cell transcriptomic data across
different conditions, technologies, and species</span>,”</span>
<em>Nature Biotechnology</em>, vol. 36, no. 5, pp. 411–420, Jun. 2018,
doi: <a href="https://doi.org/10.1038/nbt.4096" class="external-link">10.1038/nbt.4096</a>.</div>
</div>
<div id="ref-Stuart2019" class="csl-entry">
<div class="csl-left-margin">[2] </div>
<div class="csl-right-inline">T.
Stuart <em>et al.</em>, <span>“<span class="nocase">Comprehensive
Integration of Single-Cell Data</span>,”</span> <em>Cell</em>, vol. 177,
no. 7, pp. 1888–1902.e21, Jun. 2019, doi: <a href="https://doi.org/10.1016/j.cell.2019.05.031" class="external-link">10.1016/j.cell.2019.05.031</a>.</div>
</div>
<div id="ref-Satija2017" class="csl-entry">
<div class="csl-left-margin">[3] </div>
<div class="csl-right-inline">R.
Satija, J. A. Farrell, D. Gennert, A. F. Schier, and A. Regev,
<span>“Spatial reconstruction of single-cell gene expression
data,”</span> <em>Nature Biotechnology</em>, vol. 33, pp. 495–502, 2015,
doi: <a href="https://doi.org/10.1038/nbt.3192" class="external-link">10.1038/nbt.3192</a>.</div>
</div>
<div id="ref-Hao2021" class="csl-entry">
<div class="csl-left-margin">[4] </div>
<div class="csl-right-inline">Y.
Hao <em>et al.</em>, <span>“Integrated analysis of multimodal
single-cell data,”</span> <em>Cell</em>, 2021, doi: <a href="https://doi.org/10.1016/j.cell.2021.04.048" class="external-link">10.1016/j.cell.2021.04.048</a>.</div>
</div>
<div id="ref-McCarthy2017" class="csl-entry">
<div class="csl-left-margin">[5] </div>
<div class="csl-right-inline">D.
J. McCarthy, K. R. Campbell, A. T. L. Lun, and Q. F. Wills, <span>“<span class="nocase">Scater: Pre-processing, quality control, normalization
and visualization of single-cell RNA-seq data in R</span>,”</span>
<em>Bioinformatics</em>, vol. 33, no. 8, pp. 1179–1186, Apr. 2017, doi:
<a href="https://doi.org/10.1093/bioinformatics/btw777" class="external-link">10.1093/bioinformatics/btw777</a>.</div>
</div>
<div id="ref-Hafemeister2019" class="csl-entry">
<div class="csl-left-margin">[6] </div>
<div class="csl-right-inline">C.
Hafemeister and R. Satija, <span>“<span class="nocase">Normalization and
variance stabilization of single-cell RNA-seq data using regularized
negative binomial regression</span>,”</span> <em>Genome Biology</em>,
vol. 20, no. 1, pp. 1–15, Dec. 2019, doi: <a href="https://doi.org/10.1186/s13059-019-1874-1" class="external-link">10.1186/s13059-019-1874-1</a>.</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
