<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pathway Analysis • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Pathway Analysis">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/console_analysis_tutorial.html">A La Carte Workflow - PBMC3K (R Console)</a>
    </li>
    <li>
      <a href="../articles/cnsl_seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/cnsl_2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Enrichment &amp; Pathway Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Pathway Analysis</h1>
                        <h4 data-toc-skip class="author">Nida Pervaiz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/pathwayAnalysis.Rmd" class="external-link"><code>vignettes/articles/pathwayAnalysis.Rmd</code></a></small>
      <div class="hidden name"><code>pathwayAnalysis.Rmd</code></div>

    </div>

    
    

<meta name="viewport" content="width=device-width, initial-scale=1">
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><strong>VAM</strong> <span class="citation">[1]</span> and <strong>GSVA</strong> <span class="citation">[2]</span> are popular methods for GSE (Gene Set Enrichment) and allows the identification of changes in pathway activity in RNA-Seq data. Overall, gene sets can be uploaded or selected from the available databases (<a href="import_genesets.html">How to <em>Import Genesets</em>?</a>), and the pathway analysis method can be run on the selected gene sets which can eventually be visualized using Violin plot.</p>
<p>To view detailed instructions on how to use these methods, please select “Interactive Analysis” for using pathway analysis in Shiny application or “Console Analysis” for using these methods on R console from the tabs below:</p>
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p><strong>Entry of The Panel</strong></p>
<p>From anywhere of the UI, the panel for pathway analysis can be accessed from the top navigation panel at the circled tab shown below.</p>
<p><img src="ui_screenshots/pathwayAnalysis/path_ui_entry.PNG" alt="PathwayEntry"><br></p>
<p>The UI is constructed in a sidebar style, where the left-sided sidebar works for setting the parameters and running the pathway analysis, and the right part main panel is for visualization checking.</p>
<p><strong>Run Pathway Analysis</strong></p>
<p><strong>1. Choose An Algorithm</strong></p>
<p><img src="ui_screenshots/pathwayAnalysis/path_ui_algo.png" alt="PathwayAlgo"><br></p>
<p>For running any types of pathway analysis, there are always three essential inputs that users should be sure with:</p>
<ul>
<li>The data matrix to use - selection input <strong>“Select Assay”</strong>. In terms of pathway analysis, SCTK always requires a full-sized feature expression data (i.e. assay) as a valid input.</li>
<li>The pathway analysis method - selection input <strong>“Select Method”</strong>. All methods supported are listed as options.</li>
<li>The geneset collection name - selection input <strong>“Select Geneset Collection”</strong>. All geneset names are listed as options.</li>
</ul>
<p>After the pathway analysis method is confirmed, the lower part will dynamically switch to the method specific settings.</p>
<p><strong>2. Parameter Settings</strong></p>
 <style>
div.offset { background-color:inherit ; border-radius: 5px; margin-left: 15px;}
</style>
<div class="offset">
<details><summary><b>VAM</b></summary><p><img src="ui_screenshots/pathwayAnalysis/path_ui_VAM.PNG" alt="VAMParameter"><br></p>
<p>When the selected algorithm is “VAM” , the parameter settings will look like the figure above. The method specific parameters include:</p>
<ul>
<li>
<code>center</code>, If True, values will be mean centered when computating the Mahalanobis statistic.</li>
<li>
<code>gamma</code>, If True, a gamma distribution will be fit to the non-zero squared Mahalanobis distances computed from a row-permuted version of the gene expression matrix. The estimated gamma distribution will be used to compute a one-sided p-value for each cell. If False, the p-value will be computed using the standard chi-square approximation for the squared Mahalanobis distance (or non-central if center = FALSE).</li>
</ul></details><details><summary><b>GSVA</b></summary><p><img src="ui_screenshots/pathwayAnalysis/path_ui_gsva.PNG" alt="GSVAParameter"><br></p>
<p>When the selected algorithm is “GSVA,” no additional parameter settings exist other than the three essential inputs common for both VAM and GSVA.</p>
</details>
</div>
<br><p><strong>3. Visualization</strong></p>
<p><img src="ui_screenshots/pathwayAnalysis/path_ui_result.PNG" alt="Pathwayresult"><br></p>
<p>The visualization is implemented with a plotting of cell by geneset matrix score.</p>
<div class="offset">
<details><summary><b>Detail</b></summary><p><img src="ui_screenshots/pathwayAnalysis/path_ui_visualization.PNG" alt="Pathwayvisualization"><br></p>
<p>The parameter panel for visualization comprises of:</p>
<ul>
<li>The result to visualize - selection input <strong>“Select Score matrix which you want to plot:”</strong>. User should choose one of the data frames produced within the results which they want to visualize.</li>
<li>The geneset to visualize - selection input <strong>“Select Geneset:”</strong>. User should choose the geneset from the collection used for running the algorithm.</li>
<li>The additional condition of grouping the data- selection input <strong>“Select Condition(s) of interest to group data (optional)”</strong>. User may select a condition to group the data in accordance with the colData.</li>
<li>Visualization criteria - Boolean input <strong>“Boxplot”</strong>. If TRUE, will plot boxplots for each violin plot. Default TRUE</li>
<li>Visualization criteria - Boolean input <strong>“Violinplot”</strong>. If TRUE, will plot the violin plot. Default TRUE</li>
<li>Summarize stats - selection input <strong>“Select summary parameter”</strong>. Adds a summary statistic, as well as a crossbar to the violin plot. Options are “mean” or “median.” Default NULL.</li>
</ul></details>
</div>
</div>

<div id="console" class="tabcontent">
<p>The pathway analysis methods can be easily applied to a <a href="https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link"><code>SingleCellExperiment</code></a> (SCE) object. The methods return the same SCE object with results updated in <code>reducedDim</code> slot.</p>
<p><strong>Basic Parameters</strong></p>
<p>For each of the three functions, the common things are:</p>
<ul>
<li>The input SCE object - <code>inSCE</code>.</li>
<li>The user must select the gene sets they would like to upload from <code>Import Gene Sets</code> tab - <code>geneSetCollectionName</code>.</li>
<li>Users need to use <code>useAssay</code> for a full-sized expression data.</li>
</ul>
<p>As for another essential parameter, the data matrix used for running the algorithm is a must. Here, users should notice that:</p>
<ul>
<li>For <code><a href="../reference/runVAM.html">runVAM()</a></code>, there are two extra parameters: <code>center</code> and <code>gamma</code>. Please refer to the function manual pages for the details.</li>
</ul>
<p><strong>Examples</strong></p>
<p>To demonstrate simple and clear examples, here we use the “PBMC-3k” dataset from “10X” which can be easily imported with SCTK functions.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></code></pre></div>
<p>The preprocessing only includes necessary steps before implementing pathway analysis</p>
<div class="offset">
<details><summary><b>Preprocessing</b></summary><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaterlogNormCounts.html">scaterlogNormCounts</a></span><span class="op">(</span><span class="va">sce</span>, assayName <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importGeneSetsFromMSigDB.html">importGeneSetsFromMSigDB</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, categoryIDs <span class="op">=</span> <span class="st">"H"</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, mapping <span class="op">=</span> <span class="st">"gene_symbol"</span>, by <span class="op">=</span> <span class="st">"rownames"</span><span class="op">)</span></code></pre></div>
</details><details><summary><b>Example with VAM</b></summary><details><summary><b>Method specific parameters</b></summary><ul>
<li>
<code>center</code>, If True, values will be mean centered when computating the Mahalanobis statistic.</li>
<li>
<code>gamma</code>, If True, a gamma distribution will be fit to the non-zero squared Mahalanobis distances computed from a row-permuted version of the gene expression matrix. The estimated gamma distribution will be used to compute a one-sided p-value for each cell. If False, the p-value will be computed using the standard chi-square approximation for the squared Mahalanobis distance (or non-central if center = FALSE).</li>
</ul></details><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Directly use an assay</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runVAM.html">runVAM</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, geneSetCollectionName <span class="op">=</span> <span class="st">"H"</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, gamma <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</details><details><summary><b>Example with GSVA</b></summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Directly use an assay</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runGSVA.html">runGSVA</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, geneSetCollectionName <span class="op">=</span> <span class="st">"H"</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></code></pre></div>
</details>
</div>
<p><strong>Results and Visualization</strong></p>
<p>The results generated with either GSVA or VAM are stored in the <code>reducedDims</code> slot of the SCE object.</p>
<p>To visualize the scores for a specific geneset from the geneset collection that is used for running the algorithm:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotPathway.html">plotPathway</a></span><span class="op">(</span><span class="va">sce</span>, resultName <span class="op">=</span> <span class="st">"VAM_H_CDF"</span>, geneset <span class="op">=</span> <span class="st">"HALLMARK_INFLAMMATORY_RESPONSE"</span><span class="op">)</span></code></pre></div>
<p>In the example above, <code>"VAM_H_CDF"</code> is the name of the result generated by VAM, where <code>"HALLMARK_INFLAMMATORY_RESPONSE"</code> is a geneset from geneset collection <code>"H"</code>, imported at the very beginning.</p>
<p>To get the names of all pathway analysis results:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/getPathwayResultNames.html">getPathwayResultNames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
<p>To get a list of all genesets that a pathway analysis result includes:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/getGenesetNamesFromCollection.html">getGenesetNamesFromCollection</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, geneSetCollectionName <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span></code></pre></div>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body">
<div id="ref-Frost2020.02.18.954321" class="csl-entry">
<div class="csl-left-margin">[1] </div>
<div class="csl-right-inline">H. R. Frost, <span>“Variance-adjusted mahalanobis (VAM): A fast and accurate method for cell-specific gene set scoring,”</span> <em>Nucleic Acids Research</em>, 2020, doi: <a href="https://doi.org/10.1093/nar/gkaa582" class="external-link">10.1093/nar/gkaa582</a>.</div>
</div>
<div id="ref-Hanzelmann2013" class="csl-entry">
<div class="csl-left-margin">[2] </div>
<div class="csl-right-inline">S. Hänzelmann, R. Castelo, and J. Guinney, <span>“<span class="nocase">GSVA: Gene set variation analysis for microarray and RNA-Seq data</span>,”</span> <em>BMC Bioinformatics</em>, vol. 14, no. 1, p. 7, Jan. 2013, doi: <a href="https://doi.org/10.1186/1471-2105-14-7" class="external-link">10.1186/1471-2105-14-7</a>.</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
