<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Scanpy Curated Workflow • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Scanpy Curated Workflow">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.8.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Scanpy Curated Workflow</h1>
                        <h4 data-toc-skip class="author">Nida Pervaiz,
Irzam Sarfraz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/scanpy_curated_workflow.Rmd" class="external-link"><code>vignettes/articles/scanpy_curated_workflow.Rmd</code></a></small>
      <div class="hidden name"><code>scanpy_curated_workflow.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <em>singleCellTK</em> package integrates functions from <a href="https://scanpy.readthedocs.io" class="external-link"><em>Scanpy</em></a> which is a
scalable Python-based toolkit for analyzing single-cell gene expression
data built on top of the <em><a href="https://anndata.readthedocs.io/" class="external-link">AnnData</a></em> framework to
store data and results. It includes functions for finding variable
genes, dimensionality reduction, clustering, and differential
expression. The Python-based implementation efficiently deals with
datasets of more than one million cells.</p>
<p>Users can assess the Scanpy workflow via the singleCellTK package
using either the R console or interactive graphical user interface (GUI)
built with R/Shiny. The <em>singleCellTK</em> package uses the <em><a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></em>
object to store data and results. The Scanpy wrapper functions in
<em>singleCellTK</em> automatically convert the data and results back
and forth between SCE and AnnData objects so users do not have to do
this manually.</p>
<p>The Shiny app allows users to access Scanpy functions and workflows
without having to know the underlying code. Users can access all
functions in the <em>a la carte</em> workflow in combination with
functions from other packages. The Shiny app also contains a curated
workflow that lets the users run the steps of the complete Scanpy
workflow in a sequential manner and display the results with interactive
plots.</p>
<p>In this tutorial, you can select the ‘Interactive Analysis’ tabs
below for detailed instructions on how to run the Scanpy workflow with
the Shiny or the ‘Console Analysis’ tabs for instructions on running the
workflow using the R console.</p>
<blockquote>
<p><strong>Note:</strong> This tutorial assumes that the data has
already been imported, QC’ed, and filtered. Please see the <a href="%60qc.link%60">Import and QC</a> tutorial for more
information.</p>
</blockquote>

<meta name="viewport" content="width=device-width, initial-scale=1">
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p>In this tutorial example, we illustrate all the steps of the
<strong>Scanpy</strong> curated workflow and focus on the options
available to customize the individual steps per the <a href="https://scanpy.readthedocs.io/" class="external-link">Scanpy instructions</a>. To start
the <code>Scanpy</code> curated workflow in the Shiny app, click on the
‘Curated Workflows’ tab in the top nagivation menu and select
<code>Scanpy</code>:</p>
<p><strong>1. Normalize Data</strong> <br> The first major step in the
analysis is to normalize the raw counts for differences in overall
library size (i.e. total number of counts) and apply a log
transformation. For this purpose, any assay containing raw counts in the
uploaded data can be used. 1. Click on the ‘Normalize’ vertical tab. 2.
Select the <code>assay</code> to normalize from the dropdown menu. 3.
Modify optional parameters if necessary 4. Press the ‘Normalize’ button
to start the normalization process.</p>
<blockquote>
<p><strong>Note:</strong> The input raw counts matrix should only
contained filtered cells and empty droplets should already have been
removed. If you applied ambient RNA removal algorithms and want to use
the decontominated counts, make sure to specify the appropriate matrix
as input.</p>
</blockquote>
<p><img src="ui_screenshots/scanpy/scanpy_1_normalize_annotated.png"><br></p>
<p><strong>2. Highly Variable Genes</strong> <br> Identification and
selection of highly variable features/genes is important in the
<code>Scanpy</code> workflow to produce high quality clustering results.
Currently, <code>Scanpy</code> provides three methods for variable genes
identification (<code>seurat</code>, <code>cell_ranger</code> and
<code>seurat_v3</code>).</p>
<ol style="list-style-type: decimal">
<li>Click on the ‘Highly Variable Genes’ vertical tab.</li>
<li>Select method for computation of highly variable genes.</li>
<li>Input the number of genes that should be identified as
variable.</li>
<li>Modify optional parameters if necessary.</li>
<li>Press ‘Find HVG’ button to find the top highly variable genes.</li>
<li>A plot is procuded that displays the relationship between the mean
and variabiliyt or disperson of each gene. Highly variable features will
have higher levels of variability or dispersion regardless of the
mean.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_3_hvg_annotated.png"><br></p>
<p><strong>3. Dimensionality Reduction</strong> <br><code>Scanpy</code>
workflow offers <code>PCA</code> for dimensionality reduction. The
principal components (PCs) from PCA will be used in the downstream
clustering and 2-D embedding analyses. In the downstream analysis, you
will have to select the number of PCs to use. Several plots are
available for the user to inspect the output of the dimensionality
reduction and identify the appropriate number of PCs to use in the
downstream analyses.</p>
<ul>
<li>The ‘PCA Variance’ show the percentage of variability explained by
each PC. This plot can be useful for selecting the number of PCs. In
general, the “elbow” of the curve is a good selection criteria. The
elbow is where there is a sharp drop in the amount the percentage of
variance decreases after including additional PCs.</li>
<li>The ‘PCA Plot’ can be used to view cells in the reduced dimensional
space of PC1 versus PC2.</li>
<li>The ‘PCA Gene Ranking’ plot shows the highest ranked genes for each
PC.</li>
</ul>
<ol style="list-style-type: decimal">
<li>Click on the ‘Dimensionality Reduction’ vertical tab.</li>
<li>Change the method for computation of PCA if desired.</li>
<li>Input the total number of components to compute.</li>
<li>Press the ‘Run PCA’ button to start processing.</li>
<li>Select the number of PCs that should be used in the downstream
clustering and 2-D embedding analyses.</li>
<li>The plot area from where all computed plots can be viewed by the
user.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_dimensionalityReduction_annotated.png"><br></p>
<p><strong>4. 2D-Embedding</strong> <br> The relationships between cells
can be viewed on 2-D scatter plots using embedding algorithms such as
‘UMAP’ and ‘tSNE’. In these plots, each point is a cell and cells with
more similar expression profiles will be closer together in the
plot.</p>
<ol style="list-style-type: decimal">
<li>Click on the ‘2-D Embedding’ vertical tab.</li>
<li>Select the ‘tSNE’ or ‘UMAP’ sub-tab.</li>
<li>Select a reduction method. Only results that were previously
computed in the ‘Dimensionality Reduction’ tab will be available.</li>
<li>Change optional parameters if desired.</li>
<li>Press ‘Run tSNE’ or ‘Run UMAP’ button to start processing.</li>
<li>A 2-D embedding scatter plot will be generated on the right.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_tsneUmap.png"><br></p>
<p><strong>5. Clustering</strong> <br> Cells can be clustered using a
previously computed reduced dimensional object. Cluster labels will be
displaye on previously computed 2-D embedding or reduction plots.</p>
<ol style="list-style-type: decimal">
<li>Click on the ‘Clustering’ vertical tab.</li>
<li>Select number of components to use from the previously computed
PCA.</li>
<li>Select clustering method.</li>
<li>Set parameters.</li>
<li>Press ‘Find Clusters’ button to start processing.</li>
<li>Re-computed plots with cluster labels. Only those plots are
available that have previously been computed.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_6_clustering.png"><br></p>
<p><strong>6. Find Markers</strong> <br> ‘Find Markers’ tab can be used
to identify the marker genes that are enriched in each cluster or group.
Markers genes can be identified that are differentially expressed
between each cluster and all other cluster or between two selected
clusters/groups. Expression of individual marker genes can be visualized
in Dot plots, Violin plots, Matrix plots and via a Heatmap.</p>
<ol style="list-style-type: decimal">
<li>Click on “Find Markers” vertical tab.</li>
<li>Select if you want to identify marker genes between for each
cluster/group compared to all other cluster/groups or between two
specific clusters/groups.</li>
<li>Change optional parameters if desired.</li>
<li>Press “Find Markers” button to run the marker identification
analysis.</li>
<li>Differentially expressed features will be displayed in the table on
the right. Additional filters can be applied to further refine the list
of differentially expressed genes. For example, specifying a log fold
change cutoff in addition to an adjusted p-value cutoff can sometimes
help identify better marker genes.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_findmarkers_1.png"><br></p>
<ol start="6" style="list-style-type: decimal">
<li>A Heatmap plot of the top differentially expressed genes for each
cluster/group can be used to visualize the patterns of expression within
and across clusters/groups.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_findmarkers_2.png"><br></p>
<p><strong>7. Downstream Analysis</strong> <br> Once all steps of the
Scanpy workflow have been completed, users can further explore the data
by selecting other downstream analysis options.</p>
</div>
<div id="console" class="tabcontent">
<p>All methods Scanpy wrapper functions in <em>singleCellTK</em> take a
<code>SingleCellExperiment</code> (SCE) object as an input and will
return an SCE with additional information. For this tutorial, we will
use a small PBMC dataset:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">'pbmc3k'</span><span class="op">)</span></span></code></pre></div>
<p><strong>1. Normalize Data</strong> <br> Once raw data is uploaded and
stored in a <code>SingleCellExperiment</code> object,
<code><a href="../reference/runScanpyNormalizeData.html">runScanpyNormalizeData()</a></code> function can be used to normalize
the data. The method returns a <code>SingleCellExperiment</code> object
with normalized data stored as a new assay in the input object.</p>
<p>The main parameters to this function include <code>useAssay</code>
which is used to specify the assay with raw counts that should be
normalized and <code>normAssayName</code> which will be the name of the
new normalized assay to be created.</p>
<blockquote>
<p><strong>Note:</strong> The input raw counts matrix should only
contained filtered cells and empty droplets should already have been
removed. If you applied ambient RNA removal algorithms and want to use
the decontominated counts, make sure to specify the appropriate matrix
as input.</p>
</blockquote>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyNormalizeData.html">runScanpyNormalizeData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, normAssayName <span class="op">=</span> <span class="st">"scanpyNormData"</span><span class="op">)</span></span></code></pre></div>
<p><strong>2. Scale Data</strong> <br> Normalized data can be scaled by
z-scoring with the <code><a href="../reference/runScanpyScaleData.html">runScanpyScaleData()</a></code> function that takes
input a <code>SingleCellExperiment</code> object that has been
normalized previously by the <code><a href="../reference/runScanpyNormalizeData.html">runScanpyNormalizeData()</a></code>
function. The scaled assay is stored back in the input object.</p>
<p>The main parameters to this function include <code>useAssay</code>
which is used to specify the normalized assay from previous steps and
<code>scaledAssayName</code> which will be the name of the new scaled
assay to be created.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyScaleData.html">runScanpyScaleData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"scanpyNormData"</span>, scaledAssayName <span class="op">=</span> <span class="st">"scanpyScaledData"</span><span class="op">)</span></span></code></pre></div>
<p><strong>3. Highly Variable Genes</strong> <br> Highly variable genes
can be identified by first using the <code><a href="../reference/runScanpyFindHVG.html">runScanpyFindHVG()</a></code>
function that computes that statistics against a selected HVG method in
the rowData of input object. The variable genes can be visualized using
the <code><a href="../reference/plotScanpyHVG.html">plotScanpyHVG()</a></code> method. <br></p>
<p>The main parameters to this function include <code>useAssay</code>
which is used to specify the name of the input log normalized assay and
<code>method</code> which is used to specify the method to use for
computing the variability of each gene.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyFindHVG.html">runScanpyFindHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"scanpyNormData"</span>, method <span class="op">=</span> <span class="st">"seurat"</span>, minMean <span class="op">=</span> <span class="fl">0.0125</span>, maxMean <span class="op">=</span> <span class="fl">3</span>, maxDisp <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p>The <code>getTopHVG</code> function can be used to select the top
most variable genes and store it back in <code>rowData</code> of the SCE
object as a Boolean vector.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="../reference/getTopHVG.html">getTopHVG</a></span><span class="op">(</span><span class="va">sce</span>, method <span class="op">=</span> <span class="st">"seurat"</span>, hvgNumber <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "PPBP"         "FCRL2"        "CEP128"       "DOK3"         "NFE2L2"      </span></span>
<span><span class="co">##  [6] "YPEL2"        "ARVCF"        "UBE2D4"       "FAM210B"      "CTB-113I20.2"</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotScanpyHVG.html">plotScanpyHVG</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Using the <span style="color: #00BB00;">'counts'</span> assay as the <span style="color: #00BB00;">X matrix</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #BBBB00;">!</span> <span style="color: #00BB00;">skip_assays</span> is <span style="color: #00BB00;">TRUE</span> so <span style="color: #00BB00;">X/layers</span> will be empty</span></span></code></pre>
<p><strong>4. Dimensionality Reduction</strong> <br> PCA can be computed
using the <code><a href="../reference/runScanpyPCA.html">runScanpyPCA()</a></code> function.</p>
<p>The main parameters to this function include <code>useAssay</code>
which is used to specify the name of the input log normalized assay and
<code>method</code> which is used to specify the method to use for
computing the variability of each gene.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyPCA.html">runScanpyPCA</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"scanpyScaledData"</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyPCA"</span>, nPCs <span class="op">=</span> <span class="fl">50</span>, method <span class="op">=</span> <span class="st">"auto"</span>, use_highly_variable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The results of the PCA can be visualized in several ways. The
<code><a href="../reference/plotScanpyPCAVariance.html">plotScanpyPCAVariance()</a></code> function shows the percentage of
variability explained by each PC. This plot can be useful for selecting
the number of PCs. In general, the “elbow” of the curve is a good
selection criteria. The elbow is where there is a sharp drop in the
amount the percentage of variance decreases after including additional
PCs.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotScanpyPCAVariance.html">plotScanpyPCAVariance</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, nPCs <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/plotScanpyPCA.html">plotScanpyPCA()</a></code> function can be used to view cells
in the reduced dimensional space of PC1 versus PC2.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotScanpyPCA.html">plotScanpyPCA</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyPCA"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Using the <span style="color: #00BB00;">'counts'</span> assay as the <span style="color: #00BB00;">X matrix</span></span></span></code></pre>
<p>The <code><a href="../reference/plotScanpyPCAGeneRanking.html">plotScanpyPCAGeneRanking()</a></code> function shows the
highest ranked genes for each PC.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotScanpyPCAGeneRanking.html">plotScanpyPCAGeneRanking</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p><strong>5. 2-D Embedding</strong> <br></p>
<p>The relationships between cells can be viewed on 2-D scatter plots
using embedding algorithms such as ‘UMAP’ and ‘tSNE’. In these plots,
each point is a cell and cells with more similar expression profiles
will be closer together in the plot. <code><a href="../reference/runScanpyTSNE.html">runScanpyTSNE()</a></code> and
<code><a href="../reference/runScanpyUMAP.html">runScanpyUMAP()</a></code> can be used to compute 2-D embeddings which
are stored in the <code>reducedDim</code> slot of the SCE object.</p>
<p>Parameters to both functions include <code>useReducedDim</code> which
specifies the reduced dimensional object to use as input,
<code>reducedDimName</code> which will be the name of this new
reduction, and <code>dims</code> which specifies the number of
dimensions to use from the reduced dimensional object. The
<code><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding()</a></code> function can be used to visualize the
results. Below is an example of calculating and displaying a UMAP:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyUMAP.html">runScanpyUMAP</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useReducedDim <span class="op">=</span> <span class="st">"scanpyPCA"</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyUMAP"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding</a></span><span class="op">(</span><span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyUMAP"</span><span class="op">)</span></span></code></pre></div>
<p><strong>6. Clustering</strong> <br></p>
<p>The <code><a href="../reference/runScanpyFindClusters.html">runScanpyFindClusters()</a></code> function can be used to
cluster cells. The main parameters to this function include
<code>useAssay</code> which should be the name of the scaled assay,
<code>useReducedDim</code> <code>useReducedDim</code> which specifies
the reduced dimensional object to use as input, <code>dims</code> which
specifies the number of dimensions to use from the reduced dimensional
object, and <code>method</code> which specifies the clustering
method.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyFindClusters.html">runScanpyFindClusters</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"scanpyScaledData"</span>, useReducedDim <span class="op">=</span> <span class="st">"scanpyPCA"</span>, method <span class="op">=</span> <span class="st">"leiden"</span>, resolution <span class="op">=</span> <span class="fl">1</span>, nNeighbors <span class="op">=</span> <span class="fl">10</span>, dims <span class="op">=</span> <span class="fl">10</span>, cor_method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding()</a></code> can then be used to plot the
cluster labels on the previously computed 2-D embeddings or reduced
dimensional objects. For example, we can plot the cluster labels on the
UMAP:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding</a></span><span class="op">(</span><span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyUMAP"</span>, color <span class="op">=</span> <span class="st">'Scanpy_leiden_1'</span><span class="op">)</span></span></code></pre></div>
<p><strong>7. Find Markers</strong> <br> Marker genes can be identified
using the <code><a href="../reference/runScanpyFindMarkers.html">runScanpyFindMarkers()</a></code> function. The main
parameter for this function is a vector of cluster/group labels that are
stored in the <code>colData</code> of the input SCE object. By default.
each cluster of cells will be compared to the rest of the cells in all
other clusters. Users can can find also find differentially expressed
genes between two clusters using the <code>group1</code> and
<code>group2</code> parameters. Generally, group labels defined by
clustering functions will be used as input. Here we will find markers
for the clusters we identified previously:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyFindMarkers.html">runScanpyFindMarkers</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, colDataName <span class="op">=</span> <span class="st">"Scanpy_leiden_1"</span>, nGenes <span class="op">=</span> <span class="fl">10</span>, test <span class="op">=</span> <span class="st">"t-test"</span>, corr_method <span class="op">=</span> <span class="st">"benjamini-hochberg"</span><span class="op">)</span></span></code></pre></div>
<p>The full table of differentially expressed marker genes can be
accessed in the metadata with the command
<code>metadata(sce)[["scanpyMarkersTable"]]</code>. Here are the first
few genes in the table:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[[</span><span class="st">"scanpyMarkersTable"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Gene findMarker_cluster   Log2_FC       Pvalue   zscore</span></span>
<span><span class="co">## 01  RPS27                  0 0.4768241 1.773535e-49 17.00457</span></span>
<span><span class="co">## 02 EEF1A1                  0 0.4685496 4.900430e-45 16.33228</span></span>
<span><span class="co">## 03  RPS12                  0 0.4607034 5.183866e-46 16.14429</span></span>
<span><span class="co">## 04  RPS25                  0 0.5404358 3.057087e-43 15.85425</span></span>
<span><span class="co">## 05  RPS14                  0 0.4289994 2.032707e-33 13.61009</span></span>
<span><span class="co">## 06   RPS6                  0 0.4592834 1.808564e-32 13.41268</span></span></code></pre>
<p>Expression of individual marker genes can be visualized in Dot plots,
Violin plots, Matrix plots and via a Heatmap:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotScanpyMarkerGenes.html">plotScanpyMarkerGenes</a></span><span class="op">(</span><span class="va">sce</span>, groups <span class="op">=</span> <span class="st">'0'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotScanpyMarkerGenesViolin.html">plotScanpyMarkerGenesViolin</a></span><span class="op">(</span><span class="va">sce</span>, groups <span class="op">=</span> <span class="st">'0'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotScanpyMarkerGenesHeatmap.html">plotScanpyMarkerGenesHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, groupBy <span class="op">=</span> <span class="st">'Scanpy_leiden_1'</span>, nGenes <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
      </div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
