<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to singleCellTK • singleCellTK</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">singleCellTK</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/v01-Introduction_to_singleCellTK.html">Introduction to singleCellTK</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Interactive Shiny Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/v03-tab01_Upload.html">1: Upload</a>
    </li>
    <li>
      <a href="../articles/v04-tab02_Data-Summary-and-Filtering.html">2: Data Summary and Filtering</a>
    </li>
    <li>
      <a href="../articles/v05-tab03_Dimensionality-Reduction-and-Clustering.html">3: Dimensionality Reduction and Clustering</a>
    </li>
    <li>
      <a href="../articles/v06-tab04_Batch-Correction.html">4: Batch Correction</a>
    </li>
    <li>
      <a href="../articles/v07-tab05_Differential-Expression.html">5: Differential Expression</a>
    </li>
    <li>
      <a href="../articles/v08-tab06_Pathway-Activity-Analysis.html">6: Pathway Activity Analysis</a>
    </li>
    <li>
      <a href="../articles/v09-tab07_Sample-Size.html">7: Sample Size</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/v02-Processing_and_Visualizing_Data_in_the_SingleCellTK.html">Processing and Visualizing Data</a>
    </li>
    <li>
      <a href="../articles/v10-Aligning_and_Quantifying_scRNA-Seq_Data.html">Aligning and Quantifying scRNA-Seq Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Introduction to singleCellTK</h1>
                        <h4 class="author">David Jenkins</h4>
            <address class="author_afil">
      The Section of Computational Biomedicine, Boston University School of Medicine, Boston, MA<br><a class="author_email" href="mailto:#"></a><a href="mailto:dfj@bu.edu">dfj@bu.edu</a>
      </address>
                              <h4 class="author">Tyler Faits</h4>
            <address class="author_afil">
      The Section of Computational Biomedicine, Boston University School of Medicine, Boston, MA<br><h4 class="author">W. Evan Johnson</h4>
            <address class="author_afil">
      The Section of Computational Biomedicine, Boston University School of Medicine, Boston, MA<br>
</address>
</address>
</div>

    
    
<div class="contents">
<p>The Single Cell Toolkit (SCTK) is an interactive scRNA-Seq analysis package that allows a user to upload raw scRNA-Seq count matrices and perform downstream scRNA-Seq analysis interactively through a web interface, or through a set of R functions through the command line. The package is written in R with a graphical user interface (GUI) written in Shiny. Users can perform analysis with modules for filtering raw results, clustering, batch correction, differential expression, pathway enrichment, and scRNA-Seq study design, all in a simple to use point and click interface. The toolkit also supports command line data processing, and results can be loaded into the GUI for additional exploration and downstream analysis.</p>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<blockquote>
<p>Note: Some package dependencies require Bioconductor v3.6, <a href="https://bioconductor.org/install/" class="uri">https://bioconductor.org/install/</a></p>
</blockquote>
<p>singleCellTK is under development. You can install the development version from github:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install.packages("devtools")</span>
devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"compbiomed/singleCellTK"</span>)</code></pre></div>
<div id="troubleshooting-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#troubleshooting-installation" class="anchor"></a>Troubleshooting Installation</h2>
<p>For the majority of users, the commands above will install the latest version of the singleCellTK without any errors. Rarely, you may encounter an error due to previously installed versions of some packages that are required for the singleCellTK. If you encounter an error during installation, use the commands below to check the version of Bioconductor that is installed:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocVersion</span>()</code></pre></div>
<p>If the version number is not 3.6 or higher, you must upgrade Bioconductor to install the toolkit:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biocLite</span>(<span class="st">"BiocUpgrade"</span>)</code></pre></div>
<p>After you install Bioconductor 3.6 or higher, you should be able to install the toolkit using <code><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">devtools::install_github("compbiomed/singleCellTK")</a></code>. If you still encounter an error, ensure your Bioconductor packages are up to date by running the following command.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biocValid</span>()</code></pre></div>
<p>If the command above does not return <code>TRUE</code>, run the following command to update your R packages:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biocLite</span>()</code></pre></div>
<p>Then, try to install the toolkit again:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"compbiomed/singleCellTK"</span>)</code></pre></div>
<p>If you still encounter an error, please <a href="mailto:dfj@bu.edu">contact us</a> and we’d be happy to help.</p>
</div>
</div>
<div id="data-structure" class="section level1">
<h1 class="hasAnchor">
<a href="#data-structure" class="anchor"></a>Data Structure</h1>
<p>The Single Cell Toolkit uses a <a href="https://www.bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">SingleCellExperiment</a> object to store data matrices along with annotation information, metadata, and reduced dimensionality data (PCA, t-SNE, etc.).</p>
</div>
<div id="run-the-shiny-app" class="section level1">
<h1 class="hasAnchor">
<a href="#run-the-shiny-app" class="anchor"></a>Run the Shiny App</h1>
<div id="example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example Data</h2>
<p>Example data is available within the app. To get started, simply run the singleCellTK function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(singleCellTK)
<span class="kw"><a href="../reference/singleCellTK.html">singleCellTK</a></span>()</code></pre></div>
</div>
<div id="upload-data-directly-through-the-shiny-app" class="section level2">
<h2 class="hasAnchor">
<a href="#upload-data-directly-through-the-shiny-app" class="anchor"></a>Upload data directly through the shiny app</h2>
<p>To upload count matrices and annotation information stored as text files, run the singleCellTK function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(singleCellTK)
<span class="kw"><a href="../reference/singleCellTK.html">singleCellTK</a></span>()</code></pre></div>
<p>Then, follow data upload instructions in the “Upload Tab” vignette.</p>
</div>
<div id="load-data-from-a-singlecellexperiment-object" class="section level2">
<h2 class="hasAnchor">
<a href="#load-data-from-a-singlecellexperiment-object" class="anchor"></a>Load data from a SingleCellExperiment object</h2>
<div id="creating-a-singlecellexperiment-object" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-a-singlecellexperiment-object" class="anchor"></a>Creating a SingleCellExperiment object</h3>
<p>To create a SingleCellExperiment object, we have provided the <code><a href="../reference/createSCE.html">createSCE()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(singleCellTK)
<span class="kw">data</span>(<span class="st">"GSE60361_subset_sce"</span>)
counts_mat &lt;-<span class="st"> </span><span class="kw">assay</span>(GSE60361_subset_sce, <span class="st">"counts"</span>)
sample_annot &lt;-<span class="st"> </span><span class="kw">colData</span>(GSE60361_subset_sce)
row_annot &lt;-<span class="st"> </span><span class="kw">rowData</span>(GSE60361_subset_sce)
newSCE &lt;-<span class="st"> </span><span class="kw"><a href="../reference/createSCE.html">createSCE</a></span>(<span class="dt">assayfile =</span> counts_mat, <span class="dt">annotfile =</span> sample_annot, 
                    <span class="dt">featurefile =</span> row_annot, <span class="dt">assay_name =</span> <span class="st">"counts"</span>,
                    <span class="dt">inputdataframes =</span> <span class="ot">TRUE</span>, <span class="dt">create_logcounts =</span> <span class="ot">TRUE</span>)</code></pre></div>
</div>
<div id="loading-data-stored-in-a-singlecellexperiment-object" class="section level3">
<h3 class="hasAnchor">
<a href="#loading-data-stored-in-a-singlecellexperiment-object" class="anchor"></a>Loading data stored in a SingleCellExperiment object</h3>
<p>Once a SingleCellExperiment object is created, the object can be loaded directly from the R console:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/singleCellTK.html">singleCellTK</a></span>(newSCE)</code></pre></div>
</div>
</div>
</div>
<div id="vignettes" class="section level1">
<h1 class="hasAnchor">
<a href="#vignettes" class="anchor"></a>Vignettes</h1>
<p>To help you get started with the SCTK, we have prepared several vignettes in two categories: interactive analysis and R console analysis.</p>
<div id="interactive-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#interactive-analysis" class="anchor"></a>Interactive Analysis</h2>
<ul>
<li><a href="v03-tab01_Upload.html">Upload Tab</a></li>
<li><a href="v04-tab02_Data-Summary-and-Filtering.html">Data Summary and Filtering Tab</a></li>
<li><a href="v05-tab03_Dimensionality-Reduction-and-Clustering.html">Dimensionality Reduction and Filtering Tab</a></li>
<li><a href="v06-tab04_Batch-Correction.html">Batch Correction Tab</a></li>
<li><a href="v07-tab05_Differential-Expression.html">Differential Expression Tab</a></li>
<li><a href="v08-tab06_Pathway-Activity-Analysis.html">Pathway Activity Analysis Tab</a></li>
<li><a href="v09-tab07_Sample-Size.html">Sample Size Tab</a></li>
</ul>
</div>
<div id="r-console-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#r-console-analysis" class="anchor"></a>R Console Analysis</h2>
<ul>
<li><a href="v02-Processing_and_Visualizing_Data_in_the_SingleCellTK.html">Processing and Visualizing Data in the Single Cell Toolkit</a></li>
<li><a href="v10-Aligning_and_Quantifying_scRNA-Seq_Data.html">Aligning and Quantifying scRNA-Seq Data</a></li>
</ul>
</div>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<pre><code>## R version 3.4.3 (2017-11-30)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS High Sierra 10.13.3
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] singleCellTK_0.4.8         SingleCellExperiment_1.0.0
##  [3] SummarizedExperiment_1.8.1 DelayedArray_0.4.1        
##  [5] matrixStats_0.53.1         Biobase_2.38.0            
##  [7] GenomicRanges_1.30.3       GenomeInfoDb_1.14.0       
##  [9] IRanges_2.12.0             S4Vectors_0.16.0          
## [11] BiocGenerics_0.24.0       
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.15           knitr_1.20             XVector_0.18.0        
##  [4] magrittr_1.5           zlibbioc_1.24.0        lattice_0.20-35       
##  [7] stringr_1.3.0          tools_3.4.3            grid_3.4.3            
## [10] htmltools_0.3.6        yaml_2.1.17            rprojroot_1.3-2       
## [13] digest_0.6.15          Matrix_1.2-12          GenomeInfoDbData_1.0.0
## [16] bitops_1.0-6           RCurl_1.95-4.10        evaluate_0.10.1       
## [19] rmarkdown_1.9          stringi_1.1.6          compiler_3.4.3        
## [22] backports_1.1.2</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#installation">Installation</a><ul class="nav nav-pills nav-stacked">
<li><a href="#troubleshooting-installation">Troubleshooting Installation</a></li>
      </ul>
</li>
      <li><a href="#data-structure">Data Structure</a></li>
      <li>
<a href="#run-the-shiny-app">Run the Shiny App</a><ul class="nav nav-pills nav-stacked">
<li><a href="#example-data">Example Data</a></li>
      <li><a href="#upload-data-directly-through-the-shiny-app">Upload data directly through the shiny app</a></li>
      <li><a href="#load-data-from-a-singlecellexperiment-object">Load data from a SingleCellExperiment object</a></li>
      </ul>
</li>
      <li>
<a href="#vignettes">Vignettes</a><ul class="nav nav-pills nav-stacked">
<li><a href="#interactive-analysis">Interactive Analysis</a></li>
      <li><a href="#r-console-analysis">R Console Analysis</a></li>
      </ul>
</li>
      <li><a href="#session-info">Session info</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by David Jenkins, Tyler Faits, Emma Briars, Sebastian Carrasco Pro, W. Evan Johnson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
