<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Visualization • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Visualization">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/console_analysis_tutorial.html">A La Carte Workflow - PBMC3K (R Console)</a>
    </li>
    <li>
      <a href="../articles/cnsl_seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/cnsl_2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Enrichment &amp; Pathway Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Visualization</h1>
                        <h4 data-toc-skip class="author">Yichen Wang</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/visualization.Rmd" class="external-link"><code>vignettes/articles/visualization.Rmd</code></a></small>
      <div class="hidden name"><code>visualization.Rmd</code></div>

    </div>

    
    

<meta name="viewport" content="width=device-width, initial-scale=1">
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Single Cell Toolkit (singleCellTK, SCTK) allows users to create various types of plots, including scatter plot, bar plot, violin plot and heatmap. This documentation will focus on the first three types of plot. For heatmap generation, please refer to the <a href="heatmap.html">Heatmap Documentation</a>.</p>
<p>To view detailed instructions on how to generate these visualizations, please select ‘Interactive Analysis’ for shiny application or ‘Console Analysis’ for using these visualizations on R console from the tabs below: <br></p>
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p><strong>Entry of the Panel</strong></p>
<p><img src="ui_screenshots/cellViewer/ui_cellViewer_entry.png" alt="entry"><br></p>
<p>SCTK UI provides a general portal, CellViewer, for cell level visualization, including the functionality to create <strong>scatter plot</strong>, <strong>bar plot</strong>, and <strong>violin plot</strong>. Users should find the Cell Viewer as instructed in the screenshot above. This documentation will start with a quick example and then talk about the detail.</p>
<p><strong>Quick Example</strong></p>
<p>Here we provide a common case of plotting a UMAP with clusters labeled. We take the frequently used “PBMC3k” dataset, provided by 10X, as an example. The detailed screenshot for each prerpocessing step will not be included, but users can refer to the documentation of those steps within other pages.</p>
<details><summary><b>Preprocessing Steps</b></summary><ol style="list-style-type: decimal">
<li>The example “PBMC3k” dataset can be imported at the SCTK landing page. For a quick demonstration, QC and filtering are skipped, but it is always recommended to have this done.</li>
<li>Normalize the raw counts with “Seurat lognormalize” method</li>
<li>Use the log-normalized assay to produce a UMAP by using Seurat’s method. Note that usually it is recommended to find the variable features and calculate the PCA before reaching to the UMAP, but “SeuratUMAP” has integrated all these steps at one click.</li>
<li>For the clustering, select the PCA embedding (produced when calculating the UMAP) as the matrix to perform Seurat’s louvain algorithm on.</li>
</ol></details><p><img src="ui_screenshots/cellViewer/ui_cellViewer_example.png" alt="example"><br></p>
<p>In Cell Viewer:</p>
<ol style="list-style-type: decimal">
<li>Locate at <strong>“Scatter Plot”</strong> tab</li>
<li>Select the UMAP at <strong>“Select Coordinates”</strong>.</li>
<li>In <strong>“Color”</strong>, choose to color by <strong>“Cell Annotation”</strong>
</li>
<li>Select the clustering result from the cell annotation option list, which comes out after Step 3.</li>
<li>Click on <strong>“Plot”</strong> button</li>
</ol>
<p><strong>Step Specific Introduction</strong></p>
<p>The procedure of creating a single plot in Cell Viewer can be generalized as three steps: <strong>axis selection</strong>, <strong>color specification</strong>, and <strong>peripheral setting</strong>.</p>
 <style>
div.offset { background-color:inherit ; border-radius: 5px; margin-left: 15px;}
</style>
<div class="offset">
<details><summary><b>Axis</b></summary><p><img src="ui_screenshots/cellViewer/ui_cellViewer_coord.png" alt="axis"><br></p>
<p>In <strong>“scatter plot”</strong> tab, the X-/Y-axis can be fully customized, while quick access for existing dimension reduction results will also be listed. “Fully customized” means an axis can be <strong>a dimension of a dimension reduction</strong>, <strong>the expression of a feature from an assay</strong>, or <strong>one of the cell annotations</strong>. In <strong>“bar plot”</strong> and <strong>“violin plot”</strong> tab, the X-axis is limited to cell annotation level, while the Y-axis can still be fully customized.</p>
</details><details><summary><b>Color</b></summary><p><img src="ui_screenshots/cellViewer/ui_cellViewer_color.png" alt="color"><br></p>
<p>The color setting determines the color of each dot or bar in the plot. In <strong>“Scatter Plot”</strong> tab, colors can be set by the value of <strong>a dimension from a dimension reduction embedding</strong>, <strong>the expression value of a gene from an assay</strong>, <strong>one of the cell annotation</strong>, or a <strong>single color</strong> for all dots (cells). While SCTK only supports single color plotting for bar plot and violin plot.</p>
</details><details><summary><b>Other Settings</b></summary><p><img src="ui_screenshots/cellViewer/ui_cellViewer_other.png" alt="others"><br></p>
<p>Besides the informative settings, Cell Viewer gives users the control on title, legend, axis labels, dots and etc., as shown in the screenshot above and can be freely explored.</p>
</details>
</div>
</div>

<div id="console" class="tabcontent">
<p>SCTK provides plenty of methods to visualize users analysis, including the plotting for intermediate attribute for parameter decision, traditional scatter/box/violin plot for cells, and <a href="heatmap.html">heatmap</a>. Additionally, there are also wrappers of the base functions that are used for analysis type specific visualization. (e.g. QC metric plotting, differential expression heatmap) A full list of visualization functions can be found in the <a href="../../reference/index.html#section-visualization">reference tab</a> of this documentation site.</p>
<p>Workflow specific visualization should be introduced in the documentation of the related steps (e.g. QC metric plotting, HVG plotting, or differential expression plotting). Here, we would provide some examples for the most frequent use case.</p>
<details><summary><b>Example 1: Scatter plot of cells</b></summary><p>Here we will take PBMC3k dataset, from “10X,” as an example. First we will do a quick clustering with the Seurat curated workflow implemented in SCTK. Then we will present the method to generate a scatter plot of the UMAP of this dataset, with clusters labeled by color.</p>
<details><summary><b>Preprocessing Steps</b></summary><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span>
<span class="co"># It is recommended to deduplicate the feature names, as this example data does </span>
<span class="co"># have duplicated feature names that might cause failure in downstream analysis.</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dedupRowNames.html">dedupRowNames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="co"># QC and filtering skipped for quick demonstration</span>
<span class="co"># Users are **not recommended** to skip it in practical analysis</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratNormalizeData.html">runSeuratNormalizeData</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"counts"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratScaleData.html">runSeuratScaleData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratFindHVG.html">runSeuratFindHVG</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratPCA.html">runSeuratPCA</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratFindClusters.html">runSeuratFindClusters</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="co"># With `runSeuratFindClusters()`, the clustering result would be saved to the </span>
<span class="co"># `colData` slot with the name "Seurat_louvain_Resolution0.8" by default.</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Have a check on the summary of the clustering result</span>
<span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedArray-utils.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">Seurat_louvain_Resolution0.8</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##   0   1   2   3   4   5   6   7   8   9  10 </span>
<span class="co">## 550 486 349 303 257 236 162 158 149  36  14</span></code></pre>
</details><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSeuratUMAP.html">runSeuratUMAP</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotSCEDimReduceColData.html">plotSCEDimReduceColData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"seuratUMAP"</span>, colorBy <span class="op">=</span> <span class="st">"Seurat_louvain_Resolution0.8"</span>, legendTitle <span class="op">=</span> <span class="st">"Seurat Cluster"</span>, title <span class="op">=</span> <span class="st">"Clustering on UMAP"</span><span class="op">)</span></code></pre></div>
<p><img src="visualization_files/figure-html/pbmc_cluster_plot-1.png" width="700"></p>
<p>As users might have tried, the essential input are <code>inSCE</code>, <code>reducedDimName</code>, and <code>colorBy</code>. Other arguments are all optional, but users can explore them to best clarify their intention.</p>
<p>Additionally but occasionally, <code>shape</code> argument is also useful for double level annotating.</p>
<p>In SCTK, there are plotting functions that are aimed for other attributes other than <code>colData</code> and <code>reducedDim</code>. For example, <code><a href="../reference/plotSCEDimReduceFeatures.html">plotSCEDimReduceFeatures()</a></code> focuses on <code>reducedDim</code> and feature expression, In most of base SCTK plotting functions, which starts with “plotSCE,” users can easily figure out the usage by checking the function manual and see which arguments are required.</p>
</details><details><summary><b>Example 2: Plots with multiple samples</b></summary><p>When there are multiple samples in one dataset and it is necessary to have separation on them, users can invoke <code>sample</code> or <code>groupby</code> argument in most of base SCTK plotting functions.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span>
<span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dedupRowNames.html">dedupRowNames</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span>
<span class="va">pbmc4k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">"pbmc4k"</span><span class="op">)</span>
<span class="va">pbmc4k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dedupRowNames.html">dedupRowNames</a></span><span class="op">(</span><span class="va">pbmc4k</span><span class="op">)</span>
<span class="va">sceList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pbmc3k <span class="op">=</span> <span class="va">pbmc3k</span>, pbmc4k <span class="op">=</span> <span class="va">pbmc4k</span><span class="op">)</span>
<span class="co"># By combining, a "sample" colData annotation is set by default with the </span>
<span class="co"># names of sceList.</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineSCE.html">combineSCE</a></span><span class="op">(</span>sceList <span class="op">=</span> <span class="va">sceList</span>, by.r <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span><span class="op">)</span>, by.c <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span><span class="op">)</span>, combined <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPerCellQC.html">runPerCellQC</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotSCEViolinColData.html">plotSCEViolinColData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, coldata <span class="op">=</span> <span class="st">"total"</span>, groupBy <span class="op">=</span> <span class="st">"sample"</span>, ylab <span class="op">=</span> <span class="st">"Total counts"</span><span class="op">)</span></code></pre></div>
<p><img src="visualization_files/figure-html/groupby-1.png" width="700"></p>
<p>Otherwise, if users need individual plot for each of the sample, or indeed make individual plots by any categorical groupingon cells, use <code>sample</code> argument.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSCEViolinColData.html">plotSCEViolinColData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, coldata <span class="op">=</span> <span class="st">"total"</span>, sample <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">sample</span>, ylab <span class="op">=</span> <span class="st">"Total counts"</span><span class="op">)</span></code></pre></div>
<p><img src="visualization_files/figure-html/sample-1.png" width="700"><img src="visualization_files/figure-html/sample-2.png" width="700"></p>
</details>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
