<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Convert gene IDs — convert_gene_ids • singleCellTK</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">singleCellTK</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Interactive Shiny Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/tab01_Upload.html">1: Upload</a>
    </li>
    <li>
      <a href="../articles/tab02_Data-Summary-and-Filtering.html">2: Data Summary and Filtering</a>
    </li>
    <li>
      <a href="../articles/tab03_Dimensionality-Reduction-and-Clustering.html">3: Dimensionality Reduction and Clustering</a>
    </li>
    <li>
      <a href="../articles/tab04_Batch-Correction.html">4: Batch Correction</a>
    </li>
    <li>
      <a href="../articles/tab05_Differential-Expression.html">5: Differential Expression</a>
    </li>
    <li>
      <a href="../articles/tab06_Pathway-Activity-Analysis.html">6: Pathway Activity Analysis</a>
    </li>
    <li>
      <a href="../articles/tab07_Sample-Size.html">7: Sample Size</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Aligning_and_Quantifying_scRNA-Seq_Data.html">Aligning and Quantifying scRNA-Seq Data</a>
    </li>
    <li>
      <a href="../articles/Processing_and_Visualizing_Data_in_the_SingleCellTK.html">Processing and Visualizing Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Convert gene IDs</h1>
    </div>

    
    <p>Convert the gene IDs in a SingleCellExperiment object using Bioconductor
org.*.eg.db data packages. Because annotation databases do not have a 1:1
relationship, this tool removes rows with no corresponding annotation in
your desired annotation, and remove any duplicate annotations after
conversion.</p>
    

    <pre class="usage"><span class='fu'>convert_gene_ids</span>(<span class='no'>inSCESet</span>, <span class='no'>in_symbol</span>, <span class='no'>out_symbol</span>, <span class='kw'>database</span> <span class='kw'>=</span> <span class='st'>"org.Hs.eg.db"</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>inSCESet</th>
      <td><p>Input SCtkExperiment object. Required</p></td>
    </tr>
    <tr>
      <th>in_symbol</th>
      <td><p>The input symbol type</p></td>
    </tr>
    <tr>
      <th>out_symbol</th>
      <td><p>The output symbol type</p></td>
    </tr>
    <tr>
      <th>database</th>
      <td><p>The org.*.eg.db database to use. The default is org.Hs.eg.db</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A SCtkExperiment with converted gene IDs.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span>(<span class='fu'>requireNamespace</span>(<span class='st'>"org.Mm.eg.db"</span>, <span class='kw'>quietly</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)) {
  <span class='co'>#convert mouse gene symbols to ensembl IDs</span>
  <span class='fu'>library</span>(<span class='st'>"org.Mm.eg.db"</span>)
  <span class='fu'>sample</span>(<span class='fu'>rownames</span>(<span class='no'>GSE60361_subset_sce</span>), <span class='fl'>50</span>)
  <span class='no'>GSE60361_subset_symbol</span> <span class='kw'>&lt;-</span> <span class='fu'>convert_gene_ids</span>(<span class='kw'>inSCESet</span> <span class='kw'>=</span> <span class='no'>GSE60361_subset_sce</span>,
                                             <span class='kw'>in_symbol</span> <span class='kw'>=</span> <span class='st'>"SYMBOL"</span>,
                                             <span class='kw'>out_symbol</span> <span class='kw'>=</span> <span class='st'>"ENSEMBL"</span>,
                                             <span class='kw'>database</span> <span class='kw'>=</span> <span class='st'>"org.Mm.eg.db"</span>)
  <span class='fu'>sample</span>(<span class='fu'>rownames</span>(<span class='no'>GSE60361_subset_symbol</span>), <span class='fl'>50</span>)
}</div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>'select()' returned 1:many mapping between keys and columns</span></div><div class='output co'>#&gt; <span class='message'>1773 ENSEMBL are NA after conversion. Removing</span></div><div class='output co'>#&gt; <span class='message'>51 ENSEMBL are duplicated after conversion. Removing additional copies</span></div><div class='output co'>#&gt; <span class='message'>153 SYMBOL are duplicated after conversion. Removing additional copies</span></div><div class='output co'>#&gt; <span class='message'>19972 SYMBOL originally, 18153 ENSEMBLs</span></div><div class='output co'>#&gt;  [1] "ENSMUSG00000036282" "ENSMUSG00000033416" "ENSMUSG00000024781"
#&gt;  [4] "ENSMUSG00000014773" "ENSMUSG00000061360" "ENSMUSG00000029601"
#&gt;  [7] "ENSMUSG00000052821" "ENSMUSG00000039611" "ENSMUSG00000067578"
#&gt; [10] "ENSMUSG00000042331" "ENSMUSG00000030351" "ENSMUSG00000033159"
#&gt; [13] "ENSMUSG00000039616" "ENSMUSG00000095138" "ENSMUSG00000067571"
#&gt; [16] "ENSMUSG00000043953" "ENSMUSG00000032059" "ENSMUSG00000004864"
#&gt; [19] "ENSMUSG00000024045" "ENSMUSG00000055013" "ENSMUSG00000021681"
#&gt; [22] "ENSMUSG00000018068" "ENSMUSG00000004415" "ENSMUSG00000022217"
#&gt; [25] "ENSMUSG00000050201" "ENSMUSG00000040724" "ENSMUSG00000020878"
#&gt; [28] "ENSMUSG00000039154" "ENSMUSG00000038670" "ENSMUSG00000020929"
#&gt; [31] "ENSMUSG00000086899" "ENSMUSG00000024806" "ENSMUSG00000067851"
#&gt; [34] "ENSMUSG00000033684" "ENSMUSG00000022243" "ENSMUSG00000031751"
#&gt; [37] "ENSMUSG00000112663" "ENSMUSG00000063605" "ENSMUSG00000001865"
#&gt; [40] "ENSMUSG00000030785" "ENSMUSG00000031434" "ENSMUSG00000056091"
#&gt; [43] "ENSMUSG00000022791" "ENSMUSG00000045252" "ENSMUSG00000045672"
#&gt; [46] "ENSMUSG00000022505" "ENSMUSG00000059208" "ENSMUSG00000063802"
#&gt; [49] "ENSMUSG00000025665" "ENSMUSG00000021983"</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Jenkins, Tyler Faits, Emma Briars, Sebastian Carrasco Pro, W. Evan Johnson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
