<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Find the marker gene set for each cluster
With an input SingleCellExperiment object and specifying the clustering
labels, this function iteratively call the differential expression analysis
on each cluster against all the others. — findMarkerDiffExp • singleCellTK</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Find the marker gene set for each cluster
With an input SingleCellExperiment object and specifying the clustering
labels, this function iteratively call the differential expression analysis
on each cluster against all the others. — findMarkerDiffExp" />
<meta property="og:description" content="Find the marker gene set for each cluster
With an input SingleCellExperiment object and specifying the clustering
labels, this function iteratively call the differential expression analysis
on each cluster against all the others." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/articles/introduction_to_SCTK.html">General Workflow</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shell Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../articles/articles/cmd_qc.html">Quality Control</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    R Console Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/articles/cnsl_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_dropletqc.html">Quality Control of droplet data</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_cellqc.html">Quality Control of cell data</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_batch_correction.html">Batch Correction</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_feature_selection.html">Feature Selection</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_find_marker.html">Find Marker via DE</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_heatmap.html">Heatmap Visualization</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shiny UI Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/articles/ui_import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/articles/ui_import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/articles/ui_export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/articles/ui_delete_data.html">Delete Data</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_qc.html">Quality Control</a>
        </li>
        <li>
          <a href="../articles/articles/ui_filtering.html">Filtering</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/articles/ui_batch_correction.html">Batch Correction</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/articles/ui_dimemsionality_reduction.html">Dimensionality Reduction</a>
        </li>
      </ul>
    </li>
    <li>
      <a href="../articles/articles/ui_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/articles/ui_find_marker.html">Find Marker</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_seurat_curated_workflow.html">Seurat</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_cellViewer.html">cellViewer</a>
        </li>
        <li>
          <a href="../articles/articles/ui_heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find the marker gene set for each cluster
With an input SingleCellExperiment object and specifying the clustering
labels, this function iteratively call the differential expression analysis
on each cluster against all the others.</h1>
    <small class="dont-index">Source: <a href='https://github.com/compbiomed/singleCellTK/blob/master/R/findMarkerDiffExp.R'><code>R/findMarkerDiffExp.R</code></a></small>
    <div class="hidden name"><code>findMarkerDiffExp.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Find the marker gene set for each cluster
With an input SingleCellExperiment object and specifying the clustering
labels, this function iteratively call the differential expression analysis
on each cluster against all the others.</p>
    </div>

    <pre class="usage"><span class='fu'>findMarkerDiffExp</span><span class='op'>(</span>
  <span class='va'>inSCE</span>,
  useAssay <span class='op'>=</span> <span class='st'>"logcounts"</span>,
  method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"wilcox"</span>, <span class='st'>"MAST"</span>, <span class='st'>"DESeq2"</span>, <span class='st'>"Limma"</span>, <span class='st'>"ANOVA"</span><span class='op'>)</span>,
  cluster <span class='op'>=</span> <span class='st'>"cluster"</span>,
  covariates <span class='op'>=</span> <span class='cn'>NULL</span>,
  log2fcThreshold <span class='op'>=</span> <span class='fl'>0.25</span>,
  fdrThreshold <span class='op'>=</span> <span class='fl'>0.05</span>,
  minClustExprPerc <span class='op'>=</span> <span class='fl'>0.6</span>,
  maxCtrlExprPerc <span class='op'>=</span> <span class='fl'>0.4</span>,
  minMeanExpr <span class='op'>=</span> <span class='fl'>0.5</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>inSCE</th>
      <td><p>SingleCellExperiment inherited object.</p></td>
    </tr>
    <tr>
      <th>useAssay</th>
      <td><p>character. A string specifying which assay to use for the
MAST calculations. Default <code>"logcounts"</code>.</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>A single character for specific differential expression
analysis method. Choose from <code>'wilcox'</code>, <code>'MAST'</code>, <code>'DESeq2'</code>,
<code>'Limma'</code>, and <code>'ANOVA'</code>. Default <code>"wilcox"</code>.</p></td>
    </tr>
    <tr>
      <th>cluster</th>
      <td><p>One single character to specify a column in
<code>colData(inSCE)</code> for the clustering label. Alternatively, a vector or
a factor is also acceptable. Default <code>"cluster"</code>.</p></td>
    </tr>
    <tr>
      <th>covariates</th>
      <td><p>A character vector of additional covariates to use when
building the model. All covariates must exist in
<code><a href='https://rdrr.io/r/base/names.html'>names(colData(inSCE))</a></code>. Not applicable when <code>method</code> is
<code>"MAST"</code> method. Default <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>log2fcThreshold</th>
      <td><p>Only out put DEGs with the absolute values of log2FC
larger than this value. Default <code>NULL</code></p></td>
    </tr>
    <tr>
      <th>fdrThreshold</th>
      <td><p>Only out put DEGs with FDR value smaller than this
value. Default <code>1</code></p></td>
    </tr>
    <tr>
      <th>minClustExprPerc</th>
      <td><p>A numeric scalar. The minimum cutoff of the
percentage of cells in the cluster of interests that expressed the marker
gene. Default <code>0.7</code>.</p></td>
    </tr>
    <tr>
      <th>maxCtrlExprPerc</th>
      <td><p>A numeric scalar. The maximum cutoff of the
percentage of cells out of the cluster (control group) that expressed the
marker gene. Default <code>0.4</code>.</p></td>
    </tr>
    <tr>
      <th>minMeanExpr</th>
      <td><p>A numeric scalar. The minimum cutoff of the mean
expression value of the marker in the cluster of interests. Default <code>1</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The input SingleCellExperiment object with
<code>metadata(inSCE)$findMarker</code> updated with a data.table of the up-
regulated DEGs for each cluster.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"mouseBrainSubsetSCE"</span>, package <span class='op'>=</span> <span class='st'>"singleCellTK"</span><span class='op'>)</span>
<span class='va'>mouseBrainSubsetSCE</span> <span class='op'>&lt;-</span> <span class='fu'>findMarkerDiffExp</span><span class='op'>(</span><span class='va'>mouseBrainSubsetSCE</span>,
                                         useAssay <span class='op'>=</span> <span class='st'>"logcounts"</span>,
                                         cluster <span class='op'>=</span> <span class='st'>"level1class"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Running with wilcox</span></div><div class='output co'>#&gt; <span class='message'>Computing for cluster: microglia</span></div><div class='output co'>#&gt; <span class='message'>Computing for cluster: oligodendrocytes</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


