<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Construct SCE object from Cell Ranger output — importCellRanger • singleCellTK</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Construct SCE object from Cell Ranger output — importCellRanger"><meta property="og:description" content="Read the filtered barcodes, features, and matrices for all
 samples from (preferably a single run of) Cell Ranger output. Import and
 combine them
 as one big SingleCellExperiment object."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.14.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul></li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul></li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Construct SCE object from Cell Ranger output</h1>
    <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/R/importCellRanger.R" class="external-link"><code>R/importCellRanger.R</code></a></small>
    <div class="hidden name"><code>importCellRanger.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Read the filtered barcodes, features, and matrices for all
 samples from (preferably a single run of) Cell Ranger output. Import and
 combine them
 as one big <a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a> object.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">importCellRanger</span><span class="op">(</span></span>
<span>  cellRangerDirs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sampleDirs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sampleNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellRangerOuts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dataType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"filtered"</span>, <span class="st">"raw"</span><span class="op">)</span>,</span>
<span>  matrixFileNames <span class="op">=</span> <span class="st">"matrix.mtx.gz"</span>,</span>
<span>  featuresFileNames <span class="op">=</span> <span class="st">"features.tsv.gz"</span>,</span>
<span>  barcodesFileNames <span class="op">=</span> <span class="st">"barcodes.tsv.gz"</span>,</span>
<span>  gzipped <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Matrix"</span>, <span class="st">"matrix"</span><span class="op">)</span>,</span>
<span>  delayedArray <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rowNamesDedup <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">importCellRangerV2</span><span class="op">(</span></span>
<span>  cellRangerDirs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sampleDirs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sampleNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dataTypeV2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"filtered"</span>, <span class="st">"raw"</span><span class="op">)</span>,</span>
<span>  class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Matrix"</span>, <span class="st">"matrix"</span><span class="op">)</span>,</span>
<span>  delayedArray <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellRangerOutsV2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rowNamesDedup <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">importCellRangerV3</span><span class="op">(</span></span>
<span>  cellRangerDirs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sampleDirs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sampleNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dataType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"filtered"</span>, <span class="st">"raw"</span><span class="op">)</span>,</span>
<span>  class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Matrix"</span>, <span class="st">"matrix"</span><span class="op">)</span>,</span>
<span>  delayedArray <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rowNamesDedup <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>cellRangerDirs</dt>
<dd><p>The root directories where Cell Ranger was run. These
folders should contain sample specific folders. Default <code>NULL</code>,
meaning the paths for each sample will be specified in <em>samples</em>
argument.</p></dd>


<dt>sampleDirs</dt>
<dd><p>Default <code>NULL</code>. Can be one of</p><ul><li><p><code>NULL</code>. All samples within <code>cellRangerDirs</code> will be
   imported. The order of samples will be first determined by the order of
   <code>cellRangerDirs</code> and then by <a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a>. This is only
   for the case where <code>cellRangerDirs</code> is specified.</p></li>
<li><p>A list of vectors containing the folder names for samples to import.
   Each vector in
   the list corresponds to samples from one of <code>cellRangerDirs</code>.
   These names are the same as the folder names under <code>cellRangerDirs</code>.
   This is only for the case where <code>cellRangerDirs</code> is specified.</p></li>
<li><p>A vector of folder paths for the samples to import. This is only for
   the case where <code>cellRangerDirs</code> is <code>NULL</code>.</p></li>
</ul><p>The cells in the final SCE object will be ordered in the same order of
 <code>sampleDirs</code>.</p></dd>


<dt>sampleNames</dt>
<dd><p>A vector of user-defined sample names for the samples
to be
imported. Must have the same length as <code>length(unlist(sampleDirs))</code> if
<code>sampleDirs</code> is not <code>NULL</code>. Otherwise, make sure the length and
order match the output of
<code>unlist(lapply(cellRangerDirs, list.dirs, recursive = FALSE))</code>. Default
<code>NULL</code>, in which case the folder names will be used as sample names.</p></dd>


<dt>cellRangerOuts</dt>
<dd><p>Character vector. The intermediate
paths to filtered or raw cell barcode, feature, and matrix files
for each sample. <strong>Supercedes <code>dayaType</code></strong>. If <code>NULL</code>,
<code>dataType</code>
will be used to determine Cell Ranger output directory. If not <code>NULL</code>,
<code>dataType</code> will be ingored and <code>cellRangerOuts</code> specifies the
paths. Must have length 1 or the same length as
<code>length(unlist(sampleDirs))</code> if
<code>sampleDirs</code> is not <code>NULL</code>. Otherwise, make sure the length and
order match the output of
<code>unlist(lapply(cellRangerDirs, list.dirs, recursive = FALSE))</code>.
Reference genome names might need to be
appended for CellRanger version below 3.0.0 if reads were mapped to
multiple genomes when running Cell Ranger pipeline. Probable options
include "outs/filtered_feature_bc_matrix/", "outs/raw_feature_bc_matrix/",
"outs/filtered_gene_bc_matrix/", "outs/raw_gene_bc_matrix/".</p></dd>


<dt>dataType</dt>
<dd><p>Character. The type of data to import. Can be one of
"filtered" (which is equivalent to
<code>cellRangerOuts = "outs/filtered_feature_bc_matrix/"</code> or
<code>cellRangerOuts = "outs/filtered_gene_bc_matrix/"</code>) or "raw" (which
is equivalent to
<code>cellRangerOuts = "outs/raw_feature_bc_matrix/"</code> or
<code>cellRangerOuts = "outs/raw_gene_bc_matrix/"</code>). Default
"filtered" which imports the counts for filtered cell barcodes only.</p></dd>


<dt>matrixFileNames</dt>
<dd><p>Character vector. Filenames for the Market Exchange
Format (MEX) sparse matrix files (matrix.mtx or matrix.mtx.gz files).
Must have length 1 or the same
length as <code>length(unlist(sampleDirs))</code> if
<code>sampleDirs</code> is not <code>NULL</code>. Otherwise, make sure the length and
order match the output of
<code>unlist(lapply(cellRangerDirs, list.dirs, recursive = FALSE))</code>.</p></dd>


<dt>featuresFileNames</dt>
<dd><p>Character vector. Filenames for the feature
annotation files. They are usually named <em>features.tsv.gz</em> or
<em>genes.tsv</em>. Must have length 1 or the same
length as <code>length(unlist(sampleDirs))</code> if
<code>sampleDirs</code> is not <code>NULL</code>. Otherwise, make sure the length and
order match the output of
<code>unlist(lapply(cellRangerDirs, list.dirs, recursive = FALSE))</code>.</p></dd>


<dt>barcodesFileNames</dt>
<dd><p>Character vector. Filename for the cell barcode
list files. They are usually named <em>barcodes.tsv.gz</em> or
<em>barcodes.tsv</em>. Must have length 1 or the same
length as <code>length(unlist(sampleDirs))</code> if
<code>sampleDirs</code> is not <code>NULL</code>. Otherwise, make sure the length and
order match the output of
<code>unlist(lapply(cellRangerDirs, list.dirs, recursive = FALSE))</code>.</p></dd>


<dt>gzipped</dt>
<dd><p><code>TRUE</code> if the Cell Ranger output files
(barcodes.tsv, features.tsv, and matrix.mtx) were
gzip compressed. <code>FALSE</code> otherwise. This is true after Cell Ranger
3.0.0 update. Default <code>"auto"</code> which automatically detects if the
files are gzip compressed. If not <code>"auto"</code>, <code>gzipped</code> must have
length 1 or the same
length as <code>length(unlist(sampleDirs))</code> if
<code>sampleDirs</code> is not <code>NULL</code>. Otherwise, make sure the length and
order match the output of
<code>unlist(lapply(cellRangerDirs, list.dirs, recursive = FALSE))</code>.</p></dd>


<dt>class</dt>
<dd><p>Character. The class of the expression matrix stored in the SCE
object. Can be one of "Matrix" (as returned by
readMM function), or "matrix" (as returned by
<a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a> function). Default <code>"Matrix"</code>.</p></dd>


<dt>delayedArray</dt>
<dd><p>Boolean. Whether to read the expression matrix as
DelayedArray object or not. Default <code>FALSE</code>.</p></dd>


<dt>rowNamesDedup</dt>
<dd><p>Boolean. Whether to deduplicate rownames. Default 
<code>TRUE</code>.</p></dd>


<dt>dataTypeV2</dt>
<dd><p>Character. The type of output to import for
Cellranger version below 3.0.0. Whether to import the filtered or the
raw data. Can be one of 'filtered' or 'raw'. Default 'filtered'. When
<code>cellRangerOuts</code> is specified, <code>dataTypeV2</code> and <code>reference</code> will
be ignored.</p></dd>


<dt>reference</dt>
<dd><p>Character vector. The reference genome names.
Default <code>NULL</code>. If not <code>NULL</code>, it must gave the length and order as
<code>length(unlist(sampleDirs))</code> if <code>sampleDirs</code> is not <code>NULL</code>.
Otherwise, make sure the length and order match the output of
<code>unlist(lapply(cellRangerDirs, list.dirs, recursive = FALSE))</code>. Only needed
for Cellranger version below 3.0.0.</p></dd>


<dt>cellRangerOutsV2</dt>
<dd><p>Character vector. The intermediate paths
to filtered or raw cell barcode, feature, and matrix files for each
sample for Cellranger version below 3.0.0. If <code>NULL</code>, <code>reference</code> and
<code>dataTypeV2</code> will be used to determine Cell Ranger output directory. If it has
length 1, it assumes that all samples use the same genome reference and
the function will load only filtered or raw data.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code>SingleCellExperiment</code> object containing the combined count
 matrix, the feature annotations, and the cell annotation.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><code>importCellRangerV2</code> imports output from Cell Ranger V2.
 <code>importCellRangerV2Sample</code> imports output from one sample from Cell
 Ranger V2.
 <code>importCellRangerV3</code> imports output from Cell Ranger V3.
 <code>importCellRangerV3</code> imports output from one sample from Cell Ranger
 V3.
 Some implicit
 assumptions which match the output structure of Cell Ranger V2 &amp; V3
 are made in these 4 functions including <code>cellRangerOuts</code>,
 <code>matrixFileName</code>, <code>featuresFileName</code>, <code>barcodesFileName</code>,
 and <code>gzipped</code>.
 Alternatively, user can call <code>importCellRanger</code> to explicitly
 specify these arguments.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example #1</span></span></span>
<span class="r-in"><span><span class="co"># The following filtered feature, cell, and matrix files were downloaded from</span></span></span>
<span class="r-in"><span><span class="co"># https://support.10xgenomics.com/single-cell-gene-expression/datasets/</span></span></span>
<span class="r-in"><span><span class="co"># 3.0.0/hgmm_1k_v3</span></span></span>
<span class="r-in"><span><span class="co"># The top 10 hg19 &amp; mm10 genes are included in this example.</span></span></span>
<span class="r-in"><span><span class="co"># Only the first 20 cells are included.</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">importCellRanger</span><span class="op">(</span></span></span>
<span class="r-in"><span>    cellRangerDirs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/"</span>, package <span class="op">=</span> <span class="st">"singleCellTK"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    sampleDirs <span class="op">=</span> <span class="st">"hgmm_1k_v3_20x20"</span>,</span></span>
<span class="r-in"><span>    sampleNames <span class="op">=</span> <span class="st">"hgmm1kv3"</span>,</span></span>
<span class="r-in"><span>    dataType <span class="op">=</span> <span class="st">"filtered"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Metrics summary file (metrics_summary.csv) not found for sample: hgmm1kv3</span>
<span class="r-in"><span><span class="co"># The following filtered feature, cell, and matrix files were downloaded from</span></span></span>
<span class="r-in"><span><span class="co"># https://support.10xgenomics.com/single-cell-gene-expression/datasets/</span></span></span>
<span class="r-in"><span><span class="co"># 2.1.0/pbmc4k</span></span></span>
<span class="r-in"><span><span class="co"># Top 20 genes are kept. 20 cell barcodes are extracted.</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">importCellRangerV2</span><span class="op">(</span></span></span>
<span class="r-in"><span>    cellRangerDirs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/"</span>, package <span class="op">=</span> <span class="st">"singleCellTK"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    sampleDirs <span class="op">=</span> <span class="st">"pbmc_4k_v2_20x20"</span>,</span></span>
<span class="r-in"><span>    sampleNames <span class="op">=</span> <span class="st">"pbmc4k_20"</span>,</span></span>
<span class="r-in"><span>    reference <span class="op">=</span> <span class="st">'GRCh38'</span>,</span></span>
<span class="r-in"><span>    dataTypeV2 <span class="op">=</span> <span class="st">"filtered"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Metrics summary file (metrics_summary.csv) not found for sample: pbmc4k_20</span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">importCellRangerV3</span><span class="op">(</span></span></span>
<span class="r-in"><span>    cellRangerDirs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/"</span>, package <span class="op">=</span> <span class="st">"singleCellTK"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    sampleDirs <span class="op">=</span> <span class="st">"hgmm_1k_v3_20x20"</span>,</span></span>
<span class="r-in"><span>    sampleNames <span class="op">=</span> <span class="st">"hgmm1kv3"</span>,</span></span>
<span class="r-in"><span>    dataType <span class="op">=</span> <span class="st">"filtered"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Metrics summary file (metrics_summary.csv) not found for sample: hgmm1kv3</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

