<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ singleCellTK</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  <script src="../extra.js"></script>

<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import Data in Commandline</a>
        </li>
        <li>
          <a href="../articles/articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cmd_qc.html">QC with commandline</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_cellqc.html">Cell QC in R console</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_dropletqc.html">Droplet QC in R console</a>
        </li>
        <li>
          <a href="../articles/articles/ui_qc.html">QC in the UI</a>
        </li>
        <li>
          <a href="../articles/articles/filtering.html">Filtering</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
    </li>
    <li>
      <a href="../articles/articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/articles/console_analysis_tutorial.html">Console Analysis Tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fas fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1">
    <span class="fas fa-twitter"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-importing-scrna-seq-data" class="hasAnchor"><a href="#section-importing-scrna-seq-data" class="anchor"></a>Importing scRNA-seq data</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="importAlevin.html">importAlevin()</a></code> </p>
        </td>
        <td><p>Construct SCE object from Salmon-Alevin output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importAnnData.html">importAnnData()</a></code> </p>
        </td>
        <td><p>Create a SingleCellExperiment Object from Python AnnData .h5ad files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importBUStools.html">importBUStools()</a></code> </p>
        </td>
        <td><p>Construct SCE object from BUStools output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importCellRanger.html">importCellRanger()</a></code> <code><a href="importCellRanger.html">importCellRangerV2()</a></code> <code><a href="importCellRanger.html">importCellRangerV3()</a></code> </p>
        </td>
        <td><p>Construct SCE object from Cell Ranger output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importCellRangerV2Sample.html">importCellRangerV2Sample()</a></code> </p>
        </td>
        <td><p>Construct SCE object from Cell Ranger V2 output for a single sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importCellRangerV3Sample.html">importCellRangerV3Sample()</a></code> </p>
        </td>
        <td><p>Construct SCE object from Cell Ranger V3 output for a single sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importDropEst.html">importDropEst()</a></code> </p>
        </td>
        <td><p>Create a SingleCellExperiment Object from DropEst output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importExampleData.html">importExampleData()</a></code> </p>
        </td>
        <td><p>Retrieve example datasets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importFromFiles.html">importFromFiles()</a></code> </p>
        </td>
        <td><p>Create a SingleCellExperiment object from files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importGeneSetsFromCollection.html">importGeneSetsFromCollection()</a></code> </p>
        </td>
        <td><p>Imports gene sets from a GeneSetCollection object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importGeneSetsFromGMT.html">importGeneSetsFromGMT()</a></code> </p>
        </td>
        <td><p>Imports gene sets from a GMT file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importGeneSetsFromList.html">importGeneSetsFromList()</a></code> </p>
        </td>
        <td><p>Imports gene sets from a list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importGeneSetsFromMSigDB.html">importGeneSetsFromMSigDB()</a></code> </p>
        </td>
        <td><p>Imports gene sets from MSigDB</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importMitoGeneSet.html">importMitoGeneSet()</a></code> </p>
        </td>
        <td><p>Import mitochondrial gene sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importMultipleSources.html">importMultipleSources()</a></code> </p>
        </td>
        <td><p>Imports samples from different sources and compiles them into a list of SCE objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importOptimus.html">importOptimus()</a></code> </p>
        </td>
        <td><p>Construct SCE object from Optimus output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importSEQC.html">importSEQC()</a></code> </p>
        </td>
        <td><p>Construct SCE object from seqc output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="importSTARsolo.html">importSTARsolo()</a></code> </p>
        </td>
        <td><p>Construct SCE object from STARsolo outputs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="readSingleCellMatrix.html">readSingleCellMatrix()</a></code> </p>
        </td>
        <td><p>Read single cell expression matrix</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-quality-control-preprocessing" class="hasAnchor"><a href="#section-quality-control-preprocessing" class="anchor"></a>Quality Control &amp; Preprocessing</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="runCellQC.html">runCellQC()</a></code> </p>
        </td>
        <td><p>Perform comprehensive single cell QC</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runDropletQC.html">runDropletQC()</a></code> </p>
        </td>
        <td><p>Perform comprehensive droplet QC</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-decontamination" class="hasAnchor"><a href="#section-decontamination" class="anchor"></a>Decontamination</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="runDecontX.html">runDecontX()</a></code> </p>
        </td>
        <td><p>Detecting contamination with DecontX.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-doublet-empty-droplet-detection" class="hasAnchor"><a href="#section-doublet-empty-droplet-detection" class="anchor"></a>Doublet/Empty Droplet Detection</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="runBarcodeRankDrops.html">runBarcodeRankDrops()</a></code> </p>
        </td>
        <td><p>Identify empty droplets using <a href='https://rdrr.io/pkg/DropletUtils/man/barcodeRanks.html'>barcodeRanks</a>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runEmptyDrops.html">runEmptyDrops()</a></code> </p>
        </td>
        <td><p>Identify empty droplets using <a href='https://rdrr.io/pkg/DropletUtils/man/emptyDrops.html'>emptyDrops</a>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runBcds.html">runBcds()</a></code> </p>
        </td>
        <td><p>Find doublets/multiplets using <a href='https://rdrr.io/pkg/scds/man/bcds.html'>bcds</a>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runCxds.html">runCxds()</a></code> </p>
        </td>
        <td><p>Find doublets/multiplets using <a href='https://rdrr.io/pkg/scds/man/cxds.html'>cxds</a>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runCxdsBcdsHybrid.html">runCxdsBcdsHybrid()</a></code> </p>
        </td>
        <td><p>Find doublets/multiplets using <a href='https://rdrr.io/pkg/scds/man/cxds_bcds_hybrid.html'>cxds_bcds_hybrid</a>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runScDblFinder.html">runScDblFinder()</a></code> </p>
        </td>
        <td><p>Detect doublet cells using <a href='https://rdrr.io/pkg/scDblFinder/man/scDblFinder.html'>scDblFinder</a>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runDoubletFinder.html">runDoubletFinder()</a></code> </p>
        </td>
        <td><p>Generates a doublet score for each cell via doubletFinder</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runScrublet.html">runScrublet()</a></code> </p>
        </td>
        <td><p>Find doublets using <code>scrublet</code>.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-normalization" class="hasAnchor"><a href="#section-normalization" class="anchor"></a>Normalization</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="runNormalization.html">runNormalization()</a></code> </p>
        </td>
        <td><p>Wrapper function to run any of the integrated normalization/transformation
methods in the singleCellTK. The available methods include 'LogNormalize',
'CLR', 'RC' and 'SCTransform' from Seurat, 'logNormCounts and 'CPM' from
Scater. Additionally, users can 'scale' using Z.Score, 'transform' using
log, log1p and sqrt, add 'pseudocounts' and trim the final matrices
between a range of values.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaterlogNormCounts.html">scaterlogNormCounts()</a></code> </p>
        </td>
        <td><p>scaterlogNormCounts
Uses logNormCounts to log normalize input data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaterCPM.html">scaterCPM()</a></code> </p>
        </td>
        <td><p>scaterCPM
Uses CPM from scater library to compute counts-per-million.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratNormalizeData.html">seuratNormalizeData()</a></code> </p>
        </td>
        <td><p>seuratNormalizeData
Wrapper for NormalizeData() function from seurat library
Normalizes the sce object according to the input parameters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratScaleData.html">seuratScaleData()</a></code> </p>
        </td>
        <td><p>seuratScaleData
Scales the input sce object according to the input parameters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="computeZScore.html">computeZScore()</a></code> </p>
        </td>
        <td><p>Compute Z-Score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trimCounts.html">trimCounts()</a></code> </p>
        </td>
        <td><p>Trim Counts</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-batch-effect-correction" class="hasAnchor"><a href="#section-batch-effect-correction" class="anchor"></a>Batch Effect Correction</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="runComBatSeq.html">runComBatSeq()</a></code> </p>
        </td>
        <td><p>Apply ComBat-Seq batch effect correction method to SingleCellExperiment
object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runBBKNN.html">runBBKNN()</a></code> </p>
        </td>
        <td><p>Apply BBKNN batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runFastMNN.html">runFastMNN()</a></code> </p>
        </td>
        <td><p>Apply a fast version of the mutual nearest neighbors (MNN) batch effect
correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runLimmaBC.html">runLimmaBC()</a></code> </p>
        </td>
        <td><p>Apply Limma's batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runMNNCorrect.html">runMNNCorrect()</a></code> </p>
        </td>
        <td><p>Apply the mutual nearest neighbors (MNN) batch effect correction method to
SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runSCANORAMA.html">runSCANORAMA()</a></code> </p>
        </td>
        <td><p>Apply the mutual nearest neighbors (MNN) batch effect correction method to
SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runSCMerge.html">runSCMerge()</a></code> </p>
        </td>
        <td><p>Apply scMerge batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratIntegration.html">seuratIntegration()</a></code> </p>
        </td>
        <td><p>seuratIntegration
A wrapper function to Seurat Batch-Correction/Integration workflow.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runZINBWaVE.html">runZINBWaVE()</a></code> </p>
        </td>
        <td><p>Apply ZINBWaVE Batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotBatchVariance.html">plotBatchVariance()</a></code> </p>
        </td>
        <td><p>Plot the percent of the variation that is explained by batch and condition
in the data</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-feature-selection" class="hasAnchor"><a href="#section-feature-selection" class="anchor"></a>Feature Selection</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="scranModelGeneVar.html">scranModelGeneVar()</a></code> </p>
        </td>
        <td><p>scranModelGeneVar
Generates and stores variability data from scran::modelGeneVar in the input singleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratFindHVG.html">seuratFindHVG()</a></code> </p>
        </td>
        <td><p>seuratFindHVG
Find highly variable genes and store in the input sce object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getTopHVG.html">getTopHVG()</a></code> </p>
        </td>
        <td><p>getTopHVG
Extracts the top variable genes from an input <code>SingleCellExperiment</code> object.
Note that the variability metrics must be computed using the `runFeatureSelection`
method before extracting the feature names of the top variable features. If
`altExp` parameter is a <code>character</code> value, this function will return the
input <code>SingleCellExperiment</code> object with the subset containing only the
top variable features stored as an <code>altExp</code> slot in returned object.
However, if this parameter is set to <code>NULL</code>, only the names of the top
variable features will be returned as a <code>character</code> vector.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratPlotHVG.html">seuratPlotHVG()</a></code> </p>
        </td>
        <td><p>seuratPlotHVG
Plot highly variable genes from input sce object (must have highly variable
genes computations stored)</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-dimensionality-reduction" class="hasAnchor"><a href="#section-dimensionality-reduction" class="anchor"></a>Dimensionality Reduction</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="scaterPCA.html">scaterPCA()</a></code> </p>
        </td>
        <td><p>Perform PCA on a SingleCellExperiment Object
A wrapper to <a href='https://rdrr.io/pkg/scater/man/runPCA.html'>runPCA</a> function to compute principal component
analysis (PCA) from a given SingleCellExperiment object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getUMAP.html">getUMAP()</a></code> </p>
        </td>
        <td><p>Uniform Manifold Approximation and Projection(UMAP) algorithm for
dimension reduction.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getTSNE.html">getTSNE()</a></code> </p>
        </td>
        <td><p>Run t-SNE dimensionality reduction method on a SingleCellExperiment Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratICA.html">seuratICA()</a></code> </p>
        </td>
        <td><p>seuratICA
Computes ICA on the input sce object and stores the calculated independent
components within the sce object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratPCA.html">seuratPCA()</a></code> </p>
        </td>
        <td><p>seuratPCA
Computes PCA on the input sce object and stores the calculated principal
components within the sce object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratRunUMAP.html">seuratRunUMAP()</a></code> </p>
        </td>
        <td><p>seuratRunUMAP
Computes UMAP from the given sce object and stores the UMAP computations back
into the sce object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratRunTSNE.html">seuratRunTSNE()</a></code> </p>
        </td>
        <td><p>seuratRunTSNE
Computes tSNE from the given sce object and stores the tSNE computations back
into the sce object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runDimReduce.html">runDimReduce()</a></code> </p>
        </td>
        <td><p>Generic Wrapper function for running dimensionality reduction</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEDimReduceColData.html">plotSCEDimReduceColData()</a></code> </p>
        </td>
        <td><p>Dimension reduction plot tool for colData</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEDimReduceFeatures.html">plotSCEDimReduceFeatures()</a></code> </p>
        </td>
        <td><p>Dimension reduction plot tool for assay data</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-clustering" class="hasAnchor"><a href="#section-clustering" class="anchor"></a>Clustering</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="runScranSNN.html">runScranSNN()</a></code> </p>
        </td>
        <td><p>Get clustering with SNN graph</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seuratFindClusters.html">seuratFindClusters()</a></code> </p>
        </td>
        <td><p>seuratFindClusters
Computes the clusters from the input sce object and stores them back in sce
object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runKMeans.html">runKMeans()</a></code> </p>
        </td>
        <td><p>Get clustering with KMeans</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-differential-expression" class="hasAnchor"><a href="#section-differential-expression" class="anchor"></a>Differential Expression</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="runDEAnalysis.html">runDEAnalysis()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis on SCE with specified method
Method supported: 'MAST', 'DESeq2', 'Limma', 'ANOVA'</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runWilcox.html">runWilcox()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis on SCE with Wilcoxon test</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runMAST.html">runMAST()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis on SCE with MAST</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runDESeq2.html">runDESeq2()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis on SCE with DESeq2.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runLimmaDE.html">runLimmaDE()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis on SCE with Limma.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runANOVA.html">runANOVA()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis on SCE with ANOVA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGViolin.html">plotDEGViolin()</a></code> </p>
        </td>
        <td><p>plot the violin plot to show visualize the expression distribution of DEGs
identified by differential expression analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGRegression.html">plotDEGRegression()</a></code> </p>
        </td>
        <td><p>plot the linear regression to show visualize the expression the of DEGs
identified by differential expression analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGHeatmap.html">plotDEGHeatmap()</a></code> </p>
        </td>
        <td><p>Heatmap visualization of DEG result</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMASTThresholdGenes.html">plotMASTThresholdGenes()</a></code> </p>
        </td>
        <td><p>MAST Identify adaptive thresholds</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-find-marker" class="hasAnchor"><a href="#section-find-marker" class="anchor"></a>Find Marker</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="findMarkerDiffExp.html">findMarkerDiffExp()</a></code> </p>
        </td>
        <td><p>Find the marker gene set for each cluster
With an input SingleCellExperiment object and specifying the clustering
labels, this function iteratively call the differential expression analysis
on each cluster against all the others.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findMarkerTopTable.html">findMarkerTopTable()</a></code> </p>
        </td>
        <td><p>Fetch the table of top markers that pass the filtering</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMarkerDiffExp.html">plotMarkerDiffExp()</a></code> <code><a href="plotMarkerDiffExp.html">plotMarkerDiffExp()</a></code> </p>
        </td>
        <td><p>Plot a heatmap to visualize the result of <code>findMarkerDiffExp</code></p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-enrichment-pathway-analysis" class="hasAnchor"><a href="#section-enrichment-pathway-analysis" class="anchor"></a>Enrichment &amp; Pathway Analysis</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="getMSigDBTable.html">getMSigDBTable()</a></code> </p>
        </td>
        <td><p>Shows MSigDB categories</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="enrichRSCE.html">enrichRSCE()</a></code> </p>
        </td>
        <td><p>enrichR
Given a list of genes this function runs the enrichR() to perform Gene
enrichment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runGSVA.html">runGSVA()</a></code> </p>
        </td>
        <td><p>Run GSVA analysis on a SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runVAM.html">runVAM()</a></code> </p>
        </td>
        <td><p>Run VAM to score gene sets in single cell data</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-visualization" class="hasAnchor"><a href="#section-visualization" class="anchor"></a>Visualization</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="plotBarcodeRankDropsResults.html">plotBarcodeRankDropsResults()</a></code> </p>
        </td>
        <td><p>Plots for runEmptyDrops outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotBarcodeRankScatter.html">plotBarcodeRankScatter()</a></code> </p>
        </td>
        <td><p>Plots for runBarcodeRankDrops outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotBatchVariance.html">plotBatchVariance()</a></code> </p>
        </td>
        <td><p>Plot the percent of the variation that is explained by batch and condition
in the data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotBcdsResults.html">plotBcdsResults()</a></code> </p>
        </td>
        <td><p>Plots for runBcds outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotBiomarker.html">plotBiomarker()</a></code> </p>
        </td>
        <td><p>Given a set of genes, return a ggplot of expression
values.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotClusterAbundance.html">plotClusterAbundance()</a></code> </p>
        </td>
        <td><p>Plot the differential Abundance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCxdsResults.html">plotCxdsResults()</a></code> </p>
        </td>
        <td><p>Plots for runCxds outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDecontXResults.html">plotDecontXResults()</a></code> </p>
        </td>
        <td><p>Plots for runDecontX outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGHeatmap.html">plotDEGHeatmap()</a></code> </p>
        </td>
        <td><p>Heatmap visualization of DEG result</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGRegression.html">plotDEGRegression()</a></code> </p>
        </td>
        <td><p>plot the linear regression to show visualize the expression the of DEGs
identified by differential expression analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGViolin.html">plotDEGViolin()</a></code> </p>
        </td>
        <td><p>plot the violin plot to show visualize the expression distribution of DEGs
identified by differential expression analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDimRed.html">plotDimRed()</a></code> </p>
        </td>
        <td><p>Plot dimensionality reduction from computed metrics including PCA, ICA, tSNE
and UMAP</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDoubletFinderResults.html">plotDoubletFinderResults()</a></code> </p>
        </td>
        <td><p>Plots for runDoubletFinder outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotEmptyDropsResults.html">plotEmptyDropsResults()</a></code> </p>
        </td>
        <td><p>Plots for runEmptyDrops outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotEmptyDropsScatter.html">plotEmptyDropsScatter()</a></code> </p>
        </td>
        <td><p>Plots for runEmptyDrops outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMarkerDiffExp.html">plotMarkerDiffExp()</a></code> <code><a href="plotMarkerDiffExp.html">plotMarkerDiffExp()</a></code> </p>
        </td>
        <td><p>Plot a heatmap to visualize the result of <code>findMarkerDiffExp</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMASTThresholdGenes.html">plotMASTThresholdGenes()</a></code> </p>
        </td>
        <td><p>MAST Identify adaptive thresholds</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotPCA.html">plotPCA()</a></code> </p>
        </td>
        <td><p>Plot PCA run data from its components.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotRunPerCellQCResults.html">plotRunPerCellQCResults()</a></code> </p>
        </td>
        <td><p>Plots for runPerCellQC outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotScDblFinderResults.html">plotScDblFinderResults()</a></code> </p>
        </td>
        <td><p>Plots for runScDblFinder outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotScdsHybridResults.html">plotScdsHybridResults()</a></code> </p>
        </td>
        <td><p>Plots for runCxdsBcdsHybrid outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEBarAssayData.html">plotSCEBarAssayData()</a></code> </p>
        </td>
        <td><p>Bar plot of assay data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEBarColData.html">plotSCEBarColData()</a></code> </p>
        </td>
        <td><p>Bar plot of colData.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEBatchFeatureMean.html">plotSCEBatchFeatureMean()</a></code> </p>
        </td>
        <td><p>Plot mean feature value in each batch of a SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEDensity.html">plotSCEDensity()</a></code> </p>
        </td>
        <td><p>Density plot of any data stored in the SingleCellExperiment object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEDensityAssayData.html">plotSCEDensityAssayData()</a></code> </p>
        </td>
        <td><p>Density plot of assay data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEDensityColData.html">plotSCEDensityColData()</a></code> </p>
        </td>
        <td><p>Density plot of colData.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEDimReduceColData.html">plotSCEDimReduceColData()</a></code> </p>
        </td>
        <td><p>Dimension reduction plot tool for colData</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEDimReduceFeatures.html">plotSCEDimReduceFeatures()</a></code> </p>
        </td>
        <td><p>Dimension reduction plot tool for assay data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEHeatmap.html">plotSCEHeatmap()</a></code> </p>
        </td>
        <td><p>Plot heatmap of using data stored in SingleCellExperiment Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEScatter.html">plotSCEScatter()</a></code> </p>
        </td>
        <td><p>Dimension reduction plot tool for all types of data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEViolin.html">plotSCEViolin()</a></code> </p>
        </td>
        <td><p>Violin plot of any data stored in the SingleCellExperiment object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEViolinAssayData.html">plotSCEViolinAssayData()</a></code> </p>
        </td>
        <td><p>Violin plot of assay data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEViolinColData.html">plotSCEViolinColData()</a></code> </p>
        </td>
        <td><p>Violin plot of colData.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotScrubletResults.html">plotScrubletResults()</a></code> </p>
        </td>
        <td><p>Plots for runScrublet outputs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotTopHVG.html">plotTopHVG()</a></code> </p>
        </td>
        <td><p>Plot highly variable genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotTSNE.html">plotTSNE()</a></code> </p>
        </td>
        <td><p>Plot t-SNE plot on dimensionality reduction data run from t-SNE method.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotUMAP.html">plotUMAP()</a></code> </p>
        </td>
        <td><p>Plot UMAP results either on already run results or run first and then plot.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-report-generation" class="hasAnchor"><a href="#section-report-generation" class="anchor"></a>Report Generation</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="reportCellQC.html">reportCellQC()</a></code> </p>
        </td>
        <td><p>Get runCellQC .html report</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reportDiffExp.html">reportDiffExp()</a></code> </p>
        </td>
        <td><p>Get runDEAnalysis .html report</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reportDropletQC.html">reportDropletQC()</a></code> </p>
        </td>
        <td><p>Get runDropletQC .html report</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reportFindMarker.html">reportFindMarker()</a></code> </p>
        </td>
        <td><p>Get findMarkerDiffExp .html report</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reportQCTool.html">reportQCTool()</a></code> </p>
        </td>
        <td><p>Get .html report of the output of the selected QC algorithm</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-exporting-results" class="hasAnchor"><a href="#section-exporting-results" class="anchor"></a>Exporting Results</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="exportSCE.html">exportSCE()</a></code> </p>
        </td>
        <td><p>Export data in SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="exportSCEtoAnnData.html">exportSCEtoAnnData()</a></code> </p>
        </td>
        <td><p>Export a <a href='https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html'>SingleCellExperiment</a> R object as
Python annData object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="exportSCEtoFlatFile.html">exportSCEtoFlatFile()</a></code> </p>
        </td>
        <td><p>Export a <a href='https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html'>SingleCellExperiment</a> object to flat text files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="exportSCEToSeurat.html">exportSCEToSeurat()</a></code> </p>
        </td>
        <td><p>Export data in Seurat object</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-datasets" class="hasAnchor"><a href="#section-datasets" class="anchor"></a>Datasets</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="mouseBrainSubsetSCE.html">mouseBrainSubsetSCE</a></code> </p>
        </td>
        <td><p>Example Single Cell RNA-Seq data in SingleCellExperiment Object, GSE60361
subset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sceBatches.html">sceBatches</a></code> </p>
        </td>
        <td><p>Example Single Cell RNA-Seq data in SingleCellExperiment object, with
different batches annotated</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MitoGenes.html">MitoGenes</a></code> </p>
        </td>
        <td><p>List of mitochondrial genes of multiple reference</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="msigdb_table.html">msigdb_table</a></code> </p>
        </td>
        <td><p>MSigDB gene get Cctegory table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SEG.html">SEG</a></code> </p>
        </td>
        <td><p>Stably Expressed Gene (SEG) list obect, with SEG sets for human and mouse.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-other-data-processing" class="hasAnchor"><a href="#section-other-data-processing" class="anchor"></a>Other Data processing</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="combineSCE.html">combineSCE()</a></code> </p>
        </td>
        <td><p>Combine a list of SingleCellExperiment objects as one SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertSCEToSeurat.html">convertSCEToSeurat()</a></code> </p>
        </td>
        <td><p>convertSCEToSeurat
Converts sce object to seurat while retaining all assays and metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertSeuratToSCE.html">convertSeuratToSCE()</a></code> </p>
        </td>
        <td><p>convertSeuratToSCE
Converts the input seurat object to a sce object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertGeneIDs.html">convertGeneIDs()</a></code> </p>
        </td>
        <td><p>Convert Gene IDs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectCellOutlier.html">detectCellOutlier()</a></code> </p>
        </td>
        <td><p>Detecting outliers within the SingleCellExperiment object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsetSCECols.html">subsetSCECols()</a></code> </p>
        </td>
        <td><p>Subset a SingleCellExperiment object by columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsetSCERows.html">subsetSCERows()</a></code> </p>
        </td>
        <td><p>Subset a SingleCellExperiment object by rows</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dedupRowNames.html">dedupRowNames()</a></code> </p>
        </td>
        <td><p>Deduplicate the rownames of a matrix or SingleCellExperiment object
Adds '-1', '-2', ... '-i' to multiple duplicated rownames, and in place
replace the unique rownames, store unique rownames in <code>rowData</code>, or
return the unique rownames as character vecetor.</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


