<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ singleCellTK</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.13.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul></li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul></li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="start-the-shiny-app">Start the Shiny APP <a href="#start-the-shiny-app" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="singleCellTK.html">singleCellTK()</a></code> </p>
        </td>
        <td><p>Run the single cell analysis app</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="importing-scrna-seq-data">Importing scRNA-seq Data <a href="#importing-scrna-seq-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="importAlevin.html">importAlevin()</a></code> </p>
        </td>
        <td><p>Construct SCE object from Salmon-Alevin output</p></td>
      </tr><tr><td>
          <p><code><a href="importAnnData.html">importAnnData()</a></code> </p>
        </td>
        <td><p>Create a SingleCellExperiment Object from Python AnnData .h5ad files</p></td>
      </tr><tr><td>
          <p><code><a href="importBUStools.html">importBUStools()</a></code> </p>
        </td>
        <td><p>Construct SCE object from BUStools output</p></td>
      </tr><tr><td>
          <p><code><a href="importCellRanger.html">importCellRanger()</a></code> <code><a href="importCellRanger.html">importCellRangerV2()</a></code> <code><a href="importCellRanger.html">importCellRangerV3()</a></code> </p>
        </td>
        <td><p>Construct SCE object from Cell Ranger output</p></td>
      </tr><tr><td>
          <p><code><a href="importCellRangerV2Sample.html">importCellRangerV2Sample()</a></code> </p>
        </td>
        <td><p>Construct SCE object from Cell Ranger V2 output for a single sample</p></td>
      </tr><tr><td>
          <p><code><a href="importCellRangerV3Sample.html">importCellRangerV3Sample()</a></code> </p>
        </td>
        <td><p>Construct SCE object from Cell Ranger V3 output for a single sample</p></td>
      </tr><tr><td>
          <p><code><a href="importDropEst.html">importDropEst()</a></code> </p>
        </td>
        <td><p>Create a SingleCellExperiment Object from DropEst output</p></td>
      </tr><tr><td>
          <p><code><a href="importExampleData.html">importExampleData()</a></code> </p>
        </td>
        <td><p>Retrieve example datasets</p></td>
      </tr><tr><td>
          <p><code><a href="importFromFiles.html">importFromFiles()</a></code> </p>
        </td>
        <td><p>Create a SingleCellExperiment object from files</p></td>
      </tr><tr><td>
          <p><code><a href="importGeneSetsFromCollection.html">importGeneSetsFromCollection()</a></code> </p>
        </td>
        <td><p>Imports gene sets from a GeneSetCollection object</p></td>
      </tr><tr><td>
          <p><code><a href="importGeneSetsFromGMT.html">importGeneSetsFromGMT()</a></code> </p>
        </td>
        <td><p>Imports gene sets from a GMT file</p></td>
      </tr><tr><td>
          <p><code><a href="importGeneSetsFromList.html">importGeneSetsFromList()</a></code> </p>
        </td>
        <td><p>Imports gene sets from a list</p></td>
      </tr><tr><td>
          <p><code><a href="importGeneSetsFromMSigDB.html">importGeneSetsFromMSigDB()</a></code> </p>
        </td>
        <td><p>Imports gene sets from MSigDB</p></td>
      </tr><tr><td>
          <p><code><a href="importMitoGeneSet.html">importMitoGeneSet()</a></code> </p>
        </td>
        <td><p>Import mitochondrial gene sets</p></td>
      </tr><tr><td>
          <p><code><a href="importMultipleSources.html">importMultipleSources()</a></code> </p>
        </td>
        <td><p>Imports samples from different sources and compiles them into a list of SCE objects</p></td>
      </tr><tr><td>
          <p><code><a href="importOptimus.html">importOptimus()</a></code> </p>
        </td>
        <td><p>Construct SCE object from Optimus output</p></td>
      </tr><tr><td>
          <p><code><a href="importSEQC.html">importSEQC()</a></code> </p>
        </td>
        <td><p>Construct SCE object from seqc output</p></td>
      </tr><tr><td>
          <p><code><a href="importSTARsolo.html">importSTARsolo()</a></code> </p>
        </td>
        <td><p>Construct SCE object from STARsolo outputs</p></td>
      </tr><tr><td>
          <p><code><a href="readSingleCellMatrix.html">readSingleCellMatrix()</a></code> </p>
        </td>
        <td><p>Read single cell expression matrix</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="quality-control-preprocessing">Quality Control &amp; Preprocessing <a href="#quality-control-preprocessing" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runCellQC.html">runCellQC()</a></code> </p>
        </td>
        <td><p>Perform comprehensive single cell QC</p></td>
      </tr><tr><td>
          <p><code><a href="runDropletQC.html">runDropletQC()</a></code> </p>
        </td>
        <td><p>Perform comprehensive droplet QC</p></td>
      </tr><tr><td>
          <p><code><a href="runPerCellQC.html">runPerCellQC()</a></code> </p>
        </td>
        <td><p>Wrapper for calculating QC metrics with scater.</p></td>
      </tr><tr><td>
          <p><code><a href="reportDropletQC.html">reportDropletQC()</a></code> </p>
        </td>
        <td><p>Get runDropletQC .html report</p></td>
      </tr><tr><td>
          <p><code><a href="reportCellQC.html">reportCellQC()</a></code> </p>
        </td>
        <td><p>Get runCellQC .html report</p></td>
      </tr><tr><td>
          <p><code><a href="plotRunPerCellQCResults.html">plotRunPerCellQCResults()</a></code> </p>
        </td>
        <td><p>Plots for runPerCellQC outputs.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="decontamination">Decontamination <a href="#decontamination" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runDecontX.html">runDecontX()</a></code> </p>
        </td>
        <td><p>Detecting contamination with DecontX.</p></td>
      </tr><tr><td>
          <p><code><a href="plotDecontXResults.html">plotDecontXResults()</a></code> </p>
        </td>
        <td><p>Plots for runDecontX outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="runSoupX.html">runSoupX()</a></code> </p>
        </td>
        <td><p>Detecting and correct contamination with SoupX</p></td>
      </tr><tr><td>
          <p><code><a href="getSoupX.html">`getSoupX&lt;-`()</a></code> <code><a href="getSoupX.html">getSoupX()</a></code> </p>
        </td>
        <td><p>Get or Set SoupX Result</p></td>
      </tr><tr><td>
          <p><code><a href="plotSoupXResults.html">plotSoupXResults()</a></code> </p>
        </td>
        <td><p>Plot SoupX Result</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="doublet-empty-droplet-detection">Doublet/Empty Droplet Detection <a href="#doublet-empty-droplet-detection" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runBarcodeRankDrops.html">runBarcodeRankDrops()</a></code> </p>
        </td>
        <td><p>Identify empty droplets using <a href="https://rdrr.io/pkg/DropletUtils/man/barcodeRanks.html" class="external-link">barcodeRanks</a>.</p></td>
      </tr><tr><td>
          <p><code><a href="runEmptyDrops.html">runEmptyDrops()</a></code> </p>
        </td>
        <td><p>Identify empty droplets using <a href="https://rdrr.io/pkg/DropletUtils/man/emptyDrops.html" class="external-link">emptyDrops</a>.</p></td>
      </tr><tr><td>
          <p><code><a href="runBcds.html">runBcds()</a></code> </p>
        </td>
        <td><p>Find doublets/multiplets using <a href="https://rdrr.io/pkg/scds/man/bcds.html" class="external-link">bcds</a>.</p></td>
      </tr><tr><td>
          <p><code><a href="runCxds.html">runCxds()</a></code> </p>
        </td>
        <td><p>Find doublets/multiplets using <a href="https://rdrr.io/pkg/scds/man/cxds.html" class="external-link">cxds</a>.</p></td>
      </tr><tr><td>
          <p><code><a href="runCxdsBcdsHybrid.html">runCxdsBcdsHybrid()</a></code> </p>
        </td>
        <td><p>Find doublets/multiplets using <a href="https://rdrr.io/pkg/scds/man/cxds_bcds_hybrid.html" class="external-link">cxds_bcds_hybrid</a>.</p></td>
      </tr><tr><td>
          <p><code><a href="runScDblFinder.html">runScDblFinder()</a></code> </p>
        </td>
        <td><p>Detect doublet cells using <a href="https://plger.github.io/scDblFinder/reference/scDblFinder.html" class="external-link">scDblFinder</a>.</p></td>
      </tr><tr><td>
          <p><code><a href="runDoubletFinder.html">runDoubletFinder()</a></code> </p>
        </td>
        <td><p>Generates a doublet score for each cell via doubletFinder</p></td>
      </tr><tr><td>
          <p><code><a href="runScrublet.html">runScrublet()</a></code> </p>
        </td>
        <td><p>Find doublets using <code>scrublet</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="plotBarcodeRankDropsResults.html">plotBarcodeRankDropsResults()</a></code> </p>
        </td>
        <td><p>Plots for runBarcodeRankDrops outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotEmptyDropsResults.html">plotEmptyDropsResults()</a></code> </p>
        </td>
        <td><p>Plots for runEmptyDrops outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotBcdsResults.html">plotBcdsResults()</a></code> </p>
        </td>
        <td><p>Plots for runBcds outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotCxdsResults.html">plotCxdsResults()</a></code> </p>
        </td>
        <td><p>Plots for runCxds outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotScdsHybridResults.html">plotScdsHybridResults()</a></code> </p>
        </td>
        <td><p>Plots for runCxdsBcdsHybrid outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotScDblFinderResults.html">plotScDblFinderResults()</a></code> </p>
        </td>
        <td><p>Plots for runScDblFinder outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotDoubletFinderResults.html">plotDoubletFinderResults()</a></code> </p>
        </td>
        <td><p>Plots for runDoubletFinder outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotScrubletResults.html">plotScrubletResults()</a></code> </p>
        </td>
        <td><p>Plots for runScrublet outputs.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="normalization">Normalization <a href="#normalization" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runNormalization.html">runNormalization()</a></code> </p>
        </td>
        <td><p>Run normalization/transformation with various methods</p></td>
      </tr><tr><td>
          <p><code><a href="scaterlogNormCounts.html">scaterlogNormCounts()</a></code> </p>
        </td>
        <td><p>scaterlogNormCounts
Uses logNormCounts to log normalize input data</p></td>
      </tr><tr><td>
          <p><code><a href="scaterCPM.html">scaterCPM()</a></code> </p>
        </td>
        <td><p>scaterCPM
Uses CPM from scater library to compute counts-per-million.</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratNormalizeData.html">runSeuratNormalizeData()</a></code> </p>
        </td>
        <td><p>runSeuratNormalizeData
Wrapper for NormalizeData() function from seurat library
Normalizes the sce object according to the input parameters</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratScaleData.html">runSeuratScaleData()</a></code> </p>
        </td>
        <td><p>runSeuratScaleData
Scales the input sce object according to the input parameters</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratSCTransform.html">runSeuratSCTransform()</a></code> </p>
        </td>
        <td><p>runSeuratSCTransform
Runs the <a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a> function to transform/normalize the input
data</p></td>
      </tr><tr><td>
          <p><code><a href="computeZScore.html">computeZScore()</a></code> </p>
        </td>
        <td><p>Compute Z-Score</p></td>
      </tr><tr><td>
          <p><code><a href="trimCounts.html">trimCounts()</a></code> </p>
        </td>
        <td><p>Trim Counts</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="batch-effect-correction">Batch Effect Correction <a href="#batch-effect-correction" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runMNNCorrect.html">runMNNCorrect()</a></code> </p>
        </td>
        <td><p>Apply the mutual nearest neighbors (MNN) batch effect correction method to
SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="runComBatSeq.html">runComBatSeq()</a></code> </p>
        </td>
        <td><p>Apply ComBat-Seq batch effect correction method to SingleCellExperiment
object</p></td>
      </tr><tr><td>
          <p><code><a href="runBBKNN.html">runBBKNN()</a></code> </p>
        </td>
        <td><p>Apply BBKNN batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="runFastMNN.html">runFastMNN()</a></code> </p>
        </td>
        <td><p>Apply a fast version of the mutual nearest neighbors (MNN) batch effect
correction method to SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="runLimmaBC.html">runLimmaBC()</a></code> </p>
        </td>
        <td><p>Apply Limma's batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="runHarmony.html">runHarmony()</a></code> </p>
        </td>
        <td><p>Apply Harmony batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="runSCANORAMA.html">runSCANORAMA()</a></code> </p>
        </td>
        <td><p>Apply the mutual nearest neighbors (MNN) batch effect correction method to
SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="runSCMerge.html">runSCMerge()</a></code> </p>
        </td>
        <td><p>Apply scMerge batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratIntegration.html">runSeuratIntegration()</a></code> </p>
        </td>
        <td><p>runSeuratIntegration
A wrapper function to Seurat Batch-Correction/Integration workflow.</p></td>
      </tr><tr><td>
          <p><code><a href="runZINBWaVE.html">runZINBWaVE()</a></code> </p>
        </td>
        <td><p>Apply ZINBWaVE Batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="plotBatchVariance.html">plotBatchVariance()</a></code> </p>
        </td>
        <td><p>Plot the percent of the variation that is explained by batch and condition
in the data</p></td>
      </tr><tr><td>
          <p><code><a href="plotBatchCorrCompare.html">plotBatchCorrCompare()</a></code> </p>
        </td>
        <td><p>Plot comparison of batch corrected result against original assay</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEBatchFeatureMean.html">plotSCEBatchFeatureMean()</a></code> </p>
        </td>
        <td><p>Plot mean feature value in each batch of a SingleCellExperiment object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="feature-selection">Feature Selection <a href="#feature-selection" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runFeatureSelection.html">runFeatureSelection()</a></code> </p>
        </td>
        <td><p>Run Variable Feature Detection Methods</p></td>
      </tr><tr><td>
          <p><code><a href="runModelGeneVar.html">runModelGeneVar()</a></code> </p>
        </td>
        <td><p>Calculate Variable Genes with Scran modelGeneVar</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratFindHVG.html">runSeuratFindHVG()</a></code> </p>
        </td>
        <td><p>runSeuratFindHVG
Find highly variable genes and store in the input sce object</p></td>
      </tr><tr><td>
          <p><code><a href="getTopHVG.html">getTopHVG()</a></code> <code><a href="getTopHVG.html">setTopHVG()</a></code> </p>
        </td>
        <td><p>Get or set top HVG after calculation</p></td>
      </tr><tr><td>
          <p><code><a href="plotTopHVG.html">plotTopHVG()</a></code> </p>
        </td>
        <td><p>Plot highly variable genes</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="dimensionality-reduction-embedding">Dimensionality Reduction &amp; Embedding <a href="#dimensionality-reduction-embedding" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runDimReduce.html">runDimReduce()</a></code> </p>
        </td>
        <td><p>Generic Wrapper function for running dimensionality reduction</p></td>
      </tr><tr><td>
          <p><code><a href="scaterPCA.html">scaterPCA()</a></code> </p>
        </td>
        <td><p>Perform scater PCA on a SingleCellExperiment Object</p></td>
      </tr><tr><td>
          <p><code><a href="runUMAP.html">runUMAP()</a></code> <code><a href="runUMAP.html">runQuickUMAP()</a></code> <code><a href="runUMAP.html">getUMAP()</a></code> </p>
        </td>
        <td><p>Run UMAP embedding with scater method</p></td>
      </tr><tr><td>
          <p><code><a href="runTSNE.html">runTSNE()</a></code> <code><a href="runTSNE.html">runQuickTSNE()</a></code> <code><a href="runTSNE.html">getTSNE()</a></code> </p>
        </td>
        <td><p>Run t-SNE embedding with Rtsne method</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratICA.html">runSeuratICA()</a></code> </p>
        </td>
        <td><p>runSeuratICA
Computes ICA on the input sce object and stores the calculated independent
components within the sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratPCA.html">runSeuratPCA()</a></code> </p>
        </td>
        <td><p>runSeuratPCA
Computes PCA on the input sce object and stores the calculated principal
components within the sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratUMAP.html">runSeuratUMAP()</a></code> </p>
        </td>
        <td><p>runSeuratUMAP
Computes UMAP from the given sce object and stores the UMAP computations back
into the sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratTSNE.html">runSeuratTSNE()</a></code> </p>
        </td>
        <td><p>runSeuratTSNE
Computes tSNE from the given sce object and stores the tSNE computations back
into the sce object</p></td>
      </tr><tr><td>
          <p><code><a href="plotPCA.html">plotPCA()</a></code> </p>
        </td>
        <td><p>Plot PCA run data from its components.</p></td>
      </tr><tr><td>
          <p><code><a href="plotUMAP.html">plotUMAP()</a></code> </p>
        </td>
        <td><p>Plot UMAP results either on already run results or run first and then plot.</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSNE.html">plotTSNE()</a></code> </p>
        </td>
        <td><p>Plot t-SNE plot on dimensionality reduction data run from t-SNE method.</p></td>
      </tr><tr><td>
          <p><code><a href="plotDimRed.html">plotDimRed()</a></code> </p>
        </td>
        <td><p>Plot dimensionality reduction from computed metrics including PCA, ICA, tSNE
and UMAP</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEDimReduceColData.html">plotSCEDimReduceColData()</a></code> </p>
        </td>
        <td><p>Dimension reduction plot tool for colData</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEDimReduceFeatures.html">plotSCEDimReduceFeatures()</a></code> </p>
        </td>
        <td><p>Dimension reduction plot tool for assay data</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="clustering">Clustering <a href="#clustering" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runScranSNN.html">runScranSNN()</a></code> </p>
        </td>
        <td><p>Get clustering with SNN graph</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratFindClusters.html">runSeuratFindClusters()</a></code> </p>
        </td>
        <td><p>runSeuratFindClusters
Computes the clusters from the input sce object and stores them back in sce
object</p></td>
      </tr><tr><td>
          <p><code><a href="runKMeans.html">runKMeans()</a></code> </p>
        </td>
        <td><p>Get clustering with KMeans</p></td>
      </tr><tr><td>
          <p><code><a href="runClusterSummaryMetrics.html">runClusterSummaryMetrics()</a></code> </p>
        </td>
        <td><p>Run Cluster Summary Metrics</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="differential-expression">Differential Expression <a href="#differential-expression" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runDEAnalysis.html">runDEAnalysis()</a></code> <code><a href="runDEAnalysis.html">runDESeq2()</a></code> <code><a href="runDEAnalysis.html">runLimmaDE()</a></code> <code><a href="runDEAnalysis.html">runANOVA()</a></code> <code><a href="runDEAnalysis.html">runMAST()</a></code> <code><a href="runDEAnalysis.html">runWilcox()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis on SCE object</p></td>
      </tr><tr><td>
          <p><code><a href="getDEGTopTable.html">getDEGTopTable()</a></code> </p>
        </td>
        <td><p>Get Top Table of a DEG analysis</p></td>
      </tr><tr><td>
          <p><code><a href="plotDEGVolcano.html">plotDEGVolcano()</a></code> </p>
        </td>
        <td><p>Generate volcano plot for DEGs</p></td>
      </tr><tr><td>
          <p><code><a href="plotDEGViolin.html">plotDEGViolin()</a></code> </p>
        </td>
        <td><p>Generate violin plot to show the expression of top DEGs</p></td>
      </tr><tr><td>
          <p><code><a href="plotDEGRegression.html">plotDEGRegression()</a></code> </p>
        </td>
        <td><p>Create linear regression plot to show the expression the of top DEGs</p></td>
      </tr><tr><td>
          <p><code><a href="plotDEGHeatmap.html">plotDEGHeatmap()</a></code> </p>
        </td>
        <td><p>Heatmap visualization of DEG result</p></td>
      </tr><tr><td>
          <p><code><a href="plotMASTThresholdGenes.html">plotMASTThresholdGenes()</a></code> </p>
        </td>
        <td><p>MAST Identify adaptive thresholds</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="find-marker">Find Marker <a href="#find-marker" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runFindMarker.html">runFindMarker()</a></code> <code><a href="runFindMarker.html">findMarkerDiffExp()</a></code> </p>
        </td>
        <td><p>Find the marker gene set for each cluster</p></td>
      </tr><tr><td>
          <p><code><a href="getFindMarkerTopTable.html">getFindMarkerTopTable()</a></code> <code><a href="getFindMarkerTopTable.html">findMarkerTopTable()</a></code> </p>
        </td>
        <td><p>Fetch the table of top markers that pass the filtering</p></td>
      </tr><tr><td>
          <p><code><a href="plotFindMarkerHeatmap.html">plotFindMarkerHeatmap()</a></code> <code><a href="plotFindMarkerHeatmap.html">plotMarkerDiffExp()</a></code> </p>
        </td>
        <td><p>Plot a heatmap to visualize the result of <code><a href="https://rdrr.io/pkg/singleCellTK/man/runFindMarker.html" class="external-link">runFindMarker</a></code></p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="differential-abundance">Differential Abundance <a href="#differential-abundance" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="diffAbundanceFET.html">diffAbundanceFET()</a></code> </p>
        </td>
        <td><p>Calculate Differential Abundance with FET</p></td>
      </tr><tr><td>
          <p><code><a href="getDiffAbundanceResults.html">getDiffAbundanceResults()</a></code> <code><a href="getDiffAbundanceResults.html">`getDiffAbundanceResults&lt;-`()</a></code> </p>
        </td>
        <td><p>Get/Set diffAbundanceFET result table</p></td>
      </tr><tr><td>
          <p><code><a href="plotClusterAbundance.html">plotClusterAbundance()</a></code> </p>
        </td>
        <td><p>Plot the differential Abundance</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="cell-type-labeling">Cell Type Labeling <a href="#cell-type-labeling" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runSingleR.html">runSingleR()</a></code> </p>
        </td>
        <td><p>Label cell types with SingleR</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="enrichment-pathway-analysis">Enrichment &amp; Pathway Analysis <a href="#enrichment-pathway-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getMSigDBTable.html">getMSigDBTable()</a></code> </p>
        </td>
        <td><p>Shows MSigDB categories</p></td>
      </tr><tr><td>
          <p><code><a href="runEnrichR.html">runEnrichR()</a></code> </p>
        </td>
        <td><p>Run EnrichR on SCE object</p></td>
      </tr><tr><td>
          <p><code><a href="getEnrichRResult.html">`getEnrichRResult&lt;-`()</a></code> <code><a href="getEnrichRResult.html">getEnrichRResult()</a></code> </p>
        </td>
        <td><p>Get or Set EnrichR Result</p></td>
      </tr><tr><td>
          <p><code><a href="runGSVA.html">runGSVA()</a></code> </p>
        </td>
        <td><p>Run GSVA analysis on a <a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a> object</p></td>
      </tr><tr><td>
          <p><code><a href="runVAM.html">runVAM()</a></code> </p>
        </td>
        <td><p>Run VAM to score gene sets in single cell data</p></td>
      </tr><tr><td>
          <p><code><a href="getPathwayResultNames.html">getPathwayResultNames()</a></code> </p>
        </td>
        <td><p>List pathway analysis result names</p></td>
      </tr><tr><td>
          <p><code><a href="plotPathway.html">plotPathway()</a></code> </p>
        </td>
        <td><p>Generate violin plots for pathway analysis results</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="trajectory-analysis">Trajectory Analysis <a href="#trajectory-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runTSCAN.html">runTSCAN()</a></code> </p>
        </td>
        <td><p>Run TSCAN to obtain pseudotime values for cells</p></td>
      </tr><tr><td>
          <p><code><a href="runTSCANClusterDEAnalysis.html">runTSCANClusterDEAnalysis()</a></code> </p>
        </td>
        <td><p>Find DE genes between all TSCAN paths rooted from given cluster</p></td>
      </tr><tr><td>
          <p><code><a href="runTSCANDEG.html">runTSCANDEG()</a></code> </p>
        </td>
        <td><p>Test gene expression changes along a TSCAN trajectory path</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANClusterDEG.html">plotTSCANClusterDEG()</a></code> </p>
        </td>
        <td><p>Plot features identified by <code><a href="https://rdrr.io/pkg/singleCellTK/man/runTSCANClusterDEAnalysis.html" class="external-link">runTSCANClusterDEAnalysis</a></code> on
cell 2D embedding with MST overlaid</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANClusterPseudo.html">plotTSCANClusterPseudo()</a></code> </p>
        </td>
        <td><p>Plot TSCAN pseudotime rooted from given cluster</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANDimReduceFeatures.html">plotTSCANDimReduceFeatures()</a></code> </p>
        </td>
        <td><p>Plot feature expression on cell 2D embedding with MST overlaid</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANPseudotimeGenes.html">plotTSCANPseudotimeGenes()</a></code> </p>
        </td>
        <td><p>Plot expression changes of top features along a TSCAN pseudotime path</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANPseudotimeHeatmap.html">plotTSCANPseudotimeHeatmap()</a></code> </p>
        </td>
        <td><p>Plot heatmap of genes with expression change along TSCAN pseudotime</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANResults.html">plotTSCANResults()</a></code> </p>
        </td>
        <td><p>Plot MST pseudotime values on cell 2D embedding</p></td>
      </tr><tr><td>
          <p><code><a href="getTSCANResults.html">getTSCANResults()</a></code> <code><a href="getTSCANResults.html">`getTSCANResults&lt;-`()</a></code> <code><a href="getTSCANResults.html">listTSCANResults()</a></code> <code><a href="getTSCANResults.html">listTSCANTerminalNodes()</a></code> </p>
        </td>
        <td><p>getTSCANResults accessor function</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="seurat-curated-workflow">Seurat Curated Workflow <a href="#seurat-curated-workflow" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runSeuratFindClusters.html">runSeuratFindClusters()</a></code> </p>
        </td>
        <td><p>runSeuratFindClusters
Computes the clusters from the input sce object and stores them back in sce
object</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratFindHVG.html">runSeuratFindHVG()</a></code> </p>
        </td>
        <td><p>runSeuratFindHVG
Find highly variable genes and store in the input sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratFindMarkers.html">runSeuratFindMarkers()</a></code> </p>
        </td>
        <td><p>runSeuratFindMarkers</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratHeatmap.html">runSeuratHeatmap()</a></code> </p>
        </td>
        <td><p>runSeuratHeatmap
Computes the heatmap plot object from the pca slot in the input sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratICA.html">runSeuratICA()</a></code> </p>
        </td>
        <td><p>runSeuratICA
Computes ICA on the input sce object and stores the calculated independent
components within the sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratIntegration.html">runSeuratIntegration()</a></code> </p>
        </td>
        <td><p>runSeuratIntegration
A wrapper function to Seurat Batch-Correction/Integration workflow.</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratJackStraw.html">runSeuratJackStraw()</a></code> </p>
        </td>
        <td><p>runSeuratJackStraw
Compute jackstraw plot and store the computations in the input sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratNormalizeData.html">runSeuratNormalizeData()</a></code> </p>
        </td>
        <td><p>runSeuratNormalizeData
Wrapper for NormalizeData() function from seurat library
Normalizes the sce object according to the input parameters</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratPCA.html">runSeuratPCA()</a></code> </p>
        </td>
        <td><p>runSeuratPCA
Computes PCA on the input sce object and stores the calculated principal
components within the sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratSCTransform.html">runSeuratSCTransform()</a></code> </p>
        </td>
        <td><p>runSeuratSCTransform
Runs the <a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a> function to transform/normalize the input
data</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratScaleData.html">runSeuratScaleData()</a></code> </p>
        </td>
        <td><p>runSeuratScaleData
Scales the input sce object according to the input parameters</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratTSNE.html">runSeuratTSNE()</a></code> </p>
        </td>
        <td><p>runSeuratTSNE
Computes tSNE from the given sce object and stores the tSNE computations back
into the sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runSeuratUMAP.html">runSeuratUMAP()</a></code> </p>
        </td>
        <td><p>runSeuratUMAP
Computes UMAP from the given sce object and stores the UMAP computations back
into the sce object</p></td>
      </tr><tr><td>
          <p><code><a href="computeHeatmap.html">computeHeatmap()</a></code> </p>
        </td>
        <td><p>Computes heatmap for a set of features against dimensionality reduction 
components</p></td>
      </tr><tr><td>
          <p><code><a href="getSeuratVariableFeatures.html">getSeuratVariableFeatures()</a></code> </p>
        </td>
        <td><p>Get variable feature names after running runSeuratFindHVG function</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="scanpy-curated-workflow">Scanpy Curated Workflow <a href="#scanpy-curated-workflow" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="runScanpyFindClusters.html">runScanpyFindClusters()</a></code> </p>
        </td>
        <td><p>runScanpyFindClusters
Computes the clusters from the input sce object and stores them back in sce
object</p></td>
      </tr><tr><td>
          <p><code><a href="runScanpyFindHVG.html">runScanpyFindHVG()</a></code> </p>
        </td>
        <td><p>runScanpyFindHVG
Find highly variable genes and store in the input sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runScanpyFindMarkers.html">runScanpyFindMarkers()</a></code> </p>
        </td>
        <td><p>runScanpyFindMarkers</p></td>
      </tr><tr><td>
          <p><code><a href="runScanpyNormalizeData.html">runScanpyNormalizeData()</a></code> </p>
        </td>
        <td><p>runScanpyNormalizeData
Wrapper for NormalizeData() function from scanpy library
Normalizes the sce object according to the input parameters</p></td>
      </tr><tr><td>
          <p><code><a href="runScanpyPCA.html">runScanpyPCA()</a></code> </p>
        </td>
        <td><p>runScanpyPCA
Computes PCA on the input sce object and stores the calculated principal
components within the sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runScanpyScaleData.html">runScanpyScaleData()</a></code> </p>
        </td>
        <td><p>runScanpyScaleData
Scales the input sce object according to the input parameters</p></td>
      </tr><tr><td>
          <p><code><a href="runScanpyTSNE.html">runScanpyTSNE()</a></code> </p>
        </td>
        <td><p>runScanpyTSNE
Computes tSNE from the given sce object and stores the tSNE computations back
into the sce object</p></td>
      </tr><tr><td>
          <p><code><a href="runScanpyUMAP.html">runScanpyUMAP()</a></code> </p>
        </td>
        <td><p>runScanpyUMAP
Computes UMAP from the given sce object and stores the UMAP computations back
into the sce object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualization">Visualization <a href="#visualization" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotBarcodeRankDropsResults.html">plotBarcodeRankDropsResults()</a></code> </p>
        </td>
        <td><p>Plots for runBarcodeRankDrops outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotBarcodeRankScatter.html">plotBarcodeRankScatter()</a></code> </p>
        </td>
        <td><p>Plots for runBarcodeRankDrops outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotBatchCorrCompare.html">plotBatchCorrCompare()</a></code> </p>
        </td>
        <td><p>Plot comparison of batch corrected result against original assay</p></td>
      </tr><tr><td>
          <p><code><a href="plotBatchVariance.html">plotBatchVariance()</a></code> </p>
        </td>
        <td><p>Plot the percent of the variation that is explained by batch and condition
in the data</p></td>
      </tr><tr><td>
          <p><code><a href="plotBcdsResults.html">plotBcdsResults()</a></code> </p>
        </td>
        <td><p>Plots for runBcds outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotBubble.html">plotBubble()</a></code> </p>
        </td>
        <td><p>Plot Bubble plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotClusterAbundance.html">plotClusterAbundance()</a></code> </p>
        </td>
        <td><p>Plot the differential Abundance</p></td>
      </tr><tr><td>
          <p><code><a href="plotCxdsResults.html">plotCxdsResults()</a></code> </p>
        </td>
        <td><p>Plots for runCxds outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotDEGHeatmap.html">plotDEGHeatmap()</a></code> </p>
        </td>
        <td><p>Heatmap visualization of DEG result</p></td>
      </tr><tr><td>
          <p><code><a href="plotDEGRegression.html">plotDEGRegression()</a></code> </p>
        </td>
        <td><p>Create linear regression plot to show the expression the of top DEGs</p></td>
      </tr><tr><td>
          <p><code><a href="plotDEGViolin.html">plotDEGViolin()</a></code> </p>
        </td>
        <td><p>Generate violin plot to show the expression of top DEGs</p></td>
      </tr><tr><td>
          <p><code><a href="plotDEGVolcano.html">plotDEGVolcano()</a></code> </p>
        </td>
        <td><p>Generate volcano plot for DEGs</p></td>
      </tr><tr><td>
          <p><code><a href="plotDecontXResults.html">plotDecontXResults()</a></code> </p>
        </td>
        <td><p>Plots for runDecontX outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotDimRed.html">plotDimRed()</a></code> </p>
        </td>
        <td><p>Plot dimensionality reduction from computed metrics including PCA, ICA, tSNE
and UMAP</p></td>
      </tr><tr><td>
          <p><code><a href="plotDoubletFinderResults.html">plotDoubletFinderResults()</a></code> </p>
        </td>
        <td><p>Plots for runDoubletFinder outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotEmptyDropsResults.html">plotEmptyDropsResults()</a></code> </p>
        </td>
        <td><p>Plots for runEmptyDrops outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotEmptyDropsScatter.html">plotEmptyDropsScatter()</a></code> </p>
        </td>
        <td><p>Plots for runEmptyDrops outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotFindMarkerHeatmap.html">plotFindMarkerHeatmap()</a></code> <code><a href="plotFindMarkerHeatmap.html">plotMarkerDiffExp()</a></code> </p>
        </td>
        <td><p>Plot a heatmap to visualize the result of <code><a href="https://rdrr.io/pkg/singleCellTK/man/runFindMarker.html" class="external-link">runFindMarker</a></code></p></td>
      </tr><tr><td>
          <p><code><a href="plotMASTThresholdGenes.html">plotMASTThresholdGenes()</a></code> </p>
        </td>
        <td><p>MAST Identify adaptive thresholds</p></td>
      </tr><tr><td>
          <p><code><a href="plotPCA.html">plotPCA()</a></code> </p>
        </td>
        <td><p>Plot PCA run data from its components.</p></td>
      </tr><tr><td>
          <p><code><a href="plotPathway.html">plotPathway()</a></code> </p>
        </td>
        <td><p>Generate violin plots for pathway analysis results</p></td>
      </tr><tr><td>
          <p><code><a href="plotRunPerCellQCResults.html">plotRunPerCellQCResults()</a></code> </p>
        </td>
        <td><p>Plots for runPerCellQC outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEBarAssayData.html">plotSCEBarAssayData()</a></code> </p>
        </td>
        <td><p>Bar plot of assay data.</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEBarColData.html">plotSCEBarColData()</a></code> </p>
        </td>
        <td><p>Bar plot of colData.</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEBatchFeatureMean.html">plotSCEBatchFeatureMean()</a></code> </p>
        </td>
        <td><p>Plot mean feature value in each batch of a SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEDensity.html">plotSCEDensity()</a></code> </p>
        </td>
        <td><p>Density plot of any data stored in the SingleCellExperiment object.</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEDensityAssayData.html">plotSCEDensityAssayData()</a></code> </p>
        </td>
        <td><p>Density plot of assay data.</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEDensityColData.html">plotSCEDensityColData()</a></code> </p>
        </td>
        <td><p>Density plot of colData.</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEDimReduceColData.html">plotSCEDimReduceColData()</a></code> </p>
        </td>
        <td><p>Dimension reduction plot tool for colData</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEDimReduceFeatures.html">plotSCEDimReduceFeatures()</a></code> </p>
        </td>
        <td><p>Dimension reduction plot tool for assay data</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEHeatmap.html">plotSCEHeatmap()</a></code> </p>
        </td>
        <td><p>Plot heatmap of using data stored in SingleCellExperiment Object</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEScatter.html">plotSCEScatter()</a></code> </p>
        </td>
        <td><p>Dimension reduction plot tool for all types of data</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEViolin.html">plotSCEViolin()</a></code> </p>
        </td>
        <td><p>Violin plot of any data stored in the SingleCellExperiment object.</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEViolinAssayData.html">plotSCEViolinAssayData()</a></code> </p>
        </td>
        <td><p>Violin plot of assay data.</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCEViolinColData.html">plotSCEViolinColData()</a></code> </p>
        </td>
        <td><p>Violin plot of colData.</p></td>
      </tr><tr><td>
          <p><code><a href="plotScDblFinderResults.html">plotScDblFinderResults()</a></code> </p>
        </td>
        <td><p>Plots for runScDblFinder outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyDotPlot.html">plotScanpyDotPlot()</a></code> </p>
        </td>
        <td><p>plotScanpyDotPlot</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyEmbedding.html">plotScanpyEmbedding()</a></code> </p>
        </td>
        <td><p>plotScanpyEmbedding</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyHVG.html">plotScanpyHVG()</a></code> </p>
        </td>
        <td><p>plotScanpyHVG</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyHeatmap.html">plotScanpyHeatmap()</a></code> </p>
        </td>
        <td><p>plotScanpyHeatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyMarkerGenes.html">plotScanpyMarkerGenes()</a></code> </p>
        </td>
        <td><p>plotScanpyMarkerGenes</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyMarkerGenesDotPlot.html">plotScanpyMarkerGenesDotPlot()</a></code> </p>
        </td>
        <td><p>plotScanpyMarkerGenesDotPlot</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyMarkerGenesHeatmap.html">plotScanpyMarkerGenesHeatmap()</a></code> </p>
        </td>
        <td><p>plotScanpyMarkerGenesHeatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyMarkerGenesMatrixPlot.html">plotScanpyMarkerGenesMatrixPlot()</a></code> </p>
        </td>
        <td><p>plotScanpyMarkerGenesMatrixPlot</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyMarkerGenesViolin.html">plotScanpyMarkerGenesViolin()</a></code> </p>
        </td>
        <td><p>plotScanpyMarkerGenesViolin</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyMatrixPlot.html">plotScanpyMatrixPlot()</a></code> </p>
        </td>
        <td><p>plotScanpyMatrixPlot</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyPCA.html">plotScanpyPCA()</a></code> </p>
        </td>
        <td><p>plotScanpyPCA</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyPCAGeneRanking.html">plotScanpyPCAGeneRanking()</a></code> </p>
        </td>
        <td><p>plotScanpyPCAGeneRanking</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyPCAVariance.html">plotScanpyPCAVariance()</a></code> </p>
        </td>
        <td><p>plotScanpyPCAVariance</p></td>
      </tr><tr><td>
          <p><code><a href="plotScanpyViolin.html">plotScanpyViolin()</a></code> </p>
        </td>
        <td><p>plotScanpyViolin</p></td>
      </tr><tr><td>
          <p><code><a href="plotScdsHybridResults.html">plotScdsHybridResults()</a></code> </p>
        </td>
        <td><p>Plots for runCxdsBcdsHybrid outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotScrubletResults.html">plotScrubletResults()</a></code> </p>
        </td>
        <td><p>Plots for runScrublet outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeuratElbow.html">plotSeuratElbow()</a></code> </p>
        </td>
        <td><p>plotSeuratElbow
Computes the plot object for elbow plot from the pca slot in the input sce
object</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeuratGenes.html">plotSeuratGenes()</a></code> </p>
        </td>
        <td><p>Compute and plot visualizations for marker genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeuratHVG.html">plotSeuratHVG()</a></code> </p>
        </td>
        <td><p>plotSeuratHVG
Plot highly variable genes from input sce object (must have highly variable
genes computations stored)</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeuratHeatmap.html">plotSeuratHeatmap()</a></code> </p>
        </td>
        <td><p>plotSeuratHeatmap
Modifies the heatmap plot object so it contains specified number of heatmaps
in a single plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeuratJackStraw.html">plotSeuratJackStraw()</a></code> </p>
        </td>
        <td><p>plotSeuratJackStraw
Computes the plot object for jackstraw plot from the pca slot in the input
sce object</p></td>
      </tr><tr><td>
          <p><code><a href="plotSeuratReduction.html">plotSeuratReduction()</a></code> </p>
        </td>
        <td><p>plotSeuratReduction
Plots the selected dimensionality reduction method</p></td>
      </tr><tr><td>
          <p><code><a href="plotSoupXResults.html">plotSoupXResults()</a></code> </p>
        </td>
        <td><p>Plot SoupX Result</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANClusterDEG.html">plotTSCANClusterDEG()</a></code> </p>
        </td>
        <td><p>Plot features identified by <code><a href="https://rdrr.io/pkg/singleCellTK/man/runTSCANClusterDEAnalysis.html" class="external-link">runTSCANClusterDEAnalysis</a></code> on
cell 2D embedding with MST overlaid</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANClusterPseudo.html">plotTSCANClusterPseudo()</a></code> </p>
        </td>
        <td><p>Plot TSCAN pseudotime rooted from given cluster</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANDimReduceFeatures.html">plotTSCANDimReduceFeatures()</a></code> </p>
        </td>
        <td><p>Plot feature expression on cell 2D embedding with MST overlaid</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANPseudotimeGenes.html">plotTSCANPseudotimeGenes()</a></code> </p>
        </td>
        <td><p>Plot expression changes of top features along a TSCAN pseudotime path</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANPseudotimeHeatmap.html">plotTSCANPseudotimeHeatmap()</a></code> </p>
        </td>
        <td><p>Plot heatmap of genes with expression change along TSCAN pseudotime</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSCANResults.html">plotTSCANResults()</a></code> </p>
        </td>
        <td><p>Plot MST pseudotime values on cell 2D embedding</p></td>
      </tr><tr><td>
          <p><code><a href="plotTSNE.html">plotTSNE()</a></code> </p>
        </td>
        <td><p>Plot t-SNE plot on dimensionality reduction data run from t-SNE method.</p></td>
      </tr><tr><td>
          <p><code><a href="plotTopHVG.html">plotTopHVG()</a></code> </p>
        </td>
        <td><p>Plot highly variable genes</p></td>
      </tr><tr><td>
          <p><code><a href="plotUMAP.html">plotUMAP()</a></code> </p>
        </td>
        <td><p>Plot UMAP results either on already run results or run first and then plot.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="report-generation">Report Generation <a href="#report-generation" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="reportCellQC.html">reportCellQC()</a></code> </p>
        </td>
        <td><p>Get runCellQC .html report</p></td>
      </tr><tr><td>
          <p><code><a href="reportClusterAbundance.html">reportClusterAbundance()</a></code> </p>
        </td>
        <td><p>Get plotClusterAbundance .html report</p></td>
      </tr><tr><td>
          <p><code><a href="reportDiffAbundanceFET.html">reportDiffAbundanceFET()</a></code> </p>
        </td>
        <td><p>Get diffAbundanceFET .html report</p></td>
      </tr><tr><td>
          <p><code><a href="reportDiffExp.html">reportDiffExp()</a></code> </p>
        </td>
        <td><p>Get runDEAnalysis .html report</p></td>
      </tr><tr><td>
          <p><code><a href="reportDropletQC.html">reportDropletQC()</a></code> </p>
        </td>
        <td><p>Get runDropletQC .html report</p></td>
      </tr><tr><td>
          <p><code><a href="reportFindMarker.html">reportFindMarker()</a></code> </p>
        </td>
        <td><p>Get runFindMarker .html report</p></td>
      </tr><tr><td>
          <p><code><a href="reportQCTool.html">reportQCTool()</a></code> </p>
        </td>
        <td><p>Get .html report of the output of the selected QC algorithm</p></td>
      </tr><tr><td>
          <p><code><a href="reportSeurat.html">reportSeurat()</a></code> </p>
        </td>
        <td><p>Generates an HTML report for the complete Seurat workflow and returns the
 SCE object with the results computed and stored inside the object.</p></td>
      </tr><tr><td>
          <p><code><a href="reportSeuratClustering.html">reportSeuratClustering()</a></code> </p>
        </td>
        <td><p>Generates an HTML report for Seurat Clustering and returns the SCE object
 with the results computed and stored inside the object.</p></td>
      </tr><tr><td>
          <p><code><a href="reportSeuratDimRed.html">reportSeuratDimRed()</a></code> </p>
        </td>
        <td><p>Generates an HTML report for Seurat Dimensionality Reduction
 and returns the SCE object with the results computed and stored
 inside the object.</p></td>
      </tr><tr><td>
          <p><code><a href="reportSeuratFeatureSelection.html">reportSeuratFeatureSelection()</a></code> </p>
        </td>
        <td><p>Generates an HTML report for Seurat Feature Selection and returns the
SCE object with the results computed and stored inside the object.</p></td>
      </tr><tr><td>
          <p><code><a href="reportSeuratMarkerSelection.html">reportSeuratMarkerSelection()</a></code> </p>
        </td>
        <td><p>Generates an HTML report for Seurat Results (including Clustering &amp; Marker
 Selection) and returns the SCE object with the results computed and stored
 inside the object.</p></td>
      </tr><tr><td>
          <p><code><a href="reportSeuratNormalization.html">reportSeuratNormalization()</a></code> </p>
        </td>
        <td><p>Generates an HTML report for Seurat Normalization
 and returns the SCE object with the results computed and stored
 inside the object.</p></td>
      </tr><tr><td>
          <p><code><a href="reportSeuratResults.html">reportSeuratResults()</a></code> </p>
        </td>
        <td><p>Generates an HTML report for Seurat Results (including Clustering &amp; Marker
 Selection) and returns the SCE object with the results computed and stored
 inside the object.</p></td>
      </tr><tr><td>
          <p><code><a href="reportSeuratRun.html">reportSeuratRun()</a></code> </p>
        </td>
        <td><p>Generates an HTML report for Seurat Run (including Normalization,
 Feature Selection, Dimensionality Reduction &amp; Clustering) and returns the
 SCE object with the results computed and stored inside the object.</p></td>
      </tr><tr><td>
          <p><code><a href="reportSeuratScaling.html">reportSeuratScaling()</a></code> </p>
        </td>
        <td><p>Generates an HTML report for Seurat Scaling
 and returns the SCE object with the results computed and stored
 inside the object.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="exporting-results">Exporting Results <a href="#exporting-results" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="exportSCE.html">exportSCE()</a></code> </p>
        </td>
        <td><p>Export data in SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="exportSCEToSeurat.html">exportSCEToSeurat()</a></code> </p>
        </td>
        <td><p>Export data in Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="exportSCEtoAnnData.html">exportSCEtoAnnData()</a></code> </p>
        </td>
        <td><p>Export a <a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a> R object as
Python annData object</p></td>
      </tr><tr><td>
          <p><code><a href="exportSCEtoFlatFile.html">exportSCEtoFlatFile()</a></code> </p>
        </td>
        <td><p>Export a <a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a> object to flat text files</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="datasets">Datasets <a href="#datasets" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="mouseBrainSubsetSCE.html">mouseBrainSubsetSCE</a></code> </p>
        </td>
        <td><p>Example Single Cell RNA-Seq data in SingleCellExperiment Object, GSE60361
subset</p></td>
      </tr><tr><td>
          <p><code><a href="sceBatches.html">sceBatches</a></code> </p>
        </td>
        <td><p>Example Single Cell RNA-Seq data in SingleCellExperiment object, with
different batches annotated</p></td>
      </tr><tr><td>
          <p><code><a href="MitoGenes.html">MitoGenes</a></code> </p>
        </td>
        <td><p>List of mitochondrial genes of multiple reference</p></td>
      </tr><tr><td>
          <p><code><a href="msigdb_table.html">msigdb_table</a></code> </p>
        </td>
        <td><p>MSigDB gene get Category table</p></td>
      </tr><tr><td>
          <p><code><a href="sce.html">scExample</a></code> </p>
        </td>
        <td><p>Example Single Cell RNA-Seq data in SingleCellExperiment Object,
subset of 10x public dataset</p></td>
      </tr><tr><td>
          <p><code><a href="SEG.html">SEG</a></code> </p>
        </td>
        <td><p>Stably Expressed Gene (SEG) list obect, with SEG sets for human and mouse.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="other-data-processing">Other Data Processing <a href="#other-data-processing" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="expData-ANY-character-method.html">expData(<i>&lt;ANY&gt;</i>,<i>&lt;character&gt;</i>)</a></code> </p>
        </td>
        <td><p>expData
Get data item from an input <code>SingleCellExperiment</code> object. The data item can be an <code>assay</code>, <code>altExp</code> (subset) or a <code>reducedDim</code>, which is retrieved based on the name of the data item.</p></td>
      </tr><tr><td>
          <p><code><a href="expData-set-ANY-character-CharacterOrNullOrMissing-logical-method.html">`expData&lt;-`(<i>&lt;ANY&gt;</i>,<i>&lt;character&gt;</i>,<i>&lt;CharacterOrNullOrMissing&gt;</i>,<i>&lt;logical&gt;</i>)</a></code> </p>
        </td>
        <td><p>expData
Store data items using tags to identify the type of data item stored. To be used as a replacement for assay&lt;- setter function but with additional parameter to set a tag to a data item.</p></td>
      </tr><tr><td>
          <p><code><a href="expData-set.html">`expData&lt;-`()</a></code> </p>
        </td>
        <td><p>expData
Store data items using tags to identify the type of data item stored. To be used as a replacement for assay&lt;- setter function but with additional parameter to set a tag to a data item.</p></td>
      </tr><tr><td>
          <p><code><a href="expData.html">expData()</a></code> </p>
        </td>
        <td><p>expData
Get data item from an input <code>SingleCellExperiment</code> object. The data item can be an <code>assay</code>, <code>altExp</code> (subset) or a <code>reducedDim</code>, which is retrieved based on the name of the data item.</p></td>
      </tr><tr><td>
          <p><code><a href="expDataNames-ANY-method.html">expDataNames(<i>&lt;ANY&gt;</i>)</a></code> </p>
        </td>
        <td><p>expDataNames
Get names of all the data items in the input <code>SingleCellExperiment</code> object including assays, altExps and reducedDims.</p></td>
      </tr><tr><td>
          <p><code><a href="expDataNames.html">expDataNames()</a></code> </p>
        </td>
        <td><p>expDataNames
Get names of all the data items in the input <code>SingleCellExperiment</code> object including assays, altExps and reducedDims.</p></td>
      </tr><tr><td>
          <p><code><a href="expDeleteDataTag.html">expDeleteDataTag()</a></code> </p>
        </td>
        <td><p>expDeleteDataTag
Remove tag against an input data from the stored tag information in the metadata of the input object.</p></td>
      </tr><tr><td>
          <p><code><a href="expSetDataTag.html">expSetDataTag()</a></code> </p>
        </td>
        <td><p>expSetDataTag
Set tag to an assay or a data item in the input SCE object.</p></td>
      </tr><tr><td>
          <p><code><a href="expTaggedData.html">expTaggedData()</a></code> </p>
        </td>
        <td><p>expTaggedData
Returns a list of names of data items from the 
input <code>SingleCellExperiment</code> object based upon the input parameters.</p></td>
      </tr><tr><td>
          <p><code><a href="calcEffectSizes.html">calcEffectSizes()</a></code> </p>
        </td>
        <td><p>Finds the effect sizes for all genes in the original dataset, regardless of
significance.</p></td>
      </tr><tr><td>
          <p><code><a href="combineSCE.html">combineSCE()</a></code> </p>
        </td>
        <td><p>Combine a list of SingleCellExperiment objects as one SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="convertSCEToSeurat.html">convertSCEToSeurat()</a></code> </p>
        </td>
        <td><p>convertSCEToSeurat
Converts sce object to seurat while retaining all assays and metadata</p></td>
      </tr><tr><td>
          <p><code><a href="convertSeuratToSCE.html">convertSeuratToSCE()</a></code> </p>
        </td>
        <td><p>convertSeuratToSCE
Converts the input seurat object to a sce object</p></td>
      </tr><tr><td>
          <p><code><a href="constructSCE.html">constructSCE()</a></code> </p>
        </td>
        <td><p>Create SingleCellExperiment object from csv or txt input</p></td>
      </tr><tr><td>
          <p><code><a href="dedupRowNames.html">dedupRowNames()</a></code> </p>
        </td>
        <td><p>Deduplicate the rownames of a matrix or SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="detectCellOutlier.html">detectCellOutlier()</a></code> </p>
        </td>
        <td><p>Detecting outliers within the SingleCellExperiment object.</p></td>
      </tr><tr><td>
          <p><code><a href="discreteColorPalette.html">discreteColorPalette()</a></code> </p>
        </td>
        <td><p>Generate given number of color codes</p></td>
      </tr><tr><td>
          <p><code><a href="distinctColors.html">distinctColors()</a></code> </p>
        </td>
        <td><p>Generate a distinct palette for coloring different clusters</p></td>
      </tr><tr><td>
          <p><code><a href="downSampleCells.html">downSampleCells()</a></code> </p>
        </td>
        <td><p>Estimate numbers of detected genes, significantly differentially expressed
genes, and median significant effect size</p></td>
      </tr><tr><td>
          <p><code><a href="downSampleDepth.html">downSampleDepth()</a></code> </p>
        </td>
        <td><p>Estimate numbers of detected genes, significantly differentially expressed
genes, and median significant effect size</p></td>
      </tr><tr><td>
          <p><code><a href="featureIndex.html">featureIndex()</a></code> </p>
        </td>
        <td><p>Retrieve row index for a set of features</p></td>
      </tr><tr><td>
          <p><code><a href="retrieveSCEIndex.html">retrieveSCEIndex()</a></code> </p>
        </td>
        <td><p>Retrieve cell/feature index by giving identifiers saved in col/rowData</p></td>
      </tr><tr><td>
          <p><code><a href="sampleSummaryStats.html">sampleSummaryStats()</a></code> </p>
        </td>
        <td><p>Generate table of SCTK QC outputs.</p></td>
      </tr><tr><td>
          <p><code><a href="getSampleSummaryStatsTable.html">getSampleSummaryStatsTable()</a></code> <code><a href="getSampleSummaryStatsTable.html">`setSampleSummaryStatsTable&lt;-`()</a></code> </p>
        </td>
        <td><p>Stores and returns table of SCTK QC outputs to metadata.</p></td>
      </tr><tr><td>
          <p><code><a href="listSampleSummaryStatsTables.html">listSampleSummaryStatsTables()</a></code> </p>
        </td>
        <td><p>Lists the table of SCTK QC outputs stored within the metadata.</p></td>
      </tr><tr><td>
          <p><code><a href="sctkListGeneSetCollections.html">sctkListGeneSetCollections()</a></code> </p>
        </td>
        <td><p>Lists imported GeneSetCollections</p></td>
      </tr><tr><td>
          <p><code><a href="getGenesetNamesFromCollection.html">getGenesetNamesFromCollection()</a></code> </p>
        </td>
        <td><p>List geneset names from geneSetCollection</p></td>
      </tr><tr><td>
          <p><code><a href="setSCTKDisplayRow.html">setSCTKDisplayRow()</a></code> </p>
        </td>
        <td><p>Indicates which rowData to use for visualization</p></td>
      </tr><tr><td>
          <p><code><a href="subsetSCECols.html">subsetSCECols()</a></code> </p>
        </td>
        <td><p>Subset a SingleCellExperiment object by columns</p></td>
      </tr><tr><td>
          <p><code><a href="subsetSCERows.html">subsetSCERows()</a></code> </p>
        </td>
        <td><p>Subset a SingleCellExperiment object by rows</p></td>
      </tr><tr><td>
          <p><code><a href="generateHTANMeta.html">generateHTANMeta()</a></code> </p>
        </td>
        <td><p>Generate HTAN manifest file for droplet and cell count data</p></td>
      </tr><tr><td>
          <p><code><a href="generateMeta.html">generateMeta()</a></code> </p>
        </td>
        <td><p>Generate HTAN manifest file for droplet and cell count data</p></td>
      </tr><tr><td>
          <p><code><a href="generateSimulatedData.html">generateSimulatedData()</a></code> </p>
        </td>
        <td><p>Generates a single simulated dataset, bootstrapping from the input counts
matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="getBiomarker.html">getBiomarker()</a></code> </p>
        </td>
        <td><p>Given a list of genes and a SingleCellExperiment object, return
the binary or continuous expression of the genes.</p></td>
      </tr><tr><td>
          <p><code><a href="getSceParams.html">getSceParams()</a></code> </p>
        </td>
        <td><p>Extract QC parameters from the SingleCellExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="iterateSimulations.html">iterateSimulations()</a></code> </p>
        </td>
        <td><p>Returns significance data from a snapshot.</p></td>
      </tr><tr><td>
          <p><code><a href="mergeSCEColData.html">mergeSCEColData()</a></code> </p>
        </td>
        <td><p>Merging colData from two singleCellExperiment objects</p></td>
      </tr><tr><td>
          <p><code><a href="qcInputProcess.html">qcInputProcess()</a></code> </p>
        </td>
        <td><p>Create SingleCellExperiment object from command line input arguments</p></td>
      </tr><tr><td>
          <p><code><a href="setRowNames.html">setRowNames()</a></code> </p>
        </td>
        <td><p>Set rownames of SCE with a character vector or a rowData column</p></td>
      </tr><tr><td>
          <p><code><a href="subDiffEx.html">subDiffEx()</a></code> <code><a href="subDiffEx.html">subDiffExttest()</a></code> <code><a href="subDiffEx.html">subDiffExANOVA()</a></code> </p>
        </td>
        <td><p>Passes the output of generateSimulatedData() to differential expression
tests, picking either t-tests or ANOVA for data with only two conditions or
multiple conditions, respectively.</p></td>
      </tr><tr><td>
          <p><code><a href="summarizeSCE.html">summarizeSCE()</a></code> </p>
        </td>
        <td><p>Summarize an assay in a <a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="python-environment-setting">Python Environment Setting <a href="#python-environment-setting" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="sctkPythonInstallConda.html">sctkPythonInstallConda()</a></code> </p>
        </td>
        <td><p>Installs Python packages into a Conda environment</p></td>
      </tr><tr><td>
          <p><code><a href="selectSCTKConda.html">selectSCTKConda()</a></code> </p>
        </td>
        <td><p>Selects a Conda environment</p></td>
      </tr><tr><td>
          <p><code><a href="sctkPythonInstallVirtualEnv.html">sctkPythonInstallVirtualEnv()</a></code> </p>
        </td>
        <td><p>Installs Python packages into a virtual environment</p></td>
      </tr><tr><td>
          <p><code><a href="selectSCTKVirtualEnvironment.html">selectSCTKVirtualEnvironment()</a></code> </p>
        </td>
        <td><p>Selects a virtual environment</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

