<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot a heatmap to visualize the result of findMarkerDiffExp — plotMarkerDiffExp • singleCellTK</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Plot a heatmap to visualize the result of findMarkerDiffExp — plotMarkerDiffExp"><meta property="og:description" content='This function will first reads the result saved in
metadata slot, named by "findMarker" and generated by
findMarkerDiffExp. Then it do the filtering on the statistics
based on the input parameters and get unique genes to plot. We choose the
genes that are identified as up-regulated only. As for the genes identified
as up-regulated for multiple clusters, we only keep the belonging towards the
one they have the highest Log2FC value.
In the heatmap, there will always be a cell annotation for the cluster
labeling used when finding the markers, and a feature annotation for which
cluster each gene belongs to. And by default we split the heatmap by these
two annotations. Additional legends can be added and the splitting can be
canceled.
This function will first reads the result saved in
metadata slot, named by "findMarker" and generated by
findMarkerDiffExp. Then it do the filtering on the statistics
based on the input parameters and get unique genes to plot. We choose the
genes that are identified as up-regulated only. As for the genes identified
as up-regulated for multiple clusters, we only keep the belonging towards the
one they have the highest Log2FC value.
In the heatmap, there will always be a cell annotation for the cluster
labeling used when finding the markers, and a feature annotation for which
cluster each gene belongs to. And by default we split the heatmap by these
two annotations. Additional legends can be added and the splitting can be
canceled.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/console_analysis_tutorial.html">A La Carte Workflow - PBMC3K (R Console)</a>
    </li>
    <li>
      <a href="../articles/cnsl_seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/cnsl_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/cnsl_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/cnsl_2d_embedding.html">2D Embedding</a>
        </li>
      </ul></li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Enrichment &amp; Pathway Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/cnsl_enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul></li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot a heatmap to visualize the result of <code><a href="findMarkerDiffExp.html">findMarkerDiffExp</a></code></h1>
    <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/R/findMarker.R" class="external-link"><code>R/findMarker.R</code></a>, <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/R/plotMarkerDiffExp.R" class="external-link"><code>R/plotMarkerDiffExp.R</code></a></small>
    <div class="hidden name"><code>plotMarkerDiffExp.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function will first reads the result saved in
<code>metadata</code> slot, named by <code>"findMarker"</code> and generated by
<code><a href="findMarkerDiffExp.html">findMarkerDiffExp</a></code>. Then it do the filtering on the statistics
based on the input parameters and get unique genes to plot. We choose the
genes that are identified as up-regulated only. As for the genes identified
as up-regulated for multiple clusters, we only keep the belonging towards the
one they have the highest Log2FC value.
In the heatmap, there will always be a cell annotation for the cluster
labeling used when finding the markers, and a feature annotation for which
cluster each gene belongs to. And by default we split the heatmap by these
two annotations. Additional legends can be added and the splitting can be
canceled.</p>
<p>This function will first reads the result saved in
<code>metadata</code> slot, named by <code>"findMarker"</code> and generated by
<code><a href="findMarkerDiffExp.html">findMarkerDiffExp</a></code>. Then it do the filtering on the statistics
based on the input parameters and get unique genes to plot. We choose the
genes that are identified as up-regulated only. As for the genes identified
as up-regulated for multiple clusters, we only keep the belonging towards the
one they have the highest Log2FC value.
In the heatmap, there will always be a cell annotation for the cluster
labeling used when finding the markers, and a feature annotation for which
cluster each gene belongs to. And by default we split the heatmap by these
two annotations. Additional legends can be added and the splitting can be
canceled.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">plotMarkerDiffExp</span><span class="op">(</span>
  <span class="va">inSCE</span>,
  orderBy <span class="op">=</span> <span class="st">"size"</span>,
  log2fcThreshold <span class="op">=</span> <span class="fl">1</span>,
  fdrThreshold <span class="op">=</span> <span class="fl">0.05</span>,
  minClustExprPerc <span class="op">=</span> <span class="fl">0.7</span>,
  maxCtrlExprPerc <span class="op">=</span> <span class="fl">0.4</span>,
  minMeanExpr <span class="op">=</span> <span class="fl">1</span>,
  topN <span class="op">=</span> <span class="fl">10</span>,
  decreasing <span class="op">=</span> <span class="cn">TRUE</span>,
  rowDataName <span class="op">=</span> <span class="cn">NULL</span>,
  colDataName <span class="op">=</span> <span class="cn">NULL</span>,
  featureAnnotations <span class="op">=</span> <span class="cn">NULL</span>,
  cellAnnotations <span class="op">=</span> <span class="cn">NULL</span>,
  featureAnnotationColor <span class="op">=</span> <span class="cn">NULL</span>,
  cellAnnotationColor <span class="op">=</span> <span class="cn">NULL</span>,
  colSplitBy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">orderBy</span><span class="op">)</span>, <span class="cn">NULL</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">inSCE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">findMarker</span><span class="op">)</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span>,
  rowSplitBy <span class="op">=</span> <span class="st">"marker"</span>,
  rowDend <span class="op">=</span> <span class="cn">FALSE</span>,
  colDend <span class="op">=</span> <span class="cn">FALSE</span>,
  title <span class="op">=</span> <span class="st">"Top Marker Heatmap"</span>,
  <span class="va">...</span>
<span class="op">)</span>

<span class="fu">plotMarkerDiffExp</span><span class="op">(</span>
  <span class="va">inSCE</span>,
  orderBy <span class="op">=</span> <span class="st">"size"</span>,
  log2fcThreshold <span class="op">=</span> <span class="fl">1</span>,
  fdrThreshold <span class="op">=</span> <span class="fl">0.05</span>,
  minClustExprPerc <span class="op">=</span> <span class="fl">0.7</span>,
  maxCtrlExprPerc <span class="op">=</span> <span class="fl">0.4</span>,
  minMeanExpr <span class="op">=</span> <span class="fl">1</span>,
  topN <span class="op">=</span> <span class="fl">10</span>,
  decreasing <span class="op">=</span> <span class="cn">TRUE</span>,
  rowDataName <span class="op">=</span> <span class="cn">NULL</span>,
  colDataName <span class="op">=</span> <span class="cn">NULL</span>,
  featureAnnotations <span class="op">=</span> <span class="cn">NULL</span>,
  cellAnnotations <span class="op">=</span> <span class="cn">NULL</span>,
  featureAnnotationColor <span class="op">=</span> <span class="cn">NULL</span>,
  cellAnnotationColor <span class="op">=</span> <span class="cn">NULL</span>,
  colSplitBy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">orderBy</span><span class="op">)</span>, <span class="cn">NULL</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">inSCE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">findMarker</span><span class="op">)</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span>,
  rowSplitBy <span class="op">=</span> <span class="st">"marker"</span>,
  rowDend <span class="op">=</span> <span class="cn">FALSE</span>,
  colDend <span class="op">=</span> <span class="cn">FALSE</span>,
  title <span class="op">=</span> <span class="st">"Top Marker Heatmap"</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>inSCE</dt>
<dd><p>SingleCellExperiment inherited object.</p></dd>
<dt>orderBy</dt>
<dd><p>The ordering method of the clusters on the splitted heatmap.
Can be chosen from <code>"size"</code> or <code>"name"</code>, specified with vector of
ordered unique cluster labels, or set as <code>NULL</code> for unsplitted heatmap.
Default <code>"size"</code>.</p></dd>
<dt>log2fcThreshold</dt>
<dd><p>Only use DEGs with the absolute values of log2FC
larger than this value. Default <code>1</code></p></dd>
<dt>fdrThreshold</dt>
<dd><p>Only use DEGs with FDR value smaller than this value.
Default <code>0.05</code></p></dd>
<dt>minClustExprPerc</dt>
<dd><p>A numeric scalar. The minimum cutoff of the
percentage of cells in the cluster of interests that expressed the marker
gene. Default <code>0.7</code>.</p></dd>
<dt>maxCtrlExprPerc</dt>
<dd><p>A numeric scalar. The maximum cutoff of the
percentage of cells out of the cluster (control group) that expressed the
marker gene. Default <code>0.4</code>.</p></dd>
<dt>minMeanExpr</dt>
<dd><p>A numeric scalar. The minimum cutoff of the mean
expression value of the marker in the cluster of interests. Default <code>1</code>.</p></dd>
<dt>topN</dt>
<dd><p>An integer. Only to plot this number of top markers for each
cluster in maximum, in terms of log2FC value. Use <code>NULL</code> to cancel the
top N subscription. Default <code>10</code>.</p></dd>
<dt>decreasing</dt>
<dd><p>Order the cluster decreasingly. Default <code>TRUE</code>.</p></dd>
<dt>rowDataName</dt>
<dd><p>character. The column name(s) in <code>rowData</code> that need
to be added to the annotation. Default <code>NULL</code>.</p></dd>
<dt>colDataName</dt>
<dd><p>character. The column name(s) in <code>colData</code> that need
to be added to the annotation. Default <code>NULL</code>.</p></dd>
<dt>featureAnnotations</dt>
<dd><p><code>data.frame</code>, with <code>rownames</code> containing
all the features going to be plotted. Character columns should be factors.
Default <code>NULL</code>.</p></dd>
<dt>cellAnnotations</dt>
<dd><p><code>data.frame</code>, with <code>rownames</code> containing
all the cells going to be plotted. Character columns should be factors.
Default <code>NULL</code>.</p></dd>
<dt>featureAnnotationColor</dt>
<dd><p>A named list. Customized color settings for
feature labeling. Should match the entries in the <code>featureAnnotations</code>
or <code>rowDataName</code>. For each entry, there should be a list/vector of
colors named with categories. Default <code>NULL</code>.</p></dd>
<dt>cellAnnotationColor</dt>
<dd><p>A named list. Customized color settings for
cell labeling. Should match the entries in the <code>cellAnnotations</code> or
<code>colDataName</code>. For each entry, there should be a list/vector of colors
named with categories. Default <code>NULL</code>.</p></dd>
<dt>colSplitBy</dt>
<dd><p>character vector. Do semi-heatmap based on the grouping of
this(these) annotation(s). Should exist in either <code>colDataName</code> or
<code>names(cellAnnotations)</code>. Default is the value of <code>cluster</code> in
<code><a href="findMarkerDiffExp.html">findMarkerDiffExp</a></code> when <code>orderBy</code> is not <code>NULL</code>, or
<code>NULL</code> otherwise.</p></dd>
<dt>rowSplitBy</dt>
<dd><p>character vector. Do semi-heatmap based on the grouping of
this(these) annotation(s). Should exist in either <code>rowDataName</code> or
<code>names(featureAnnotations)</code>. Default <code>"marker"</code>, which indicates an
auto generated annotation for this plot.</p></dd>
<dt>rowDend</dt>
<dd><p>Whether to display row dendrogram. Default <code>FALSE</code>.</p></dd>
<dt>colDend</dt>
<dd><p>Whether to display column dendrogram. Default <code>FALSE</code>.</p></dd>
<dt>title</dt>
<dd><p>Text of the title, at the top of the heatmap. Default
<code>"Top Marker Heatmap"</code>.</p></dd>
<dt>...</dt>
<dd><p>Other arguments passed to <code><a href="plotSCEHeatmap.html">plotSCEHeatmap</a></code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></code> object
A <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></code> object</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Yichen Wang</p>
<p>Yichen Wang</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sceBatches"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu">logcounts</span><span class="op">(</span><span class="va">sceBatches</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sceBatches</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span class="r-in"><span class="va">sce.w</span> <span class="op">&lt;-</span> <span class="fu"><a href="subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">sceBatches</span>, colData <span class="op">=</span> <span class="st">"batch == 'w'"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">sce.w</span> <span class="op">&lt;-</span> <span class="fu"><a href="findMarkerDiffExp.html">findMarkerDiffExp</a></span><span class="op">(</span><span class="va">sce.w</span>, method <span class="op">=</span> <span class="st">"wilcox"</span>, cluster <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running with wilcox</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing for cluster: beta</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing for cluster: alpha</span>
<span class="r-in"><span class="fu">plotMarkerDiffExp</span><span class="op">(</span><span class="va">sce.w</span><span class="op">)</span></span>
<span class="r-plt img"><img src="plotMarkerDiffExp-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sceBatches"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu">logcounts</span><span class="op">(</span><span class="va">sceBatches</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sceBatches</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span class="r-in"><span class="va">sce.w</span> <span class="op">&lt;-</span> <span class="fu"><a href="subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">sceBatches</span>, colData <span class="op">=</span> <span class="st">"batch == 'w'"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">sce.w</span> <span class="op">&lt;-</span> <span class="fu"><a href="findMarkerDiffExp.html">findMarkerDiffExp</a></span><span class="op">(</span><span class="va">sce.w</span>, method <span class="op">=</span> <span class="st">"wilcox"</span>, cluster <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running with wilcox</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing for cluster: beta</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing for cluster: alpha</span>
<span class="r-in"><span class="fu">plotMarkerDiffExp</span><span class="op">(</span><span class="va">sce.w</span><span class="op">)</span></span>
<span class="r-plt img"><img src="plotMarkerDiffExp-2.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

