<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Detecting contamination with DecontX. — runDecontX • singleCellTK</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Detecting contamination with DecontX. — runDecontX"><meta property="og:description" content="A wrapper function for decontX. Identify
 potential contamination from experimental factors such as ambient RNA."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.18.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>

  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul></li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul></li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Detecting contamination with DecontX.</h1>
    <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/R/celda_decontX.R" class="external-link"><code>R/celda_decontX.R</code></a></small>
    <div class="hidden name"><code>runDecontX.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A wrapper function for <a href="https://rdrr.io/pkg/celda/man/decontX.html" class="external-link">decontX</a>. Identify
 potential contamination from experimental factors such as ambient RNA.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runDecontX</span><span class="op">(</span></span>
<span>  <span class="va">inSCE</span>,</span>
<span>  sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useAssay <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  background <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bgAssayName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bgBatch <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  z <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxIter <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  delta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  estimateDelta <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  convergence <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  iterLogLik <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  varGenes <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  dbscanEps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">12345</span>,</span>
<span>  logfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-insce">inSCE<a class="anchor" aria-label="anchor" href="#arg-insce"></a></dt>
<dd><p>A <a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a> object.</p></dd>


<dt id="arg-sample">sample<a class="anchor" aria-label="anchor" href="#arg-sample"></a></dt>
<dd><p>A single character specifying a name that can be found in
<code>colData(inSCE)</code> to directly use the cell annotation; or a character
vector with as many elements as cells to indicates which sample each cell
belongs to. Default NULL. <a href="https://rdrr.io/pkg/celda/man/decontX.html" class="external-link">decontX</a> will be run on cells from
each sample separately.</p></dd>


<dt id="arg-useassay">useAssay<a class="anchor" aria-label="anchor" href="#arg-useassay"></a></dt>
<dd><p>A string specifying which assay in the SCE to use. Default
'counts'.</p></dd>


<dt id="arg-background">background<a class="anchor" aria-label="anchor" href="#arg-background"></a></dt>
<dd><p>A <a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>
with the matrix located in the assay slot under <code>bgAssayName</code>. It should have
the same structure as inSCE except it contains the matrix of empty droplets instead
of cells. When supplied, empirical distribution of transcripts from these
empty droplets will be used as the contamination distribution. Default NULL.</p></dd>


<dt id="arg-bgassayname">bgAssayName<a class="anchor" aria-label="anchor" href="#arg-bgassayname"></a></dt>
<dd><p>Character. Name of the assay to use if background is a
<a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>. If NULL, the function
will use the same value as <code>useAssay</code>. Default is NULL.</p></dd>


<dt id="arg-bgbatch">bgBatch<a class="anchor" aria-label="anchor" href="#arg-bgbatch"></a></dt>
<dd><p>Batch labels for <code>background</code>. If <code>background</code> is a
<a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a> object, this can be a single
character specifying a name that can be found in <code>colData(background)</code>
to directly use the barcode annotation; or a numeric / character vector that has
as many elements as barcodes to indicate which sample each barcode belongs to. Its
unique values should be the same as those in <code>sample</code>, such that each
batch of cells have their corresponding batch of empty droplets as background,
pointed by this parameter. Default to NULL.</p></dd>


<dt id="arg-z">z<a class="anchor" aria-label="anchor" href="#arg-z"></a></dt>
<dd><p>Numeric or character vector. Cell cluster labels. If NULL,
PCA will be used to reduce the dimensionality of the dataset initially,
'<a href="https://jlmelville.github.io/uwot/reference/umap.html" class="external-link">umap</a>' from the 'uwot' package
will be used to further reduce the dataset to 2 dimenions and
the '<a href="https://rdrr.io/pkg/dbscan/man/dbscan.html" class="external-link">dbscan</a>' function from the 'dbscan' package
will be used to identify clusters of broad cell types. Default NULL.</p></dd>


<dt id="arg-maxiter">maxIter<a class="anchor" aria-label="anchor" href="#arg-maxiter"></a></dt>
<dd><p>Integer. Maximum iterations of the EM algorithm. Default 500.</p></dd>


<dt id="arg-delta">delta<a class="anchor" aria-label="anchor" href="#arg-delta"></a></dt>
<dd><p>Numeric Vector of length 2. Concentration parameters for
the Dirichlet prior for the contamination in each cell. The first element
is the prior for the native counts while the second element is the prior for
the contamination counts. These essentially act as pseudocounts for the
native and contamination in each cell. If <code>estimateDelta = TRUE</code>,
this is only used to produce a random sample of proportions for an initial
value of contamination in each cell. Then
<code><a href="https://rdrr.io/pkg/MCMCprecision/man/fit_dirichlet.html" class="external-link">fit_dirichlet</a></code> is used to update
<code>delta</code> in each iteration.
If <code>estimateDelta = FALSE</code>, then <code>delta</code> is fixed with these
values for the entire inference procedure. Fixing <code>delta</code> and
setting a high number in the second element will force <code>decontX</code>
to be more aggressive and estimate higher levels of contamination at
the expense of potentially removing native expression.
Default <code>c(10, 10)</code>.</p></dd>


<dt id="arg-estimatedelta">estimateDelta<a class="anchor" aria-label="anchor" href="#arg-estimatedelta"></a></dt>
<dd><p>Boolean. Whether to update <code>delta</code> at each
iteration.</p></dd>


<dt id="arg-convergence">convergence<a class="anchor" aria-label="anchor" href="#arg-convergence"></a></dt>
<dd><p>Numeric. The EM algorithm will be stopped if the maximum
difference in the contamination estimates between the previous and
current iterations is less than this. Default 0.001.</p></dd>


<dt id="arg-iterloglik">iterLogLik<a class="anchor" aria-label="anchor" href="#arg-iterloglik"></a></dt>
<dd><p>Integer. Calculate log likelihood every <code>iterLogLik</code>
iteration. Default 10.</p></dd>


<dt id="arg-vargenes">varGenes<a class="anchor" aria-label="anchor" href="#arg-vargenes"></a></dt>
<dd><p>Integer. The number of variable genes to use in
dimensionality reduction before clustering. Variability is calcualted using
<code><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></code> function from the 'scran' package.
Used only when z is not provided. Default 5000.</p></dd>


<dt id="arg-dbscaneps">dbscanEps<a class="anchor" aria-label="anchor" href="#arg-dbscaneps"></a></dt>
<dd><p>Numeric. The clustering resolution parameter
used in '<a href="https://rdrr.io/pkg/dbscan/man/dbscan.html" class="external-link">dbscan</a>' to estimate broad cell clusters.
Used only when z is not provided. Default 1.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Integer. Passed to <a href="https://withr.r-lib.org/reference/with_seed.html" class="external-link">with_seed</a>. For reproducibility,
a default value of 12345 is used. If NULL, no calls to
<a href="https://withr.r-lib.org/reference/with_seed.html" class="external-link">with_seed</a> are made.</p></dd>


<dt id="arg-logfile">logfile<a class="anchor" aria-label="anchor" href="#arg-logfile"></a></dt>
<dd><p>Character. Messages will be redirected to a file named
`logfile`. If NULL, messages will be printed to stdout.  Default NULL.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical. Whether to print log messages. Default TRUE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a> object with
 'decontX_Contamination' and 'decontX_Clusters' added to the
 colData slot. Additionally, the
decontaminated counts will be added as an assay called 'decontXCounts'.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">scExample</span>, package <span class="op">=</span> <span class="st">"singleCellTK"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">sce</span>, colData <span class="op">=</span> <span class="st">"type != 'EmptyDroplet'"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runDecontX</span><span class="op">(</span><span class="va">sce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,<span class="fl">20</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:38 2025 ... Running 'DecontX'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> --------------------------------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Starting DecontX</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> --------------------------------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:38 2025 .. Analyzing all cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:39 2025 .... Generating UMAP and estimating cell types</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 .... Estimating contamination</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 10 | converge: 0.01985</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 20 | converge: 0.01116</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 30 | converge: 0.006593</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 40 | converge: 0.003919</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 50 | converge: 0.002691</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 60 | converge: 0.002093</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 70 | converge: 0.001932</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 80 | converge: 0.001769</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 90 | converge: 0.001605</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 100 | converge: 0.00145</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 110 | converge: 0.001304</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 120 | converge: 0.001168</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 130 | converge: 0.001041</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 ...... Completed iteration: 134 | converge: 0.0009927</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:40:42 2025 .. Calculating final decontaminated matrix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> --------------------------------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Completed DecontX. Total time: 3.68077 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> --------------------------------------------------</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

