<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Find the marker gene set for each cluster — runFindMarker • singleCellTK</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Find the marker gene set for each cluster — runFindMarker"><meta property="og:description" content="With an input SingleCellExperiment object and specifying the
clustering labels, this function iteratively call the differential expression
analysis on each cluster against all the others. runFindMarker
will be deprecated in the future."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.12.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul></li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul></li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find the marker gene set for each cluster</h1>
    <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/R/runFindMarker.R" class="external-link"><code>R/runFindMarker.R</code></a></small>
    <div class="hidden name"><code>runFindMarker.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>With an input SingleCellExperiment object and specifying the
clustering labels, this function iteratively call the differential expression
analysis on each cluster against all the others. <code>runFindMarker</code>
will be deprecated in the future.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runFindMarker</span><span class="op">(</span></span>
<span>  <span class="va">inSCE</span>,</span>
<span>  useAssay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  useReducedDim <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"wilcox"</span>, <span class="st">"MAST"</span>, <span class="st">"DESeq2"</span>, <span class="st">"Limma"</span>, <span class="st">"ANOVA"</span><span class="op">)</span>,</span>
<span>  cluster <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span>  covariates <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  log2fcThreshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fdrThreshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  minClustExprPerc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxCtrlExprPerc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minMeanExpr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  detectThresh <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">findMarkerDiffExp</span><span class="op">(</span></span>
<span>  <span class="va">inSCE</span>,</span>
<span>  useAssay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  useReducedDim <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"wilcox"</span>, <span class="st">"MAST"</span>, <span class="st">"DESeq2"</span>, <span class="st">"Limma"</span>, <span class="st">"ANOVA"</span><span class="op">)</span>,</span>
<span>  cluster <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span>  covariates <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  log2fcThreshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fdrThreshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  minClustExprPerc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxCtrlExprPerc <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minMeanExpr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  detectThresh <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>inSCE</dt>
<dd><p>SingleCellExperiment inherited object.</p></dd>


<dt>useAssay</dt>
<dd><p>character. A string specifying which assay to use for the
MAST calculations. Default <code>"logcounts"</code>.</p></dd>


<dt>useReducedDim</dt>
<dd><p>character. A string specifying which reducedDim to use
for MAST calculations. Set <code>useAssay</code> to <code>NULL</code> when using.
Required.</p></dd>


<dt>method</dt>
<dd><p>A single character for specific differential expression
analysis method. Choose from <code>'wilcox'</code>, <code>'MAST'</code>, <code>'DESeq2'</code>,
<code>'Limma'</code>, and <code>'ANOVA'</code>. Default <code>"wilcox"</code>.</p></dd>


<dt>cluster</dt>
<dd><p>One single character to specify a column in
<code>colData(inSCE)</code> for the clustering label. Alternatively, a vector or
a factor is also acceptable. Default <code>"cluster"</code>.</p></dd>


<dt>covariates</dt>
<dd><p>A character vector of additional covariates to use when
building the model. All covariates must exist in
<code>names(colData(inSCE))</code>. Not applicable when <code>method</code> is
<code>"MAST"</code> method. Default <code>NULL</code>.</p></dd>


<dt>log2fcThreshold</dt>
<dd><p>Only out put DEGs with the absolute values of log2FC
larger than this value. Default <code>NULL</code></p></dd>


<dt>fdrThreshold</dt>
<dd><p>Only out put DEGs with FDR value smaller than this
value. Default <code>NULL</code></p></dd>


<dt>minClustExprPerc</dt>
<dd><p>A numeric scalar. The minimum cutoff of the
percentage of cells in the cluster of interests that expressed the marker
gene. From 0 to 1. Default <code>NULL</code>.</p></dd>


<dt>maxCtrlExprPerc</dt>
<dd><p>A numeric scalar. The maximum cutoff of the
percentage of cells out of the cluster (control group) that expressed the
marker gene. From 0 to 1. Default <code>NULL</code>.</p></dd>


<dt>minMeanExpr</dt>
<dd><p>A numeric scalar. The minimum cutoff of the mean
expression value of the marker in the cluster of interests. Default
<code>NULL</code>.</p></dd>


<dt>detectThresh</dt>
<dd><p>A numeric scalar, above which a matrix value will be
treated as expressed when calculating cluster/control expression percentage.
Default <code>0</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>The input SingleCellExperiment object with
<code>metadata(inSCE)$findMarker</code> updated with a data.table of the up-
regulated DEGs for each cluster.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The returned marker table, in the <code>metadata</code> slot, consists of 8
columns: <code>"Gene"</code>, <code>"Log2_FC"</code>, <code>"Pvalue"</code>, <code>"FDR"</code>,
<code>cluster</code>, <code>"clusterExprPerc"</code>, <code>"ControlExprPerc"</code> and
<code>"clusterAveExpr"</code>.</p>
<p><code>"clusterExprPerc"</code> is the fraction of cells,
that has marker value (e.g. gene expression counts) larger than
<code>detectThresh</code>, in the cell population of the cluster. As for each
cluster, we set all cells out of this cluster as control. Similarly,
<code>"ControlExprPerc"</code> is the fraction of cells with marker value larger
than <code>detectThresh</code> in the control cell group.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="runDEAnalysis.html">runDEAnalysis</a></code>, <code><a href="getFindMarkerTopTable.html">getFindMarkerTopTable</a></code>,
<code><a href="plotFindMarkerHeatmap.html">plotFindMarkerHeatmap</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"mouseBrainSubsetSCE"</span>, package <span class="op">=</span> <span class="st">"singleCellTK"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mouseBrainSubsetSCE</span> <span class="op">&lt;-</span> <span class="fu">runFindMarker</span><span class="op">(</span><span class="va">mouseBrainSubsetSCE</span>,</span></span>
<span class="r-in"><span>                                     useAssay <span class="op">=</span> <span class="st">"logcounts"</span>,</span></span>
<span class="r-in"><span>                                     cluster <span class="op">=</span> <span class="st">"level1class"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Fri Feb  2 14:21:45 2024 ... Identifying markers for cluster 'microglia', using DE method 'wilcox'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Fri Feb  2 14:21:46 2024 ... Identifying markers for cluster 'oligodendrocytes', using DE method 'wilcox'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Fri Feb  2 14:21:47 2024 ... Organizing findMarker result</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

