<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Wrapper for calculating QC metrics with scater. — runPerCellQC • singleCellTK</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Wrapper for calculating QC metrics with scater. — runPerCellQC" />
<meta property="og:description" content="A wrapper function for addPerCellQC. Calculate
 general quality control metrics for each cell in the count matrix." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/articles/introduction_to_SCTK.html">General Workflow</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shell Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../articles/articles/cmd_qc.html">Quality Control</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    R Console Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/articles/cnsl_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_dropletqc.html">Quality Control of droplet data</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_cellqc.html">Quality Control of cell data</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_batch_correction.html">Batch Correction</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_feature_selection.html">Feature Selection</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../articles/articles/cnsl_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_find_marker.html">Find Marker via DE</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_heatmap.html">Heatmap Visualization</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/articles/cnsl_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shiny UI Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/articles/ui_import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/articles/ui_import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/articles/ui_export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/articles/ui_delete_data.html">Delete Data</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_qc.html">Quality Control</a>
        </li>
        <li>
          <a href="../articles/articles/ui_filtering.html">Filtering</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/articles/ui_batch_correction.html">Batch Correction</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/articles/ui_dimemsionality_reduction.html">Dimensionality Reduction</a>
        </li>
      </ul>
    </li>
    <li>
      <a href="../articles/articles/ui_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/articles/ui_find_marker.html">Find Marker</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_seurat_curated_workflow.html">Seurat</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="../articles/articles/ui_cellViewer.html">cellViewer</a>
        </li>
        <li>
          <a href="../articles/articles/ui_heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Wrapper for calculating QC metrics with scater.</h1>
    <small class="dont-index">Source: <a href='https://github.com/compbiomed/singleCellTK/blob/master/R/scater_addPerCellQC.R'><code>R/scater_addPerCellQC.R</code></a></small>
    <div class="hidden name"><code>runPerCellQC.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A wrapper function for <a href='https://rdrr.io/pkg/scuttle/man/addPerCellQC.html'>addPerCellQC</a>. Calculate
 general quality control metrics for each cell in the count matrix.</p>
    </div>

    <pre class="usage"><span class='fu'>runPerCellQC</span><span class='op'>(</span>
  <span class='va'>inSCE</span>,
  useAssay <span class='op'>=</span> <span class='st'>"counts"</span>,
  collectionName <span class='op'>=</span> <span class='cn'>NULL</span>,
  geneSetList <span class='op'>=</span> <span class='cn'>NULL</span>,
  geneSetListLocation <span class='op'>=</span> <span class='st'>"rownames"</span>,
  geneSetCollection <span class='op'>=</span> <span class='cn'>NULL</span>,
  percent_top <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>50</span>, <span class='fl'>100</span>, <span class='fl'>200</span>, <span class='fl'>500</span><span class='op'>)</span>,
  use_altexps <span class='op'>=</span> <span class='cn'>FALSE</span>,
  flatten <span class='op'>=</span> <span class='cn'>TRUE</span>,
  detectionLimit <span class='op'>=</span> <span class='fl'>0</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'>BiocParallel</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html'>SerialParam</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>inSCE</th>
      <td><p>Input <a href='https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html'>SingleCellExperiment</a> object.</p></td>
    </tr>
    <tr>
      <th>useAssay</th>
      <td><p>A string specifying which assay in the SCE to use. Default
<code>"counts"</code>.</p></td>
    </tr>
    <tr>
      <th>collectionName</th>
      <td><p>Character. Name of a <code>GeneSetCollection</code> obtained
by using one of the importGeneSet* functions. Default <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>geneSetList</th>
      <td><p>List of gene sets to be quantified. The genes in the
assays will be matched to the genes in the list based on
<code>geneSetListLocation</code>. Default <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>geneSetListLocation</th>
      <td><p>Character or numeric vector. If set to 'rownames',
then the genes in 'geneSetList' will be looked up in <code><a href='https://rdrr.io/r/base/colnames.html'>rownames(inSCE)</a></code>.
If another character is supplied, then genes will be looked up in the column
 names of <code>rowData(inSCE)</code>. A character vector with the same length as
 <code>geneSetList</code> can be supplied if the IDs for different
gene sets are found in different places, including a mixture of 'rownames'
and <code>rowData(inSCE)</code>. An integer or integer vector can be supplied to
denote the column index in <code>rowData(inSCE)</code>. Default 'rownames'.</p></td>
    </tr>
    <tr>
      <th>geneSetCollection</th>
      <td><p>Class of <code>GeneSetCollection</code> from package
<code>GSEAbase</code>. The location of the gene IDs in <code>inSCE</code> should be in
the <code>description</code> slot of each gene set and should follow the
same notation as <code>geneSetListLocation</code>. The function <a href='https://rdrr.io/pkg/GSEABase/man/getObjects.html'>getGmt</a> can
be used to read in gene sets from a GMT file. If reading a GMT file, the
second column for each gene set should be the description denoting the
location of the gene IDs in <code>inSCE</code>. These gene sets will be included
with those from <code>geneSetList</code> if both parameters are provided.</p></td>
    </tr>
    <tr>
      <th>percent_top</th>
      <td><p>An integer vector. Each element is treated as a
number of top genes to compute the percentage of library size occupied by
the most highly expressed genes in each cell.</p></td>
    </tr>
    <tr>
      <th>use_altexps</th>
      <td><p>Logical scalar indicating whether QC statistics should
be computed for alternative Experiments in x. If TRUE, statistics are
computed
for all alternative experiments.
Alternatively, an integer or character vector specifying the alternative
Experiments to use to compute QC statistics.
Alternatively NULL, in which case alternative experiments are not used.</p></td>
    </tr>
    <tr>
      <th>flatten</th>
      <td><p>Logical scalar indicating whether the nested
<a href='https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html'>DataFrame-class</a>
in the output should be flattened.</p></td>
    </tr>
    <tr>
      <th>detectionLimit</th>
      <td><p>A numeric scalar specifying the lower detection limit
for expression.</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>A BiocParallelParam object specifying
whether the QC calculations should be parallelized.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <a href='https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html'>SingleCellExperiment</a> object with
 cell QC metrics added to the colData slot. If <code>geneSetList</code> or
 <code>geneSetCollection</code> are provided, then the rownames for each gene set
 will be saved in <code>metadata(inSCE)$scater$addPerCellQC$geneSets</code>.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>scExample</span>, package <span class='op'>=</span> <span class='st'>"singleCellTK"</span><span class='op'>)</span>
<span class='va'>mito.ix</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grep</a></span><span class='op'>(</span><span class='st'>"^MT-"</span>, <span class='fu'>rowData</span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>$</span><span class='va'>feature_name</span><span class='op'>)</span>
<span class='va'>geneSet</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='st'>"Mito"</span><span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>sce</span><span class='op'>)</span><span class='op'>[</span><span class='va'>mito.ix</span><span class='op'>]</span><span class='op'>)</span>
<span class='va'>sce</span> <span class='op'>&lt;-</span> <span class='fu'>runPerCellQC</span><span class='op'>(</span><span class='va'>sce</span>, geneSetList <span class='op'>=</span> <span class='va'>geneSet</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Tue May 11 18:06:09 2021 ... Running 'perCellQCMetrics'</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


