<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Find doublets using scrublet. — runScrublet • singleCellTK</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Find doublets using scrublet. — runScrublet"><meta property="og:description" content="A wrapper function that calls scrub_doublets from python
 module scrublet. Simulates doublets from the observed data and uses
 a k-nearest-neighbor classifier to calculate a continuous
 scrublet_score (between 0 and 1) for each transcriptome. The score
 is automatically thresholded to generate scrublet_call, a boolean
 array that is TRUE for predicted doublets and FALSE
 otherwise."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.18.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>

  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul></li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul></li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find doublets using <code>scrublet</code>.</h1>
    <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/R/scrublet_doubletDetection.R" class="external-link"><code>R/scrublet_doubletDetection.R</code></a></small>
    <div class="hidden name"><code>runScrublet.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A wrapper function that calls <code>scrub_doublets</code> from python
 module <code>scrublet</code>. Simulates doublets from the observed data and uses
 a k-nearest-neighbor classifier to calculate a continuous
 <code>scrublet_score</code> (between 0 and 1) for each transcriptome. The score
 is automatically thresholded to generate <code>scrublet_call</code>, a boolean
 array that is <code>TRUE</code> for predicted doublets and <code>FALSE</code>
 otherwise.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runScrublet</span><span class="op">(</span></span>
<span>  <span class="va">inSCE</span>,</span>
<span>  sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useAssay <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  simDoubletRatio <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  nNeighbors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minDist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  expectedDoubletRate <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  stdevDoubletRate <span class="op">=</span> <span class="fl">0.02</span>,</span>
<span>  syntheticDoubletUmiSubsampling <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  useApproxNeighbors <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  distanceMetric <span class="op">=</span> <span class="st">"euclidean"</span>,</span>
<span>  getDoubletNeighborParents <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  minCounts <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  minCells <span class="op">=</span> <span class="fl">3L</span>,</span>
<span>  minGeneVariabilityPctl <span class="op">=</span> <span class="fl">85</span>,</span>
<span>  logTransform <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  meanCenter <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  normalizeVariance <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  nPrinComps <span class="op">=</span> <span class="fl">30L</span>,</span>
<span>  tsneAngle <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tsnePerplexity <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">12345</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-insce">inSCE<a class="anchor" aria-label="anchor" href="#arg-insce"></a></dt>
<dd><p>A SingleCellExperiment object.</p></dd>


<dt id="arg-sample">sample<a class="anchor" aria-label="anchor" href="#arg-sample"></a></dt>
<dd><p>Character vector or colData variable name. Indicates which
sample each cell belongs to. Default <code>NULL</code>.</p></dd>


<dt id="arg-useassay">useAssay<a class="anchor" aria-label="anchor" href="#arg-useassay"></a></dt>
<dd><p>A string specifying which assay in the SCE to use. Default
<code>"counts"</code>.</p></dd>


<dt id="arg-simdoubletratio">simDoubletRatio<a class="anchor" aria-label="anchor" href="#arg-simdoubletratio"></a></dt>
<dd><p>Numeric. Number of doublets to simulate relative to
the number of observed transcriptomes. Default <code>2.0</code>.</p></dd>


<dt id="arg-nneighbors">nNeighbors<a class="anchor" aria-label="anchor" href="#arg-nneighbors"></a></dt>
<dd><p>Integer. Number of neighbors used to construct the KNN
graph of observed transcriptomes and simulated doublets. If <code>NULL</code>,
this is set to <code>round(0.5 * sqrt(n_cells))</code>. Default <code>NULL</code>.</p></dd>


<dt id="arg-mindist">minDist<a class="anchor" aria-label="anchor" href="#arg-mindist"></a></dt>
<dd><p>Float Determines how tightly UMAP packs points together. If
<code>NULL</code>, this is set to <code>0.1</code>. Default <code>NULL</code>.</p></dd>


<dt id="arg-expecteddoubletrate">expectedDoubletRate<a class="anchor" aria-label="anchor" href="#arg-expecteddoubletrate"></a></dt>
<dd><p>The estimated doublet rate for the experiment.
Default <code>0.1</code>.</p></dd>


<dt id="arg-stdevdoubletrate">stdevDoubletRate<a class="anchor" aria-label="anchor" href="#arg-stdevdoubletrate"></a></dt>
<dd><p>Uncertainty in the expected doublet rate. Default
<code>0.02</code>.</p></dd>


<dt id="arg-syntheticdoubletumisubsampling">syntheticDoubletUmiSubsampling<a class="anchor" aria-label="anchor" href="#arg-syntheticdoubletumisubsampling"></a></dt>
<dd><p>Numeric. Rate for sampling UMIs when
creating synthetic doublets. If <code>1.0</code>, each doublet is created by simply
adding the UMIs from two randomly sampled observed transcriptomes. For
values less than 1, the UMI counts are added and then randomly sampled at
the specified rate. Defuault <code>1.0</code>.</p></dd>


<dt id="arg-useapproxneighbors">useApproxNeighbors<a class="anchor" aria-label="anchor" href="#arg-useapproxneighbors"></a></dt>
<dd><p>Boolean. Use approximate nearest neighbor method
(annoy) for the KNN classifier. Default <code>TRUE</code>.</p></dd>


<dt id="arg-distancemetric">distanceMetric<a class="anchor" aria-label="anchor" href="#arg-distancemetric"></a></dt>
<dd><p>Character. Distance metric used when finding nearest
neighbors. See detail. Default <code>"euclidean"</code>.</p></dd>


<dt id="arg-getdoubletneighborparents">getDoubletNeighborParents<a class="anchor" aria-label="anchor" href="#arg-getdoubletneighborparents"></a></dt>
<dd><p>Boolean. If <code>TRUE</code>, return the
parent transcriptomes that generated the doublet neighbors of each
observed transcriptome. This information can be used to infer the cell states
that generated a given doublet state. Default <code>FALSE</code>.</p></dd>


<dt id="arg-mincounts">minCounts<a class="anchor" aria-label="anchor" href="#arg-mincounts"></a></dt>
<dd><p>Numeric. Used for gene filtering prior to PCA. Genes
expressed at fewer than <code>minCounts</code> in fewer than <code>minCells</code> are
excluded. Default <code>3</code>.</p></dd>


<dt id="arg-mincells">minCells<a class="anchor" aria-label="anchor" href="#arg-mincells"></a></dt>
<dd><p>Integer. Used for gene filtering prior to PCA. Genes
expressed at fewer than <code>minCounts</code> in fewer than <code>minCells</code> are
excluded. Default <code>3</code>.</p></dd>


<dt id="arg-mingenevariabilitypctl">minGeneVariabilityPctl<a class="anchor" aria-label="anchor" href="#arg-mingenevariabilitypctl"></a></dt>
<dd><p>Numeric. Used for gene filtering prior to
PCA. Keep the most highly variable genes (in the top
<code>minGeneVariabilityPctl</code> percentile), as measured by the v-statistic
(Klein et al., Cell 2015). Default <code>85</code>.</p></dd>


<dt id="arg-logtransform">logTransform<a class="anchor" aria-label="anchor" href="#arg-logtransform"></a></dt>
<dd><p>Boolean. If <code>TRUE</code>, log-transform the counts matrix
(log1p(TPM)). <code>sklearn.decomposition.TruncatedSVD</code> will be used for
dimensionality reduction, unless <code>meanCenter</code> is <code>TRUE</code>. Default
<code>FALSE</code>.</p></dd>


<dt id="arg-meancenter">meanCenter<a class="anchor" aria-label="anchor" href="#arg-meancenter"></a></dt>
<dd><p>If <code>TRUE</code>, center the data such that each gene has a
mean of <code>0</code>. <code>sklearn.decomposition.PCA</code> will be used for
dimensionality reduction. Default <code>TRUE</code>.</p></dd>


<dt id="arg-normalizevariance">normalizeVariance<a class="anchor" aria-label="anchor" href="#arg-normalizevariance"></a></dt>
<dd><p>Boolean. If <code>TRUE</code>, normalize the data such
that each gene has a variance of 1.
<code>sklearn.decomposition.TruncatedSVD</code> will be used for dimensionality
reduction, unless <code>meanCenter</code> is <code>TRUE</code>. Default <code>TRUE</code>.</p></dd>


<dt id="arg-nprincomps">nPrinComps<a class="anchor" aria-label="anchor" href="#arg-nprincomps"></a></dt>
<dd><p>Integer. Number of principal components used to embed
the transcriptomes prior to k-nearest-neighbor graph construction.
Default <code>30</code>.</p></dd>


<dt id="arg-tsneangle">tsneAngle<a class="anchor" aria-label="anchor" href="#arg-tsneangle"></a></dt>
<dd><p>Float. Determines angular size of a distant node as measured
from a point in the t-SNE plot. If <code>NULL</code>, it is set to <code>0.5</code>.
Default <code>NULL</code>.</p></dd>


<dt id="arg-tsneperplexity">tsnePerplexity<a class="anchor" aria-label="anchor" href="#arg-tsneperplexity"></a></dt>
<dd><p>Integer. The number of nearest neighbors that is used
in other manifold learning algorithms. If <code>NULL</code>, it is set to 30.
Default <code>NULL</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Boolean. If <code>TRUE</code>, print progress updates. Default
<code>TRUE</code>.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Seed for the random number generator, can be set to <code>NULL</code>.
Default <code>12345</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A SingleCellExperiment object with
<code>scrub_doublets</code> output appended to the colData slot. The columns
include <code>scrublet_score</code> and <code>scrublet_call</code>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>For the list of valid values for <code>distanceMetric</code>, see the
documentation for
<a href="https://github.com/spotify/annoy#full-python-api" class="external-link">annoy</a> (if
<code>useApproxNeighbors</code> is <code>TRUE</code>) or
<a href="https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html" class="external-link">sklearn.neighbors.NearestNeighbors</a>
(if <code>useApproxNeighbors</code> is  <code>FALSE</code>).</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="plotScrubletResults.html">plotScrubletResults</a></code>, <code><a href="runCellQC.html">runCellQC</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">scExample</span>, package <span class="op">=</span> <span class="st">"singleCellTK"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">sce</span>, colData <span class="op">=</span> <span class="st">"type != 'EmptyDroplet'"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runScrublet</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

