<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run t-SNE embedding with Rtsne method — runTSNE • singleCellTK</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Run t-SNE embedding with Rtsne method — runTSNE"><meta property="og:description" content="T-Stochastic Neighbour Embedding (t-SNE) algorithm is commonly
for 2D visualization of single-cell data. This function wraps the
Rtsne Rtsne function.
With this funciton, users can create tSNE embedding directly from raw count
matrix, with necessary preprocessing including normalization, scaling,
dimension reduction all automated. Yet we still recommend having the PCA as
input, so that the result can match with the clustering based on the same
input PCA, and will be much faster."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.18.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>

  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul></li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul></li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Web Application</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/webApp.html">FAQs</a>
        </li>
      </ul></li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run t-SNE embedding with Rtsne method</h1>
    <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/R/runTSNE.R" class="external-link"><code>R/runTSNE.R</code></a></small>
    <div class="hidden name"><code>runTSNE.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>T-Stochastic Neighbour Embedding (t-SNE) algorithm is commonly
for 2D visualization of single-cell data. This function wraps the
Rtsne <code><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html" class="external-link">Rtsne</a></code> function.</p>
<p>With this funciton, users can create tSNE embedding directly from raw count
matrix, with necessary preprocessing including normalization, scaling,
dimension reduction all automated. Yet we still recommend having the PCA as
input, so that the result can match with the clustering based on the same
input PCA, and will be much faster.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runTSNE</span><span class="op">(</span></span>
<span>  <span class="va">inSCE</span>,</span>
<span>  useReducedDim <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>  useAssay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useAltExp <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reducedDimName <span class="op">=</span> <span class="st">"TSNE"</span>,</span>
<span>  logNorm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  useFeatureSubset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nTop <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  pca <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  partialPCA <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  initialDims <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  theta <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  perplexity <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  nIterations <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  numThreads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">12345</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">runQuickTSNE</span><span class="op">(</span><span class="va">inSCE</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">getTSNE</span><span class="op">(</span></span>
<span>  <span class="va">inSCE</span>,</span>
<span>  useReducedDim <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>  useAssay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useAltExp <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reducedDimName <span class="op">=</span> <span class="st">"TSNE"</span>,</span>
<span>  logNorm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  useFeatureSubset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nTop <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  pca <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  partialPCA <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  initialDims <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  theta <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  perplexity <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  nIterations <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  numThreads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">12345</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-insce">inSCE<a class="anchor" aria-label="anchor" href="#arg-insce"></a></dt>
<dd><p>Input SingleCellExperiment object.</p></dd>


<dt id="arg-usereduceddim">useReducedDim<a class="anchor" aria-label="anchor" href="#arg-usereduceddim"></a></dt>
<dd><p>The low dimension representation to use for UMAP
computation. Default <code>"PCA"</code>.</p></dd>


<dt id="arg-useassay">useAssay<a class="anchor" aria-label="anchor" href="#arg-useassay"></a></dt>
<dd><p>Assay to use for tSNE computation. If <code>useAltExp</code> is
specified, <code>useAssay</code> has to exist in
<code>assays(altExp(inSCE, useAltExp))</code>. Default <code>NULL</code>.</p></dd>


<dt id="arg-usealtexp">useAltExp<a class="anchor" aria-label="anchor" href="#arg-usealtexp"></a></dt>
<dd><p>The subset to use for tSNE computation, usually for the
selected.variable features. Default <code>NULL</code>.</p></dd>


<dt id="arg-reduceddimname">reducedDimName<a class="anchor" aria-label="anchor" href="#arg-reduceddimname"></a></dt>
<dd><p>a name to store the results of the dimension
reductions. Default <code>"TSNE"</code>.</p></dd>


<dt id="arg-lognorm">logNorm<a class="anchor" aria-label="anchor" href="#arg-lognorm"></a></dt>
<dd><p>Whether the counts will need to be log-normalized prior to
generating the tSNE via <code><a href="scaterlogNormCounts.html">scaterlogNormCounts</a></code>. Ignored when using
<code>useReducedDim</code>. Default <code>TRUE</code>.</p></dd>


<dt id="arg-usefeaturesubset">useFeatureSubset<a class="anchor" aria-label="anchor" href="#arg-usefeaturesubset"></a></dt>
<dd><p>Subset of feature to use for dimension reduction. A
character string indicating a <code>rowData</code> variable that stores the logical
vector of HVG selection, or a vector that can subset the rows of
<code>inSCE</code>. Default <code>NULL</code>.</p></dd>


<dt id="arg-ntop">nTop<a class="anchor" aria-label="anchor" href="#arg-ntop"></a></dt>
<dd><p>Automatically detect this number of variable features to use for
dimension reduction. Ignored when using <code>useReducedDim</code> or using
<code>useFeatureSubset</code>. Default <code>2000</code>.</p></dd>


<dt id="arg-center">center<a class="anchor" aria-label="anchor" href="#arg-center"></a></dt>
<dd><p>Whether data should be centered before PCA is applied. Ignored
when using <code>useReducedDim</code>. Default <code>TRUE</code>.</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>Whether data should be scaled before PCA is applied. Ignored
when using <code>useReducedDim</code>. Default <code>TRUE</code>.</p></dd>


<dt id="arg-pca">pca<a class="anchor" aria-label="anchor" href="#arg-pca"></a></dt>
<dd><p>Whether an initial PCA step should be performed. Ignored when
using <code>useReducedDim</code>. Default <code>TRUE</code>.</p></dd>


<dt id="arg-partialpca">partialPCA<a class="anchor" aria-label="anchor" href="#arg-partialpca"></a></dt>
<dd><p>Whether truncated PCA should be used to calculate principal
components (requires the irlba package). This is faster for large input
matrices. Ignored when using <code>useReducedDim</code>. Default <code>FALSE</code>.</p></dd>


<dt id="arg-initialdims">initialDims<a class="anchor" aria-label="anchor" href="#arg-initialdims"></a></dt>
<dd><p>Number of dimensions from PCA to use as input in tSNE.
Default <code>25</code>.</p></dd>


<dt id="arg-theta">theta<a class="anchor" aria-label="anchor" href="#arg-theta"></a></dt>
<dd><p>Numeric value for speed/accuracy trade-off (increase for less
accuracy), set to <code>0.0</code> for exact TSNE. Default <code>0.5</code>.</p></dd>


<dt id="arg-perplexity">perplexity<a class="anchor" aria-label="anchor" href="#arg-perplexity"></a></dt>
<dd><p>perplexity parameter. Should not be bigger than
<code>3 * perplexity &lt; ncol(inSCE) - 1</code>. Default <code>30</code>. See
<code><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html" class="external-link">Rtsne</a></code> details for interpretation.</p></dd>


<dt id="arg-niterations">nIterations<a class="anchor" aria-label="anchor" href="#arg-niterations"></a></dt>
<dd><p>maximum iterations. Default <code>1000</code>.</p></dd>


<dt id="arg-numthreads">numThreads<a class="anchor" aria-label="anchor" href="#arg-numthreads"></a></dt>
<dd><p>Integer, number of threads to use using OpenMP, Default
<code>1</code>. <code>0</code> corresponds to using all available cores.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Random seed for reproducibility of tSNE results.
Default <code>NULL</code> will use global seed in use by the R environment.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other parameters to be passed to <code>runTSNE</code></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A SingleCellExperiment object with tSNE computation
updated in <code>reducedDim(inSCE, reducedDimName)</code>.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">scExample</span>, package <span class="op">=</span> <span class="st">"singleCellTK"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">sce</span>, colData <span class="op">=</span> <span class="st">"type != 'EmptyDroplet'"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Run from raw counts</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runQuickTSNE</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wed Apr 23 11:42:25 2025 ... Computing Rtsne.</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Run from PCA</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="scaterlogNormCounts.html">scaterlogNormCounts</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"logcounts"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="runModelGeneVar.html">runModelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="getTopHVG.html">setTopHVG</a></span><span class="op">(</span><span class="va">sce</span>, method <span class="op">=</span> <span class="st">"modelGeneVar"</span>, hvgNumber <span class="op">=</span> <span class="fl">2000</span>, </span></span>
<span class="r-in"><span>                 featureSubsetName <span class="op">=</span> <span class="st">"HVG_modelGeneVar2000"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="scaterPCA.html">scaterPCA</a></span><span class="op">(</span><span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"logcounts"</span>,</span></span>
<span class="r-in"><span>                 useFeatureSubset <span class="op">=</span> <span class="st">"HVG_modelGeneVar2000"</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">runTSNE</span><span class="op">(</span><span class="va">sce</span>, useReducedDim <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Ming Liu, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

